/**
This is the CARD Antibiotic Resistance Ontology, which is released under
the Creative Commons CC-BY license version 4.0 by McMaster University.
https://creativecommons.org/licenses/by/4.0/
https://card.mcmaster.ca/download
**/

export const CARD_AMR_ONTOLOGY = [
  {
    accession: "ARO:0000000",
    name: "macrolide antibiotic",
    description:
      "Macrolides are a group of drugs (typically antibiotics) that have a large macrocyclic lactone ring of 12-16 carbons to which one or more deoxy sugars, usually cladinose and desosamine, may be attached. Macrolides bind to the 50S-subunit of bacterial ribosomes, inhibiting the synthesis of vital proteins.",
  },
  {
    accession: "ARO:0000001",
    name: "fluoroquinolone antibiotic",
    description:
      "The fluoroquinolones are a family of synthetic broad-spectrum antibiotics that are 4-quinolone-3-carboxylates. These compounds interact with topoisomerase II (DNA gyrase) to disrupt bacterial DNA replication, damage DNA, and cause cell death.",
  },
  {
    accession: "ARO:0000002",
    name: "tetracycline-resistant ribosomal protection protein",
    description:
      "A family of proteins known to bind to the 30S ribosomal subunit. This interaction prevents tetracycline and tetracycline derivatives from inhibiting ribosomal function. Thus, these proteins confer elevated resistance to tetracycline derivatives as a ribosomal protection protein.",
  },
  {
    accession: "ARO:0000003",
    name: "astromicin",
    description:
      "Astromicin is an aminoglycoside antibiotic used to treat different types of bacterial infections. Astromicin works by binding to the bacterial 30S ribosomal subunit, causing misreading of mRNA and leaving the bacterium unable to synthesize proteins vital to its growth.",
  },
  {
    accession: "ARO:0000004",
    name: "monobactam",
    description:
      "Monobactams are a class of beta-lactam antibiotics with a broad spectrum of antibacterial activity, and have a structure which renders them highly resistant to beta-lactamases. Unlike penams and cephems, monobactams do not have any ring fused to its four-member lactam structure. Monobactam antibiotics are bactericidal, and act by inhibiting the synthesis of the peptidoglycan layer of bacterial cell walls. The peptidoglycan layer is important for cell wall structural integrity, especially in Gram-positive organisms.",
  },
  {
    accession: "ARO:0000005",
    name: "neomycin",
    description:
      "Neomycin is an aminoglycoside antibiotic used to treat different types of bacterial infections. Neomycin works by binding to the bacterial 30S ribosomal subunit, causing misreading of mRNA and leaving the bacterium unable to synthesize proteins vital to its growth.",
  },
  {
    accession: "ARO:0000006",
    name: "erythromycin",
    description:
      "Erythromycin is a macrolide antibiotic with a 14-carbon ring that has an antimicrobial spectrum similar to or slightly wider than that of penicillin, and is often used for people that have an allergy to penicillins. Erythromycin may possess bacteriocidal activity, particularly at higher concentrations by binding to the 50S subunit of the bacterial 70S rRNA complex, inhibiting peptidyl-tRNA translocation. Thus, protein synthesis and subsequently structure/function processes critical for life or replication are inhibited.",
  },
  {
    accession: "ARO:0000007",
    name: "dibekacin",
    description:
      "Dibekacin is an aminoglycoside antibiotic used to treat different types of bacterial infections. Dibekacin works by binding to the bacterial 30S ribosomal subunit, causing misreading of mRNA and leaving the bacterium unable to synthesize proteins vital to its growth.",
  },
  {
    accession: "ARO:0000008",
    name: "cefoxitin",
    description:
      "Cefoxitin is a cephamycin antibiotic often grouped with the second generation cephalosporins. Cefoxitin is bactericidal and acts by inhibiting the synthesis of the peptidoglycan layer of bacterial cell walls. The peptidoglycan layer is important for cell wall structural integrity, especially in Gram-positive organisms. Cefoxitin's 7-alpha-methoxy group and 3' leaving group make it a poor substrate for most beta-lactamases.",
  },
  {
    accession: "ARO:0000009",
    name: "tunicamycin",
    description:
      "Tunicamycin is mixture of homologous nucleoside antibiotics that block the reaction of UDP-N-acetylglucosamine and dolichyl phosphate in the first step of glycoprotein synthesis.",
  },
  {
    accession: "ARO:0000010",
    name: "antibiotic resistance gene cluster, cassette, or operon",
    description:
      "Clusters of antibiotic resistance genes. May be regulated by a shared promoter or repressor.",
  },
  {
    accession: "ARO:0000011",
    name: "cloxacillin",
    description:
      "Cloxacillin is a semisynthetic, isoxazolyl penicillin derivative in the beta-lactam class of antibiotics. It interferes with peptidogylcan synthesis and is commonly used for treating penicillin-resistant Staphylococcus aureus infections.",
  },
  {
    accession: "ARO:0000012",
    name: "streptothricin",
    description:
      "Streptothricins are a group of N-glycoside antibiotics that include a carbamoylated D-glucosamine to which are attached a series of L-beta-lysine residues at position 2 and a streptolidine at position 1.  Streptothricins vary by the number of beta-lysine residues (from 1 (nourseothricin) to 7) and target protein synthesis in bacteria and eukaryotes.",
  },
  {
    accession: "ARO:0000013",
    name: "amikacin",
    description:
      "Amikacin is an aminoglycoside antibiotic that works by binding to the bacterial 30S ribosomal subunit, causing misreading of mRNA and leaving the bacterium unable to synthesize proteins vital to its growth.",
  },
  {
    accession: "ARO:0000014",
    name: "gentamicin C",
    description:
      "Gentamicin C is a mixture of gentamicin C1, gentamicin C1a, and gentamicin C2 (these differ in substituents at position C6'). Gentamicin works by binding to the bacterial 30S ribosomal subunit, causing misreading of mRNA and leaving the bacterium unable to synthesize proteins vital to its growth.",
  },
  {
    accession: "ARO:0000015",
    name: "methicillin",
    description:
      "Derived from penicillin to combat penicillin-resistance, methicillin is insensitive to beta-lactamases (also known as penicillinases) secreted by many penicillin-resistant bacteria. Methicillin is bactericidal, and acts by inhibiting the synthesis of the peptidoglycan layer of bacterial cell walls. The peptidoglycan layer is important for cell wall structural integrity, especially in Gram-positive organisms.",
  },
  {
    accession: "ARO:0000016",
    name: "aminoglycoside antibiotic",
    description:
      "Aminoglycosides are a group of antibiotics that are mostly effective against Gram-negative bacteria. These molecules consist of aminated sugars attached to a dibasic cyclitol. Aminoglycosides work by binding to the bacterial 30S ribosomal subunit (some work by binding to the 50S subunit), inhibiting the translocation of the peptidyl-tRNA from the A-site to the P-site and also causing misreading of mRNA, leaving the bacterium unable to synthesize proteins vital to its growth.",
  },
  {
    accession: "ARO:0000017",
    name: "lincosamide antibiotic",
    description:
      "Lincosamides (e.g. lincomycin, clindamycin) are a class of drugs which bind to the 23s portion of the 50S subunit of bacterial ribosomes. This interaction inhibits early elongation of peptide chains by inhibiting the transpeptidase reaction, acting similarly to macrolides.",
  },
  {
    accession: "ARO:0000018",
    name: "viomycin",
    description:
      "Viomycin sulfate (Viocin) is an polypeptide antibiotic used in the treatment of tuberculosis. It is produced by the actinomycete Streptomyces puniceus and binds to the bacterial ribosome, inhibiting prokaryotic protein synthesis and certain forms of RNA splicing.",
  },
  {
    accession: "ARO:0000020",
    name: "carbapenem",
    description:
      "Carbapenems are a class of beta-lactam antibiotics with a broad spectrum of antibacterial activity, and have a structure which renders them highly resistant to beta-lactamases. Carbapenem antibiotics are bactericidal, and act by inhibiting the synthesis of the peptidoglycan layer of bacterial cell walls. The peptidoglycan layer is important for cell wall structural integrity, especially in Gram-positive organisms.",
  },
  {
    accession: "ARO:0000021",
    name: "ribostamycin",
    description:
      "Ribostamycin is an aminoglycoside antibiotic used to treat different types of bacterial infections. Ribostamycin works by binding to the bacterial 30S ribosomal subunit, causing misreading of mRNA and leaving the bacterium unable to synthesize proteins vital to its growth.",
  },
  {
    accession: "ARO:0000022",
    name: "polymyxin antibiotic",
    description:
      "Polymyxins are cationic detergent antibiotics, with a general structure of a cyclic peptide with a long hydrophobic tail. They disrupt the structure of the bacterial cell membrane by interacting with its phospholipids. Polymyxins have a bactericidal effect on Gram-negative bacilli, especially on Pseudomonas and coliform organisms.",
  },
  {
    accession: "ARO:0000023",
    name: "enoxacin",
    description:
      "Enoxacin belongs to a group called fluoroquinolones. Its mode of action depends upon blocking bacterial DNA replication by binding itself to DNA gyrase and causing double-stranded breaks in the bacterial chromosome.",
  },
  {
    accession: "ARO:0000024",
    name: "butirosin",
    description:
      "Butirosin is an aminoglycoside antibiotic used to treat different types of bacterial infections. Butirosin works by binding to the bacterial 30S ribosomal subunit, causing misreading of mRNA and leaving the bacterium unable to synthesize proteins vital to its growth.",
  },
  {
    accession: "ARO:0000025",
    name: "fosfomycin",
    description:
      "Fosfomycin (also known as phosphomycin and phosphonomycin) is a broad-spectrum antibiotic produced by certain Streptomyces species. It is effective on gram positive and negative bacteria as it targets the cell wall, an essential feature shared by both bacteria. Its specific target is MurA (MurZ in E.coli), which attaches phosphoenolpyruvate (PEP) to UDP-N-acetylglucosamine, a step of commitment to cell wall synthesis. In the active site of MurA, the active cysteine molecule is alkylated which stops the catalytic reaction.",
  },
  {
    accession: "ARO:0000026",
    name: "streptogramin antibiotic",
    description:
      "Streptogramin antibiotics are natural products produced by various members of the Streptomyces genus. These antibiotics bind to the P site of the 50S subunit of bacterial ribosomes to inhibit protein synthesis. The family consists of two subgroups, type A and type B, which are simultaneously produced by the same bacterial species in a ratio of roughly 70:30.",
  },
  {
    accession: "ARO:0000027",
    name: "roxithromycin",
    description:
      "Roxithromycin is a semi-synthetic, 14-carbon ring macrolide antibiotic derived from erythromycin. It is used to treat respiratory tract, urinary and soft tissue infections. Roxithromycin may possess bacteriocidal activity, particularly at higher concentrations by binding to the 50S subunit of the bacterial 70S rRNA complex, protein synthesis and subsequently structure/function processes critical for life or replication are inhibited.",
  },
  {
    accession: "ARO:0000028",
    name: "vancomycin",
    description:
      "Vancomycin is a glycopeptide antibiotic used in the prophylaxis and treatment of infections caused by Gram-positive bacteria. Vancomycin inhibits the synthesis of peptidoglycan, the major component of the cell wall of gram-positive bacteria. Its mechanism of action is unusual in that it acts by binding precursors of peptidoglycan, rather than by interacting with an enzyme.",
  },
  {
    accession: "ARO:0000029",
    name: "teicoplanin",
    description:
      "Teicoplanin is a glycopeptide antibiotic used in the prophylaxis and treatment of serious infections caused by Gram-positive bacteria. Teicoplanin has a unique acyl-aliphatic chain, and binds to cell wall precursors to inhibit transglycosylation  and transpeptidation.",
  },
  {
    accession: "ARO:0000030",
    name: "tigecycline",
    description:
      "Tigecycline is an glycylcycline antibiotic. It works by inhibiting action of the prokaryotic 30S ribosome.",
  },
  {
    accession: "ARO:0000031",
    name: "antibiotic resistant gene variant or mutant",
    description:
      "Resistance to antibiotics is often conferred by single nucleotide polymorphisms (SNPs) and other mutations in target genes.",
  },
  {
    accession: "ARO:0000032",
    name: "cephalosporin",
    description:
      "Cephalosporins are a class of beta-lactam antibiotics, containing the beta-lactam ring fused with a dihydrothiazolidine ring. Together with cephamycins they belong to a sub-group called cephems. Cephalosporin are bactericidal, and act by inhibiting the synthesis of the peptidoglycan layer of bacterial cell walls. The peptidoglycan layer is important for cell wall structural integrity, especially in Gram-positive organisms.",
  },
  {
    accession: "ARO:0000034",
    name: "streptogramin A antibiotic",
    description:
      "Streptogramin A antibiotics are cyclic polyketide peptide hybrids that bind to the ribosomal peptidyl transfer centre. Structural variation arises from substituting a proline for its desaturated derivative and by its substitution for Ala or Cys. Used alone, streptogramin A antibiotics are bacteriostatic, but is bactericidal when used with streptogramin B antibiotics.",
  },
  {
    accession: "ARO:0000035",
    name: "sisomicin",
    description:
      "Sisomicin is an aminoglycoside antibiotic used to treat different types of bacterial infections. Sisomicin works by binding to the bacterial 30S ribosomal subunit, causing misreading of mRNA and leaving the bacterium unable to synthesize proteins vital to its growth.",
  },
  {
    accession: "ARO:0000036",
    name: "ciprofloxacin",
    description:
      "Ciprofloxacin is a bacteriocidal fluoroquinolone. It blocks bacterial DNA replication by binding to the toposiomerase II or IV-DNA complex (or cleavable complex), thereby causing double-stranded breaks in the bacterial chromosome.",
  },
  {
    accession: "ARO:0000037",
    name: "apramycin",
    description:
      "Apramycin is an aminoglycoside antibiotic used to treat different types of bacterial infections in animals. Apramycin works by binding to the bacterial 30S ribosomal subunit, causing misreading of mRNA and leaving the bacterium unable to synthesize proteins vital to its growth.",
  },
  {
    accession: "ARO:0000038",
    name: "netilmicin",
    description:
      "Netilmicin is a member of the aminoglycoside family of antibiotics. These antibiotics have the ability to kill a wide variety of bacteria by binding to the bacterial 30S ribosomal subunit, causing misreading of mRNA and leaving the bacterium unable to synthesize proteins vital to its growth. Netilmicin is not absorbed from the gut and is therefore only given by injection or infusion. It is only used in the treatment of serious infections particularly those resistant to gentamicin.",
  },
  {
    accession: "ARO:0000039",
    name: "spectinomycin",
    description:
      "Spectinomycin is an aminoglycoside antibiotic used to treat different types of bacterial infections. Spectinomycin works by binding to the bacterial 30S ribosomal subunit inhibiting translation.",
  },
  {
    accession: "ARO:0000040",
    name: "streptomycin",
    description:
      "Streptomycin is an aminoglycoside antibiotic used to treat different types of bacterial infections. Streptomycin works by binding to the bacterial 30S ribosomal subunit, causing misreading of mRNA and leaving the bacterium unable to synthesize proteins vital to its growth.",
  },
  {
    accession: "ARO:0000041",
    name: "bacitracin",
    description:
      "Bacitracin is a mixture of related cyclic polypeptides produced by organisms of the licheniformis group of Bacillus subtilis var Tracy. Bacitracin interferes with the dephosphorylation of the C55-isoprenyl pyrophosphate, a molecule which carries the building blocks of the peptidoglycan bacterial cell wall outside of the inner membrane.",
  },
  {
    accession: "ARO:0000042",
    name: "glycylcycline",
    description:
      "Glycylcyclines are a new class of antibiotics derived from tetracycline. These tetracycline analogues are specifically designed to overcome two common mechanisms of tetracycline resistance. Presently, there is only one glycylcycline antibiotic for clinical use: tigecycline. It works by inhibiting action of the prokaryotic 30S ribosome, preventing the binding of aminoacyl-tRNA.",
  },
  {
    accession: "ARO:0000043",
    name: "carbenicillin",
    description:
      "Carbenicillin is a semi-synthetic antibiotic belonging to the carboxypenicillin subgroup of the penicillins. It has gram-negative coverage which includes Pseudomonas aeruginosa but limited gram-positive coverage. The carboxypenicillins are susceptible to degradation by beta-lactamase enzymes. Carbenicillin antibiotics are bactericidal, and act by inhibiting the synthesis of the peptidoglycan layer of bacterial cell walls. The peptidoglycan layer is important for cell wall structural integrity, especially in Gram-positive organisms.",
  },
  {
    accession: "ARO:0000044",
    name: "cephamycin",
    description:
      "Cephamycins are a group of beta-lactam antibiotics, very similar to cephalosporins. Together with cephalosporins, they form a sub-group of antibiotics known as cephems. Cephamycins are bactericidal, and act by inhibiting the synthesis of the peptidoglycan layer of bacterial cell walls. The peptidoglycan layer is important for cell wall structural integrity, especially in Gram-positive organisms. The 7-alpha-methoxy group increases resistance to beta-lactamases.",
  },
  {
    accession: "ARO:0000045",
    name: "acriflavine",
    description:
      "Acriflavine is a topical antiseptic. It has the form of an orange or brown powder. It may be harmful in the eyes or if inhaled. Acriflavine is also used as treatment for external fungal infections of aquarium fish.",
  },
  {
    accession: "ARO:0000046",
    name: "lincomycin",
    description:
      "Lincomycin is a lincosamide antibiotic that comes from the actinomyces Streptomyces lincolnensis. It binds to the 23s portion of the 50S subunit of bacterial ribosomes and inhibit early elongation of peptide chain by inhibiting transpeptidase reaction.",
  },
  {
    accession: "ARO:0000047",
    name: "puromycin",
    description:
      "Puromycin is an aminonucleoside antibiotic, derived from Streptomyces alboniger, that causes premature chain termination during ribosomal protein translation.",
  },
  {
    accession: "ARO:0000049",
    name: "kanamycin A",
    description:
      "Kanamycin is an aminoglycoside antibiotic used to treat different types of bacterial infections. Kanamycin works by binding to the bacterial 30S ribosomal subunit, causing misreading of mRNA and leaving the bacterium unable to synthesize proteins vital to its growth.",
  },
  {
    accession: "ARO:0000050",
    name: "streptogramin B antibiotic",
    description:
      "Streptogramin B antibiotics are are cyclic hepta- or hexa-depsipeptides.  Type B streptogramins block the peptide exit tunnel of the 50S bacterial ribosome. The general composition of group B streptogramins is 3-hydroxypicolinic acid-L-Thr-D-aminobutyric acid (or D-Ala)-L-Pro-L-Phe (or 4-N-,N-(dimethylamino)-L-Phe)-X-L-phenylglycine. Used alone, streptogramin B antibiotics are bacteriostatic, but is bactericidal when used with streptogramin A antibiotics.",
  },
  {
    accession: "ARO:0000051",
    name: "tetracycline",
    description:
      "Tetracycline is a broad-spectrum polyketide antibiotic produced by many Streptomyces. It works by inhibiting action of the prokaryotic 30S ribosome.",
  },
  {
    accession: "ARO:0000052",
    name: "tobramycin",
    description:
      "Tobramycin is an aminoglycoside antibiotic used to treat different types of bacterial infections. Tobramycin works by binding to the bacterial 30S ribosomal subunit, causing misreading of mRNA and leaving the bacterium unable to synthesize proteins vital to its growth.",
  },
  {
    accession: "ARO:0000053",
    name: "bleomycin",
    description:
      "Bleomycin is a family of glycopeptide antibiotics produced by the bacterium Streptomyces verticillus. Bleomycins, taken as a mixture, act by the induction of DNA and RNA strand breaks. In addition to its antibacterial activity, bleomycin is also used as an anticancer agent.",
  },
  {
    accession: "ARO:0000054",
    name: "penicillin",
    description:
      "Penicillin (sometimes abbreviated PCN) is a beta-lactam antibiotic used in the treatment of bacterial infections caused by susceptible, usually Gram-positive, organisms. It works by inhibiting the synthesis of the peptidoglycan layer of bacterial cell walls. The peptidoglycan layer is important for cell wall structural integrity, especially in Gram-positive organisms.",
  },
  {
    accession: "ARO:0000055",
    name: "bicyclomycin",
    description:
      "Bicyclomycin represents a unique class of antibiotics, discovered in 1972. It is obtained by the fermentation of Streptomyces sapporonensis. In the crystalline form bicyclomycin is observed to be rhombic or monoclinic, depending on the solvent used. This antibiotic kills bacteria by inhibiting the Rho transcription terminator factor, halting ribonucleic acid (RNA) synthesis.",
  },
  {
    accession: "ARO:0000056",
    name: "oxacillin",
    description:
      "Oxacillin is a penicillinase-resistant beta-lactam. It is similar to methicillin, and has replaced methicillin in clinical use. Oxacillin, especially in combination with other antibiotics, is effective against many penicillinase-producing strains of Staphylococcus aureus and Staphylococcus epidermidis.",
  },
  {
    accession: "ARO:0000057",
    name: "telithromycin",
    description:
      "Telithromycin is a semi-synthetic derivative of erythromycin. It is a 14-membered macrolide and is the first ketolide antibiotic to be used in clinics. Telithromycin binds the 50S subunit of the bacterial ribosome to inhibit protein synthesis.",
  },
  {
    accession: "ARO:0000058",
    name: "cefazolin",
    description:
      "Cefazolin (INN), also known as cefazoline or cephazolin, is a first generation cephalosporin antibiotic. It is administered parenterally, and is active against a broad spectrum of bacteria.",
  },
  {
    accession: "ARO:0000059",
    name: "cefepime",
    description:
      "Cefepime (INN) is a fourth-generation cephalosporin antibiotic developed in 1994. It contains an aminothiazolyl group that decreases its affinity with beta-lactamases. Cefepime shows high binding affinity with penicillin-binding proteins and has an extended spectrum of activity against Gram-positive and Gram-negative bacteria, with greater activity against both Gram-negative and Gram-positive organisms than third-generation agents.",
  },
  {
    accession: "ARO:0000060",
    name: "ceftazidime",
    description:
      "Ceftazidime is a third-generation cephalosporin antibiotic. Like other third-generation cephalosporins, it has broad spectrum activity against Gram-positive and Gram-negative bacteria. Unlike most third-generation agents, it is active against Pseudomonas aeruginosa, however it has weaker activity against Gram-positive microorganisms and is not used for such infections.",
  },
  {
    accession: "ARO:0000061",
    name: "ceftobiprole",
    description:
      "Ceftobiprole (Zeftera/Zevtera) is a next generation (5th generation) cephalosporin antibiotic with activity against methicillin-resistant Staphylococcus aureus, penicillin-resistant Streptococcus pneumoniae, Pseudomonas aeruginosa, and Enterococci. Ceftobiprole inhibits transpeptidases essential to building cell walls, and is a poor substrate for most beta-lactamases.",
  },
  {
    accession: "ARO:0000062",
    name: "ceftriaxone",
    description:
      "Ceftriaxone is a third-generation cephalosporin antibiotic. The presence of an aminothiazolyl sidechain increases ceftriazone's resistance to beta-lactamases. Like other third-generation cephalosporins, it has broad spectrum activity against Gram-positive and Gram-negative bacteria.",
  },
  {
    accession: "ARO:0000063",
    name: "cefuroxime",
    description:
      "Cefuroxime is a second-generation cephalosporin antibiotic with increased stability with beta-lactamases than first-generation cephalosporins. Cefuroxime is active against Gram-positive organisms but less active against methicillin-resistant strains.",
  },
  {
    accession: "ARO:0000064",
    name: "amoxicillin",
    description:
      "Amoxicillin is a moderate-spectrum, bacteriolytic, beta-lactam antibiotic used to treat bacterial infections caused by susceptible microorganisms. A derivative of penicillin, it has a wider range of treatment but remains relatively ineffective against Gram-negative bacteria. It is commonly taken with clavulanic acid, a beta-lactamase inhibitor. Like other beta-lactams, amoxicillin interferes with the synthesis of peptidoglycan.",
  },
  {
    accession: "ARO:0000065",
    name: "clarithromycin",
    description:
      "Clarithromycin is a methyl derivative of erythromycin, sharing the 14-carbon macrolide ring. The antibiotic binds to the 50S subunit of the ribosome and is used to treat pharyngitis, tonsillitis, acute maxillary sinusitis, acute bacterial exacerbation of chronic bronchitis, pneumonia (especially atypical pneumonias associated with Chlamydia pneumoniae or TWAR), and skin structure infections.",
  },
  {
    accession: "ARO:0000066",
    name: "clindamycin",
    description:
      "Clindamycin is a lincosamide antibiotic that blocks A-site aminoacyl-tRNA binding. It is usually used to treat infections with anaerobic bacteria but can also be used to treat some protozoal diseases, such as malaria.",
  },
  {
    accession: "ARO:0000067",
    name: "colistin",
    description:
      "Colistins are polymyxin antibiotics produced by certain strains of Bacillus polymyxa var. colistinus. Colistin, also referred to as polymyxin E, is a mixture of cyclic polypeptides colistin A and B which disrupt the bacterial cell membrane and is effective against Gram-negative bacteria.",
  },
  {
    accession: "ARO:0000068",
    name: "daptomycin",
    description:
      "Daptomycin is a novel lipopeptide antibiotic used in the treatment of certain infections caused by Gram-positive organisms. Daptomycin interferes with the bacterial cell membrane, reducing membrane potential and inhibiting cell wall synthesis.",
  },
  {
    accession: "ARO:0000069",
    name: "doxycycline",
    description:
      "Doxycycline is second generation semi-synthetic derivative of the tetracycline group of antibiotics. It inhibits bacterial protein synthesis by binding to the 30S subunit of the bacterial ribosome and preventing the aminotransferase-tRNA from associating with the ribosome.",
  },
  {
    accession: "ARO:0000070",
    name: "ertapenem",
    description:
      "Ertapenem is a carbapenem antibiotic and is highly resistant to beta-lactamases like other carbapenems. It inhibits bacterial cell wall synthesis.",
  },
  {
    accession: "ARO:0000071",
    name: "levofloxacin",
    description:
      "Levofloxacin is a synthetic chemotherapeutic antibiotic of the fluoroquinolone drug class. Its main target is topoisomerase IV, inhibiting its function and disrupting DNA replication.",
  },
  {
    accession: "ARO:0000072",
    name: "linezolid",
    description:
      "Linezolid is a synthetic antibiotic used for the treatment of serious infections caused by Gram-positive bacteria that are resistant to several other antibiotics. It inhibits protein synthesis by binding to domain V of the 23S rRNA of the 50S subunit of bacterial ribosomes.",
  },
  {
    accession: "ARO:0000073",
    name: "meropenem",
    description:
      "Meropenem is an ultra-broad spectrum injectable antibiotic used to treat a wide variety of infections, including meningitis and pneumonia. It is a beta-lactam and belongs to the subgroup of carbapenem, similar to imipenem and ertapenem.",
  },
  {
    accession: "ARO:0000074",
    name: "moxifloxacin",
    description:
      "Moxifloxacin is a fourth generation synthetic fluoroquinolone chemotherapeutic agent, and has been shown to be significantly more active than levofloxacin (4 to 8 times more) against Streptococcus pneumoniae. It acts by inhibiting bacterial DNA topoisomerases.",
  },
  {
    accession: "ARO:0000075",
    name: "nitrofurantoin",
    description:
      "Nitrofurantoin is an antibiotic used to treat urinary tract infections. It inhibits enzyme synthesis by inhibiting essential enzymes involved in the citric acid cycle, as well as those involved in DNA, RNA, and protein synthesis. It is marketed under the following brand names: Furadantin, Macrobid, Macrodantin, Nitro Macro and Urantoin.",
  },
  {
    accession: "ARO:0000076",
    name: "resistance-modifying agents",
    description:
      "Resistance-modifying agents (RMA) include antibiotic adjuvants and other inhibitors of antibiotic resistance, as well as antibiotic potentiators. These are non-antibiotic compounds which act to block resistance mechanisms or enhance antibiotic action. These are often delivered in combination with an antibiotic (e.g. amoxicillin-clavulanic acid) and may either affect the host organism or the pathogen. Adjuvants and potentiators are therefore used to rescue the activity of existing antibiotic drugs, and are researched as an alternative solution to the antibiotic resistance crisis.",
  },
  {
    accession: "ARO:0000077",
    name: "tazobactam",
    description:
      "Tazobactam is a compound which inhibits the action of bacterial beta-lactamases.",
  },
  {
    accession: "ARO:0000078",
    name: "piperacillin",
    description:
      "Piperacillin is an acetylureidopenicillin and has an extended spectrum of targets relative to other beta-lactam antibiotics. It inhibits cell wall synthesis in bacteria, and is usually taken with the beta-lactamase inhibitor tazobactam to overcome penicillin-resistant bacteria.",
  },
  {
    accession: "ARO:0000079",
    name: "clavulanate",
    description:
      "Clavulanic acid is a beta-lactamase inhibitor (marketed by GlaxoSmithKline, formerly Beecham) combined with penicillin group antibiotics to overcome certain types of antibiotic resistance. It is used to overcome resistance in bacteria that secrete beta-lactamase, which otherwise inactivates most penicillins.",
  },
  {
    accession: "ARO:0000092",
    name: "skin lesion",
    description:
      "Skin lesion is a skin and integumentary tissue symptom characterized as an abnormal change in structure of the skin that is especially circumscribed and well defined due to injury or disease.",
  },
  {
    accession: "ARO:0001001",
    name: "antibiotic target alteration",
    description:
      "Mutational alteration or enzymatic modification of antibiotic target which results in antibiotic resistance.",
  },
  {
    accession: "ARO:0001002",
    name: "antibiotic target replacement",
    description:
      "Replacement or substitution of antibiotic action target, which process will result in antibiotic resistance.",
  },
  {
    accession: "ARO:0001003",
    name: "antibiotic target protection",
    description:
      "Protection of antibiotic action target from antibiotic binding, which process will result in antibiotic resistance.",
  },
  {
    accession: "ARO:0001003",
    name: "skin epidermis",
    description:
      "The outer epithelial layer of the skin that is superficial to the dermis.",
  },
  {
    accession: "ARO:0001004",
    name: "antibiotic inactivation",
    description:
      "Enzymatic inactivation of antibiotic to confer drug resistance.",
  },
  {
    accession: "ARO:0003134",
    name: "female reproductive organ",
    description: "A female organ involved in reproduction.",
  },
  {
    accession: "ARO:0003135",
    name: "male reproductive organ",
    description: "A male organ involved in reproduction.",
  },
  {
    accession: "ARO:0010000",
    name: "antibiotic efflux",
    description:
      "Antibiotic resistance via the transport of antibiotics out of the cell.",
  },
  {
    accession: "ARO:0010001",
    name: "ATP-binding cassette (ABC) antibiotic efflux pump",
    description:
      "Directed pumping of antibiotic out of a cell to confer resistance. ATP-binding cassette (ABC) transporters are present in all cells of all organisms and use the energy of ATP binding/hydrolysis to transport substrates across cell membranes.",
  },
  {
    accession: "ARO:0010002",
    name: "major facilitator superfamily (MFS) antibiotic efflux pump",
    description:
      "Directed pumping of antibiotic out of a cell to confer resistance. Major facilitator superfamily (MFS) transporters and ABC transporters comprise the two largest and most functionally diverse of the transporter superfamilies. However, MFS transporters are distinct from ABC transporters in both their primary sequence and structure and in the mechanism of energy coupling. As secondary transporters they are, like RND and SMR transporters, energized by the electrochemical proton gradient.",
  },
  {
    accession: "ARO:0010003",
    name: "small multidrug resistance (SMR) antibiotic efflux pump",
    description:
      "Directed pumping of antibiotic out of a cell to confer resistance. Small multidrug resistance (SMR) proteins are a relatively small family of transporters, restricted to prokaryotic cells. They are also the smallest multidrug transporters, with only four transmembrane alpha-helices and no significant extramembrane domain.",
  },
  {
    accession: "ARO:0010004",
    name: "resistance-nodulation-cell division (RND) antibiotic efflux pump",
    description:
      "Directed pumping of antibiotic out of a cell to confer resistance. Resistance-nodulation-division (RND) proteins are found in both prokaryotic and eukaryotic cells and have diverse substrate specificities and physiological roles. However, there are relatively few RND transporters and they are secondary transporters, energized not by ATP binding/hydrolysis but by proton movement down the transmembrane electrochemical gradient.",
  },
  {
    accession: "ARO:0031292",
    name: "cutaneous abscess",
    description: "A circumscribed area of pus or necrotic debris in the skin.",
  },
  {
    accession: "ARO:0050161",
    name: "lower respiratory tract disease",
    description:
      "A respiratory system disease which involves the lower respiratory tract.",
  },
  {
    accession: "ARO:1000001",
    name: "process or component of antibiotic biology or chemistry",
    description:
      "Antibiotic resistance is the ability of a microorganism to withstand the effects of an antibiotic. It is a specific type of drug resistance.",
  },
  {
    accession: "ARO:1000002",
    name: "mechanism of antibiotic resistance",
    description:
      "Antibiotic resistance mechanisms evolve naturally via natural selection through random mutation, but it could also be engineered by applying an evolutionary stress on a population.",
  },
  {
    accession: "ARO:1000003",
    name: "antibiotic molecule",
    description:
      "Antibiotics are commonly classified based on their mechanism of action, chemical structure or spectrum of activity.",
  },
  {
    accession: "ARO:11162",
    name: "respiratory failure",
    description:
      "A lung disease characterized by inadequate gas exchange by the respiratory system.",
  },
  {
    accession: "ARO:11394",
    name: "adult respiratory distress sydrome",
    description:
      "A respiratory failure that results from diffuse injury to the endothelium of the lung (as in sepsis, chest trauma, massive blood transfusion, aspiration of the gastric contents, or pneumonia) and is characterized by pulmonary edema with an abnormally high amount of protein in the edematous fluid and by difficult rapid breathing and hypoxemia.",
  },
  {
    accession: "ARO:12716",
    name: "newborn respiratory distress syndrome",
    description:
      "A respiratory failure that is characterized by deficiency of the surfactant coating the inner surface of the lungs, by failure of the lungs to expand and contract properly during breathing with resulting collapse, and by the accumulation of a protein-containing film lining the alveoli and their ducts.",
  },
  {
    accession: "ARO:13148",
    name: "acute cystitis",
    description:
      "A cystitis characterized by a sudden onset or severe symptoms.",
  },
  {
    accession: "ARO:1579",
    name: "respiratory system disease",
    description:
      "A disease of anatomical entity that located_in the respiratory system which extends from the nasal sinuses to the diaphragm.",
  },
  {
    accession: "ARO:3000000",
    name: "determinant of antibiotic resistance",
    description:
      "A mutation, single nucleotide polymorphism, gene, or gene product that confers antibiotic resistance.",
  },
  {
    accession: "ARO:3000001",
    name: "beta-lactamase",
    description:
      "Beta-lactamases are enzymes (EC 3.5.2.6) produced by some bacteria and are responsible for their resistance to beta-lactam antibiotics like penicillins, cephalosporins (are relatively resistant to beta-lactamase), cephamycins, and carbapenems (ertapenem). These antibiotics have a common element in their molecular structure: a four-atom ring known as a beta-lactam. The lactamase enzyme breaks that ring open, deactivating the molecule's antibacterial properties. Beta-lactam antibiotics are typically used to treat a broad spectrum of gram-positive and gram-negative bacteria. Beta-lactamases produced by gram-negative organisms are usually secreted.",
  },
  {
    accession: "ARO:3000002",
    name: "vanW",
    description:
      "vanW is an accessory gene, with unknown function, found on vancomycin resistance operons.",
  },
  {
    accession: "ARO:3000003",
    name: "antibiotic without defined classification",
    description:
      "These compounds are antibiotics of unique structure or origin, without a defined classification.",
  },
  {
    accession: "ARO:3000004",
    name: "class B (metallo-) beta-lactamase",
    description:
      "Ambler Class B beta-lactamases are the metallo-beta-lactamases.  These enzymes possess one or two zinc ions in the active site, which are used to orient a hydroxide nucleophile to attack the carbonyl on a beta-lactam ring.  There are currently no inhibitors in late-stage development for these relatively new beta-lactamases.",
  },
  {
    accession: "ARO:3000005",
    name: "vanD",
    description:
      "VanD is a D-Ala-D-Ala ligase homolog similar to VanA, and can synthesize D-Ala-D-Lac, an alternative substrate for peptidoglycan synthesis that reduces vancomycin binding affinity. It is associated with both vancomycin and teicoplanin resistance.",
  },
  {
    accession: "ARO:3000006",
    name: "vanH",
    description:
      "VanH is a D-specific alpha-ketoacid dehydrogenase that synthesizes D-lactate. D-lactate is incorporated into the end of the peptidoglycan subunits, decreasing vancomycin binding affinity.",
  },
  {
    accession: "ARO:3000007",
    name: "beta-lactam antibiotic",
    description:
      "Beta-lactam antibiotics are a broad class of antibiotics that include penams (penicillin derivatives), cephems (cephalosporins), monobactams, and carbapenems. These antibiotic agents contain a beta-lactam nucleus in its molecular structure. They are the most widely-used group of antibiotics.",
  },
  {
    accession: "ARO:3000008",
    name: "penam",
    description:
      "Penams, often referred to as penicillins, are a group of antibiotics derived from Penicillium fungi. Penicillin antibiotics are historically significant because they are the first drugs that were effective against many previously serious diseases such as syphilis and Staphylococcus infections. Penicillins are still widely used today, though many types of bacteria are now resistant. All penicillins are beta-lactam antibiotics in the penam sub-group, and are used in the treatment of bacterial infections caused by susceptible, usually Gram-positive, organisms.",
  },
  {
    accession: "ARO:3000010",
    name: "vanA",
    description:
      "VanA is a D-Ala-D-Ala ligase homolog that synthesizes D-Ala-D-Lac, an alternative substrate for peptidoglycan synthesis that reduces vancomycin binding affinity. It has been isolated from VREs. It is associated with both vancomycin and teicoplanin resistance.",
  },
  {
    accession: "ARO:3000011",
    name: "vanX",
    description:
      "VanX is a D,D-dipeptidase that cleaves D-Ala-D-Ala but not D-Ala-D-Lac, ensuring that the latter dipeptide that has reduced binding affinity with vancomycin is used to synthesize peptidoglycan substrate.",
  },
  {
    accession: "ARO:3000012",
    name: "protein(s) conferring antibiotic resistance via molecular bypass",
    description:
      "Proteins involved in restructuring of the cell wall, causing antibiotic resistance.",
  },
  {
    accession: "ARO:3000013",
    name: "vanB",
    description:
      "VanB is a D-Ala-D-Ala ligase homolog similar to VanA, and can synthesize D-Ala-D-Lac, an alternative substrate for peptidoglycan synthesis that reduces vancomycin binding affinity. It has been isolated from VREs. It is associated with vancomycin resistance, but not teicoplanin resistance.",
  },
  {
    accession: "ARO:3000014",
    name: "TEM beta-lactamase",
    description:
      "TEM-1 is the most commonly-encountered beta-lactamase in gram-negative bacteria. Up to 90% of ampicillin resistance in E. coli is due to the production of TEM-1. Also responsible for the ampicillin and penicillin resistance that is seen in H. influenzae and N. gonorrhoeae in increasing numbers. Although TEM-type beta-lactamases are most often found in E. coli and K. pneumoniae, they are also found in other species of gram-negative bacteria with increasing frequency. The amino acid substitutions responsible for the ESBL phenotype cluster around the active site of the enzyme and change its configuration, allowing access to oxyimino-beta-lactam substrates. Opening the active site to beta-lactam substrates also typically enhances the susceptibility of the enzyme to b-lactamase inhibitors, such as clavulanic acid.  Although the inhibitor-resistant beta-lactamases are not ESBLs, they are often discussed with ESBLs because they are also derivatives of the classical TEM- or SHV-type enzymes. These enzymes were at first given the designation IRT for inhibitor-resistant TEM beta-lactamase; however, all have subsequently been renamed with numerical TEM designations. There are at least 19 distinct inhibitor-resistant TEM beta-lactamases. Inhibitor-resistant TEM beta-lactamases have been found mainly in clinical isolates of E. coli, but also some strains of K. pneumoniae, Klebsiella oxytoca, P. mirabilis, and Citrobacter freundii. Although the inhibitor-resistant TEM variants are resistant to inhibition by clavulanic acid and sulbactam, thereby showing clinical resistance to the beta-lactam-lactamase inhibitor combinations of amoxicillin-clavulanate (Co-amoxiclav), ticarcillin-clavulanate, and ampicillin/sulbactam, they normally remain susceptible to inhibition by tazobactam and subsequently the combination of piperacillin/tazobactam, although resistance has been described.",
  },
  {
    accession: "ARO:3000015",
    name: "SHV beta-lactamase",
    description:
      "SHV-1 shares 68 percent of its amino acids with TEM-1 and has a similar overall structure. The SHV-1 beta-lactamase is most commonly found in K. pneumoniae and is responsible for up to 20% of the plasmid-mediated ampicillin resistance in this species. ESBLs in this family also have amino acid changes around the active site, most commonly at positions 238 or 238 and 240. More than 60 SHV varieties are known.",
  },
  {
    accession: "ARO:3000016",
    name: "CTX-M beta-lactamase",
    description:
      "These enzymes were named for their greater activity against cefotaxime than other oxyimino-beta-lactam substrates (eg, ceftazidime, ceftriaxone, or cefepime). Rather than arising by mutation, they represent examples of plasmid acquisition of beta-lactamase genes normally found on the chromosome of Kluyvera species, a group of rarely pathogenic commensal organisms. These enzymes are not very closely related to TEM or SHV beta-lactamases in that they show only approximately 40% identity with these two commonly isolated beta-lactamases. Despite their name, a few are more active on ceftazidime than cefotaxime.  CTX-M-15 was recently found in bacterial strains expressing NDM-1 and were responsible for resistance to aztreonam.",
  },
  {
    accession: "ARO:3000017",
    name: "OXA beta-lactamase",
    description:
      "OXA beta-lactamases were long recognized as a less common but also plasmid-mediated beta-lactamase variety that could hydrolyze oxacillin and related anti-staphylococcal penicillins. These beta-lactamases differ from the TEM and SHV enzymes in that they belong to molecular class D and functional group 2d. The OXA-type beta-lactamases confer resistance to ampicillin and cephalothin and are characterized by their high hydrolytic activity against oxacillin and cloxacillin and the fact that they are poorly inhibited by clavulanic acid. Amino acid substitutions in OXA enzymes can also give the ESBL phenotype. The OXA beta-lactamase family was originally created as a phenotypic rather than a genotypic group for a few beta-lactamases that had a specific hydrolysis profile. Therefore, there is as little as 20% sequence homology among some of the members of this family. However, recent additions to this family show some degree of homology to one or more of the existing members of the OXA beta-lactamase family. Some confer resistance predominantly to ceftazidime, but OXA-17 confers greater resistance to cefotaxime and cefepime than it does resistance to ceftazidime.",
  },
  {
    accession: "ARO:3000018",
    name: "IMI beta-lactamase",
    description:
      "IMI beta-lactamases are a group of TEM-1-like beta-lactamase that are known to hydrolyze imipenem. IMI beta-lactamases are inhibited by clavulanic acid and tazobactam.",
  },
  {
    accession: "ARO:3000020",
    name: "IMP beta-lactamase",
    description:
      "Plasmid mediated IMP-type carbapenemases, of which at least 26 varieties are currently known, became established in Japan in the 1990s in enteric gram-negative organisms, Pseudomonas and Acinetobacter species. Integron-associated, sometimes within plasmids. Hydrolyses all beta-lactams except monobactams, and evades all beta-lactam inhibitors.",
  },
  {
    accession: "ARO:3000021",
    name: "VIM beta-lactamase",
    description:
      "The Verone integron-encoded metallo-beta-lactamase (VIM) family was reported from Italy in 1999. There are, to date, 23 reported variants.  VIM enzymes mostly occur in P. aeruginosa, also P. putida and, very rarely, Enterobacteriaceae. Integron-associated, sometimes within plasmids. Hydrolyses all beta-lactams except monobactams, and evades all beta-lactam inhibitors.  There is a strong incidence of these in East Asia.",
  },
  {
    accession: "ARO:3000022",
    name: "ristocetin",
    description:
      "Ristocetin is a glycopeptide antibiotic similar to vancomycin but positively charged. It is not used clinically because it induces platelet agglutination.",
  },
  {
    accession: "ARO:3000023",
    name: "aminocoumarin sensitive parE",
    description: "Subunit of the topoisomerase IV sensitive to aminocoumarins",
  },
  {
    accession: "ARO:3000024",
    name: "patA",
    description:
      "PatA is an ABC transporter of Streptococcus pneumoniae that interacts with PatB to confer fluoroquinolone resistance.",
  },
  {
    accession: "ARO:3000025",
    name: "patB",
    description:
      "PatB is an ABC transporter of Streptococcus pneumoniae that interacts with PatA to confer fluoroquinolone resistance..",
  },
  {
    accession: "ARO:3000026",
    name: "mepA",
    description:
      "MepA is an efflux protein regulated by MepR and part of the MepRAB cluster. Its presence in Staphylococcus aureus led to multidrug resistance, while it has also been shown to decrease tigecycline susceptibility.",
  },
  {
    accession: "ARO:3000027",
    name: "emrA",
    description:
      "EmrA is a membrane fusion protein, providing an efflux pathway with EmrB and TolC between the inner and outer membranes of E. coli, a Gram-negative bacterium.",
  },
  {
    accession: "ARO:3000034",
    name: "nucleoside antibiotic",
    description:
      "Nucleoside antibiotics are made of modified nucleosides and nucleotides with wide-ranging activities and means of antibacterial effects. This drug class includes aminonucleoside antibiotics, which contain an amino group.",
  },
  {
    accession: "ARO:3000035",
    name: "lipopeptide antibiotic",
    description:
      "Lipopeptide antibiotics are aliphatic, with their hydrophobic components interacting with the bacterial cell membrane.",
  },
  {
    accession: "ARO:3000036",
    name: "tetracycline inactivation enzyme",
    description:
      "Enzymes or other gene products which hydroxylate tetracycline and other tetracycline derivatives. Hydroxylation inactivates tetracycline-like antibiotics, thus conferring resistance to these compounds.",
  },
  {
    accession: "ARO:3000040",
    name: "hydrolysis of antibiotic conferring resistance",
    description: "Hydrolysis of an antibiotic molecule to confer resistance.",
  },
  {
    accession: "ARO:3000042",
    name: "beta-lactamase inhibitor",
    description:
      "Antibiotic adjuvants shown to inhibit the action of a beta-lactamase enzyme or enhance the ability of a beta-lactam antibiotic. These compounds are used along with antibiotics to treat beta-lactam-resistant infectious pathogens.",
  },
  {
    accession: "ARO:3000043",
    name: "VEB beta-lactamase",
    description:
      "VEB beta-lactamases or Vietnamese extended-spectrum beta-lactamases are class A beta-lactamases that confer high-level resistance to oxyimino cephalosporins and to aztreonam",
  },
  {
    accession: "ARO:3000045",
    name: "component of AMR phenotype terminology",
    description:
      "A collection of curated phenotypic terms relating to antibiotic resistance and the Comprehensive Antibiotic Resistance Database.",
  },
  {
    accession: "ARO:3000050",
    name: "tetracycline antibiotic",
    description:
      "These antibiotics are derived from tetracycline, a polyketide antibiotic that inhibits the 30S subunit of bacterial ribosomes.",
  },
  {
    accession: "ARO:3000052",
    name: "determinant of phenicol resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to phenicol (chloramphenicol) antibiotics. These include chloramphenicol acetyltransferase (CAT) enzymes, which are found in a large number of species.",
  },
  {
    accession: "ARO:3000053",
    name: "peptide antibiotic",
    description:
      "Peptide antibiotics have a wide range of antibacterial mechanisms, depending on the amino acids that make up the antibiotic, although most act to disrupt the cell membrane in some manner. Subclasses of peptide antibiotics can include additional sidechains of other types, such as lipids in the case of the lipopeptide antibiotics.",
  },
  {
    accession: "ARO:3000054",
    name: "acridine dye",
    description:
      "Acridine dyes are cell permeable, basic molecules with an acridine chromophore. These compounds intercalate DNA. The image shown represents the core structure of the acridine family, with specific dyes containing varying substituents.",
  },
  {
    accession: "ARO:3000055",
    name: "SME beta-lactamase",
    description:
      "SME beta-lactamases are chromosome-mediated class A beta-lactamases that hydrolyze carbapenems in Serratia marcescens.",
  },
  {
    accession: "ARO:3000056",
    name: "PER beta-lactamase",
    description:
      "PER beta-lactamases are plasmid-mediated extended spectrum beta-lactamases found in the Enterobacteriaceae family.",
  },
  {
    accession: "ARO:3000057",
    name: "NDM beta-lactamase",
    description:
      "NDM beta-lactamases or New Delhi metallo-beta-lactamases are class B beta-lactamases that confer resistance to a broad range of antibiotics including carbapenems, cephalosporins and penicillins.",
  },
  {
    accession: "ARO:3000058",
    name: "MIR beta-lactamase",
    description:
      "MIR beta-lactamases are plasmid-mediated beta-lactamases that confer resistance to oxyimino- and alpha-methoxy beta-lactams",
  },
  {
    accession: "ARO:3000059",
    name: "KPC beta-lactamase",
    description:
      "Klebsiella pneumoniae carbapenem resistant (KPC) beta-lactamases are notorious for their ability to efficiently hydrolyze carbapenems, unlike other Ambler Class A beta-lactamases.  There are currently 9 variants reported worldwide.  These enzymes were first isolated from Klebsiella pneumoniae strains in 2001 in the United States.  Hospital outbreaks have since been reported in Greece and Israel and KPC carrying strains are now endemic to New York facilities.  KPC-1 and KPC-2 have been shown to be identical and are now referred to as KPC-2.",
  },
  {
    accession: "ARO:3000060",
    name: "IND beta-lactamase",
    description:
      "IND beta-lactamases are class B carbapenem-hydrolyzing beta-lactamases",
  },
  {
    accession: "ARO:3000066",
    name: "GES beta-lactamase",
    description:
      "GES beta-lactamases or Guiana extended-spectrum beta-lactamases are related to the other plasmid-located class A beta-lactamases",
  },
  {
    accession: "ARO:3000067",
    name: "FOX beta-lactamase",
    description:
      "FOX beta-lactamases are plasmid-encoded AmpC-type beta-lactamase which conferred resistance to broad-spectrum cephalosporins and cephamycins",
  },
  {
    accession: "ARO:3000068",
    name: "DHA beta-lactamase",
    description:
      "DHA beta-lactamases are plasmid-mediated AmpC \u03b2-lactamases that confer resistance to cephamycins and oxyimino-cephalosporins.",
  },
  {
    accession: "ARO:3000069",
    name: "CMY beta-lactamase",
    description:
      "CMY beta-lactamases are plasmid-mediated class C beta-lactamases that encodes for resistance to cephamycins.",
  },
  {
    accession: "ARO:3000071",
    name: "vanS",
    description:
      "VanS is similar to histidine protein kinases like EnvZ and acts as a response regulator by activating VanR. VanS is required for high level transcription of other van glycopeptide resistance genes.",
  },
  {
    accession: "ARO:3000072",
    name: "ACT beta-lactamase",
    description:
      "ACT beta-lactamases, also known as AmpC beta-lactamases, are cephalosporinases that cannot be inhibited by clavulanate. These enzymes are encoded by genes located on the chromosome and can be induced by the presence of beta-lactam antibiotics. However recently, these genes have been found on plasmids and expressed at high constitutive levels in Escherichia coli and Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000073",
    name: "ACC beta-lactamase",
    description:
      "ACC beta-lactamases or Ambler class C beta-lactamases are AmpC beta-lactamases. They possess an interesting resistance phenotype due to their low activity against cephamycins.",
  },
  {
    accession: "ARO:3000074",
    name: "emrB",
    description:
      "emrB is a translocase in the emrB -TolC efflux protein in E. coli. It recognizes substrates including carbonyl cyanide m-chlorophenylhydrazone (CCCP), nalidixic acid, and thioloactomycin.",
  },
  {
    accession: "ARO:3000075",
    name: "class D beta-lactamase",
    description:
      "Class D beta-lactamases are one of the subgroups of beta-lactamases that are classified as serine enzymes.",
  },
  {
    accession: "ARO:3000076",
    name: "class C beta-lactamase",
    description:
      "AmpC type beta-lactamases are commonly isolated from extended-spectrum cephalosporin-resistant Gram-negative bacteria. AmpC beta-lactamases (also termed class C or group 1) are typically encoded on the chromosome of many Gram-negative bacteria including Citrobacter, Serratia, Enterobacter species, and P. aeruginosa where its expression is usually inducible; it may also occur on Escherichia coli but is not usually inducible, although it can be hyperexpressed. AmpC type beta-lactamases may also be carried on plasmids. AmpC beta-lactamases, in contrast to ESBLs, hydrolyse broad and extended-spectrum cephalosporins (cephamycins as well as to oxyimino-beta-lactams) but are not inhibited by beta-lactamase inhibitors such as clavulanic acid.",
  },
  {
    accession: "ARO:3000077",
    name: "vanY",
    description:
      "VanY is a D,D-carboxypeptidase that cleaves removes the terminal D-Ala from peptidoglycan for the addition of D-Lactate. The D-Ala-D-Lac peptidoglycan subunits have reduced binding affinity with vancomycin compared to D-Ala-D-Ala.",
  },
  {
    accession: "ARO:3000078",
    name: "class A beta-lactamase",
    description:
      "The Class A beta-lactamases are one of the subgroups of beta-lactamases that are classified as serine enzymes. Class A beta-lactamases exhibit a large degree of variability and are known to hydrolyze penicillins.",
  },
  {
    accession: "ARO:3000079",
    name: "oxazolidinone antibiotic",
    description:
      "Oxazolidinones are a class of synthetic antibiotics discovered the the 1980's.  They inhibit protein synthesis by binding to domain V of the 23S rRNA of the 50S subunit of bacterial ribosomes.  Linezolid is the only member of this class currently in clinical use.",
  },
  {
    accession: "ARO:3000080",
    name: "aminonucleoside antibiotic",
    description:
      "Aminonucleoside antibiotics are nucleoside antibiotics that contain an amino group. This amino group is often acylated, e.g. puromycin.",
  },
  {
    accession: "ARO:3000081",
    name: "glycopeptide antibiotic",
    description:
      "Glycopeptide antibiotics are natural products produced non-ribosomally by Actinomycetales bacteria. With the exception of bleomycins, they act by binding the terminal D-Ala-D-Ala in peptidoglycan precursors of the growing bacterial cell wall and are generally active against Gram-positive bacteria.  This inhibits transglycosylation leading to cell death due to osmotic stress.",
  },
  {
    accession: "ARO:3000082",
    name: "antibiotic biosynthesis",
    description: "The biological synthesis of antibiotics.",
  },
  {
    accession: "ARO:3000083",
    name: "MOX beta-lactamase",
    description:
      "MOX beta-lactamases are plasmid-mediated AmpC-type beta-lactamases.",
  },
  {
    accession: "ARO:3000085",
    name: "CMY-LAT-MOX beta-lactamase",
    description: "A grouping of the related CMY, LAT, and MOX beta-lactamases",
  },
  {
    accession: "ARO:3000086",
    name: "CMY-LAT beta-lactamase",
    description: "A grouping of the related CMY and LAT beta-lactamases.",
  },
  {
    accession: "ARO:3000087",
    name: "CMY-MOX beta-lactamase",
    description: "A grouping of the related CMY and MOX beta-lactamases",
  },
  {
    accession: "ARO:3000089",
    name: "AER beta-lactamase",
    description: "AER beta-lactamases are capable of hydrolyzing arbenicillin.",
  },
  {
    accession: "ARO:3000090",
    name: "Bla1",
    description:
      "Bla1 is a chromosomal-encoded beta-lactamase, found in Bacillus anthracis, which hydrolyzes penicillins.",
  },
  {
    accession: "ARO:3000091",
    name: "CARB beta-lactamase",
    description:
      "CARB beta-lactamases are class A lactamases that can hydrolyze carbenicillin. Many of the PSE beta-lactamases have been renamed as CARB-lactamases with the notable exception of PSE-2 which is now OXA-10.",
  },
  {
    accession: "ARO:3000094",
    name: "OCH beta-lactamase",
    description:
      "OCH beta-lactamases are Ambler class C chromosomal-encoded beta-lactamases in Ochrobactrum anthropi",
  },
  {
    accession: "ARO:3000095",
    name: "SRT beta-lactamase",
    description: "SRT beta-lactamases.",
  },
  {
    accession: "ARO:3000096",
    name: "SHV-LEN beta-lactamase",
    description: "A grouping of the related SHV and LEN beta-lactamases.",
  },
  {
    accession: "ARO:3000097",
    name: "LEN beta-lactamase",
    description:
      "LEN beta-lactamases are chromosomal class A beta-lactamases that confer resistance to ampicillin, amoxicillin, carbenicillin, and ticarcillin but not to extended-spectrum beta-lactams.",
  },
  {
    accession: "ARO:3000098",
    name: "PDC beta-lactamase",
    description:
      "PDC beta-lactamases are class C beta-lactamases that are found in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3000100",
    name: "gene modulating beta-lactam resistance",
    description:
      "Genes that directly or indirectly modulate beta-lactam resistance.",
  },
  {
    accession: "ARO:3000102",
    name: "determinant of fluoroquinolone resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to fluoroquinolone antibiotics",
  },
  {
    accession: "ARO:3000103",
    name: "aminocoumarin antibiotic",
    description:
      "Aminocoumarin antibiotics bind DNA gyrase subunit B to inhibit ATP-dependent DNA supercoiling.",
  },
  {
    accession: "ARO:3000104",
    name: "determinant of aminoglycoside resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to aminoglycoside antibiotics.",
  },
  {
    accession: "ARO:3000105",
    name: "phosphorylation of antibiotic conferring resistance",
    description: "Phosphorylation of antibiotic usually by ATP, sometimes GTP",
  },
  {
    accession: "ARO:3000106",
    name: "acylation of antibiotic conferring resistance",
    description:
      "Addition of an acyl group to an antibiotic, often via acetylation by acetylCoA.",
  },
  {
    accession: "ARO:3000107",
    name: "nucleotidylation of antibiotic conferring resistance",
    description: "Modification by NMP, usually AMP.",
  },
  {
    accession: "ARO:3000111",
    name: "novobiocin",
    description:
      "Novobiocin is an aminocoumarin antibiotic produced by Streptomyces spheroides and Streptomyces niveus, and binds DNA gyrase subunit B inhibiting ATP-dependent DNA supercoiling.",
  },
  {
    accession: "ARO:3000112",
    name: "multidrug and toxic compound extrusion (MATE) transporter",
    description:
      "Directed pumping of antibiotic out of a cell to confer resistance. Multidrug and toxic compound extrusion (MATE) transporters utilize the cationic gradient across the membrane as an energy source. Although there is a diverse substrate specificity, almost all MATE transporters recognize fluoroquinolones. Arciflavine, ethidium and aminoglycosides are also good substrates.",
  },
  {
    accession: "ARO:3000113",
    name: "vga-type ABC-F protein",
    description:
      "Vga-type plasmid-borne ABC-F proteins, expressed in staphylococci that confer resistance to streptogramin A antibiotics through ribosomal protection",
  },
  {
    accession: "ARO:3000114",
    name: "aminoglycoside phosphotransferase (APH)",
    description:
      "Kinases that modify aminoglycoside antibiotics by phosphorylation using NTPs as cofactor",
  },
  {
    accession: "ARO:3000115",
    name: "antibiotic resistant DNA topoisomerase subunit parE",
    description:
      "ParE is a subunit of topoisomerase IV, which decatenates and relaxes DNA to allow access to genes for transcription or translation. Mutations in ParE prevents antibiotics from inhibiting DNA synthesis, thus conferring resistance.",
  },
  {
    accession: "ARO:3000116",
    name: "vanZ",
    description:
      "VanZ is a teicoplanin resistance gene that is an accessory protein. VanZ prevents the incorporation of the terminal D-Ala into peptidoglycan subunits.",
  },
  {
    accession: "ARO:3000117",
    name: "antibiotic A47934",
    description:
      "A47934 is an 'aglycone' glycopeptide antibiotic produced by Streptomyces toyocaensis. It is a teicoplanin-like glycopeptide.",
  },
  {
    accession: "ARO:3000118",
    name: "vgaB",
    description:
      "Vga(B) is an ABC-F protein expressed in staphylococci that confers resistance to streptogramin A antibiotics and related compounds. It is associated with plasmid DNA.",
  },
  {
    accession: "ARO:3000119",
    name: "edeine",
    description:
      "Edeines are basic linear peptides produced by Bacillus brevis Vm4. They have antibacterial as well as antifungal, antiviral, and anticancer properties. Edeines are bacteriostatic and bacteriocidal at low and high concentrations, respectively. They are able to inhibit DNA synthesis and protein translation. These compounds are synthesized by nonribosomal peptide synthetases and contain numerous unusual amino acids.",
  },
  {
    accession: "ARO:3000120",
    name: "balhimycin",
    description:
      "Balhimycin is a vancomycin-like glycopeptide antibiotic produced by Amycolatopsis balhimycina. It binds to the terminal Lys-D-Ala-D-Ala of peptidoglycan precursors.",
  },
  {
    accession: "ARO:3000121",
    name: "aminoglycoside acetyltransferase (AAC)",
    description:
      "Aminoglycoside acetyltransferase enzymes modify aminoglycoside antibiotics by catalyzing the transfer of an acetyl group to one of the amino groups present in aminoglycosides, using acetyl coenzyme A as a donor substrate.",
  },
  {
    accession: "ARO:3000122",
    name: "chloramphenicol acetyltransferase (CAT)",
    description:
      "Inactivates chloramphenicol by addition of an acyl group. cat is used to describe many variants of the chloramphenicol acetyltransferase gene in a range of organisms including Acinetobacter calcoaceticus, Agrobacterium tumefaciens, Bacillus clausii, Bacillus subtilis, Campylobacter coli, Enterococcus faecalis, Enterococcus faecium, Lactococcus lactis, Listeria monocytogenes, Listonella anguillarum Morganella morganii, Photobacterium damselae subsp. piscicida, Proteus mirabilis, Salmonella typhi, Serratia marcescens, Shigella flexneri, Staphylococcus aureus, Staphylococcus haemolyticus, Staphylococcus intermedius, Streptococcus agalactiae, Streptococcus suis and Streptomyces acrimycini",
  },
  {
    accession: "ARO:3000123",
    name: "gramicidin",
    description:
      "Gramicidins are a family of antibiotics synthesized by Bacillus brevis. It includes the linear pentadecapeptides gramicidin A, B and C that make up the mixture gramicidin D. Gramicidin S is a cyclic peptide chain. Gramicidins are also components of tyrothricins, another mixture of antibiotics produced by Bacillus brevis.",
  },
  {
    accession: "ARO:3000124",
    name: "mecI",
    description:
      "mecI acts as a repressor of transcription of the mecA/mecR1/mecI operon.",
  },
  {
    accession: "ARO:3000125",
    name: "hydrolysis of fosfomycin epoxide ring",
    description:
      "The use of different nucleophilic molecules by enzymes can break up the epoxide ring of fosfomycin and render the molecule ineffective.",
  },
  {
    accession: "ARO:3000126",
    name: "APH(3')",
    description:
      "Phosphorylation of 2-deoxystreptamine aminoglycosides on the hydroxyl group at position 3'",
  },
  {
    accession: "ARO:3000127",
    name: "APH(3'')",
    description:
      "Phosphorylation of streptomycin on the hydroxyl group at position 3''",
  },
  {
    accession: "ARO:3000128",
    name: "APH(2'')",
    description:
      "Phosphorylation of 2-deoxystreptamine aminoglycosides on the hydroxyl group at position 2''",
  },
  {
    accession: "ARO:3000129",
    name: "determinant of beta-lactam resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to beta-lactam antibiotics.",
  },
  {
    accession: "ARO:3000130",
    name: "edeine A",
    description:
      "Edeine A is a subtype of the peptide antibiotic edeine, composed of beta-tyr, beta-ser, diaminopropionic acid, diaminohydroxyazelaic acid, glycine, and spermidine. Edeine A is a mixture of edeine A1 and its inactive isomer, edeine A2. Edeines bind to the 30S subunit to block fMet-tRNA interaction at the P site, inhibiting protein synthesis and subsequent structure/function processes critical for life or replication.",
  },
  {
    accession: "ARO:3000132",
    name: "clorobiocin",
    description:
      "Clorobiocin is an aminocoumarin antibiotic produced by Streptomyces roseochromogenes, and binds DNA gyrase subunit B to inhibit ATP-dependent DNA supercoiling.",
  },
  {
    accession: "ARO:3000133",
    name: "fosfomycin thiol transferase",
    description:
      "Catalyzes the addition of a thiol group from a nucleophilic molecule to fosfomycin.",
  },
  {
    accession: "ARO:3000134",
    name: "edeine B",
    description:
      "Edeine B is a subtype of the peptide antibiotic edeine, composed of beta-tyr, beta-ser, diaminopropionic acid, diaminohydroxyazelaic acid, glycine, and guanylspermidine. Edeine B is a mixture of edeine B1 and its inactive isomer, edeine B2.  Edeines bind to the 30S subunit to block fMet-tRNA interaction at the P site, inhibiting protein synthesis and subsequent structure/function processes critical for life or replication. Edeine B has also been shown to inhibit septation and cause filamentous morphology, also leading to cell death.",
  },
  {
    accession: "ARO:3000135",
    name: "edeine D",
    description:
      "Edeine D is a subtype of edeine similar to edeine A with the beta-tyr replaced by beta-phe-beta-ala. Edeines bind to the 30S subunit to block fMet-tRNA interaction at the P site, inhibiting protein synthesis and subsequent structure/function processes critical for life or replication.",
  },
  {
    accession: "ARO:3000136",
    name: "edeine F",
    description:
      "Edeine F is a subtype of edeine similar to edeine B with beta-tyr replaced by beta-phe-beta-ala. Edeines bind to the 30S subunit to block fMet-tRNA interaction at the P site, inhibiting protein synthesis and subsequent structure/function processes critical for life or replication.",
  },
  {
    accession: "ARO:3000145",
    name: "tylosin",
    description:
      "Tylosin is a 16-membered macrolide, naturally produced by Streptomyces fradiae. It interacts with the bacterial ribosome 50S subunit to inhibit protein synthesis.",
  },
  {
    accession: "ARO:3000149",
    name: "FosA",
    description:
      "An enzyme that confers resistance to fosfomycin in Serratia marcescens by breaking the epoxide ring of the molecule. It depends on the cofactors Manganese (II) and Potassium and uses Glutathione (GSH) as the nucleophilic molecule. In Pseudomonas aeruginosa, FosA catalyzes the conjugation of glutathione to carbon-1 of fosfomycin, rendering it ineffective as an antibacterial drug.",
  },
  {
    accession: "ARO:3000150",
    name: "coumermycin A1",
    description:
      "Coumermycin A1 is an antibiotic produced by Streptomyces rishiriensis, and binds DNA gyrase subunit B to inhibit ATP-dependent DNA supercoiling.",
  },
  {
    accession: "ARO:3000151",
    name: "APH(6)",
    description:
      "Phosphorylation of streptomycin on the hydroxyl group at position 6",
  },
  {
    accession: "ARO:3000152",
    name: "minocycline",
    description:
      "Minocycline is second generation semi-synthetic derivative of the tetracycline group of antibiotics. It inhibits bacterial protein synthesis by binding to the 30S subunit of the bacterial ribosome and preventing the aminotransferase-tRNA from associating with the ribosome.",
  },
  {
    accession: "ARO:3000153",
    name: "APH(9)",
    description:
      "Phosphorylation of spectinomycin on the hydroxyl group at position 9",
  },
  {
    accession: "ARO:3000154",
    name: "APH(7'')",
    description:
      "Phosphorylation of hygromycin on the hydroxyl group at position 7''",
  },
  {
    accession: "ARO:3000155",
    name: "APH(4)",
    description:
      "Phosphorylation of hygromycin on the hydroxyl group at position 4",
  },
  {
    accession: "ARO:3000156",
    name: "spiramycin",
    description:
      "Spiramycin is a 16-membered macrolide and is natural product produced by Streptomyces ambofaciens. It binds to the 50S subunit of bacterial ribosomes and inhibits peptidyl transfer activity to disrupt protein synthesis.",
  },
  {
    accession: "ARO:3000157",
    name: "rifamycin antibiotic",
    description:
      "Rifamycin antibiotics are a group of broad-spectrum ansamycin antibiotics that inhibit bacterial RNA polymerase by binding to a highly conserved region, blocking the oligonucleotide exit tunnel, and preventing the extension of nascent mRNAs.",
  },
  {
    accession: "ARO:3000158",
    name: "azithromycin",
    description:
      "Azithromycin is a 15-membered macrolide and falls under the subclass of azalide. Like other macrolides, azithromycin binds bacterial ribosomes to inhibit protein synthesis. The nitrogen substitution at the C-9a position prevents its degradation.",
  },
  {
    accession: "ARO:3000159",
    name: "efflux pump complex or subunit conferring antibiotic resistance",
    description:
      "Efflux proteins that pump antibiotic out of a cell to confer resistance.",
  },
  {
    accession: "ARO:3000160",
    name: "blaI",
    description:
      "blaI acts as a repressor of transcription of the blaZ/blaR1/blaI operon.",
  },
  {
    accession: "ARO:3000164",
    name: "rRNA methyltransferase conferring antibiotic resistance",
    description: "Catalyzes methylation of rRNA",
  },
  {
    accession: "ARO:3000165",
    name: "tet(A)",
    description:
      "TetA is a tetracycline efflux pump found in many species of Gram-negative bacteria.",
  },
  {
    accession: "ARO:3000166",
    name: "tet(B)",
    description:
      "Tet(B) is a tetracycline efflux protein expressed in many Gram-negative bacteria. It confers resistance to tetracycline, doxycycline, and minocycline, but not tigecycline.",
  },
  {
    accession: "ARO:3000167",
    name: "tet(C)",
    description:
      "Tet(C) is a tetracycline efflux pump found in many species of Gram-negative bacteria. It is typically found in plasmid DNA.",
  },
  {
    accession: "ARO:3000168",
    name: "tet(D)",
    description:
      "TetD is a tetracycline efflux pump found exclusively in Gram-negative bacteria.",
  },
  {
    accession: "ARO:3000169",
    name: "rifampin",
    description:
      "Rifampin is a semi-synthetic rifamycin, and inhibits RNA synthesis by binding to RNA polymerase. Rifampin is the mainstay agent for the treatment of tuberculosis, leprosy and complicated Gram-positive infections.",
  },
  {
    accession: "ARO:3000170",
    name: "imipenem",
    description:
      "Imipenem is a broad-spectrum antibiotic and is usually taken with cilastatin, which prevents hydrolysis of imipenem by renal dehydropeptidase-I. It is resistant to hydrolysis by most other beta-lactamases. Notable exceptions are the KPC beta-lactamases and Ambler Class B enzymes.",
  },
  {
    accession: "ARO:3000171",
    name: "diaminopyrimidine antibiotic",
    description:
      "Diaminopyrimidines are a class of organic compounds containing a pyrimidine ring substituted by two amine groups.  They are inhibitors of dihydrofolate reductase, an enzyme critical for DNA synthesis.",
  },
  {
    accession: "ARO:3000172",
    name: "FosB",
    description:
      "A thiol transferase that leads to the resistance of fosfomycin. Contrasting FosA, FosB is dependent on the cofactor Magnesium (II) and uses either bacillithiol or  L-cysteine to open up the epoxide ring of fosfomycin.",
  },
  {
    accession: "ARO:3000173",
    name: "tet(E)",
    description:
      "TetE is a tetracycline efflux pump found in many Gram-negative bacteria, especially those in water environments. The gene is found on large plasmids.",
  },
  {
    accession: "ARO:3000174",
    name: "tet(G)",
    description:
      "TetG is a tetracycline efflux protein found in Gram-negative bacteria. The encoding gene is found in both chromosomal and plasmid DNA where it is frequently linked to the floR, sul1, and cmlA9 genes which encode proteins that can confer florfenicol/chloramphenicol, sulfamethoxazole, and chloramphenicol resistance, respectively.",
  },
  {
    accession: "ARO:3000175",
    name: "tet(H)",
    description:
      "TetH is a tetracycline efflux protein expressed in Gram-negative bacteria (Actinobacillus, Acinetobacter, Gallibacterium, Histophilus, Mannheimia, Moraxella, Pasteurella, and Psychrobacter). Its gene is linked to the resistance genes sul2, and strAB, which confer resistance to sulfamethoxazole and streptomycin, respectively.",
  },
  {
    accession: "ARO:3000176",
    name: "dirithromycin",
    description:
      "Dirithromycin is an oxazine derivative of erythromycin, sharing the 14-carbon macrolide ring. The antibiotic binds to the 50S subunit of the ribosome to inhibit bacterial protein synthesis.",
  },
  {
    accession: "ARO:3000177",
    name: "tet(J)",
    description:
      "TetJ is a tetracycline efflux protein expressed in Gram-negative bacteria (Escherichia, Morganella, and Proteus).",
  },
  {
    accession: "ARO:3000178",
    name: "tet(K)",
    description:
      "TetK is a tetracycline efflux protein found in both Gram-negative (Haemophilus and Gallibacterium) and Gram-positive (many species, including mycobacteria) bacteria.",
  },
  {
    accession: "ARO:3000179",
    name: "tet(L)",
    description:
      "TetL is a tetracycline efflux protein found in many species of Gram-negative and Gram-positive bacteria.",
  },
  {
    accession: "ARO:3000180",
    name: "tetA(P)",
    description:
      "TetA(P) is a inner membrane tetracycline efflux protein found on the same operon as the ribosomal protection protein TetB(P). It is found in Clostridium, a Gram-positive bacterium.",
  },
  {
    accession: "ARO:3000181",
    name: "tet(V)",
    description:
      "TetV is a tetracycline efflux protein that has been found in Mycobacterium smegmatis and M. fortuitum.",
  },
  {
    accession: "ARO:3000182",
    name: "tet(Y)",
    description:
      "TetY is a tetracycline efflux pump found in Gram-negative bacteria (Aeromonas and Escherichia). It is associated with plasmid DNA.",
  },
  {
    accession: "ARO:3000183",
    name: "tet(Z)",
    description:
      "TetZ is a tetracycline efflux protein found in Gram-positive bacteria (Corynebacterium and Lactobacillus). It is associated with plasmid DNA.",
  },
  {
    accession: "ARO:3000184",
    name: "chloroeremomycin",
    description:
      "Chloroeremomycin is a vancomycin-like glycopeptide, with three sugars instead of two in vancomycin and balhimycin. Chloroeremomycin dimerizes and binds to the terminus of peptidoglycan precursors.",
  },
  {
    accession: "ARO:3000185",
    name: "antibiotic target protection protein",
    description:
      "These proteins confer antibiotic resistance by bind the antibiotic target to prevent antibiotic binding.",
  },
  {
    accession: "ARO:3000186",
    name: "tetM",
    description:
      "TetM is a ribosomal protection protein that confers tetracycline resistance. It is found on transposable DNA elements and its horizontal transfer between bacterial species has been documented.",
  },
  {
    accession: "ARO:3000187",
    name: "hydrolysis of beta-lactam antibiotic by serine beta-lactamase",
    description:
      "Mechanism of enzymatic degradation common to Ambler Class A, C and D beta-lactamases.  A serine residue located in the active site is used to form an acyl-enzyme intermediate and subsequent hydrolysis renders the beta-lactam inactive.",
  },
  {
    accession: "ARO:3000188",
    name: "trimethoprim",
    description:
      "Trimethoprim is a synthetic 5-(3,4,5- trimethoxybenzyl) pyrimidine inhibitor of dihydrofolate reductase, inhibiting synthesis of tetrahydrofolic acid. Tetrahydrofolic acid is an essential precursor in the de novo synthesis of the DNA nucleotide thymidine. Trimethoprim is a bacteriostatic antibiotic mainly used in the prophylaxis and treatment of urinary tract infections in combination with sulfamethoxazole, a sulfonamide antibiotic.",
  },
  {
    accession: "ARO:3000189",
    name: "oritavancin",
    description:
      "Oritavancin is a semi-synthetic derivative of chloroeremomycin, a vancomycin-like glycopeptide. Oritavancin inhibits both transglycosylation and transpeptidation, by binding both the D-Ala-D-Ala and pentaglycine bridge segments of peptidoglycan to inhibit cell wall formation.",
  },
  {
    accession: "ARO:3000190",
    name: "tetO",
    description:
      "TetO is a ribosomal protection protein. It is associated with conjugative plasmids.",
  },
  {
    accession: "ARO:3000191",
    name: "tetQ",
    description:
      "TetQ is a ribosomal protection protein. Its gene is associated with a conjugative transposon and has been found in both Gram-positive and Gram-negative bacteria.",
  },
  {
    accession: "ARO:3000192",
    name: "tetS",
    description:
      "Tet(S) is a ribosomal protection protein found in Gram-positive and Gram-negative strains. It is similar to tet(M) and tet(O).",
  },
  {
    accession: "ARO:3000193",
    name: "tetT",
    description:
      "Tet(T) is a ribosomal protection protein of streptococci. It is similar to Tet(Q).",
  },
  {
    accession: "ARO:3000194",
    name: "tetW",
    description:
      "TetW is a ribosomal protection protein. It is associated with both conjugative and non conjugative DNA and has been found strains of C. difficile.",
  },
  {
    accession: "ARO:3000195",
    name: "tetB(P)",
    description:
      "TetB(P) is a tetracycline ribosomal protection protein found on the same operon as tetA(P), a tetracycline efflux protein.",
  },
  {
    accession: "ARO:3000196",
    name: "tet32",
    description:
      "Tet32 is a tetracycline resistance gene similar to Tet(O), and binds to the ribosome to confer tetracycline resistance as a ribosomal protection protein.",
  },
  {
    accession: "ARO:3000197",
    name: "tet36",
    description:
      "Tet36 is a tetracycline resistance gene found in Bacteroides similar to Tet(Q), and binds to the ribosome to confer antibiotic resistance as a ribosomal protection protein.",
  },
  {
    accession: "ARO:3000198",
    name: "FosX",
    description:
      "FosX is an enzyme used to confer resistance to fosfomycin. It's dependent on the cofactor, manganese (II), and uses water to generate a vicinal diol.",
  },
  {
    accession: "ARO:3000199",
    name: "gramicidin D",
    description:
      "Gramicidin D is a mixture of the linear peptides gramicidin A, B, and C, each with 15 alternating L- and D-amino acids. They are active against most gram-positive bacteria and select gram-negative bacteria. These compounds create channels in the bacterial membrane and increase the permeability to cations.",
  },
  {
    accession: "ARO:3000200",
    name: "gramicidin S",
    description:
      "Gramicidin S is a cyclical decapeptide with two pentapeptides (Val-Orn-Leu-D-Phe-Pro) joined head to tail. Like other gramicidins, Gramicidin S disrupts membrane permeability of cations while also destabilizing the membrane at higher concentrations.",
  },
  {
    accession: "ARO:3000201",
    name: "macrolide inactivation enzyme",
    description:
      "Enzymes shown to inactivate macrolide antibiotics by chemical modification, thereby conferring resistance to macrolides.",
  },
  {
    accession: "ARO:3000202",
    name: "Cfr 23S ribosomal RNA methyltransferase",
    description:
      "Cfr genes produce enzymes which catalyze the methylation of the 23S rRNA subunit at position 8 of adenine-2503. Methylation of 23S rRNA at this site confers resistance to some classes of antibiotics, including streptogramins, chloramphenicols, florfenicols, linezolids and clindamycin.",
  },
  {
    accession: "ARO:3000203",
    name: "hydrolysis of beta-lactam antibiotic by metallo-beta-lactamase",
    description:
      "Mechanism of enzymatic degradation common to Ambler Class B beta-lactamases. One or two zinc atoms are used to orient a hydroxide nucleophile for attack of the beta-lactam ring.  In contrast to serine beta-lactamases, no acyl-enzyme intermediate is formed.",
  },
  {
    accession: "ARO:3000205",
    name: "tetX",
    description:
      "TetX is a flavin-dependent monooxygenase conferring resistance to tetracycline antibiotics. TetX hydroxylates position 11a of the tetraketide group thus inactivating the antibiotic.",
  },
  {
    accession: "ARO:3000206",
    name: "emrK",
    description:
      "emrK is a membrane fusion protein that is a homolog of EmrA. Together with the inner membrane transporter EmrY and the outer membrane channel TolC, it mediates multidrug efflux.",
  },
  {
    accession: "ARO:3000207",
    name: "acrA",
    description:
      "Protein subunit of AcrA-AcrB-TolC multidrug efflux complex. AcrA represents the periplasmic portion of the transport protein.",
  },
  {
    accession: "ARO:3000208",
    name: "glycosylation of antibiotic conferring resistance",
    description:
      "Addition of glycosyl moiety to antibiotics thereby inactivating them.",
  },
  {
    accession: "ARO:3000209",
    name: "antibiotic resistant DNA topoisomerase subunit parY",
    description:
      "ParY is part of a topoisomerase IV that is resistant to antibiotics that affect other topoisomerases.",
  },
  {
    accession: "ARO:3000210",
    name: "rifamycin-resistant beta-subunit of RNA polymerase (rpoB)",
    description:
      "Rifampin resistant RNA polymerases include amino acids substitutions which disrupt the affinity of rifampin for its binding site. These mutations are frequently concentrated in the rif I region of the beta-subunit and most often involve amino acids which make direct interactions with rifampin. However, mutations which also confer resistance can occur outside this region and may involve amino acids which do not directly make contact with rifampin.",
  },
  {
    accession: "ARO:3000211",
    name: "ribosomal alteration conferring antibiotic resistance",
    description:
      "Chemical alteration of the ribosome results in modification of an antibiotic's target leading to resistance.",
  },
  {
    accession: "ARO:3000212",
    name: "mutation conferring antibiotic resistance",
    description:
      "Point mutations in the DNA may lead to an altered gene product that may result in antibiotic resistance. Examples included modified antibiotic targets with lower binding affinities and the deactivation of repressors that result in increased expression of genes that inactivate or pump out antibiotics.",
  },
  {
    accession: "ARO:3000213",
    name:
      "restructuring of bacterial cell wall conferring antibiotic resistance",
    description:
      "Peptidoglycan precursors ending in D-Ala-D-Lac or D-Ala-D-Ser instead of D-Ala-D-Ala conferring high level glycopeptide resistance.",
  },
  {
    accession: "ARO:3000214",
    name: "hygromycin B",
    description:
      "Hygromycin B is an aminoglycoside antibiotic used to treat different types of bacterial infections. Hygromycin B works by binding to the bacterial 30S ribosomal subunit, causing misreading of mRNA and leaving the bacterium unable to synthesize proteins vital to its growth. Hygromycin B has also been shown to interact with eukaryotic cells.",
  },
  {
    accession: "ARO:3000215",
    name: "mecR1",
    description:
      "mecR1 is a transmembrane spanning and signal transducing protein which in response to interaction with beta-lactam antibiotics results in upregulation of the mecA/mecR1/mecI operon.",
  },
  {
    accession: "ARO:3000216",
    name: "acrB",
    description:
      "Protein subunit of AcrA-AcrB-TolC multidrug efflux complex. AcrB functions as a herterotrimer which forms the inner membrane component and is primarily responsible for substrate recognition and energy transduction by acting as a drug/proton antiporter.",
  },
  {
    accession: "ARO:3000217",
    name: "blaR1",
    description:
      "blaR1 is a transmembrane spanning and signal transducing protein which in response to interaction with beta-lactam antibiotics results in upregulation of the blaZ/blaR1/blaI operon.",
  },
  {
    accession: "ARO:3000218",
    name: "aminoglycoside nucleotidyltransferase (ANT)",
    description:
      "Covalent modification of aminoglycoside antibiotic hydroxyl group by ATP-dependent transfer of AMP.",
  },
  {
    accession: "ARO:3000219",
    name: "mutant efflux regulatory protein conferring antibiotic resistance",
    description:
      "Efflux regulatory proteins with mutations that result in increased expression of efflux proteins.",
  },
  {
    accession: "ARO:3000220",
    name: "antibiotic resistant DNA topoisomerase subunit gyrB",
    description:
      "Point mutations in gyrB confer antibiotic resistance by preventing drugs from binding the beta-subunit of gyrase, essential for DNA supercoiling.",
  },
  {
    accession: "ARO:3000221",
    name: "lincosamide nucleotidyltransferase (LNU)",
    description:
      "Resistance to the lincosamide antibiotic by ATP-dependent modification of the 3' and/or 4'-hydroxyl groups of the methylthiolincosamide sugar.",
  },
  {
    accession: "ARO:3000222",
    name: "gramicidin A",
    description:
      "Gramicidin A is the most abundant (more than 80%) of the three gramicidins in gramicidin D, with a tryptophan in position 11. Gramicidins form dimers in the bacterial membrane that increase the permeability of cations.",
  },
  {
    accession: "ARO:3000223",
    name: "gramicidin B",
    description:
      "Gramicidin B is one of the three gramicidins in gramicidin D, with a phenylalanine in position 11. It is structurally similar to gramicidin A, but its ability to induce the assembly of bilayers is reduced. Gramicidins form dimers in the bacterial membrane that increase the permeability of cations.",
  },
  {
    accession: "ARO:3000224",
    name: "gramicidin C",
    description:
      "Gramicidin C is one of the three gramicidins in gramicidin D, with a tyrosine in position 11. It is structurally similar to gramicidin A, but its ability to induce the assembly of bilayers is reduced. Gramicidins form dimers in the bacterial membrane that increase the permeability of cations.",
  },
  {
    accession: "ARO:3000225",
    name: "ANT(6)",
    description:
      "Nucelotidylylation of streptomycin at the hydroxyl group at position 6",
  },
  {
    accession: "ARO:3000226",
    name: "antibiotic resistant folP",
    description:
      "Point mutations in dihydropteroate synthase folP prevent sulfonamide antibiotics from inhibiting its role in folate synthesis, thus conferring sulfonamide resistance.",
  },
  {
    accession: "ARO:3000227",
    name: "tyrothricin",
    description:
      "Tyrothricin is a mixture of antibiotics including tyrocidines and gramicidins, isolated from Bacillus brevis.",
  },
  {
    accession: "ARO:3000228",
    name: "ANT(9)",
    description:
      "Nucleotidylylation of spectinomycin at the hydroxyl group at position 9",
  },
  {
    accession: "ARO:3000229",
    name: "ANT(4')",
    description:
      "Nucleotidylylation of 2-deoxystreptamine aminoglycosides at the hydroxyl group at position 4'",
  },
  {
    accession: "ARO:3000230",
    name: "ANT(2'')-Ia",
    description:
      "Plasmid or integron-encoded nucleotidylylation of 2-deoxystreptamine aminoglycosides at the hydroxyl group at position 2'' in P. aeruginosa, K. pneumoniae, Morganella morganii, E. coli, S. typhimurium, C. freundii and A. baumannii",
  },
  {
    accession: "ARO:3000231",
    name: "beta-lactam resistance operon",
    description: "An operon conferring resistance to beta-lactam antibiotics.",
  },
  {
    accession: "ARO:3000232",
    name: "ANT(3'')-Ia",
    description:
      "Nucleotidylylation of streptomycin at the hydroxyl group at position 3''",
  },
  {
    accession: "ARO:3000233",
    name: "streptogramin inactivation enzyme",
    description:
      "Resistance to streptogramin antibiotics may be conferred through enzymatic inactivation. There are two known mechanisms of streptogramin inactivation shown clinically to confer resistance: 1) vgB lyase enzymes linearize type B streptogramin antibiotics by breaking the ester linkage; 2) vat acetyltransferase enzymes modify type A streptogramin antibiotics by transferring an acetyl group from acetyl-CoA to the secondary streptogramin hydroxyl. Both mechanisms result in antibiotic inactivation thus conferring resistance.",
  },
  {
    accession: "ARO:3000234",
    name: "glycopeptide resistance gene cluster",
    description:
      "Genes that when expressed confer resistance to vancomycin and teicoplanin type antibiotics.",
  },
  {
    accession: "ARO:3000235",
    name: "ethambutol resistant embB",
    description:
      "embB gene encodes for an arabinosyl transferase in the arabinogalactan synthesis pathway. It is inhibited by ethambutol. Mutations within the ERDR region of embB confers resistance to ethambutol.",
  },
  {
    accession: "ARO:3000236",
    name: "glycopeptide resistance gene cluster VanA",
    description:
      "This inducible cluster confers high resistance to both vancomycin and teicoplanin by allowing restructuring of peptidoglycan precursors to end in D-Ala-D-Lac. The vanA gene cluster can be located either on plasmids or on the chromosome.\n\nGene orientation: vanRSHAXYZ",
  },
  {
    accession: "ARO:3000237",
    name: "TolC",
    description:
      "TolC is a protein subunit of many multidrug efflux complexes in Gram negative bacteria. It is an outer membrane efflux protein and is constitutively open. Regulation of efflux activity is often at its periplasmic entrance by other components of the efflux complex.",
  },
  {
    accession: "ARO:3000238",
    name: "glycopeptide resistance gene cluster VanB",
    description:
      "This inducible cluster confers resistance to vancomycin but organisms remain sensitive to teicoplanin by allowing restructuring of peptidoglycan precursors to end in D-Ala-D-Lac.  Sensitivity to teicoplanin is due to lack of binding to the sensor kinase VanS. The vanB gene cluster can be located either on plasmids or on the chromosome.\n\nGene orientation: vanRSYWHBX",
  },
  {
    accession: "ARO:3000240",
    name: "determinant of streptogramin resistance",
    description:
      "Ezymes, other proteins or other gene products shown clinically to confer resistance to streptogramin antibiotics.",
  },
  {
    accession: "ARO:3000241",
    name: "determinant of lincosamide resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to lincosamide antibiotics.",
  },
  {
    accession: "ARO:3000244",
    name: "reduced permeability to antibiotic",
    description:
      "Reduction in permeability to antibiotic, generally through reduced production of porins, can provide resistance.",
  },
  {
    accession: "ARO:3000245",
    name: "RbpA",
    description:
      "RNA-polymerase binding protein which confers resistance to rifampin.",
  },
  {
    accession: "ARO:3000246",
    name: "glycopeptide resistance gene cluster VanC",
    description:
      "Confers low vancomycin resistance by engineering peptidoglycan precursors ending in D-Ala-D-Ser in an inducible or constitutive manner.  The vanC cluster is intrinsic to the Enterococcus gallinarum chromosome. vanC organisms remain susceptible to teicoplanin. Gene orientation: vanC(XY)TRS",
  },
  {
    accession: "ARO:3000247",
    name: "phosphoenolpyruvate (PEP) mutase",
    description:
      "A tetrameric protein that converts phosphoenolpyruvate (PEP) to phosponopyruvate (Ppyr).",
  },
  {
    accession: "ARO:3000248",
    name: "DnaA",
    description:
      "DnaA is a chromosomal replication initiation protein which binds and interacts with RNA polymerase in Escherichia coli. A surplus of DnaA present in a cell has been shown to confer resistance to the antibiotic Rifampicin. Normally, rifampicin inhibits initiation of transcription by RNA polymerase, but a surplus of DnaA available at the origin has been shown to disrupt Rifampicin activity and confer resistance.",
  },
  {
    accession: "ARO:3000249",
    name: "chloramphenicol phosphotransferase",
    description:
      "ATP-dependent kinase modifies the C-3 hydroxyl group of chloramphenicol.  Source is the chloramphenicol producer Streptomyces venezuelae.",
  },
  {
    accession: "ARO:3000250",
    name: "ErmC",
    description:
      "ErmC is a methyltransferase that catalyzes the methylation of A2058 of the 23S ribosomal RNA in two steps. Expression of ErmC is inducible by erythromycin. The leader peptide causes attenuation of the mRNA and stabilizes the structure preventing further translation. When erythromycin is present, it binds the leader peptide causing a change in conformation allowing for the expression of ErmC.",
  },
  {
    accession: "ARO:3000251",
    name: "msrA",
    description:
      "MsrA is an ABC-F subfamily ribosomal protection protein expressed in Staphylococcus species which confers resistance to erythromycin and streptogramin B antibiotics through antibiotic target protection mechanisms. It is associated with plasmid DNA.",
  },
  {
    accession: "ARO:3000253",
    name: "glycopeptide resistance gene cluster VanD",
    description:
      "Homologous to vanA, contains a D-Ala-D-Lac ligase.  This cluster is constitutively expressed in the chromosome due to a dysfunctional D-ala-D-ala ligase and confers moderate resistance to both vancomycin and teicoplanin.\n\nGene orientation: vanRSYHDX",
  },
  {
    accession: "ARO:3000254",
    name: "emrY",
    description:
      "emrY is a multidrug transport that moves substrates across the inner membrane of the Gram-negative E. coli. It is a homolog of emrB.",
  },
  {
    accession: "ARO:3000255",
    name: "glycopeptide resistance gene cluster VanF",
    description:
      "Homologous to vanA, contains a D-Ala-D-Lac ligase. The vanF gene cluster is inducible and confers high resistance to vancomycin in Paenibacillus popilliae. vanF organisms remain susceptible to teicoplanin. Gene orientation: RSYZHFX",
  },
  {
    accession: "ARO:3000256",
    name: "glycopeptide resistance gene cluster VanM",
    description:
      "Homologous to vanA, contains a D-Ala-D-Lac ligase. The plasmid-located vanM gene cluster is inducible and confers high resistance to vancomycin and teicoplanin. Gene orientation: RSYHMX",
  },
  {
    accession: "ARO:3000257",
    name: "glycopeptide resistance gene cluster VanG",
    description:
      "Contains a D-Ala-D-Ser ligase. The vanG gene cluster is inducible and confers low resistance to vancomycin. vanG organisms remain susceptible to teicoplanin. It is the only van gene cluster that contains two vanY genes.\n\nGene orientation: vanRSYWGYT",
  },
  {
    accession: "ARO:3000258",
    name: "ErmC'",
    description:
      "ErmC' is ErmC but without the leader peptide and therefore expression is constitutive. Originally found on the plasmid pIM13 from Bacillus subtilis. It confers the MLSb phenotype.",
  },
  {
    accession: "ARO:3000259",
    name: "glycopeptide resistance gene cluster VanE",
    description:
      "Homologous to VanC, contains a D-Ala-D-Ser ligase. The chromosome-located vanE gene cluster is inducible and confers low resistance to vancomycin. vanE organisms remain susceptible to teicoplanin. Gene orientation: E(XY)TRS",
  },
  {
    accession: "ARO:3000260",
    name: "glycopeptide resistance gene cluster VanL",
    description:
      "Homologous to VanC, contains a D-Ala-D-Ser ligase. The chromosome-located vanL gene cluster is inducible and confers low resistance to vancomycin. vanL organisms remain susceptible to teicoplanin. It is the only van gene cluster with two vanT genes.\n\nGene orientation: vanL(XY)TmTrRS",
  },
  {
    accession: "ARO:3000261",
    name: "sigma factor conferring resistance to rifampin",
    description:
      "When bound to different sigma factors, RNA-polymerase may possess an altered sensitivity to rifampin-mediated inhibition.",
  },
  {
    accession: "ARO:3000262",
    name: "bla operon",
    description: "The bla operon is composed of blaZ/blaR1/blaI.",
  },
  {
    accession: "ARO:3000263",
    name: "marA",
    description:
      "In the presence of antibiotic stress, E. coli overexpresses the global activator protein MarA, which besides inducing MDR efflux pump AcrAB, also down- regulates synthesis of the porin OmpF.",
  },
  {
    accession: "ARO:3000264",
    name: "emrE",
    description:
      "EmrE is a small multidrug transporter that functions as a homodimer and that couples the efflux of small polyaromatic cations from the cell with the import of protons down an electrochemical gradient. EmrE is found in E. coli and P. aeruginosa.",
  },
  {
    accession: "ARO:3000265",
    name: "porin OmpF",
    description:
      "In the presence of antibiotic stress, E. coli overexpresses the global activator protein MarA, which besides inducing MDR efflux pump AcrAB, also down- regulates synthesis of the porin OmpF.",
  },
  {
    accession: "ARO:3000266",
    name: "ADP-ribosylation of antibiotic conferring resistance",
    description:
      "The inactivation of antibiotics by the enzymatic addition of ADP-ribose from NAD+.",
  },
  {
    accession: "ARO:3000267",
    name: "determinant of linezolid resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to linezolid antibiotics.",
  },
  {
    accession: "ARO:3000268",
    name: "mec operon",
    description: "The mec operon is composed of mecA/mecR1/mecI.",
  },
  {
    accession: "ARO:3000269",
    name: "brodimoprim",
    description:
      "Brodimoprim is a structural derivative of trimethoprim and an inhibitor of bacterial dihydrofolate reductase. The 4-methoxy group of trimethoprim is replaced with a bromine atom.",
  },
  {
    accession: "ARO:3000270",
    name: "protein modulating permeability to antibiotic",
    description:
      "Enzymes or other proteins either directly or indirectly reducing overall permeability to antibiotics.",
  },
  {
    accession: "ARO:3000271",
    name: "determinant of fosfomycin resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to fosfomycin antibiotics.",
  },
  {
    accession: "ARO:3000273",
    name: "antibiotic resistant DNA topoisomerase subunit gyrA",
    description:
      "Point mutations in gyrA confer antibiotic resistance by preventing drugs from binding the alpha-subunit of gyrase, essential for DNA supercoiling.",
  },
  {
    accession: "ARO:3000274",
    name: "antibiotic resistant DNA topoisomerase subunit parC",
    description:
      "Point mutations in parC confer antibiotic resistance by preventing drugs from binding the parC subunit of topoisomerase IV, essential for DNA decatanation and relaxation.",
  },
  {
    accession: "ARO:3000276",
    name: "cephem",
    description:
      "Cephems have a six-membered dihydrothiazine ring with a sulfur atom and double bond fused with its beta-lactam ring. This group includes the cephalosporins and cephamycins, the latter containing an additional alpha-methoxy group.",
  },
  {
    accession: "ARO:3000282",
    name: "sulfonamide antibiotic",
    description:
      "Sulfonamides are broad spectrum, synthetic antibiotics that contain the sulfonamide group. Sulfonamides inhibit dihydropteroate synthase, which catalyzes the conversion of p-aminobenzoic acid to dihydropteroic acid as part of the tetrahydrofolic acid biosynthetic pathway. Tetrahydrofolic acid is essential for folate synthesis, a precursor of many nucleotides and amino acids. Many sulfamides are taken with trimethoprim, an inhibitor of dihydrofolate reductase, also disturbing the trihydrofolic acid synthesis pathway.",
  },
  {
    accession: "ARO:3000284",
    name: "tetroxoprim",
    description:
      "Tetroxoprim is a trimethoprim derivative that inhibits bacterial dihydrofolate reductase.",
  },
  {
    accession: "ARO:3000299",
    name: "phosphonopyruvate decarboxylase",
    description:
      "Stabilizes the C-P bond in phosphonopyruvate formed by phophoenolpyruvate mutase by catalyzing the self-removal of the carboxyl group.",
  },
  {
    accession: "ARO:3000300",
    name: "lsaA",
    description:
      "LsaA is an ABC-F subfamily protein expressed in Enterococcus faecalis. It confers resistance to clindamycin, quinupristin-dalfopristin, and dalfopristin.",
  },
  {
    accession: "ARO:3000309",
    name: "emrD",
    description:
      "EmrD is a multidrug transporter from the Major Facilitator Superfamily (MFS) primarily found in Escherichia coli. EmrD couples efflux of amphipathic compounds with proton import across the plasma membrane.",
  },
  {
    accession: "ARO:3000315",
    name: "determinant of macrolide resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to macrolide antibiotics.",
  },
  {
    accession: "ARO:3000316",
    name: "mphA",
    description:
      "The mphA gene encodes for resistance enzyme MPH(2')-I which preferentially inactivate 14-membered macrolides (e.g.erythromycin, telithromycin, roxithromycin) over 16-membered macrolides (e.g.tylosin, spiramycin). It phosphorylates macrolides at 2'-OH hydroxyl of desosamine sugar of macrolides in a GTP-dependent manner.",
  },
  {
    accession: "ARO:3000317",
    name: "AAC(1)-I",
    description:
      "Acetylation of paromomycin, and apramycin, on the amino group at position 1 in E. coli, Actinomycete, Campylobacter spp.",
  },
  {
    accession: "ARO:3000318",
    name: "mphB",
    description:
      "The mphB gene encodes for MPH(2')-II. This enzymes phosphorylates 14-membered and 16-membered macrolides.  It phosphorylates macrolides in GTP- dependent manner at 2'-OH hydroxyl of desosamine sugar of macrolides.",
  },
  {
    accession: "ARO:3000319",
    name: "mphC",
    description:
      "The mphC gene was identified from Staphylococcus aureus.  This gene shows similarity to mphB gene from Escherchia coli.",
  },
  {
    accession: "ARO:3000320",
    name: "macrolide esterase",
    description:
      "Hydrolytic enzymes that cleave the macrocycle lactone ring of macrolide antibiotics.",
  },
  {
    accession: "ARO:3000321",
    name: "hydrolysis of macrolide macrocycle lactone ring",
    description:
      "Hydrolysis of the the macrocycle lactone ring of macrolide antibiotics.",
  },
  {
    accession: "ARO:3000322",
    name: "AAC(3)",
    description:
      "Acetylation of the aminoglycoside antibiotic on the amino group at position 3.",
  },
  {
    accession: "ARO:3000324",
    name: "sulfadiazine",
    description:
      "Sulfadiazine is a potent inhibitor of dihydropteroate synthase, interfering with the tetrahydrofolic biosynthesis pathway. Tetrahydrofolic acid is essential for folate synthesis, a precursor to many nucleotides and amino acids.",
  },
  {
    accession: "ARO:3000325",
    name: "sulfadimidine",
    description:
      "Sulfadimidine is an alkaline sulfonamide antibiotic that inhibits dihydropteroate synthase, and enzyme in the tetrahydrofolic acid biosynthesis pathway. This interferes with the production of folate, which is a precursor to many amino acids and nucleotides.",
  },
  {
    accession: "ARO:3000326",
    name: "ErmE",
    description:
      "ErmE is a methyltransferase found in the erythromycin producer Saccharopolyspora erythraea. Like other Erm enzymes, it catalyzes the methylation of A2058 of the 23S ribosomal RNA. The gene is found within the erythromycin biosynthetic cluster and is responsible for self-resistance.",
  },
  {
    accession: "ARO:3000327",
    name: "sulfadoxine",
    description:
      "Sulfadoxine is an inhibitor of dihydropteroate synthase, interfering with the tetrahydrofolic biosynthesis pathway. Tetrahydrofolic acid is essential for folate synthesis, a precursor to many nucleotides and amino acids.",
  },
  {
    accession: "ARO:3000328",
    name: "rRNA with mutation conferring antibiotic resistance",
    description:
      "Single nucleotide polymorphisms (SNPs) in rRNA can confer antibiotic resistance to drugs that target the bacterial ribosome.",
  },
  {
    accession: "ARO:3000329",
    name: "sulfamethoxazole",
    description:
      "Sulfamethoxazole is a sulfonamide antibiotic usually taken with trimethoprim, a diaminopyrimidine antibiotic. Sulfamethoxazole inhibits dihydropteroate synthase, essential to tetrahydrofolic acid biosynthesis. This pathway generates compounds used in the synthesis of many amino acids and nucleotides.",
  },
  {
    accession: "ARO:3000330",
    name: "sulfisoxazole",
    description:
      "Sulfisoxazole is an inhibitor of dihydropteroate synthase, interfering with the tetrahydrofolic biosynthesis pathway. Tetrahydrofolic acid is essential for folate synthesis, a precursor to many nucleotides and amino acids.",
  },
  {
    accession: "ARO:3000333",
    name: "macrolide phosphotransferase (MPH)",
    description:
      "Macrolide phosphotransferases (MPH) are enzymes encoded by macrolide phosphotransferase genes (mph genes). These enzymes phosphorylate macrolides in GTP dependent manner at 2'-OH of desosamine sugar thereby inactivating them. Characterized MPH's are differentiated based on their substrate specificity.",
  },
  {
    accession: "ARO:3000336",
    name: "23S rRNA with mutation conferring antibiotic resistance",
    description:
      "Point mutations in bacterial 23S rRNA from the large ribosomal subunit that confer resistance to antibiotics. Antibiotics such as linezolid block peptide synthesis through peptidyl transferase activity. Mutations in the 23S rRNA subunit reduce antibiotic binding affinity at specific sites, conferring resistance.",
  },
  {
    accession: "ARO:3000337",
    name: "iclaprim",
    description:
      "Iclaprim is a bactericidal compound that inhibits dihydrofolate reductase. It is used against clinically important Gram-positive pathogens, including methicillin-sensitive Staphylococcus aureus and methicillin-resistant S. aureus.",
  },
  {
    accession: "ARO:3000338",
    name: "linearization of antibiotic conferring resistance",
    description:
      "Genes that confer antibiotic resistance by hydrolyzing bonds to linearize and deactivate antibiotics.",
  },
  {
    accession: "ARO:3000341",
    name: "AAC(2')",
    description:
      "Acetylation of the aminoglycoside antibiotic on the amino group at position 2'.",
  },
  {
    accession: "ARO:3000342",
    name: "fosfomycin inactivation enzyme",
    description: "Enzymes that inactivate fosfomycin by chemical modification.",
  },
  {
    accession: "ARO:3000343",
    name: "tap",
    description:
      "Efflux pump proteins contained within Mycobacterial genomes which confer resistance to a number of different antibiotics including aminoglycosides, and tetracyclines.",
  },
  {
    accession: "ARO:3000344",
    name: "EmrAB-TolC",
    description:
      "EmrAB-TolC is a multidrug efflux system found in E. coli. EmrB is the electrochemical-gradient powered transporter; EmrA is the linker; and TolC is the outer membrane channel. It confers resistance to nalidixic acid and thiolactomycin.",
  },
  {
    accession: "ARO:3000345",
    name: "AAC(6')",
    description:
      "Acetylation of the aminoglycoside antibiotic on the amino group at position 6'.",
  },
  {
    accession: "ARO:3000347",
    name: "ErmA",
    description:
      "ErmA confers the MLSb phenotype. Similar to ErmC, Expression of ErmA is inducible by erythromycin. The leader peptide causes attenuation of the mRNA and stabilizes the structure preventing further translation. When erythromycin is present, it binds the leader peptide causing a change in conformation allowing for the expression of ErmA.",
  },
  {
    accession: "ARO:3000359",
    name: "fosfomycin phosphotransferase",
    description:
      "In the presence of ATP and magnesium (II), fosfomycin gets phosphorylated at the phosphate group resulting in a diphosphate group which inactivates the antibiotic.",
  },
  {
    accession: "ARO:3000361",
    name: "EreA",
    description:
      "EreA is an erythromycin esterase that hydrolyses the drug's lactone ring.",
  },
  {
    accession: "ARO:3000363",
    name: "EreB",
    description:
      "EreB is an erythromycin esterase-like protein that hydrolyses the drug's lactone ring.",
  },
  {
    accession: "ARO:3000368",
    name: "vanC",
    description:
      "VanC is a D-Ala-D-Ala ligase homolog that synthesizes D-Ala-D-Ser, an alternative substrate for peptidoglycan synthesis that reduces vancomycin binding affinity. It is specific to Enterococcus gallinarum and E. casseliflavus, providing intrinsic resistance.",
  },
  {
    accession: "ARO:3000369",
    name: "MtrCDE",
    description:
      "The mtr (multiple transferable resistance) system of Neisseria gonorrhoeae confers resistance to many hydrophobic agents including antibiotics, fatty-acids and detergents. MtrCDE is homologous to AcrAB-TolC, where MtrC is the membrane fusion protein, MtrD is the inner membrane transporter, and MtrE is the outer membrane channel protein.",
  },
  {
    accession: "ARO:3000370",
    name: "antibiotic resistant DNA topoisomerase subunit",
    description:
      "Many drugs target topoisomerases to inhibit DNA synthesis. Resistant DNA topoisomerase subunits prevent antibiotic binding and thus confer resistance.",
  },
  {
    accession: "ARO:3000372",
    name: "vanT",
    description:
      "VanT is a membrane bound serine racemase, converting L-serine to D-serine. It is associated with VanC, which incorporated D-serine into D-Ala-D-Ser terminal end of peptidoglycan subunits that have a decreased binding affinity with vancomycin. It was isolated from Enterococcus gallinarum.",
  },
  {
    accession: "ARO:3000373",
    name: "EmrKY-TolC",
    description:
      "EmrKY is a homolog of EmrAB found in E. coli. Together with TolC, it is a tripartite multidrug transporter.",
  },
  {
    accession: "ARO:3000375",
    name: "ErmB",
    description:
      "ErmB confers the MLSb phenotype. Similar to ErmC, expression of ErmB is inducible by erythromycin. The leader peptide causes attenuation of the mRNA and stabilizes the structure preventing further translation. When erythromycin is present, it binds the leader peptide causing a change in conformation allowing for the expression of ErmB.",
  },
  {
    accession: "ARO:3000376",
    name: "streptogramin vgb lyase",
    description:
      "vgb (Virginiamycin B) lyase inactivates type B streptogramin antibiotics by linearizing the streptogramin lactone ring at the ester linkage through an elimination mechanism, thus conferring resistance to these compounds.",
  },
  {
    accession: "ARO:3000377",
    name: "MexA",
    description:
      "MexA is the membrane fusion protein of the MexAB-OprM multidrug efflux complex.",
  },
  {
    accession: "ARO:3000378",
    name: "MexB",
    description:
      "MexB is the inner membrane multidrug exporter of the efflux complex MexAB-OprM.",
  },
  {
    accession: "ARO:3000379",
    name: "OprM",
    description:
      "OprM is an outer membrane factor protein found in Pseudomonas aeruginosa and Burkholderia vietnamiensis. It is part of the MexAB-OprM, MexVW-OprM, MexXY-OprM and the AmrAB-OprM complex.",
  },
  {
    accession: "ARO:3000380",
    name: "FosC",
    description:
      "FosC is an enzyme that phosphorylates fosfomycin to confer resistance",
  },
  {
    accession: "ARO:3000381",
    name: "antibiotic target replacement protein",
    description:
      "Alternate proteins that have the same functions as other antibiotic target proteins, but are structurally different and thus resistant to antibiotics. These can replace the activity of other antibiotic-sensitive proteins in the presence of antibiotics.",
  },
  {
    accession: "ARO:3000382",
    name: "azidamfenicol",
    description:
      "Azidamfenicol is a water soluble derivative of chloramphenicol, sharing the same mode of action of inhibiting peptide synthesis by interacting with the 23S RNA of the 50S ribosomal subunit.",
  },
  {
    accession: "ARO:3000383",
    name: "determinant of rifamycin resistance",
    description:
      "Enzymes, other proteins, or other gene products shown clinically to confer resistance to rifamycin (rifampin) antibiotics.",
  },
  {
    accession: "ARO:3000384",
    name: "AcrAB-TolC",
    description:
      "AcrAB-TolC is a tripartite RND efflux system that confers resistance to tetracycline, chloramphenicol, ampicillin, nalidixic acid, and rifampin  in Gram-negative bacteria. The system spans the cell membrane (AcrB) and the outer-membrane (TolC), and is linked together in the periplasm by AcrA.",
  },
  {
    accession: "ARO:3000385",
    name: "chloramphenicol",
    description:
      "Chloramphenicol is a bacteriostatic antimicrobial originally derived from the bacterium Streptomyces venezuelae. It was the first antibiotic to be manufactured synthetically on a large scale. It functions by inhibiting peptidyl transferase activity of the bacterial ribosome, binding to A2451 and A2452 residues in the 23S rRNA of the 50S ribosomal subunit and preventing peptide bond formation.",
  },
  {
    accession: "ARO:3000386",
    name: "MexAB-OprM",
    description:
      "MexAB-OprM is a multidrug efflux protein expressed in the Gram-negative Pseudomonas aeruginosa. MexA is the membrane fusion protein; MexB is the inner membrane transporter; and OprM is the outer membrane channel. MexAB-OprM is associated with resistance to fluoroquinolones, chloramphenicol, erythromycin, azithromycin, novobiocin, and certain \u03b2-lactams and lastly over-expression is linked to colistin resistance.",
  },
  {
    accession: "ARO:3000387",
    name: "phenicol antibiotic",
    description:
      "Phenicols are broad spectrum bacteriostatic antibiotics acting on bacterial protein synthesis. More specifically, the phenicols block peptide elongation by binding to the peptidyltansferase centre of the 70S ribosome.",
  },
  {
    accession: "ARO:3000388",
    name: "streptogramin biosynthesis",
    description: "Biosynthesis of streptoramin antibiotics.",
  },
  {
    accession: "ARO:3000389",
    name: "streptogramin B biosynthesis",
    description:
      "Type B streptogramins are cyclic hexa- or hepta-depsipeptides produced by NRPSs in Streptomyces.  The NRPS contains 6 or 7 modules arranged on one or more genes. SnbA catalyzes the activation of the first residue, 3-hydroxypicolynic acid. SnbC then activates and incorporates threonine and aminobutyric acid. Lastly, SnbDE activates and incoporates the last four amino acids. The thioesterase domain is responsible for peptide cyclization, and it is located at the end of the assembly line.",
  },
  {
    accession: "ARO:3000390",
    name: "rifampin ADP-ribosyltransferase (Arr)",
    description:
      "Enzyme responsible for the ADP-ribosylative inactivation of rifampin at the 23-OH position using NAD+.",
  },
  {
    accession: "ARO:3000391",
    name: "norA",
    description:
      "NorA is a multidrug efflux pump in Staphylococcus aureus that confers resistance to fluoroquinolones and other structurally unrelated antibiotics like acriflavine. It shares 30% similarity with NorA, and is a structural homolog of Bmr of Bacillus subtilis. It is regulated by arlRS and mgrA, the latter also known as NorR.",
  },
  {
    accession: "ARO:3000392",
    name: "Erm(37)",
    description:
      "Erm(37) is found in Mycobacterium species and confers the MLSb phenotype. In addition to methylation of A2058 this Erm methylates adjacent adenosines (A2057 and A2059) as well.",
  },
  {
    accession: "ARO:3000393",
    name: "glycopeptide biosynthesis",
    description: "Synthesis of glycopeptide antibiotics.",
  },
  {
    accession: "ARO:3000394",
    name: "Actinoplanes teichomyceticus teicoplanin gene cluster",
    description: "Biosynthesis of teicoplanin by Actinoplanes teichomyceticus",
  },
  {
    accession: "ARO:3000408",
    name: "determinant of sulfonamide resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to sulfonamide antibiotics.",
  },
  {
    accession: "ARO:3000410",
    name: "sul1",
    description:
      "Sul1 is a sulfonamide resistant dihydropteroate synthase of Gram-negative bacteria. It is linked to other resistance genes of class 1 integrons.",
  },
  {
    accession: "ARO:3000412",
    name: "sul2",
    description:
      "Sul2 is a sulfonamide resistant dihydropteroate synthase of Gram-negative bacteria, usually found on small plasmids.",
  },
  {
    accession: "ARO:3000413",
    name: "sul3",
    description:
      "Sul3 is a sulfonamide resistant dihydropteroate synthase similar to Sul1 and Sul2. Its resistance gene was found encoded in E. coli plasmid DNA of sulfonamide resistant isolates.",
  },
  {
    accession: "ARO:3000419",
    name: "quinolone resistance protein (qnr)",
    description:
      "Qnr proteins are pentapeptide repeat proteins that mimic DNA and protect the cell from the activity of fluoroquinolone antibiotics",
  },
  {
    accession: "ARO:3000421",
    name: "norB",
    description:
      "NorB is a multidrug efflux pump in Staphylococcus aureus that confers resistance to fluoroquinolones and other structurally unrelated antibiotics like tetracycline. It shares 30% similarity with NorB, and is a structural homolog of Blt of Bacillus subtilis. It is regulated by mgrA, also known as NorR.",
  },
  {
    accession: "ARO:3000423",
    name: "FomA",
    description:
      "In the presence of ATP and magnesium (II), fosfomycin gets phosphorylated at the phosphate group resulting in a diphosphate group which inactivates the antibiotic.",
  },
  {
    accession: "ARO:3000443",
    name: "rifampin glycosyltransferase",
    description:
      "The enzymatic inactivation of rifampin by glycosylation at the 23-OH position.",
  },
  {
    accession: "ARO:3000444",
    name: "rphA",
    description:
      "The enzymatic inactivation of rifampin by phosphorylation at the 21-OH position.",
  },
  {
    accession: "ARO:3000445",
    name: "rifampin monooxygenase",
    description:
      "Enzyme responsible for the decolorization of rifampin by monoxygenation.",
  },
  {
    accession: "ARO:3000446",
    name: "antibiotic resistant isoleucyl-tRNA synthetase (ileS)",
    description:
      "Mupirocin inhibits protein synthesis by interfering with isoleucyl-tRNA synthetase (ileS).  Mutations in ileS can confer low-level mupirocin resistance.",
  },
  {
    accession: "ARO:3000448",
    name: "QepA1",
    description:
      "QepA1 is a plasmid-mediated efflux pump in E. coli, shown to contribute to fluoroquinolone resistance. It is regulated by sox genes, also known as global stress regulators.",
  },
  {
    accession: "ARO:3000449",
    name: "FomB",
    description:
      "An enzyme which on its own cannot provide fosfomycin resistance, however in conjunction with FomA, it leads to the formation of fosfomycin with three phosphates total, which makes it inactive.",
  },
  {
    accession: "ARO:3000450",
    name: "hydroxylation of antibiotic conferring resistance",
    description:
      "Inactivation of an antibiotic via introduction a hydroxyl group (-OH).",
  },
  {
    accession: "ARO:3000451",
    name:
      "protein(s) and two-component regulatory system modulating antibiotic efflux",
    description:
      "Protein(s) and two component regulatory systems that directly or indirectly change rates of antibiotic efflux.",
  },
  {
    accession: "ARO:3000452",
    name: "fluoroquinolone resistant DNA topoisomerase",
    description:
      "Fluoroquinolones inhibit type II and type IV topoisomerases (2 strand breaking enzymes) such as GyrA/GyrB and ParC/ParE. Point mutations in the associated gyrA and parC genes, in particular in the 'quinolone resistance determining region' (QRDR), give rise to resistance to the class.",
  },
  {
    accession: "ARO:3000453",
    name: "streptogramin vat acetyltransferase",
    description:
      "vat (Virginiamycin acetyltransferases) enzymes catalyze the transfer of an acetyl group from acetyl-CoA to the secondary alcohol of streptogramin A compounds, thus inactivating virginiamycin-like antibiotics and conferring resistance to these compounds.",
  },
  {
    accession: "ARO:3000454",
    name: "polymyxin B",
    description:
      "Polymyxin B is mixture of mostly polymyxins B1 and B2, mainly used for resistant gram-negative infections. They are polypeptides with cationic detergent action on cell membranes.",
  },
  {
    accession: "ARO:3000455",
    name: "streptogramin A biosynthesis",
    description:
      "Type A streptogramins are produced by a hybrid NRPS/PKS composed of 8 NRPS modules and 2 PKS modules.",
  },
  {
    accession: "ARO:3000456",
    name: "thiamphenicol",
    description:
      "Derivative of Chloramphenicol. The nitro group (-NO2) is substituted by a sulfomethyl group (-SO2CH3).",
  },
  {
    accession: "ARO:3000457",
    name: "aminocoumarin resistant parE",
    description:
      "ParE is a subunit of topoisomerase IV, which decatenates and relaxes DNA to allow access to genes for transcription or translation. Point mutations in ParE prevent anticoumarin antibiotics from inhibiting DNA synthesis, thus conferring resistance.",
  },
  {
    accession: "ARO:3000458",
    name: "macrolide glycosyltransferase",
    description:
      "Macrolide glycosyltransferases are enzymes encoded by macrolide glycosyltransferase genes and inactivate macrolides by glycosylating them at 2'-OH of desosamine sugar moiety. They are predominantly found in macrolide producers and are also found in non-producers and are used as a resistance mechanism. Different variants of this enzyme has been reported.",
  },
  {
    accession: "ARO:3000459",
    name: "fosfomycin biosynthesis",
    description: "Biosynthesis of fosfomycin.",
  },
  {
    accession: "ARO:3000461",
    name: "florfenicol",
    description:
      "Florfenicol is a fluorine derivative of chloramphenicol, where the nitro group (-NO2) is substituted by a sulfomethyl group (-SO2CH3) and the hydroxyl group (-OH), by a fluorine group (-F). The action mechanism is the same as chloramphenicol's, where the antibiotic binds to the 23S RNA of the 50S subunit of bacterial ribosomes to inhibit protein synthesis.",
  },
  {
    accession: "ARO:3000462",
    name: "mgtA",
    description:
      "A macrolide glycosyltransferase encoded by the mgtA gene in Streptomyces lividans. This enzyme inactivates macrolides using UDP-glucose as a cofactor. Its optimal substrates are lankamycin, calcomycin, rosaramicin, methymycin, and pikromycin, while interactions with erythomycin, oldeandomycin, azithromycin, and tylosin were weaker. It is inactive against spiramycin and carbomycin. Mechanism first described by Cundliffe, 1992.",
  },
  {
    accession: "ARO:3000463",
    name: "gimA",
    description:
      "A macrolide glycosyltransferase encoded by the gimA gene in Streptomyces ambofaciens, a natural producer of the macrolide antibiotic spiramycin. Chalcomycin, methymycin, tylosin, pikromycin, rosaramicin, oleandomycin, josamycin, and carbomycin are preferred substrates of gimA glycosyltransferase, while erythromycin and spiramycin have notably low binding affinities. GimA may be able to inactivate spiramycin precursors. Described by Gourmelen et al. 1998.",
  },
  {
    accession: "ARO:3000464",
    name: "Neisseria gonorrhoeae porin PIB (por)",
    description:
      "Mutant forms of the porin Por result in reduced permeability to antibiotics, particularly tetracyclines and beta-lactams.",
  },
  {
    accession: "ARO:3000465",
    name: "ole glycosyltransferase",
    description:
      "OleI and OleD are glycosyltransferases found in Streptomyces antibioticus which is a natural producer of antibiotic oleandomycin. OleI glycosylates antibiotic oleandomycin whereas OleD can glycosylate a wide variety of macrolides.",
  },
  {
    accession: "ARO:3000466",
    name: "dalbavancin",
    description:
      "Dalbavancin is a semisynthetic second-generation lipoglycopeptide derived from teicoplanin. It binds to the D-Ala-D-Ala terminus of peptidoglycan precursors. It is used to treat Gram-positive bacteria and can be used to treat methicillin-resistant Staphylococcus aureus and vancomycin-resistant enterococci.",
  },
  {
    accession: "ARO:3000467",
    name: "NDM-5",
    description: "New Delhi beta-lactamase NDM-5.",
  },
  {
    accession: "ARO:3000469",
    name: "hydroxypropylphosphonic acid epoxidase",
    description:
      "The enzyme responsible for the final step of fosfomycin biosynthesis. It converts S-2-hydroxypropylphosphonic acid (S-HPP) into fosfomycin via an oxidative cyclalization reaction. It uses Iron (II) or Zinc (II) as cofactors.",
  },
  {
    accession: "ARO:3000472",
    name: "determinant of tetracycline resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to tetracycline antibiotics or tetracycline-like derivatives.",
  },
  {
    accession: "ARO:3000476",
    name: "tet(31)",
    description:
      "Tet31 is a tetracycline efflux pump found in Aeromonas salmonicida, a Gram-negative bacteria. It has also been shown to be expressed in Gallibacterium anatis.",
  },
  {
    accession: "ARO:3000477",
    name: "determinant of aminocoumarin resistance",
    description:
      "Point mutations to DNA topoisomerase enzymes clinically shown to confer resistance to aminocoumarins. In some cases, expression of parY(R), which encodes an aminocoumarin resistant topoisomerase IV, can also confer aminocoumarin resistance.",
  },
  {
    accession: "ARO:3000478",
    name: "tet(33)",
    description:
      "Tet33 is a tetracycline efflux pump found in Gram-positive bacteria, including Arthrobacter and Corynebacterium.",
  },
  {
    accession: "ARO:3000479",
    name: "aminocoumarin resistant gyrB",
    description:
      "Point mutations in DNA gyrase subunit B (gyrB) can result in resistance to aminocoumarins. These mutations usually involve arginine residues in organisms.",
  },
  {
    accession: "ARO:3000480",
    name: "aminocoumarin resistant parY",
    description:
      "Expression of parY(R), which encodes an aminocoumarin resistant topoisomerase IV, can confer aminocoumarin resistance.",
  },
  {
    accession: "ARO:3000481",
    name: "tet(35)",
    description:
      "Tet35 is a tetracycline efflux pump found in the Gram-negative Vibrio and Stenotrophomonas. It is unrelated to other tet resistance genes.",
  },
  {
    accession: "ARO:3000488",
    name: "telavancin",
    description:
      "Telavancin is a semi-synthetic derivative of vancomycin and is a second-generation lipoglycopeptide antibiotic. Telavancin inhibits cell wall synthesis by forming a complex with the D-Ala-D-Ala terminus of peptidoglycan precursors and preventing transglycosylation.",
  },
  {
    accession: "ARO:3000489",
    name: "sav1866",
    description:
      "Sav1866 is a multidrug efflux pump in the Gram-positive Staphylococcus aureus. It is a homolog of the human ABC transporter Mdr1 and pumps out toxic compounds including verapamil, tetraphenylphosphorchloride, and Hoechst 33342.",
  },
  {
    accession: "ARO:3000490",
    name: "tuberactinomycin",
    description:
      "Tuberactinomycins are a family of cyclic peptide antibiotics that are important in the treatment of tuberculosis. Tuberactinomycins contain nonproteinogenic amino acids and inhibit group I self-splicing RNA to disrupt prokaryotic protein synthesis.",
  },
  {
    accession: "ARO:3000491",
    name: "acrD",
    description:
      "AcrD is an aminoglycoside efflux pump expressed in E. coli. Its expression can be induced by indole, and is regulated by baeRS and cpxAR.",
  },
  {
    accession: "ARO:3000492",
    name: "gene involved in self-resistance to antibiotic",
    description:
      "Genes that are involved in conferring self resistance to antibiotic",
  },
  {
    accession: "ARO:3000494",
    name: "determinant of resistance to glycopeptide antibiotics",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to glycopeptide antibiotics.",
  },
  {
    accession: "ARO:3000495",
    name: "ErmD",
    description: "ErmD confers MLSb phenotype.",
  },
  {
    accession: "ARO:3000496",
    name: "vanXY",
    description:
      "VanXY is a protein with both D,D-carboxypeptidase and D,D-dipeptidase activity, found in Enterococcus gallinarum. It cleaves and removes the terminal D-Ala of peptidoglycan subunits for the incorporation of D-Ser by VanC. D-Ala-D-Ser has low binding affinity with vancomycin.",
  },
  {
    accession: "ARO:3000497",
    name: "ethambutol",
    description:
      "Ethambutol is an antimycobacterial drug prescribed to treat tuberculosis. It is usually given in combination with other tuberculosis drugs, such as isoniazid, rifampicin, and pyrazinamide. Ethambutol inhibits arabinosyl biosynthesis, disrupting mycobacterial cell wall formation.",
  },
  {
    accession: "ARO:3000498",
    name: "ErmF",
    description: "ErmF confers the MLSb phenotype.",
  },
  {
    accession: "ARO:3000499",
    name: "AcrE",
    description: "AcrE is a membrane fusion protein, similar to AcrA.",
  },
  {
    accession: "ARO:3000500",
    name: "phosphonoacetaldehyde methyltransfererase",
    description:
      "The enzyme responsible for converting phosphonoacetaldehyde to (S)-2-/nhydroxypropylphosphonic acid (S-HPP) via an anionic methyl attack, which also reduces the aldehyde to an alcohol.",
  },
  {
    accession: "ARO:3000501",
    name: "rpoB2",
    description:
      "Due to gene duplication, the genomes of Nocardia species include both rifampin-sensitive beta-subunit of RNA polymerase (rpoB) and rifampin-resistant beta-subunit of RNA polymerase (rpoB2) genes, with ~88% similarity between the two gene products. Expression of the rpoB2 variant results in replacement of rifampin sensitivity with rifampin resistance.",
  },
  {
    accession: "ARO:3000502",
    name: "AcrF",
    description: "AcrF is a inner membrane transporter, similar to AcrB.",
  },
  {
    accession: "ARO:3000503",
    name: "AcrEF-TolC",
    description:
      "AcrEF-TolC is a tripartite multidrug efflux system similar to AcrAB-TolC and found in Gram-negative bacteria. AcrE is the membrane fusion protein, AcrF is the inner membrane transporter, and TolC is the outer membrane channel protein.",
  },
  {
    accession: "ARO:3000504",
    name: "golS",
    description:
      "GolS is a regulator activated by the presence of gold, and promotes the expression of the MdsABC efflux pump.",
  },
  {
    accession: "ARO:3000506",
    name: "MexR",
    description:
      "MexR is the repressor of the MexRAB-OprM operon. Mutant forms of mexR result in up-regulation of efflux pump system MexAB-OprM.",
  },
  {
    accession: "ARO:3000507",
    name: "rifampin-resistant RNA polymerase-binding protein",
    description:
      "Proteins which have been experimentally shown to protect RNA-polymerase from rifampin inhibition.",
  },
  {
    accession: "ARO:3000508",
    name: "gadX",
    description:
      "GadX is an AraC-family regulator that promotes mdtEF expression to confer multidrug resistance.",
  },
  {
    accession: "ARO:3000510",
    name: "Staphylococcus mupB conferring resistance to mupirocin",
    description:
      "An alternative isoleucyl-tRNA synthetase conferring resistance to mupirocin.",
  },
  {
    accession: "ARO:3000515",
    name: "evgSA",
    description:
      "EvgSA is a two-component regulatory system that regulates MdtEF and EmrKY expression for multidrug resistance. EvgS is a sensor protein that phosphorylates the regulatory protein EvgA, though EvgA can be phosphorylated by other methods when it is overexpressed.",
  },
  {
    accession: "ARO:3000516",
    name: "emrR",
    description:
      "EmrR is a negative regulator for the EmrAB-TolC multidrug efflux pump in E. coli. Mutations lead to EmrAB-TolC overexpression.",
  },
  {
    accession: "ARO:3000517",
    name: "rifaximin",
    description:
      "Rifaximin is a semi-synthetic rifamycin used to treat traveller's diarrhea. Rifaximin inhibits RNA synthesis by binding to the beta subunit of bacterial RNA polymerase.",
  },
  {
    accession: "ARO:3000518",
    name: "CRP",
    description:
      "CRP is a global regulator that represses MdtEF multidrug efflux pump expression.",
  },
  {
    accession: "ARO:3000519",
    name: "antibiotic target modifying enzyme",
    description:
      "Enzymes that confer resistance by modifying antibiotic targets.",
  },
  {
    accession: "ARO:3000520",
    name: "isoniazid",
    description:
      "Isoniazid is an organic compound that is the first-line anti tuberculosis medication in prevention and treatment. As a prodrug, it is activated by mycobacterial catalase-peroxidases such as M. tuberculosis KatG. Isoniazid inhibits mycolic acid synthesis, which prevents cell wall synthesis in mycobacteria.",
  },
  {
    accession: "ARO:3000521",
    name: "Staphylococcus mupA conferring resistance to mupirocin",
    description:
      "An alternative isoleucyl-tRNA synthetase conferring resistance to mupirocin.",
  },
  {
    accession: "ARO:3000522",
    name: "ErmG",
    description:
      "ErmG is a rRNA adenine N-6-methyltransferase that protects the ribosome from inactivation due to antibiotic binding.",
  },
  {
    accession: "ARO:3000524",
    name: "cpxAR",
    description:
      "CpxAR is a two-component regulatory system that involves a sensor kinase, CpxA, and the regulator CpxR. When the membrane envelope is stressed, CpxAR promotes acrD and mdtABC expression to confer multidrug resistance through efflux.",
  },
  {
    accession: "ARO:3000525",
    name: "antibiotic A40926",
    description:
      "A40926 is a glycopeptide antibiotic produced by Nonomuraea sp. ATCC 39727. It is precusor of the second-generation glycopeptide antibiotic dalbavancin.",
  },
  {
    accession: "ARO:3000526",
    name: "cmeR",
    description:
      "CmeR is a repressor for the CmeABC multidrug efflux pump, binding to the cmeABC promoter region.",
  },
  {
    accession: "ARO:3000527",
    name: "polyamine antibiotic",
    description:
      "Polyamine antibiotics are organic compounds having two or more primary amino groups.",
  },
  {
    accession: "ARO:3000528",
    name: "chlortetracycline",
    description:
      "Chlortetracycline was an early, first-generation tetracycline antibiotic developed in the 1940's. It inhibits bacterial protein synthesis by binding to the 30S subunit of bacterial ribosomes, preventing the aminoacyl-tRNA from binding to the ribosome.",
  },
  {
    accession: "ARO:3000529",
    name: "determinant of mupirocin resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to mupirocin antibiotics. Mutations in ileS is shown to confer low-level mupirocin resistance, while high-level resistance is conferred by the presence of alternative isoleucyl-tRNA synthetases.",
  },
  {
    accession: "ARO:3000530",
    name: "rifabutin",
    description:
      "Rifabutin is a semisynthetic rifamycin used in tuberculosis therapy. It inhibits DNA-dependent RNA synthesis.",
  },
  {
    accession: "ARO:3000531",
    name: "baeSR",
    description:
      "BaeSR is a two component regulatory system for efflux proteins in Gram-negative bacteria. BaeR is a response regulator, while BaeS is a sensor kinase.",
  },
  {
    accession: "ARO:3000532",
    name: "aminocoumarin biosynthesis",
    description: "Biosynthesis of aminocoumarin antibiotics.",
  },
  {
    accession: "ARO:3000533",
    name: "macA",
    description:
      "MacA is a membrane fusion protein that forms an antibiotic efflux complex with MacB and TolC. macA corresponds to 1 locus in Pseudomonas aeruginosa PAO1 and 1 locus in Pseudomonas aeruginosa LESB58.",
  },
  {
    accession: "ARO:3000534",
    name: "rifapentine",
    description:
      "Rifapentine is a semisynthetic rifamycin that inhibits DNA-dependent RNA synthesis. It is often used in the treatment of tuberculosis and leprosy.",
  },
  {
    accession: "ARO:3000535",
    name: "macB",
    description:
      "MacB is an ATP-binding cassette (ABC) transporter that exports macrolides with 14- or 15- membered lactones. It forms an antibiotic efflux complex with MacA and TolC. macB corresponds to 1 locus in Pseudomonas aeruginosa PAO1 and 1 locus in Pseudomonas aeruginosa LESB58.",
  },
  {
    accession: "ARO:3000536",
    name: "Streptomyces spheroides novobiocin biosynthetic gene cluster",
    description:
      "Novobiocin is produced from biosynthetic clusters in Streptomyces spheroides and Streptomyces niveus.",
  },
  {
    accession: "ARO:3000545",
    name: "MacAB-TolC",
    description:
      "MacAB-TolC is an ABC efflux pump complex expressed in E. coli and Salmonella enterica. It confers resistance to macrolides, including erythromycin.",
  },
  {
    accession: "ARO:3000546",
    name:
      "Streptomyces roseochromogenes subsp. oscitans clorobiocin biosynthetic gene cluster",
    description:
      "Clorobiocin is produced by a biosynthetic cluster in Streptomyces roseochromogenes.",
  },
  {
    accession: "ARO:3000547",
    name: "arlRS",
    description:
      "ArlRS is a two-component regulatory system for NorA. ArlS phosphorylates ArlR to promote NorA expression.",
  },
  {
    accession: "ARO:3000548",
    name:
      "Streptomyces rishiriensis strain DSM 40489 coumermycin A1 biosynthetic gene cluster",
    description:
      "Coumermycin A1 is produced by a biosynthetic cluster in Streptomyces rishiriensis.",
  },
  {
    accession: "ARO:3000549",
    name: "adeS",
    description:
      "AdeS is a sensor kinase in the AdeRS regulatory system of AdeABC. It is essential for AdeABC expression.",
  },
  {
    accession: "ARO:3000550",
    name: "aztreonam",
    description:
      "Aztreonam was the first monobactam discovered, and is greatly effective against Gram-negative bacteria while inactive against Gram-positive bacteria. Artreonam is a poor substrate for beta-lactamases, and may even act as an inhibitor. In Gram-negative bacteria, Aztreonam interferes with filamentation, inhibiting cell division and leading to cell death.",
  },
  {
    accession: "ARO:3000551",
    name: "organoarsenic antibiotic",
    description:
      "Organoarsenic antibiotics are arsenic-containing compounds with antibacterial effects. The organoarsenic antibiotic arsphenamine and its derivatives were developed in the 1910s as the first modern chemotherapeutic agents.",
  },
  {
    accession: "ARO:3000552",
    name: "arsphenamine",
    description:
      "Arsphenamine, also known as Salvarsan and 606, is a drug that was used beginning in the 1910s to treat syphilis and trypanosomiasis. It is an organoarsenic compound and was the first modern chemotherapeutic agent.",
  },
  {
    accession: "ARO:3000553",
    name: "adeR",
    description:
      "AdeR is a positive regulator of AdeABC efflux system. AdeR inactivation leads to susceptibility to aminoglycoside antibiotics.",
  },
  {
    accession: "ARO:3000554",
    name: "mupirocin",
    description:
      "Mupirocin, also known as pseudomonic acid, is a bacteriostatic polyketide antibiotic from Pseudomonas fluorescens used to treat S. aureus and MRSA. It inhibits Ile tRNA synthetase.",
  },
  {
    accession: "ARO:3000555",
    name: "macrolide biosynthesis",
    description: "Biosynthesis of macrolide antibiotics.",
  },
  {
    accession: "ARO:3000556",
    name: "tet44",
    description:
      "Tet44 is a tetracycline resistance gene found in Campylobacter fetus, and binds to the ribosome to confer antibiotic resistance as a ribosomal protection protein.",
  },
  {
    accession: "ARO:3000557",
    name: "antibiotic inactivation enzyme",
    description:
      "Enzyme that catalyzes the inactivation of an antibiotic resulting in resistance.  Inactivation includes chemical modification, destruction, etc.",
  },
  {
    accession: "ARO:3000558",
    name: "sulfonamide resistant dihydropteroate synthase",
    description:
      "Folate is an essential precursor for many nucleotides and amino acids, making folate biosynthesis a desirable target for antibiotics. Mutation to enzymes within the folate biosynthetic pathway can reverse antibiotic action. Modification of dihydropteroate synthase (folP) can result in resistance to sulfonamide antibiotics. Resistance can be conferred by mutation to the native folP gene or by acquiring a resistant DHPS (sul) gene through gene transfer. The sul gene has three known variants (sul1, sul2, sul3) and encodes a highly resistant dihydropteroate synthase that can be acquired via transposons or plasmids.",
  },
  {
    accession: "ARO:3000559",
    name: "adeN",
    description:
      "AdeN is a repressor of AdeIJK, a RND-type efflux pump in Acinetobacter baumannii. Its inactivation increases expression of AdeJ.",
  },
  {
    accession: "ARO:3000560",
    name: "Erm 23S ribosomal RNA methyltransferase",
    description:
      "Erm proteins are part of the RNA methyltransferase family and methylate A2058 (E. coli nomenclature) of the 23S ribosomal RNA conferring degrees of resistance to Macrolides, Lincosamides and Streptogramin b. This is called the MLSb phenotype.",
  },
  {
    accession: "ARO:3000561",
    name: "tet(30)",
    description:
      "Tet30 is a tetracycline efflux pump found in agrobacterium, a Gram-negative bacterium.",
  },
  {
    accession: "ARO:3000565",
    name: "tet(38)",
    description:
      "Tet38 is a tetracycline efflux pump found in the Gram-positive Staphylococcus aureus. It is regulated by mgrA, which also regulates NorB.",
  },
  {
    accession: "ARO:3000566",
    name: "tet(39)",
    description:
      "Tet39 is a tetracycline efflux pump found in Gram-negative bacteria, including Brevundimonas, Stenotrophomonas, Enterobacter, Alcaligenes, Acinetobacter, and Providencia.",
  },
  {
    accession: "ARO:3000567",
    name: "tet(40)",
    description:
      "Tet40 is a tetracycline efflux pump found in the Gram-positive Clostridium. It is similar to tetA(P).",
  },
  {
    accession: "ARO:3000568",
    name: "subclass B1 (metallo-) beta-lactamase",
    description:
      "Subclass B1 possess a binuclear active site.  Within this active site can be either one or two Zn(II) ions.  This subclass is able to hydrolyze penicillins, cephalosporins and carbapenems.  This is the most clinically relevant subclass of MBLs.",
  },
  {
    accession: "ARO:3000569",
    name: "tet(41)",
    description:
      "Tet41 is a tetracycline efflux pump found in Serratia, a Gram-negative bacterium.  It is related to Acinetobacter Tet(39).",
  },
  {
    accession: "ARO:3000570",
    name: "subclass B2 (metallo-) beta-lactamase",
    description:
      "Metallo-beta-lactmases of subclass B2 contain only one Zn ion in their active site and selectively hydrolyze carbapenems.",
  },
  {
    accession: "ARO:3000571",
    name: "subclass B3 (metallo-) beta-lactamase",
    description:
      "Metallo-beta-lactamases of subclass B3 are similar to B1 in that they have activity against penicillins, cephalosporins and carbapenems; however, the are only active with two Zn(II) ions in the active site.",
  },
  {
    accession: "ARO:3000572",
    name: "tet(42)",
    description:
      "Tet42 is a tetracycline efflux pump found in both Gram-negative (Pseudomonas) and Gram-positive (Microbacterium, Bacillus, Staphylococcus, Paenibacillus) bacteria.",
  },
  {
    accession: "ARO:3000573",
    name: "tet(43)",
    description:
      "Tet(43) is a tetracycline resistance gene with unknown origins, isolated from metagenomic DNA.",
  },
  {
    accession: "ARO:3000574",
    name: "vanR",
    description:
      "VanR is a OmpR-family transcriptional activator in the VanSR regulatory system. When activated by VanS, it promotes cotranscription of VanA, VanH, and VanX.",
  },
  {
    accession: "ARO:3000575",
    name: "vanU",
    description:
      "VanU is a transcriptional activator of vancomycin resistance genes.",
  },
  {
    accession: "ARO:3000576",
    name: "rifampin inactivation enzyme",
    description:
      "Enzymes that inactivate rifampin antibiotics by chemical modification.",
  },
  {
    accession: "ARO:3000577",
    name: "Bc beta-lactamase",
    description:
      "Bacillus cereus beta-lactamases are zinc metallo-beta-lactamases that hydrolyze a large number of penicillins and cephalosporins.",
  },
  {
    accession: "ARO:3000578",
    name: "CcrA",
    description: "This CcrA beta-lactamases is found in Bacteroides fragilis.",
  },
  {
    accession: "ARO:3000579",
    name: "Chryseobacterium meningosepticum BlaB",
    description:
      "This BlaB specific to Chryseobacterium meningosepticum mediates resistance against many beta-lactam antibiotics, notably penams and carbapenems.",
  },
  {
    accession: "ARO:3000580",
    name: "SPM beta-lactamase",
    description:
      "Sao Paulo metallo-beta-lactamase (SPM-1) confers resistance to carbapenem in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3000581",
    name: "CphA beta-lactamase",
    description:
      "CphA is an Ambler Class B MBL; subclass B2 originally isolated from Aeromonas hydrophilia.  This enzyme has specific activity against carbapenems and is active as a mono-zinc protein.",
  },
  {
    accession: "ARO:3000582",
    name: "L1 beta-lactamase",
    description:
      "L1 is an Ambler class B MBL; subclass B3 originally isolated from Stenotrophomonas maltophilia.  It has activity against a broad range of beta-lactams and is only active with two Zn(II) ions in the active site.",
  },
  {
    accession: "ARO:3000583",
    name: "pristinamycin IA",
    description:
      "Pristinamycin 1A is a type B streptogramin antibiotic produced by Streptomyces pristinaespiralis. It binds to the P site of the 50S subunit of the bacterial ribosome, preventing the extension of protein chains.",
  },
  {
    accession: "ARO:3000584",
    name: "quinupristin",
    description:
      "Quinupristin is a type B streptogramin and a semisynthetic derivative of pristinamycin 1A. It is a component of the drug Synercid and interacts with the 50S subunit of the bacterial ribosome to inhibit protein synthesis.",
  },
  {
    accession: "ARO:3000586",
    name: "pulvomycin",
    description:
      "Pulvomycin is a polyketide antibiotic that binds elongation factor Tu (EF-Tu) to inhibit protein biosynthesis by preventing the formation of the ternary complex (EF-Tu*GTP*aa-tRNA). Phenotypically, it was shown that pulvomycin sensitivity is dominant over resistance.",
  },
  {
    accession: "ARO:3000587",
    name: "sulbactam",
    description: "inhibitor of non-AmpC Ser beta-lactmases",
  },
  {
    accession: "ARO:3000588",
    name: "avibactam",
    description:
      "Ser beta-lactamase inhibitor targeting class A, class C, and some class D enzymes.",
  },
  {
    accession: "ARO:3000589",
    name: "NDM-1",
    description:
      "NDM-1 is a metallo-beta-lactamase isolated from Klebsiella pneumoniae with nearly complete resistance to all beta-lactam antibiotics.",
  },
  {
    accession: "ARO:3000590",
    name: "NDM-2",
    description:
      "NDM-2 was isolated from a strain of Acinetobacter baumannii in Germany from a patient hospitalized in Cairo.  A single amino acid substitution (P28R) differentiates this gene from NDM-1 and the two enzymes appear to have an identical spectrum of hydrolysis.",
  },
  {
    accession: "ARO:3000591",
    name: "ErmI",
    description:
      "ErmI is a methyltransferase found in the midecamycin producer Streptomyces mycarofaciens. Like other Erm enzymes, it catalyzes the methylation of A2058 of the 23S ribosomal RNA. The gene is found within the midecamycin biosynthetic cluster and is responsible for self-resistance",
  },
  {
    accession: "ARO:3000592",
    name: "ErmN",
    description:
      "ErmN is a methyltransferase found in the tylosin producer Streptomyces fradiae. Like other Erm enzymes, it catalyzes the methylation of A2058 of the 23S ribosomal RNA. Specifically, this enzyme transfers only one methyl group. The gene is found in the tylosin biosynthetic cluster and is responsible for self-resistance to tylosin.",
  },
  {
    accession: "ARO:3000593",
    name: "ErmQ",
    description: "ErmQ confers MLSb phenotype.",
  },
  {
    accession: "ARO:3000594",
    name: "ErmR",
    description:
      "ErmR is a methyltransferase found in the erythromycin producer Aeromicrobium erythreum. Like other Erm enzymes, it catalyzes the methylation of A2058 of the 23S ribosomal RNA. The gene is found within the erythromycin biosynthetic cluster and is responsible for self-resistance.",
  },
  {
    accession: "ARO:3000595",
    name: "ErmT",
    description: "ErmT confers MLSb phenotype.",
  },
  {
    accession: "ARO:3000596",
    name: "ErmX",
    description:
      "ErmX is a rRNA methyltransferase that protects the ribosome from inactivation due to antibiotic binding.",
  },
  {
    accession: "ARO:3000598",
    name: "Erm(31)",
    description:
      "Erm(31) confers a MLSb resistant phenotype. Along with erm(30), these genes are responsible for self-resistance in the pikromycin/narbomycin/methymycin/neomethymycin producer, Streptomyces venezuelae.",
  },
  {
    accession: "ARO:3000599",
    name: "Erm(33)",
    description: "ErmD confers MLSb phenotype.",
  },
  {
    accession: "ARO:3000600",
    name: "Erm(34)",
    description: "ErmD confers MLSb phenotype.",
  },
  {
    accession: "ARO:3000601",
    name: "Erm(38)",
    description: "ErmD confers MLSb phenotype.",
  },
  {
    accession: "ARO:3000602",
    name: "Erm(39)",
    description: "ErmD confers MLSb phenotype.",
  },
  {
    accession: "ARO:3000603",
    name: "Erm(41)",
    description: "Erm41 confers MLSb phenotype.",
  },
  {
    accession: "ARO:3000604",
    name: "Erm(35)",
    description: "ErmD confers MLSb phenotype.",
  },
  {
    accession: "ARO:3000605",
    name: "Erm(36)",
    description: "ErmD confers MLSb phenotype.",
  },
  {
    accession: "ARO:3000606",
    name: "FEZ-1",
    description:
      "FEZ-1 is an Ambler class B MBL; subclass B3 first isolated from Legionella gormanii.  It has activity against a broad range of beta-lactams and is only active with two Zn(II) ions in the active site.",
  },
  {
    accession: "ARO:3000610",
    name: "gene detected by antibiotic resistance screening microarray",
    description:
      "Many genes involved in antibiotic resistance are mobile and microarray probes for these genes can thus provide positive results for any number of organisms. CARD stores the coding sequences of genes targeted by the probe sets of available antibiotic resistance microarrays.",
  },
  {
    accession: "ARO:3000614",
    name: "mefE",
    description:
      "mefE is a proton motive efflux pump in Streptococcus pneumoniae that confers resistance to macrolides. It is found on the same operon as mefA and the ABC-efflux pump mel.",
  },
  {
    accession: "ARO:3000615",
    name: "mefA",
    description:
      "mefA is a motive efflux pump in Streptococcus pneumoniae that confers resistance to macrolides. It is found on the same operon as mefE and the ABC-efflux pump mel.",
  },
  {
    accession: "ARO:3000616",
    name: "mel",
    description:
      "Mel, a homolog of MsrA, is an ABC-F subfamily protein associated with macrolide resistance. It is expressed on the same operon as mefA and mefE, both MFS-type efflux proteins that confer macrolide resistance.",
  },
  {
    accession: "ARO:3000617",
    name: "mecA",
    description:
      "A foreign PBP2a acquired by lateral gene transfer that able to perform peptidoglycan synthesis in the presence of beta-lactams.",
  },
  {
    accession: "ARO:3000619",
    name: "fluoroquinolone resistant parC",
    description:
      "ParC is a subunit of topoisomerase IV, which decatenates and relaxes DNA to allow access to genes for transcription or translation. Point mutations in ParC prevent fluoroquinolone antibiotics from inhibiting DNA synthesis, and confer low-level resistance. Higher-level resistance results from both gyrA and parC mutations.",
  },
  {
    accession: "ARO:3000620",
    name: "adeL",
    description:
      "AdeL is a regulator of AdeFGH in Acinetobacter baumannii. AdeL mutations are associated with AdeFGH overexpression and multidrug resistance.",
  },
  {
    accession: "ARO:3000621",
    name: "PC1 beta-lactamase (blaZ)",
    description:
      "The blaZ beta-lactamase is found in Bacillus subtilis and Staphylococcus aureus.",
  },
  {
    accession: "ARO:3000622",
    name: "colistin A",
    description:
      "Colistin A, or polymyxin E1, has a 6-octanoic acid lipid tail. Polymyxins disrupt the cell membrane of Gram-negative bacteria.",
  },
  {
    accession: "ARO:3000624",
    name: "colistin B",
    description:
      "Colistin B, or polymyxin E2, has a 6-heptanoic acid lipid tail. Polymyxins disrupt the cell membrane of Gram-negative bacteria.",
  },
  {
    accession: "ARO:3000625",
    name: "polymyxin B1",
    description:
      "Polymyxin B1 is in the family of polymyxin lipopeptides with a 6-methyloctanoic acid acyl group. These antibiotics disrupt the cell membrane of Gram-negative bacteria.",
  },
  {
    accession: "ARO:3000626",
    name: "polymyxin B2",
    description:
      "Polymyxin B2 is in the family of polymyxin lipopeptides with a 6-methylheptanoic acid acyl group. These antibiotics disrupt the cell membrane of Gram-negative bacteria.",
  },
  {
    accession: "ARO:3000627",
    name: "polymyxin B3",
    description:
      "Polymyxin B3 is in the family of polymyxin lipopeptides with an octanoic acid acyl group. These antibiotics disrupt the cell membrane of Gram-negative bacteria.",
  },
  {
    accession: "ARO:3000628",
    name: "polymyxin B4",
    description:
      "Polymyxin B4 is in the family of polymyxin lipopeptides with a heptanoic acid acyl group. These antibiotics disrupt the cell membrane of Gram-negative bacteria.",
  },
  {
    accession: "ARO:3000629",
    name: "bacitracin A",
    description:
      "Bacitracin A is the primary component of bacitracin. It contains many uncommon amino acids and interferes with bacterial cell wall synthesis.",
  },
  {
    accession: "ARO:3000630",
    name: "bacitracin B",
    description:
      "Bacitracin B is a component of bacitracin, an antibiotic mixture that interferes with bacterial cell wall synthesis. It differs from Bacitracin A with a valine instead of an isoleucine in its peptide.",
  },
  {
    accession: "ARO:3000631",
    name: "bacitracin F",
    description:
      "Bacitracin F is a component of bacitracin, an antibiotic mixture that interferes with bacterial cell wall synthesis. It is formed when the thiazoline ring of bacitracin A is oxidatively deaminated.",
  },
  {
    accession: "ARO:3000632",
    name: "benzylpenicillin",
    description:
      "Benzylpenicillin, commonly referred to as penicillin G, is effective against both Gram-positive and Gram-negative bacteria. It is unstable in acid.",
  },
  {
    accession: "ARO:3000633",
    name: "phenoxymethylpenicillin",
    description:
      "Phenoxymethylpenicillin, or penicillin V, is a penicillin derivative that is acid stable but less active than benzylpenicillin (penicillin G).",
  },
  {
    accession: "ARO:3000634",
    name: "propicillin",
    description:
      "Propicillin is an orally taken penicillin derivative that has high absorption but poor activity.",
  },
  {
    accession: "ARO:3000635",
    name: "dicloxacillin",
    description:
      "Dicloxacillin is a penicillin derivative that has an extra chlorine atom in comparison to cloxacillin. While more active than cloxacillin, its high affinity for serum protein reduces its activity in human serum in vitro.",
  },
  {
    accession: "ARO:3000636",
    name: "flucloxacillin",
    description:
      "Flucloxacillin is similar to cloxacillin, with an extra additional fluorine atom.",
  },
  {
    accession: "ARO:3000637",
    name: "ampicillin",
    description:
      "Ampicillin is a penicillin derivative that is highly acid stable, with its activity similar to benzylpenicillin.",
  },
  {
    accession: "ARO:3000638",
    name: "azlocillin",
    description:
      "Azlocillin is a semisynthetic derivative of penicillin that is notably active against Ps. aeruginosa and other Gram-negative bacteria.",
  },
  {
    accession: "ARO:3000639",
    name: "mezlocillin",
    description: "Mezlocillin is a penicillin derivative taken parenterally.",
  },
  {
    accession: "ARO:3000640",
    name: "doripenem",
    description:
      "Doripenem is a carbapenem with a broad range of activity against Gram-positive and Gram-negative bacteria, and along with meropenem, it is the most active beta-lactam antibiotic against Pseudomonas aeruginosa. It inhibits bacterial cell wall synthesis.",
  },
  {
    accession: "ARO:3000641",
    name: "cefalexin",
    description:
      "Cefalexin is a cephalosporin antibiotic that causes filamentation. It is resistant to staphylococcal beta-lactamase, but degraded by enterobacterial beta-lactamases.",
  },
  {
    accession: "ARO:3000642",
    name: "cefadroxil",
    description:
      "Cefadroxil, or p-hydroxycephalexin, is an cephalosporin antibiotic similar to cefalexin.",
  },
  {
    accession: "ARO:3000643",
    name: "cefotiam",
    description:
      "Cefotiam is a cephalosporin antibiotic with similar activity to cefuroxime but more active against enterobacteria. It is consumed orally as the prodrug cefotiam hexetil.",
  },
  {
    accession: "ARO:3000644",
    name: "cefaclor",
    description:
      "Cefaclor is a semisynthetic cephalosporin derived from cephalexin. It has broad-spectrum antibiotic activity.",
  },
  {
    accession: "ARO:3000645",
    name: "cefotaxime",
    description:
      "Cefotaxime is a semisynthetic cephalosporin taken parenterally. It is resistant to most beta-lactamases and active against Gram-negative rods and cocci due to its aminothiazoyl and methoximino functional groups.",
  },
  {
    accession: "ARO:3000646",
    name: "cefixime",
    description:
      "Cefixime is a cephalosporin resistant to most beta-lactamases. It is active against many enterobacteria, but activity against staphylococci is poor.",
  },
  {
    accession: "ARO:3000647",
    name: "cefpodoxime",
    description:
      "Cefpodoxime is a semisynthetic cephalosporin that acts similarly to cefotaxime with broad-spectrum activity. It is stable to many plasmid-mediated beta-lactamses. Cefpodoxime is consumed as the prodrug cefpodoxime proxetil.",
  },
  {
    accession: "ARO:3000648",
    name: "ceftibuten",
    description:
      "Ceftibuten is a semisynthetic cephalosporin active against Gram-negative bacilli. It is resistant against many plasmid-mediated beta-lactamases.",
  },
  {
    accession: "ARO:3000649",
    name: "cefditoren",
    description:
      "Cefditoren is a semisynthetic cephalosporin active against staphylococci, streptococci, and and most enterobacteria. It is resistant to staphylococcal and most enterobacterial beta-lactamases, and is usually taken as the prodrug cefditoren pivoxil.",
  },
  {
    accession: "ARO:3000650",
    name: "cefdinir",
    description:
      "Cefdinir is similar to cefixime with a modified side-chain at its 7-amino position. It also shares similar activity with cefixime but is more active against staphylococci. It has also be shown to enhance phagocytosis.",
  },
  {
    accession: "ARO:3000651",
    name: "ceftaroline",
    description:
      "Ceftaroline is a novel cephalosporin active against methicillin resistant Staphylococcus aureus. Like other cephalosporins it binds penicillin-binding proteins to inhibit cell wall synthesis. It strongly binds with PBP2a, associated with methicillin resistance. It is taken orally as the prodrug ceftaroline fosamil.",
  },
  {
    accession: "ARO:3000652",
    name: "isepamicin",
    description:
      "A semi-synthetic derivative of gentamicin B (hydroxyamino propionyl genamicin B). It is modified to combat microbial inactivation and has a slightly larger spectrum of activity compared to other aminoglycosides, including Ser marcescens, Enterobacteria, and K pneumoniae.",
  },
  {
    accession: "ARO:3000653",
    name: "G418",
    description:
      "A gentamicin class aminoglycoside antibiotic often used in mammalian cell culture work as a selectable marker for the neo cassette (APH3').",
  },
  {
    accession: "ARO:3000654",
    name: "arbekacin",
    description:
      "A synthetic derivative (1-N-(4-amino-2-hydroxybutyryl) of dibekacin used in Japan. It is active against methicillin-resistant Staph. aureus and shows synergy with ampicillin when treating gentamicin and vancomycin resistant enterocci.",
  },
  {
    accession: "ARO:3000655",
    name: "gentamicin B",
    description:
      "Gentamicin B is a semisynthetic aminoglycoside antibacterial.",
  },
  {
    accession: "ARO:3000656",
    name: "AcrS",
    description:
      "AcrS is a repressor of the AcrAB efflux complex and is associated with the expression of AcrEF. AcrS is believed to regulate a switch between AcrAB and AcrEF efflux.",
  },
  {
    accession: "ARO:3000657",
    name: "paromomycin",
    description:
      "An aminoglycoside antibiotic used for the treatment of parasitic infections. It is similar to neomycin sharing a similar spectrum of activity, but its hydroxyl group at the 6'-position instead of an amino group makes it resistant to AAC(6') modifying enzymes.",
  },
  {
    accession: "ARO:3000658",
    name: "lividomycin",
    description:
      "Lividomycins are aminoglycosidic antibiotics produced by Streptomyces lividus. They contain 2-amino-2,3-dideoxy-D-glucose.",
  },
  {
    accession: "ARO:3000659",
    name: "gatifloxacin",
    description:
      "Gatifloxacin is an 8-methoxy, 7-piperazinyl, 6-fluoroquinolone that can be taken orally or by intravenous administration. It is active against most Gram-positive and Gram-negative bacteria, but inactive against non-fermenting Gram-negative rods including Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3000660",
    name: "lomefloxacin",
    description:
      "Lomefloxacin is a difluoropiperazinyl quinolone, sharing similar activities with other fluoroquinolones. It is used to treat urinary tract infections. Relative to other fluoroquinolones, it has a longer half life and has higher serum concentrations.",
  },
  {
    accession: "ARO:3000661",
    name: "nalidixic acid",
    description:
      "Nalidixic acid is a quinolone derivative of naphthyridine active against many enterobacteria, but ineffective against Ps aeruginosa, Gram-positive bacteria, and anaerobes. Acquired resistance is common in nalidixic acid treatments.",
  },
  {
    accession: "ARO:3000662",
    name: "norfloxacin",
    description:
      "Norfloxacin is a 6-fluoro, 7-piperazinyl quinolone with a wide range of activity against Gram-negative bacteria. It is inactive against most anaerobes.",
  },
  {
    accession: "ARO:3000663",
    name: "ofloxacin",
    description:
      "Ofloxacin is a 6-fluoro, 7-piperazinyl quinolone with a methyl-substituted oxazine ring. It has a broad spectrum of activity including many enterobacteria and mycoplasma but most anaerobes are resistant.",
  },
  {
    accession: "ARO:3000664",
    name: "trovafloxacin",
    description:
      "Trovafloxacin is a trifluoroquinalone with a broad spectrum of activity that acts by inhibiting the uncoiling of supercoiled DNA. While potent against many Gram-positive and Gram-negative bacteria, it is less active against pseudomonads and Cl. difficile. It is usually taken as the prodrug trovafloxacin mesylate or alatrofloxacin mesylate for oral or intravenous administration, respectively.",
  },
  {
    accession: "ARO:3000665",
    name: "grepafloxacin",
    description:
      "Grepafloxacin is a broad-spectrum antibacterial quinoline. It is no longer taken due to its high toxicity.",
  },
  {
    accession: "ARO:3000666",
    name: "sparfloxacin",
    description:
      "Sparfloxacin is a dimethylpiperazinyl difluoroquinolone that acts by inhibiting DNA gyrase. It is active against aerobic Gram-positive and Gram-negative bacteria, as well as some mycobacteria. It has moderate activity against some anaerobes.",
  },
  {
    accession: "ARO:3000667",
    name: "demeclocycline",
    description:
      "Demeclocycline is a tetracycline analog with 7-chloro and 6-methyl groups. Due to its fast absorption and slow excretion, it maintains higher effective blood levels compared to other tetracyclines.",
  },
  {
    accession: "ARO:3000668",
    name: "oxytetracycline",
    description:
      "Oxytetracycline is a derivative of tetracycline with a 5-hydroxyl group. Its activity is similar to other tetracyclines.",
  },
  {
    accession: "ARO:3000669",
    name: "pristinamycin IIA",
    description: "Pristinamycin IIA is a streptogramin A antibiotic.",
  },
  {
    accession: "ARO:3000670",
    name: "pleuromutilin antibiotic",
    description:
      "Pleuromutilins are natural fungal products that target bacterial protein translation by binding the the 23S rRNA, blocking the ribosome P site at the 50S subunit. They are mostly used for agriculture and veterinary purposes.",
  },
  {
    accession: "ARO:3000671",
    name: "tiamulin",
    description:
      "Tiamulin is a pleuromutilin derivative currently used in veterinary medicine. It binds to the 23 rRNA of the 50S ribosomal subunit to inhibit protein translation.",
  },
  {
    accession: "ARO:3000672",
    name: "madumycin II",
    description: "Madumycin II is a streptogramin A antibiotic.",
  },
  {
    accession: "ARO:3000673",
    name: "griseoviridin",
    description: "Griseoviridin is a streptogramin A antibiotic.",
  },
  {
    accession: "ARO:3000674",
    name: "dalfopristin",
    description:
      "Dalfopristin is a water-soluble semi-synthetic derivative of pristinamycin IIA. It is produced by Streptomyces pristinaespiralis and is used in combination with quinupristin in a 7:3 ratio. Both work together to inhibit protein synthesis, and is active against Gram-positive bacteria.",
  },
  {
    accession: "ARO:3000675",
    name: "pristinamycin IB",
    description:
      "Pristinamycin IB is a class B streptogramin similar to pristinamycin IA, the former containing a N-methyl-4-(methylamino)phenylalanine instead of a N-methyl-4-(dimethylamino)phenylalanine in its class A streptogramin counterpart (one less methyl group).",
  },
  {
    accession: "ARO:3000676",
    name: "H-NS",
    description:
      "H-NS is a histone-like protein involved in global gene regulation in Gram-negative bacteria. It is a repressor of the membrane fusion protein genes acrE, mdtE, and emrK as well as nearby genes of many RND-type multidrug exporters.",
  },
  {
    accession: "ARO:3000677",
    name: "virginiamycin S2",
    description: "Virginiamycin S2 is a streptogramin B antibiotic.",
  },
  {
    accession: "ARO:3000678",
    name: "vernamycin B-gamma",
    description:
      "Vernamycin B-gamma is a class B streptogramin derived from virginiamycin S1.",
  },
  {
    accession: "ARO:3000679",
    name: "vernamycin C",
    description: "Vernamycin C is a streptogramin B antibiotic.",
  },
  {
    accession: "ARO:3000680",
    name: "patricin A",
    description: "Patricin A is a streptogramin B antibiotic.",
  },
  {
    accession: "ARO:3000681",
    name: "patricin B",
    description: "Patricin B is a streptogramin B antibiotic.",
  },
  {
    accession: "ARO:3000682",
    name: "ostreogrycin B3",
    description:
      "Ostreogrycin B3 is a derivative of pristinamycin IA, with an additional 3-hydroxy group on its 4-oxopipecolic acid.",
  },
  {
    accession: "ARO:3000683",
    name: "sulfacetamide",
    description:
      "Sulfacetamide is a very soluable sulfonamide antibiotic previously used to treat urinary tract infections. Its relatively low activity and toxicity to those with Stevens-Johnson syndrome have reduced its use and availability.",
  },
  {
    accession: "ARO:3000684",
    name: "mafenide",
    description: "Mafenide is a sulfonamide used topically for treating burns.",
  },
  {
    accession: "ARO:3000685",
    name: "arylomycin",
    description:
      "Arylomycins are lipopeptide antibiotics, sharing a lipohexapeptide with a C-terminal tripeptide macrolide and varying by the length of its alkyl chains.",
  },
  {
    accession: "ARO:3000686",
    name: "thiostrepton",
    description:
      "Thiostrepton is a cyclic peptide active against Gram-positive bacteria. It is produced by streptomyces bacteria.",
  },
  {
    accession: "ARO:3000687",
    name: "moenomycin",
    description:
      "Moemomycin antibiotics are phosphoglycolipids that contain a 3-phosphoglyceric acid, often used as a mixture. They inhibit petidoglycan glycosyltransferases to prevent bacterial cell wall biosynthesis.",
  },
  {
    accession: "ARO:3000688",
    name: "furazolidone",
    description:
      "Furazolidone is a nitrofuran antibiotic that is a non-ionic synthetic compound. It is taken orally and is active against many enteric pathogens. Furazolidone also has antiprotozoal activity and acts by inhibiting monoamine oxidase.",
  },
  {
    accession: "ARO:3000689",
    name: "metronidazole",
    description:
      "Metronidazole is a nitroimidazole that is active against anaerobic bacteria and protozoa. It is not effective against aerobic bacteria. Nitroimidazoles act by oxidizing DNA causing strand breaks and cell death.",
  },
  {
    accession: "ARO:3000690",
    name: "bleomycinic acid",
    description:
      "Bleomycinic acid is a glycopeptide antibiotic produced by Streptomyces verticillus taken as a mixture of bleomycins. It induces stand breaks in bacterial nucleic acids.",
  },
  {
    accession: "ARO:3000691",
    name: "bleomycin A2",
    description:
      "Bleomycin A2 is a glycopeptide antibiotic produced by Streptomyces verticillus taken as a mixture of bleomycins. It induces stand breaks in bacterial nucleic acids.",
  },
  {
    accession: "ARO:3000692",
    name: "bleomycin B2",
    description:
      "Bleomycin B2 is a glycopeptide antibiotic produced by Streptomyces verticillus taken as a mixture of bleomycins. It induces stand breaks in bacterial nucleic acids.",
  },
  {
    accession: "ARO:3000693",
    name: "defensin",
    description:
      "Defensins are natural cationic peptides that have antibiotic properties. It is part of the innate immune system of plants and animals.",
  },
  {
    accession: "ARO:3000694",
    name: "magainin",
    description:
      "Magainins are antimicrobial peptides produced by amphibians skil cells as part of their innate immune system. It is active against a broad range of gram-negative and gram-positive bacteria, fungi and protozoa.",
  },
  {
    accession: "ARO:3000695",
    name: "actinomycin",
    description:
      "Actinomycins are peptide antibiotics that interact with DNA. This prevents RNA synthesis, thus preventing the production of proteins vital to the cell.",
  },
  {
    accession: "ARO:3000696",
    name: "Dactinomycin",
    description:
      "Dactinomycin, also known as Actinomycin D, is a cyclic peptide antibiotic that interacts with DNA to inhibit RNA synthesis.",
  },
  {
    accession: "ARO:3000697",
    name: "moenomycin A1",
    description:
      "Moenomycin A1 is a major component of moenomycin mixtures. It is produced by Streptomyces ghanaensis, S. bambergiensis, S. ederensis, and S. geysiriensis.",
  },
  {
    accession: "ARO:3000698",
    name: "sulfasalazine",
    description:
      "Sulfasalazine is a derivative of the early sulfonamide sulfapyridine (salicylazosulfapyridine). It was developed to increase water solubility and is taken orally for ulcerative colitis.",
  },
  {
    accession: "ARO:3000699",
    name: "sulfamethizole",
    description:
      "Sulfamethizole is a short-acting sulfonamide that inhibits dihydropteroate synthetase.",
  },
  {
    accession: "ARO:3000700",
    name: "lividomycin A",
    description:
      "Lividomycin A is a pentasaccharide antibiotic which interferes with bacterial protein synthesis.",
  },
  {
    accession: "ARO:3000701",
    name: "lividomycin B",
    description:
      "Lividomycin B is a derivative of lividomycin A with a removed mannose group (demannosyllividomycin A). Livodomycins interfere with bacterial protein synthesis.",
  },
  {
    accession: "ARO:3000702",
    name: "acrR",
    description:
      "AcrR is a repressor of the AcrAB-TolC multidrug efflux complex. AcrR mutations result in high level antibiotic resistance.",
  },
  {
    accession: "ARO:3000704",
    name: "cefalotin",
    description:
      "Cefalotin is a semisynthetic cephalosporin antibiotic activate against staphylococci. It is resistant to staphylococci beta-lactamases but hydrolyzed by enterobacterial beta-lactamases.",
  },
  {
    accession: "ARO:3000705",
    name: "isopenicillin N",
    description:
      "Isopenicillin N is a natural penicillin derivative produced by Penicillium chrysogenum with activity similar to penicillin N.",
  },
  {
    accession: "ARO:3000706",
    name: "penicillin N",
    description:
      "Penicillin N is a penicillin derivative produced by Cephalosporium acremonium.",
  },
  {
    accession: "ARO:3000707",
    name: "antibiotic mixture",
    description:
      "Mixtures of antibiotics are commonly used clinically. These mixtures link to antibiotic molecules that are also listed in their respective families under antibiotic molecules. i.e. 'Gramicidin A' is part of the mixture 'gramicidin D', while 'Gramicidin A' is also a 'gramicidin', which is a 'peptide antibiotic'.",
  },
  {
    accession: "ARO:3000708",
    name: "antibiotic target",
    description:
      "Antibiotic targets are components of the bacterial cell usually unique to bacteria or significantly different from their eukaryotic counterparts to allow for selective antibacterial activity. Exceptions include antibiotics that are converted to an active form inside the bacterium. Resistance mechanisms may modify these targets to protect bacteria from the effects of antibiotics.",
  },
  {
    accession: "ARO:3000709",
    name: "cell wall synthesis enzyme targeted by antibiotic",
    description:
      "Cell wall biosynthesis provides bacteria with a layer of peptidoglycan that provides rigidity and protects the cell from lysis due to osmotic pressure. Cell wall structures vary significantly between Gram-positive and Gram-negative bacteria, while mycobacteria have a unique cell wall composition.",
  },
  {
    accession: "ARO:3000710",
    name: "cell membrane component targeted by antibiotic",
    description:
      "Interaction with the cell membrane phospholipids interferes with membrane permeability and integrity, leading to the leakage of cell contents. The outer membrane of Gram-negative bacteria is particularly sensitive.",
  },
  {
    accession: "ARO:3000711",
    name: "protein synthesis machinery targeted by antibiotic",
    description:
      "Protein synthesis in bacteria is accomplished by cytosolic ribosomes that are structurally different from eukaryotic ribosomes, and thus a target for many antibiotics. By inhibiting ribosomal activity, either by directly binding to the ribosome or interacting with co-factors, antibiotics prevent protein synthesis.",
  },
  {
    accession: "ARO:3000712",
    name: "nucleic acid synthesis machinery targeted by antibiotic",
    description:
      "Bacterial DNA and RNA synthesis includes many enzymes unique to bacteria that are targeted by antibiotics.",
  },
  {
    accession: "ARO:3000713",
    name: "nucleic acid targeted by antibiotic",
    description:
      "Nucleic acids can also be targeted by antibiotics. These drugs are usually converted to an active form by enzymes unique to bacteria to prevent toxicity to the host.",
  },
  {
    accession: "ARO:3000714",
    name: "folate synthesis enzyme targeted by antibiotic",
    description:
      "Enzymes in the folate synthesis pathway are targeted by two main classes of antibiotics: sulfonamides and diaminopyrimidines. Inhibition of these enzymes block the production of key precursors for nucleotides and amino acids.",
  },
  {
    accession: "ARO:3000715",
    name: "isoprenoid synthesis enzyme targeted by antibiotic",
    description:
      "Enzymes in this pathway generate important precursors for macromolecules. While eukaryotes and prokaryotes both synthesize isoprenoids, many bacteria and notably mycobacteria use an alternative pathway that is targeted by antibiotics.",
  },
  {
    accession: "ARO:3000716",
    name: "mycolic acid synthesis enzyme targeted by antibiotic",
    description:
      "Mycolic acids are produced by mycobacteria. These fatty acids are required for the construction of cell walls. Enzymes that are targeted by antibiotics include synthases, elongases, and dehydrogenases.",
  },
  {
    accession: "ARO:3000717",
    name: "antibiotic sensitive signal peptidase I",
    description:
      "Signal peptidase I is a bacterial signal peptidase that cleaves signal peptides from translated proteins. These signal peptides direct proteins to their destination, and then must be cleaved in order to function.",
  },
  {
    accession: "ARO:3000718",
    name: "marR",
    description:
      "MarR is a repressor of the mar operon marRAB, thus regulating the expression of marA, the activator of multidrug efflux pump AcrAB.",
  },
  {
    accession: "ARO:3000719",
    name: "bacitracin sensitive isoprenyl pyrophosphate",
    description:
      "Isoprepnyl pyrophosphate removes a phosphate group that is added from the peptidoglycan subunit to permit its transfer across the cell membrane. The phosphate group must be removed before its incorporation into peptidoglycan macromolecules.",
  },
  {
    accession: "ARO:3000720",
    name: "acyl-D-alanyl-D-alanine",
    description:
      "The acyl-D-alanyl-D-alanine group is found at the terminal end of the pentapeptide chains of peptidoglycan subunits. Antibiotics that bind to this group block the addition of subunits, inhibiting peptidoglycan synthesis. This group is most accessible in Gram-postive bacteria.",
  },
  {
    accession: "ARO:3000721",
    name: "cell wall component targeted by antibiotic",
    description:
      "The cell wall is unique to bacteria and provides rigidity and protects the cell from lysis due to osmotic pressure. Its characteristic component is peptidoglycan, though cell wall structures vary significantly between Gram-positive and Gram-negative bacteria, as well as mycobacteria.",
  },
  {
    accession: "ARO:3000722",
    name: "cycloserine sensitive alanine racemase",
    description:
      "Alanine racemase converts the natural L-alanine to the D-alanine required in the pentapeptide chains in peptidoglycan. It is found in both Gram-positive and Gram-negative bacteria.",
  },
  {
    accession: "ARO:3000723",
    name: "cycloserine sensitive D-alanine synthetase",
    description:
      "D-Alanine synthetase joins two D-alanines before their addition to the pentapeptide chains of peptidoglycan. It is found in both Gram-positive and Gram-negative bacteria.",
  },
  {
    accession: "ARO:3000724",
    name: "fosfomycin sensitive pyruvyl transferase (murA)",
    description:
      "Pyruvyl transferase is responsible for the conversion of N-acetylglucosamine (NAG) to N-acetylmuramic acid (NAMA). NAMA is a part of the macromolecular backbone of peptidoglycans. Pyruvyl transferase is found in both Gram-positive and Gram-negative bacteria.",
  },
  {
    accession: "ARO:3000725",
    name: "antibiotic sensitive arabinosyltransferase",
    description:
      "Arabinosyl transferases allow for the polymerization of arabinose to form arabinan. Arabanan is required for formation of mycobacterial cell walls.",
  },
  {
    accession: "ARO:3000726",
    name: "pentaglycyl bridge",
    description:
      "The pentaglycyl bridge is part of Gram-positive peptidoglycan, allowing the formation of a thick and rigid cell wall.",
  },
  {
    accession: "ARO:3000727",
    name: "50S ribosomal subunit",
    description:
      "The larger of two ribosomal subunits that make up the 70S bacterial ribosome.",
  },
  {
    accession: "ARO:3000728",
    name: "30S ribosomal subunit",
    description:
      "The smaller of two ribosomal subunits that make up the 70S bacterial ribosome.",
  },
  {
    accession: "ARO:3000729",
    name: "ribosomal complex",
    description:
      "The ribosomal complex can be targeted by antibiotics that stabilize the entire complex and prevent further translation. These antibiotics target both subunits in the process of protein translation.",
  },
  {
    accession: "ARO:3000730",
    name: "elongation factor G",
    description:
      "Elongation factor G (EF-G) provides energy for the translocation step in protein synthesis. Antibiotics that target EF-G allow one round of translocation but prevent EF-G from dissociating from the ribosome-RNA complex.",
  },
  {
    accession: "ARO:3000731",
    name: "elongation factor Tu",
    description:
      "Elongation factor Tu is required for peptide elongation in bacterial protein synthesis.",
  },
  {
    accession: "ARO:3000732",
    name: "antibiotic sensitive Isoleucyl-tRNA transferase",
    description:
      "Isoleucyl-tRNA transferase recognizes isoleucine codons and works with the ribosome to incorporate isoleucine amino acids into polypeptide chains. It can be targeted by antibiotics that have an analog of isoleucine through competitive inhibition.",
  },
  {
    accession: "ARO:3000733",
    name: "fluoroquinolone sensitive gyrA",
    description:
      "DNA gyrase is responsible for DNA supercoiling and consists of two alpha and two beta subunits. Binding the alpha-subunit is sufficient to inhibit DNA supercoiling.",
  },
  {
    accession: "ARO:3000734",
    name: "antibiotic sensitive DNA topoisomerase subunit gyrB",
    description:
      "DNA gyrase is responsible for DNA supercoiling and consists of two alpha and two beta subunits. Binding the beta-subunit is sufficient to inhibit DNA supercoiling.",
  },
  {
    accession: "ARO:3000735",
    name: "fluoroquinolone sensitive parC",
    description:
      "subunit of the DNA topoisomerase IV sensitive to fluoroquinolone",
  },
  {
    accession: "ARO:3000736",
    name: "rho transcription terminator factor",
    description:
      "The rho transcription terminator factor moves along RNA as it is transcribed, unwinding the RNA from the DNA. It is also responsible for recognizing transcription termination pause sites, slowing down RNA polymerase and allowing mRNA to be released.",
  },
  {
    accession: "ARO:3000737",
    name: "rifamycin sensitive beta-subunit of RNA polymerase (rpoB)",
    description:
      "The beta-subunit of RNA polymerase is targeted as a complex. When blocked, RNA strand elongation cannot continue after the first few nucleotides.",
  },
  {
    accession: "ARO:3000738",
    name: "deoxyribonucleic acid",
    description:
      "Antibiotics can target DNA to cause strand breaks or prevent transcription. The activity of these antibiotics can be context dependent: some antibiotics are converted to an active form only in certain species, while others target DNA as it is being transcribed.",
  },
  {
    accession: "ARO:3000739",
    name: "RNA guanosine binding site",
    description:
      "RNA intron self-splicing action can be inhibited by antibiotics that target a G-binding site to compete with the guanosine cofactor.",
  },
  {
    accession: "ARO:3000740",
    name: "ribonucleic acid",
    description:
      "Many types of ribonucleic acids (RNAs) are targeted by antibiotics, including mRNA templates and ribosomal RNA. Some antibiotics target all RNAs and cause strand breaks.",
  },
  {
    accession: "ARO:3000741",
    name: "transcription complex DNA",
    description:
      "DNA in the transcriptional complex is partially unwound, allowing antibiotics to target exposed bases and prevent DNA transcription.",
  },
  {
    accession: "ARO:3000742",
    name: "anaerobe DNA",
    description:
      "DNA in anaerobes are targeted by antibiotics that are converted to an active form in the anaerobe environment. For nitroimidazoles, the antibiotic is reduced to a highly oxidative intermediate that causes DNA oxidation and double-stranded breaks.",
  },
  {
    accession: "ARO:3000743",
    name: "1-deoxy-D-xylulose 5-phosphate reductoisomerase",
    description:
      "This reductoisomerase is found in an alternate isoprenoid synthase pathway that is mevalonate independent. Isoprenoids are important precursors to many other macromolecules. This non-mevalonate pathway is used by plants and most bacteria, including mycobacteria, but not eukaryotes, allowing for selectivea action. Staphylococcus aureus and some other Gram-positive bacteria use a mevalonate-dependent pathway, and are resistant against 1-deoxy-D-xylose 5-phosphate reductoisomerase inhibitors.",
  },
  {
    accession: "ARO:3000744",
    name: "antibiotic sensitive dihydropteroate synthase",
    description:
      "Dihydropteroate synthase catalyzes the conversion of p-aminobenzoic acid to dihydropteroic acid as part of the tetrahydrofolic acid biosynthetic pathway. Tetrahydrofolic acid is essential for folate synthesis, and a precursor of many nucleotides and amino acids. Inhibiting dihydropteroate leads to slow cell death, as bacteria can survive on existing folate storages for a few generations.",
  },
  {
    accession: "ARO:3000745",
    name: "antibiotic sensitive dihydrofolate reductase",
    description:
      "Dihydrofolate reductase synthesizes the co-enzyme tetrahydrofolic acid, which is also an essential precursor in the de novo synthesis of the DNA nucleotide thymidine.",
  },
  {
    accession: "ARO:3000746",
    name: "mepR",
    description:
      "MepR is an upstream repressor of MepA in Staphylococcus aureus. It is part of the mepRAB operon.",
  },
  {
    accession: "ARO:3000747",
    name: "mef",
    description:
      "Mef efflux pumps are MFS-type transporters that confer macrolide resistance in Streptococcus pneumoniae.",
  },
  {
    accession: "ARO:3000748",
    name: "subunit of efflux pump conferring antibiotic resistance",
    description:
      "Subunits of efflux proteins that pump antibiotic out of a cell to confer resistance.",
  },
  {
    accession: "ARO:3000749",
    name: "beta-lactam sensitive penicillin-binding protein",
    description:
      "Penicillin-binding proteins (PBPs) are responsible for transpeptidation, joining subunits to form peptidoglycan. They are numbered according to descending molecular weight, and may vary between species. These numbers represent the PBPs of E. coli. Inhibition of the larger PBPs responsible for transpeptidation (1-3) are lethal. Smaller PBPs (4-6) are involed in carboxypeptidation; inhibition of these is non-lethal.",
  },
  {
    accession: "ARO:3000750",
    name:
      "protein of two-component regulatory system modulating antibiotic efflux",
    description:
      "A protein, either a histidine kinase or a response regulator, that is part of a two-component regulatory system that directly or indirectly change rates of antibiotic efflux.",
  },
  {
    accession: "ARO:3000751",
    name: "determinant of resistance to peptide antibiotics",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to peptide antibiotics.",
  },
  {
    accession: "ARO:3000753",
    name: "abeM",
    description:
      "AbeM is an multidrug efflux pump found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3000754",
    name: "antibiotic sensitive peptidyl transferase (23S rRNA)",
    description:
      "The bacterial 23S rRNA, at the P site of the 50S subunit, has peptidyl transferase activity and is responsible for peptide synthesis and elongation.",
  },
  {
    accession: "ARO:3000755",
    name: "antibiotic sensitive 16S rRNA",
    description:
      "The 16S rRNA of the 30S ribosomal subunit helps to stabilize the aminoacyl-tRNAs for pairing with the mRNA codons. Some antibiotics (aminoglycosides) that target the 16S rRNA lead to mRNA misreading as well as premature termination. Other antibiotics (tetracyclines) that target the 16S rRNA prevent incoming charged-tRNAs from binding to the ribosome's A-site, stopping elongation.",
  },
  {
    accession: "ARO:3000756",
    name: "30S ribosomal subunit P-site",
    description:
      "The P-site of the ribosome binds the peptidyl-tRNA with the growing polypeptide chain. Antibiotics that target the P-site block fMet-tRNA from entering, thus inhibiting peptide elongation.",
  },
  {
    accession: "ARO:3000757",
    name: "50S ribosomal subunit P-site",
    description:
      "The P-site of the ribosome binds the peptidyl-tRNA with the growing polypeptide chain, and is a common target for many antibiotics that inhibit the 50S ribosomal subunit. Some antibiotics (oxazolidinones) block the 50S subunit from interacting with the 30S subunit to prevent the formation of the ribosomal complex. Other antibiotics (thiostrepton) block the 50S subunit from interacting with elongation factors G and Tu (EF-G, Ef-Tu), preventing GTP-dependent reactions of peptidyl transferase, and translocation, respectively. Other antibiotics (macrolides) prevent protein translocation.",
  },
  {
    accession: "ARO:3000758",
    name: "fosmidomycin",
    description:
      "Fosmidomycin is an inhibitor of 1-deoxy-D-xylulose 5-phosphate reductoisomerase, an enzyme in the non-mevalonate pathway for isoprenoid synthesis. It is active against enterobacteria, but not against Gram-positive bacteria or anaerobes.",
  },
  {
    accession: "ARO:3000759",
    name: "fusidic acid",
    description:
      "Fusidic acid is the only commercially available fusidane, a group of steroid-like antibiotics. It is most active against Gram-positive bacteria, and acts by inhibiting elongation factor G to  block protein synthesis.",
  },
  {
    accession: "ARO:3000760",
    name: "cycloserine",
    description:
      "Cycloserine is an anti-mycobacterial agent, and is active against enterobacteria, streptococci, M. tuberculosis, Staphylococcus aureus, and many Gram-negative and Gram-positive bacteria. It inhibits cell wall biosynthesis.",
  },
  {
    accession: "ARO:3000761",
    name: "mecillinam",
    description:
      "Mecillinam is a broad-spectrum beta-lactam antibiotic that was semi-synthetically derived to have a different drug centre, being a 6-alpha-amidinopenicillanate instead of a 6-alpha-acylaminopenicillanate. Contrasting most beta-lactam drugs, mecillinam is most active against Gram-negative bacteria. It binds specifically to penicillin binding protein 2 (PBP2).",
  },
  {
    accession: "ARO:3000762",
    name: "pefloxacin",
    description:
      "Pefloxacin is structurally and functionally similar to norfloxacin. It is poorly active against mycobacteria, while anaerobes are resistant.",
  },
  {
    accession: "ARO:3000763",
    name: "antibiotic sensitive embC",
    description:
      "EmbC is an arabinosyltransferase involved in lipoarabinomannan synthesis.",
  },
  {
    accession: "ARO:3000764",
    name: "antibiotic sensitive embA",
    description:
      "EmbA is an arabinosyltransferase that works as a heterodimer with EmbB. It is involved in the synthesis of arabinogalactan, a component of the mycobacterial cell wall.",
  },
  {
    accession: "ARO:3000765",
    name: "antibiotic sensitive embB",
    description:
      "EmbB is an arabinosyltransferase that works as a heterodimer with EmbA. It is involved in the synthesis of arabinogalactan, a component of the mycobacterial cell wall.",
  },
  {
    accession: "ARO:3000766",
    name: "antibiotic sensitive topoisomerase subunit",
    description:
      "Topoisomerases are responsible for winding and relaxing supercoiled DNA. The subunits of gyrase (gyrA, gyrB) and topoisomerase IV (parC, parE, parY) are targeted by antibiotics.",
  },
  {
    accession: "ARO:3000767",
    name: "avoparcin",
    description:
      "Avoparcin is a glycopeptide antibiotic that is an analog of vancomycin. It was formally used to treat animals, but has since been banned because of its selection for vancomycin-resistant enterococci.",
  },
  {
    accession: "ARO:3000768",
    name: "abeS",
    description:
      "AbeS in an efflux pump of the SMR family of transporters found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3000769",
    name: "PatA-PatB",
    description:
      "PatA-PatB is an efflux protein complex that is associated with fluoroquinolone resistance in Streptococcus pneumoniae. Experiments indicate that PatA and PatB form a heterodimeric transporter.",
  },
  {
    accession: "ARO:3000770",
    name: "AdeABC",
    description:
      "AdeABC is an RND multidrug efflux system in Acinetobacter species, notably A. baunmannii. AdeA is a membrane fusion protein, AdeB is the inner membrane transporter, and AdeC is the outer membrane factor. It confers resistance to tigecycline.",
  },
  {
    accession: "ARO:3000771",
    name: "AdeFGH",
    description:
      "AdeFGH is a RND multidrug efflux pump expressed in Acinetobacter baumannii. It confers resistance to fluoroquinolone, tetracycline, tigecycline, chloramphenicol, clindamycin, trimethoprim, and sulfamethoxazole.",
  },
  {
    accession: "ARO:3000772",
    name: "AdeIJK",
    description:
      "AdeIJK is a RND multidrug efflux pump expressed in Acinetobacter baumannii. It contributes to resistance for beta-lactams, chloramphenicol, tetracycline, erythromycin, lincosamides, fluoroquinolone, fusidic acid, novobiocin, rifampicin, trimethoprim, acridine, pyronine, and safranin.",
  },
  {
    accession: "ARO:3000773",
    name: "CmeABC",
    description:
      "CmeABC is a multidrug efflux pump in Campylobacter jejuni. Its overexpression led to enhanced resistance to ciprofloxacin, norfloxacin, cefotaxime, fusidic acid, and erythromycin. CmeA acts as the periplasmic fusion protein, CmeB is the inner membrane transporter, and CmeC is an outer membrane protein.",
  },
  {
    accession: "ARO:3000774",
    name: "adeA",
    description:
      "AdeA is the membrane fusion protein of the multidrug efflux complex AdeABC.",
  },
  {
    accession: "ARO:3000775",
    name: "adeB",
    description:
      "AdeB is the multidrug transporter of the adeABC efflux system.",
  },
  {
    accession: "ARO:3000777",
    name: "adeF",
    description:
      "AdeF is the membrane fusion protein of the multidrug efflux complex AdeFGH.",
  },
  {
    accession: "ARO:3000778",
    name: "adeG",
    description:
      "AdeG is the inner membrane transporter of the AdeFGH multidrug efflux complex.",
  },
  {
    accession: "ARO:3000779",
    name: "adeH",
    description:
      "AdeH is the outer membrane channel protein of the AdeFGH multidrug efflux complex.",
  },
  {
    accession: "ARO:3000780",
    name: "adeI",
    description:
      "AdeI is the membrane fusion protein of the AdeIJK multidrug efflux complex.",
  },
  {
    accession: "ARO:3000781",
    name: "adeJ",
    description:
      "AdeJ is a RND efflux protein that acts as the inner membrane transporter of the AdeIJK efflux complex. It has 57% identity with E. coli AcrB.",
  },
  {
    accession: "ARO:3000782",
    name: "adeK",
    description:
      "AdeK is the outer membrane factor protein in the adeIJK multidrug efflux complex.",
  },
  {
    accession: "ARO:3000783",
    name: "cmeA",
    description:
      "CmeA is the membrane fusion protein of the CmeABC multidrug efflux complex.",
  },
  {
    accession: "ARO:3000784",
    name: "cmeB",
    description:
      "CmeB is the inner membrane transporter the CmeABC multidrug efflux complex.",
  },
  {
    accession: "ARO:3000785",
    name: "cmeC",
    description:
      "CmeC is the outer membrane channel protein of the CmeABC multidrug efflux complex.",
  },
  {
    accession: "ARO:3000786",
    name: "MdsABC",
    description:
      "MdsABC, or GesABC, is a RND-type efflux complex found in Salmonella. In addition to its metal-efflux system, GesABC is also capable of exporting beta-lactams, chloramphenicol, and thiamphenicol. GesB is the inner membrane transporter and is similar to MexF; GesA is the membrane-fusion protein, and GesC is the outer membrane factor.",
  },
  {
    accession: "ARO:3000787",
    name: "MdtABC-TolC",
    description:
      "MdtABC-TolC is a multidrug efflux system in Gram-negative bacteria, including E. coli and Salmonella. MdtA is a membrane fusion protein; TolC is the outer membrane channel; MdtBC form a drug transporter. In the absence of MdtB, the MdtAC-TolC has narrower drug specificity, leading to the loss of novobiocin resistance, for example.",
  },
  {
    accession: "ARO:3000788",
    name: "MdtEF-TolC",
    description:
      "MdtEF-TolC is a multidrug efflux complex in Gram-negative bacteria, including E. coli. MdtE is the membrane fusion protein, MdtF is the inner membrane transporter, while TolC is the outer membrane channel.",
  },
  {
    accession: "ARO:3000789",
    name: "mdsA",
    description:
      "MdsA is the membrane fusion protein of the multidrug and metal efflux complex MdsABC.",
  },
  {
    accession: "ARO:3000790",
    name: "mdsB",
    description:
      "MdsB is the inner membrane transporter of the multidrug and metal efflux complex MdsABC. mdsB corresponds to 1 locus in Pseudomonas aeruginosa PAO1 (gene name: mexQ) and 2 loci in Pseudomonas aeruginosa LESB58.",
  },
  {
    accession: "ARO:3000791",
    name: "mdsC",
    description:
      "MdsC is the outer membrane channel of the multidrug and metal efflux complex MdsABC.",
  },
  {
    accession: "ARO:3000792",
    name: "mdtA",
    description:
      "MdtA is the membrane fusion protein of the multidrug efflux complex mdtABC.",
  },
  {
    accession: "ARO:3000793",
    name: "mdtB",
    description:
      "MdtB is a transporter that forms a heteromultimer complex with MdtC to form a multidrug transporter. MdtBC is part of the MdtABC-TolC efflux complex.",
  },
  {
    accession: "ARO:3000794",
    name: "mdtC",
    description:
      "MdtC is a transporter that forms a heteromultimer complex with MdtB to form a multidrug transporter. MdtBC is part of the MdtABC-TolC efflux complex. In the absence of MdtB, MdtC can form a homomultimer complex that results in a functioning efflux complex with a narrower drug specificity. mdtC corresponds to 3 loci in Pseudomonas aeruginosa PAO1 (gene name: muxC/muxB) and 3 loci in Pseudomonas aeruginosa LESB58.",
  },
  {
    accession: "ARO:3000795",
    name: "mdtE",
    description:
      "MdtE is the membrane fusion protein of the MdtEF multidrug efflux complex. It shares 70% sequence similarity with AcrA.",
  },
  {
    accession: "ARO:3000796",
    name: "mdtF",
    description:
      "MdtF is the multidrug inner membrane transporter for the MdtEF-TolC efflux complex.",
  },
  {
    accession: "ARO:3000797",
    name: "MexCD-OprJ",
    description:
      "MexCD-OprJ is a multidrug efflux protein expressed in the Gram-negative Pseudomonas aeruginosa. MexC is the membrane fusion protein; MexD is the inner membrane transporter; and OprJ is the outer membrane channel. MexAB-OprM is associated with resistance to fluoroquinolones, chloramphenicol, and macrolides. MexCD-OprJ is typically quiescent in wild-type cells, with expression following mutation of the nfxB gene that is divergently transcribed from the mexCD-oprJ operon and encodes a repressor of mexCD-oprJ expression. Expression of mexCD\u2013oprJ is controlled by a single known regulator, the NfxB repressor encoded by the nfxB gene, which is transcribed divergently from the efflux genes.",
  },
  {
    accession: "ARO:3000798",
    name: "MexEF-OprN",
    description:
      "MexEF-OprN is a multidrug efflux protein expressed in the Gram-negative Pseudomonas aeruginosa. MexE is the membrane fusion protein; MexF is the inner membrane transporter; and OprN is the outer membrane channel. MexEF-OprN is associated with resistance to fluoroquinolones, chloramphenicol, and trimethoprim. Regulation of mexEF-oprN operon expression is multifaceted with the MexT activator being one of the most prominent regulatory proteins. MexEF-OprN is the only positively regulated RND pump of P. aeruginosa.",
  },
  {
    accession: "ARO:3000799",
    name: "MexGHI-OpmD",
    description:
      "MexGHI-OpmD is an efflux complex expressed in Pseudomonas aeruginosa. MexG is a membrane protein required for drug export; MexH is the membrane fusion protein; MexI is the inner membrane transporter; and MexJ is the outer membrane channel protein. MexGHI-OpmD confers resistance to vanadium, norfloxacin, and acriflavin.",
  },
  {
    accession: "ARO:3000800",
    name: "MexC",
    description:
      "MexC is the membrane fusion protein of the MexCD-OprJ multidrug efflux complex.",
  },
  {
    accession: "ARO:3000801",
    name: "MexD",
    description:
      "MexD is the multidrug inner membrane transporter of the MexCD-OprJ complex.",
  },
  {
    accession: "ARO:3000802",
    name: "OprJ",
    description:
      "OprJ is the outer membrane channel component of the MexCD-OprJ multidrug efflux complex.",
  },
  {
    accession: "ARO:3000803",
    name: "MexE",
    description:
      "MexE is the membrane fusion protein of the MexEF-OprN multidrug efflux complex.",
  },
  {
    accession: "ARO:3000804",
    name: "MexF",
    description:
      "MexF is the multidrug inner membrane transporter of the MexEF-OprN complex. mexF corresponds to 2 loci in Pseudomonas aeruginosa PAO1 (gene name: mexF/mexB) and 4 loci in Pseudomonas aeruginosa LESB58 (gene name: mexD/mexB).",
  },
  {
    accession: "ARO:3000805",
    name: "OprN",
    description:
      "OprN is the outer membrane channel component of the MexEF-OprN multidrug efflux complex.",
  },
  {
    accession: "ARO:3000806",
    name: "MexG",
    description:
      "MexG is a membrane protein required for MexGHI-OpmD efflux activity.",
  },
  {
    accession: "ARO:3000807",
    name: "MexH",
    description:
      "MexH is the membrane fusion protein of the efflux complex MexGHI-OpmD.",
  },
  {
    accession: "ARO:3000808",
    name: "MexI",
    description:
      "MexI is the inner membrane transporter of the efflux complex MexGHI-OpmD.",
  },
  {
    accession: "ARO:3000809",
    name: "OpmD",
    description:
      "OpmD is the outer membrane channel protein of the efflux complex MexGHI-OpmD.",
  },
  {
    accession: "ARO:3000810",
    name: "mtrC",
    description:
      "MtrC is the membrane fusion protein of the MtrCDE multidrug efflux complex.",
  },
  {
    accession: "ARO:3000811",
    name: "mtrD",
    description:
      "MtrD is the inner membrane multidrug transporter of the MtrCDE efflux complex.",
  },
  {
    accession: "ARO:3000812",
    name: "mtrE",
    description:
      "MtrE is an outer membrane exporter protein that is part of the MtrCDE multidrug efflux complex.",
  },
  {
    accession: "ARO:3000813",
    name: "MexS",
    description:
      "MexS is a suppressor of MexT, which is an activator of the multidrug pump MexEF-OprN. Mutations in MexS lead to multidrug resistance.",
  },
  {
    accession: "ARO:3000814",
    name: "MexT",
    description:
      "MexT is a LysR-type transcriptional activator that positively regulates the expression of MexEF-OprN, OprD, and MexS.",
  },
  {
    accession: "ARO:3000815",
    name: "mgrA",
    description:
      "MgrA, also known as NorR, is a regulator for norA, norB, and tet38. It is a positive regulator for norA expression, but is a direct repressor for norB and an indirect repressor of tet38.",
  },
  {
    accession: "ARO:3000816",
    name: "mtrA",
    description:
      "MtrA is a transcriptional activator of the MtrCDE multidrug efflux pump of Neisseria gonorrhoeae.",
  },
  {
    accession: "ARO:3000817",
    name: "mtrR",
    description:
      "MtrR is a repressor of mtrCDE expression. Mutations in mtrR increase multidrug resistance.",
  },
  {
    accession: "ARO:3000818",
    name: "nalC",
    description:
      "NalC is a repressor of PA3720-PA3719, which are positive regulators of MexAB-OprM. Thus, nalC mutants confer multidrug resistance.",
  },
  {
    accession: "ARO:3000819",
    name: "nalD",
    description:
      "NalD is a repressor of MexAB-OprM. Mutations lead to multidrug resistance and MexAB-OprM overexpression.",
  },
  {
    accession: "ARO:3000820",
    name: "NfxB",
    description:
      "NfxB is a repressor of the efflux pump mexCD-oprJ and itself (NfxB binds upstream of the nfxB gene and negatively regulates its own expression). Increased expression of MexCD\u2013OprJ brought about by mutations in NfxB.",
  },
  {
    accession: "ARO:3000821",
    name: "phoPQ",
    description:
      "PhoPQ is a two-component regulatory system. phoP is phosphorylated by phoQ at a low [mg2+], activating the repressor. Its part of a two component regulating system that activates the PmrA/B system which has downstream effects, leading to Antimicrobial resistance in many pathogens. In Salmonella, phoP bind to macAB ABC transporter and represses it. In Escherichia coli, phoPQ regulates arnA expression. In Pseudomonas aeruginosa, phoPQ regulates OprH expression. phoPQ have other roles in other species.",
  },
  {
    accession: "ARO:3000822",
    name: "pmrA",
    description:
      "PmrA is a MFS-type efflux pump expressed in Streptococcus pneumoniae that confers low-level resistance to norfloxacin, ciprofloxacin, and acriflavine.",
  },
  {
    accession: "ARO:3000823",
    name: "ramA",
    description:
      "RamA (resistance antibiotic multiple) is a positive regulator of AcrAB-TolC and leads to high level multidrug resistance in Klebsiella pneumoniae, Salmonella enterica, and Enterobacter aerugenes, increasing the expression of both the mar operon as well as AcrAB. RamA also decreases OmpF expression.",
  },
  {
    accession: "ARO:3000824",
    name: "ramR",
    description:
      "RamR is a repressor that regulates RamA expression. Mutations lead to the upregulation of AcrAB, which is positively regulated by RamA.",
  },
  {
    accession: "ARO:3000825",
    name: "rob",
    description:
      "robA is a positive regulator for the acrAB efflux genes, and is structurally similar to SoxS and MarA.",
  },
  {
    accession: "ARO:3000826",
    name: "sdiA",
    description:
      "SdiA is a cell division regulator that is also a positive regulator of AcrAB only when it's expressed from a plasmid. When the sdiA gene is on the chromosome, it has no effect on expression of acrAB",
  },
  {
    accession: "ARO:3000827",
    name: "soxRS",
    description:
      "SoxRS is a two-component regulator part of the superoxide stress response. It is a positive regulator for many multidrug efflux pumps.",
  },
  {
    accession: "ARO:3000828",
    name: "baeR",
    description:
      "BaeR is a response regulator that promotes the expression of MdtABC and AcrD efflux complexes.",
  },
  {
    accession: "ARO:3000829",
    name: "baeS",
    description:
      "BaeS is a sensor kinase in the BaeSR regulatory system. While it phosphorylates BaeR to increase its activity, BaeS is not necessary for overexpressed BaeR to confer resistance.",
  },
  {
    accession: "ARO:3000830",
    name: "cpxA",
    description:
      "CpxA is a membrane-localized sensor kinase that is activated by envelope stress. It starts a kinase cascade that activates CpxR, which promotes efflux complex expression.",
  },
  {
    accession: "ARO:3000831",
    name: "CpxR",
    description:
      "CpxR is a regulator that promotes acrD expression when phosphorylated by a cascade involving CpxA, a sensor kinase. CpxR is also directly involved in activation of expression of RND efflux pump MexAB-OprM in P. aeruginosa. CpxR is required to enhance mexAB-oprM expression and drug resistance, in the absence of repressor MexR.",
  },
  {
    accession: "ARO:3000832",
    name: "evgA",
    description:
      "EvgA, when phosphorylated, is a positive regulator for efflux protein complexes emrKY and mdtEF. While usually phosphorylated in a EvgS dependent manner, it can be phosphorylated in the absence of EvgS when overexpressed.",
  },
  {
    accession: "ARO:3000833",
    name: "evgS",
    description:
      "EvgS is a sensor protein that phosphorylates the regulatory protein EvgA. evgS corresponds to 1 locus in Pseudomonas aeruginosa PAO1 and 1 locus in Pseudomonas aeruginosa LESB58.",
  },
  {
    accession: "ARO:3000834",
    name: "phoP",
    description: "PhoP is a direct repressor of the macAB efflux genes.",
  },
  {
    accession: "ARO:3000835",
    name: "phoQ",
    description:
      "PhoQ is a sensor protein that is induced by low Mg2+ concentrations. When activated, PhoQ donates a phosphate to activate PhoP.",
  },
  {
    accession: "ARO:3000836",
    name: "soxR",
    description:
      "SoxR is a sensory protein that upregulates soxS expression in the presence of redox-cycling drugs. This stress response leads to the expression many multidrug efflux pumps. In Pseudomonas aeruginosa, which lacks SoxS, SoxR is able to directly upregulate the expression of the MexGHI-OpmD efflux pump.",
  },
  {
    accession: "ARO:3000837",
    name: "soxS",
    description:
      "SoxS is a global regulator that up-regulates the expression of AcrAB efflux genes. It also reduces OmpF expression to decrease cell membrane permeability.",
  },
  {
    accession: "ARO:3000838",
    name: "arlR",
    description:
      "ArlR is a response regulator that binds to the norA promoter to activate expression. ArlR must first be phosphorylated by ArlS.",
  },
  {
    accession: "ARO:3000839",
    name: "arlS",
    description:
      "ArlS is a protein histidine kinase that phosphorylates ArlR, a promoter for norA expression.",
  },
  {
    accession: "ARO:3000840",
    name: "JOHN-1",
    description:
      "JOHN-1 is an Ambler class B carbapenem-hydrolysing beta-lactamase from Flavobacterium johnsoniae.",
  },
  {
    accession: "ARO:3000841",
    name: "CGB-1",
    description:
      "CGB-1 is an Ambler class B beta-lactamase that mediates resistance for carbapenems in Chryseobacterium gleum",
  },
  {
    accession: "ARO:3000842",
    name: "EBR-1 beta-lactamase",
    description:
      "EBR-1 is an Ambler class B beta-lactamase found in Empedobacter brevis and is known to mediate the hydrolysis of penicillins, cephalosporins, and carbapenems efficiently but not aztreonam.",
  },
  {
    accession: "ARO:3000843",
    name: "MUS-1",
    description:
      "MUS-1 is a chromosome-encoded beta-lactamase from Myroides odoratus and Myroides odoratimimus.",
  },
  {
    accession: "ARO:3000844",
    name: "TUS-1",
    description:
      "TUS-1 is a chromosome-encoded beta-lactamase from Myroides odoratus and Myroides odoratimimus",
  },
  {
    accession: "ARO:3000845",
    name: "GIM-1",
    description:
      "GIM-1 is an integron-encoded B1 beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3000846",
    name: "SIM-1",
    description:
      "SIM-1 is an integron-encoded Ambler class B beta-lactamase isolated from Acinetobacter baumannii",
  },
  {
    accession: "ARO:3000847",
    name: "KHM-1",
    description:
      "KHM-1 is a plasmid-mediated metallo-beta-lactamase found in Citrobacter freundii that confers resistance to all broad-spectrum beta-lactams, execpt for monobactams.",
  },
  {
    accession: "ARO:3000848",
    name: "DIM-1",
    description:
      "Dutch imipenemase or DIM-1 is an integron-encoded metallo-beta-lactamase from Pseudomonas stutzeri",
  },
  {
    accession: "ARO:3000849",
    name: "SFH-1",
    description:
      "SFH-1 confers resistance to carbapenems in Serratia fonticola.",
  },
  {
    accession: "ARO:3000850",
    name: "GOB-1",
    description:
      "GOB-1 confers resistance to cephalosporins in Elizabethkingia (Chryseobacterium) meningoseptica.",
  },
  {
    accession: "ARO:3000851",
    name: "THIN-B",
    description:
      "THIN-B, isolated from Janthinobacterium lividum, hydrolyzes a broad spectrum of beta-lactams including penicillins, cephalosporins, and carbapenems.",
  },
  {
    accession: "ARO:3000853",
    name: "AIM-1",
    description:
      "AIM-1 is an Ambler class B beta-lactamase that hydrolyzes most beta-lactams except aztreonam and ceftazidime. It was isolated from Pseudomonas aeruginosa and was the first subclass B3 mobile-elements encoded beta-lactamase discovered.",
  },
  {
    accession: "ARO:3000854",
    name: "SMB-1",
    description:
      "SMB-1 can hydrolyze a variety of beta-lactams, including penicillins, cephalosporins, and carbapenems. It was identified in Serratia marcescens.",
  },
  {
    accession: "ARO:3000855",
    name: "CAU-1",
    description:
      "CAU-1 is a B3 metallo-beta-lactamase that is encoded by the Caulobacter crescentus chromosome",
  },
  {
    accession: "ARO:3000856",
    name: "BJP-1",
    description:
      "BJP-1 is a subclass B3 ortholog found in Bradyrhizobium japonicum that hydrolyzes most beta-lactams except aztreonam, ticarcillin, and temocillin.",
  },
  {
    accession: "ARO:3000857",
    name: "16S ribosomal RNA methyltransferase",
    description:
      "Methyltransferases that modify the 16S rRNA of the 30S subunit of bacterial ribosomes, conferring resistance to drugs that target 16S rRNA.",
  },
  {
    accession: "ARO:3000858",
    name: "armA",
    description:
      "ArmA is a 16S rRNA methyltransferase that targets mature or nearly mature 30S subunits. It transfers a methyl group from S-adenosyl-L-methionine to N7-G1405 of the 16S rRNA, an aminoglycoside binding site.",
  },
  {
    accession: "ARO:3000859",
    name: "rmtA",
    description:
      "RmtA is a 16S rRNA methyltransferase found in Pseudomonas aeruginosa which methylates G1405 of the 16S rRNA. It confers high level resistance to many aminoglycosides.",
  },
  {
    accession: "ARO:3000860",
    name: "rmtB",
    description:
      "RmtB is a 16S rRNA methyltransferase that targets mature or nearly mature 30S subunits. It transfers a methyl group from S-adenosyl-L-methionine to N7-G1405 of the 16S rRNA, an aminoglycoside binding site.",
  },
  {
    accession: "ARO:3000861",
    name: "rmtC",
    description:
      "RmtC is a rRNA methyltransferase found in Proteus mirabilis with high level resistance to similar to aminoglycosides, with the exception of non-4,6-disubstituted deoxystreptamines (streptomycin and neomycin). It has also been isolated in Salmonella enterica ser. Virchow. It is hypothesized to methylate G1405, like related methyltransferases RmtA, RmtB, and ArmA.",
  },
  {
    accession: "ARO:3000862",
    name: "sgm",
    description:
      "Sgm, or sisomicin-gentamicin methyltransferase, methylates G1405 of 16S rRNA to confer resistance to various aminoglycosides.",
  },
  {
    accession: "ARO:3000863",
    name: "defensin resistant mprF",
    description:
      "MprF is a integral membrane protein that modifies the negatively-charged phosphatidylglycerol on the membrane surface of both Gram-positive and Gram-negative bacteria. This confers resistance to cationic peptides that disrupt the cell membrane, including defensins.",
  },
  {
    accession: "ARO:3000864",
    name: "fluoroquinolone resistant gyrB",
    description:
      "Point mutations in DNA gyrase subunit B (gyrB) observed in Mycobacterium tuberculosis can result in resistance to fluoroquinolones.",
  },
  {
    accession: "ARO:3000865",
    name: "oleD",
    description:
      "OleD is a glycotransferase found in Streptomyces antibioticus, a natural producer  of oleandomycin. OleD can glycosylate a wide range of macrolides. Unlike oleI, oleD is not found in the oleandomycin biosynthetic cluster.",
  },
  {
    accession: "ARO:3000866",
    name: "oleI",
    description:
      "OleI is a glycosyltransferase found in Streptomyces antibioticus, specifically the oleandomycin biosynthetic cluster. OleI glycosylates oleandomycin to confer self-resistance.",
  },
  {
    accession: "ARO:3000867",
    name: "oleandomycin",
    description:
      "Oleandomycin is a 14-membered macrolide produced by Streptomyces antibioticus. It is ssimilar to erythromycin, and contains a desosamine amino sugar and an oleandrose sugar. It targets the 50S ribosomal subunit to prevent protein synthesis.",
  },
  {
    accession: "ARO:3000869",
    name: "streptothricin acetyltransferase (SAT)",
    description:
      "AcetylCoA dependent acetyltransferase that acetylate streptothricins such as nourseothricin at position 16 (beta position of beta-lysine).",
  },
  {
    accession: "ARO:3000870",
    name: "triclosan",
    description:
      "Triclosan is a common antibacterial agent added to many consumer products as a biocide.  It is an inhibitor of fatty acid biosynthesis by blocking enoyl-carrier protein reductase (FabI).",
  },
  {
    accession: "ARO:3000871",
    name: "fatty acid synthesis enzyme targeted by antibiotic",
    description:
      "Antibiotics can target bacteria-specific fatty acid biosynthesis enzymes.",
  },
  {
    accession: "ARO:3000872",
    name: "antibiotic sensitive enoyl-acyl carrier reductase",
    description:
      "Enoyl-acyl carrier reductases catalyze the final enoyl reduction step in bacterial type II fatty acid elongation. Triclosan antibiotics block enoyl reduction and therefore fatty acid synthesis, functioning as a broad spectrum biocide and antibiotic.",
  },
  {
    accession: "ARO:3000873",
    name: "TEM-1",
    description:
      "TEM-1 is a broad-spectrum beta-lactamase found in many Gram-negative bacteria. Confers resistance to penicillins and first generation cephalosphorins.",
  },
  {
    accession: "ARO:3000874",
    name: "TEM-2",
    description: "TEM-2 is a broad-spectrum beta-lactamase.",
  },
  {
    accession: "ARO:3000875",
    name: "TEM-3",
    description:
      "TEM-3 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000876",
    name: "TEM-4",
    description:
      "TEM-4 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000877",
    name: "TEM-5",
    description:
      "TEM-5 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000878",
    name: "TEM-6",
    description:
      "TEM-6 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000879",
    name: "TEM-7",
    description:
      "TEM-7 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000880",
    name: "TEM-8",
    description:
      "TEM-8 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000881",
    name: "TEM-9",
    description:
      "TEM-9 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000882",
    name: "TEM-10",
    description:
      "TEM-10 is an extended-spectrum beta-lactamase round in Morganella morganii.",
  },
  {
    accession: "ARO:3000883",
    name: "TEM-11",
    description:
      "TEM-11 is an extended-spectrum beta-lactamase found in Proteus mirabilis.",
  },
  {
    accession: "ARO:3000884",
    name: "TEM-12",
    description:
      "TEM-12 is an extended-spectrum beta-lactamase found in Klebsiella oxytoca.",
  },
  {
    accession: "ARO:3000885",
    name: "TEM-13",
    description:
      "TEM-13 is a broad-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000886",
    name: "TEM-15",
    description:
      "TEM-15 is an extended-spectrum beta-lactamase that has been found in Haemophilus parainfluenzae.",
  },
  {
    accession: "ARO:3000887",
    name: "TEM-16",
    description:
      "TEM-16 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000888",
    name: "TEM-17",
    description:
      "TEM-17 is an extended-spectrum beta-lactamase found in Capnocytophaga ochracea.",
  },
  {
    accession: "ARO:3000889",
    name: "TEM-18",
    description:
      "TEM-18 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000890",
    name: "TEM-19",
    description:
      "TEM-19 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000891",
    name: "TEM-20",
    description:
      "TEM-20 is an extended-spectrum beta-lactamase found in several species of Gram-negative bacteria.",
  },
  {
    accession: "ARO:3000892",
    name: "TEM-21",
    description:
      "TEM-21 is an extended-spectrum beta-lactamase found in many species of Gram-negative bacteria.",
  },
  {
    accession: "ARO:3000893",
    name: "TEM-22",
    description:
      "TEM-22 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000894",
    name: "TEM-24",
    description:
      "TEM-24 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000895",
    name: "TEM-25",
    description:
      "TEM-25 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000896",
    name: "TEM-26",
    description:
      "TEM-26 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000897",
    name: "TEM-27",
    description:
      "TEM-27 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000898",
    name: "TEM-28",
    description:
      "TEM-28 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000899",
    name: "TEM-29",
    description:
      "TEM-29 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000900",
    name: "TEM-30",
    description:
      "TEM-30 is an inhibitor-resistant beta-lactamase found in E. coli. Confers resistance to amoxycilllin-clavulanic acid, ticarcillin-clavulanic acid, kanamycin, neomycin, and intermediate resistance to mezlocillin and piperacillin.",
  },
  {
    accession: "ARO:3000901",
    name: "TEM-31",
    description:
      "TEM-31 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000902",
    name: "TEM-32",
    description:
      "TEM-32 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000903",
    name: "TEM-33",
    description:
      "TEM-33 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000904",
    name: "TEM-34",
    description:
      "TEM-34 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000905",
    name: "TEM-35",
    description:
      "TEM-35 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000906",
    name: "TEM-36",
    description:
      "TEM-36 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000907",
    name: "TEM-37",
    description:
      "TEM-37 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000908",
    name: "TEM-38",
    description:
      "TEM-38 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000909",
    name: "TEM-39",
    description:
      "TEM-39 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000910",
    name: "TEM-40",
    description:
      "TEM-40 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000911",
    name: "TEM-42",
    description:
      "TEM-42 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000912",
    name: "TEM-43",
    description:
      "TEM-43 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000913",
    name: "TEM-44",
    description:
      "TEM-44 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000914",
    name: "TEM-45",
    description:
      "TEM-45 is an inhibitor-resistant beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000915",
    name: "TEM-46",
    description:
      "TEM-46 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000916",
    name: "TEM-47",
    description:
      "TEM-47 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000917",
    name: "TEM-48",
    description:
      "TEM-48 is an extended-spectrum beta-lactamase found Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000918",
    name: "TEM-49",
    description: "TEM-49 extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000919",
    name: "TEM-50",
    description:
      "TEM-50 is an inhibitor-resistant, extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000920",
    name: "TEM-51",
    description:
      "TEM-51 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000921",
    name: "TEM-52",
    description:
      "TEM-52 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae and Salmonella enterica.",
  },
  {
    accession: "ARO:3000922",
    name: "TEM-53",
    description:
      "TEM-53 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000923",
    name: "TEM-54",
    description:
      "TEM-54 is an inhibitor-resistant beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000924",
    name: "TEM-55",
    description: "TEM-55 is a broad-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000925",
    name: "TEM-56",
    description:
      "TEM-56 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000926",
    name: "TEM-57",
    description: "TEM-57 is a broad-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000927",
    name: "TEM-58",
    description:
      "TEM-58 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000928",
    name: "TEM-59",
    description:
      "TEM-59 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000929",
    name: "TEM-60",
    description:
      "TEM-60 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000930",
    name: "TEM-61",
    description:
      "TEM-61 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000931",
    name: "TEM-63",
    description:
      "TEM-63 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000932",
    name: "TEM-65",
    description:
      "TEM-65 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000933",
    name: "TEM-66",
    description:
      "TEM-66 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000934",
    name: "TEM-67",
    description:
      "TEM-67 is an inhibitor-resistant beta-lactamase found in Proteus mirabilis.",
  },
  {
    accession: "ARO:3000935",
    name: "TEM-68",
    description:
      "TEM-68 is an inhibitor resistant, extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000936",
    name: "TEM-70",
    description: "TEM-70 is a beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000937",
    name: "TEM-71",
    description:
      "TEM-71 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000938",
    name: "TEM-72",
    description:
      "TEM-72 is an extended-spectrum beta-lactamase found in Proteus mirabilis and Morganella morganii.",
  },
  {
    accession: "ARO:3000939",
    name: "TEM-73",
    description:
      "TEM-73 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000940",
    name: "TEM-74",
    description:
      "TEM-74 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000941",
    name: "TEM-75",
    description:
      "TEM-75 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000942",
    name: "TEM-76",
    description:
      "TEM-76 is an inhibitor-resistant beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000944",
    name: "TEM-78",
    description:
      "TEM-78 is an inhibitor-resistant beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000946",
    name: "TEM-79",
    description:
      "TEM-79 is an inhibitor-resistant beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000947",
    name: "TEM-80",
    description:
      "TEM-80 is an inhibitor-resistant beta-lactamase found in Enterobacter cloacae.",
  },
  {
    accession: "ARO:3000948",
    name: "TEM-81",
    description:
      "TEM-81 is an inhibitor-resistant beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000949",
    name: "TEM-82",
    description:
      "TEM-82 is an inhibitor-resistant beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000950",
    name: "TEM-83",
    description:
      "TEM-83 is an inhibitor-resistant beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000951",
    name: "TEM-84",
    description:
      "TEM-84 is an inhibitor-resistant beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000952",
    name: "TEM-85",
    description:
      "TEM-85 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000953",
    name: "TEM-86",
    description:
      "TEM-86 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000954",
    name: "TEM-87",
    description:
      "TEM-87 is an extended-spectrum beta-lactamase found in Proteus mirabilis.",
  },
  {
    accession: "ARO:3000955",
    name: "TEM-88",
    description:
      "TEM-88 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000956",
    name: "TEM-89",
    description:
      "TEM-89 is an inhibitor-resistant, extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000957",
    name: "TEM-90",
    description: "TEM-90 is a broad-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000958",
    name: "TEM-91",
    description:
      "TEM-91 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000959",
    name: "TEM-92",
    description:
      "TEM-92 is an extended-spectrum beta-lactamase found in Proteus mirabilis.",
  },
  {
    accession: "ARO:3000960",
    name: "TEM-93",
    description:
      "TEM-93 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000961",
    name: "TEM-94",
    description:
      "TEM-94 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000962",
    name: "TEM-95",
    description: "TEM-95 is a broad-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000963",
    name: "TEM-96",
    description: "TEM-96 is a beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000964",
    name: "TEM-101",
    description:
      "TEM-101 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000965",
    name: "TEM-102",
    description:
      "TEM-102 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000966",
    name: "TEM-103",
    description:
      "TEM-103 is an inhibitor-resistant beta-lactamase that has been found in E. coli.",
  },
  {
    accession: "ARO:3000967",
    name: "TEM-104",
    description: "TEM-104 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000968",
    name: "TEM-105",
    description: "TEM-105 is a beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000969",
    name: "TEM-106",
    description:
      "TEM-106 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000970",
    name: "TEM-107",
    description:
      "TEM-107 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000971",
    name: "TEM-108",
    description:
      "TEM-108 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000972",
    name: "TEM-109",
    description:
      "TEM-109 is an inhibitor-resistant, extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000973",
    name: "TEM-110",
    description:
      "TEM-110 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000974",
    name: "TEM-111",
    description:
      "TEM-111 is a beta-lactamase found in E. coli and P. mirabilis",
  },
  {
    accession: "ARO:3000975",
    name: "TEM-112",
    description:
      "TEM-112 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000976",
    name: "TEM-113",
    description:
      "TEM-113 is an extended-spectrum beta-lactamase found in Proteus mirabilis.",
  },
  {
    accession: "ARO:3000977",
    name: "TEM-114",
    description:
      "TEM-114 is an extended-spectrum beta-lactamase found in Enterobacter aerogenes.",
  },
  {
    accession: "ARO:3000978",
    name: "TEM-115",
    description:
      "TEM-115 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000979",
    name: "TEM-116",
    description:
      "TEM-116 is a broad-spectrum beta-lactamase found in many species of bacteria.",
  },
  {
    accession: "ARO:3000980",
    name: "TEM-117",
    description:
      "TEM-117 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000981",
    name: "TEM-118",
    description:
      "TEM-118 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3000982",
    name: "TEM-120",
    description:
      "TEM-120 is an extended-spectrum beta-lactamase found in Klebsiella oxytoca.",
  },
  {
    accession: "ARO:3000983",
    name: "TEM-121",
    description:
      "TEM-121 is an inhibitor-resistant, extended-spectrum beta-lactamase found in E. coli and Enterobacter aerogenes.",
  },
  {
    accession: "ARO:3000984",
    name: "TEM-122",
    description:
      "TEM-122 is an inhibitor-resistant beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000985",
    name: "TEM-123",
    description:
      "TEM-123 is an extended-spectrum beta-lactamase found in Proteus mirabilis.",
  },
  {
    accession: "ARO:3000986",
    name: "TEM-124",
    description:
      "TEM-124 is an extended-spectrum beta-lactamase found in Morganella morganii.",
  },
  {
    accession: "ARO:3000987",
    name: "TEM-125",
    description:
      "TEM-125 is a CMT-type, inhibitor-resistant, extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000988",
    name: "TEM-126",
    description:
      "TEM-126 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000989",
    name: "TEM-127",
    description: "TEM-127 is a broad-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000990",
    name: "TEM-128",
    description: "TEM-128 is a broad-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3000993",
    name: "TEM-129",
    description:
      "TEM-129 is an extended-spectrum beta-lactamase found in Klebsiella oxytoca.",
  },
  {
    accession: "ARO:3000994",
    name: "TEM-130",
    description:
      "TEM-130 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000995",
    name: "TEM-131",
    description:
      "TEM-131 is an extended-spectrum beta-lactamase found in Salmonella enterica.",
  },
  {
    accession: "ARO:3000996",
    name: "TEM-132",
    description:
      "TEM-132 is an extended-spectrum beta-lactamase found in E. coli and Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000997",
    name: "TEM-133",
    description:
      "TEM-133 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3000998",
    name: "TEM-134",
    description:
      "TEM-134 is an extended-spectrum beta-lactamase found in Citrobacter koseri.",
  },
  {
    accession: "ARO:3000999",
    name: "TEM-135",
    description:
      "TEM-135 is a broad-spectrum beta-lactamase found in Salmonella enterica.",
  },
  {
    accession: "ARO:3001000",
    name: "TEM-136",
    description:
      "TEM-136 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001001",
    name: "TEM-137",
    description:
      "TEM-137 is an extended-spectrum beta-lactamase found in Shigella sonnei.",
  },
  {
    accession: "ARO:3001002",
    name: "TEM-138",
    description:
      "TEM-138 is an extended-spectrum beta-lactamase found in Salmonella enterica.",
  },
  {
    accession: "ARO:3001003",
    name: "TEM-139",
    description:
      "TEM-139 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001004",
    name: "TEM-141",
    description:
      "TEM-141 is a broad-spectrum beta-lactamase found in Enterobacter cloacae.",
  },
  {
    accession: "ARO:3001005",
    name: "TEM-142",
    description: "TEM-142 is an beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3001006",
    name: "TEM-143",
    description:
      "TEM-143 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3001007",
    name: "TEM-144",
    description:
      "TEM-144 is an extended-spectrum beta-lactamase found in Salmonella enterica.",
  },
  {
    accession: "ARO:3001012",
    name: "TEM-145",
    description:
      "TEM-145 is an inhibitor-resistant beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3001013",
    name: "TEM-146",
    description: "TEM-146 is a beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3001014",
    name: "TEM-147",
    description:
      "TEM-147 is an extended-spectrum beta-lactamase found in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3001015",
    name: "TEM-148",
    description: "TEM-148 is a TEM beta-lactamase.",
  },
  {
    accession: "ARO:3001016",
    name: "TEM-149",
    description:
      "TEM-149 is an extended-spectrum beta-lactamase found in Enterobacter aerogenes.",
  },
  {
    accession: "ARO:3001017",
    name: "TEM-150",
    description:
      "TEM-150 is a beta-lactamase found in Enterobacter spp., E. coli, and Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001018",
    name: "TEM-151",
    description:
      "TEM-151 is a CMT-type, inhibitor-resistant, extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3001019",
    name: "TEM-152",
    description:
      "TEM-152 is a CMT-type, inhibitor-resistant, extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3001020",
    name: "TEM-154",
    description:
      "TEM-154 is a CMT-type, inhibitor-resistant, extended-spectrum beta-lactamase.",
  },
  {
    accession: "ARO:3001021",
    name: "TEM-155",
    description: "TEM-155 is an extended-spectrum beta-lactamase.",
  },
  {
    accession: "ARO:3001022",
    name: "TEM-156",
    description: "TEM-156 is a TEM beta-lactamase.",
  },
  {
    accession: "ARO:3001023",
    name: "TEM-157",
    description:
      "TEM-157 is an extended-spectrum beta-lactamase found in Enterobacter cloacae.",
  },
  {
    accession: "ARO:3001024",
    name: "TEM-158",
    description:
      "TEM-158 is an inhibitor-resistant, extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3001025",
    name: "TEM-159",
    description:
      "TEM-159 is an inhibitor-resistant beta-lactamase found in Proteus mirabilis.",
  },
  {
    accession: "ARO:3001026",
    name: "TEM-160",
    description:
      "TEM-160 is an inhibitor-resistant beta-lactamase found in Proteus mirabilis.",
  },
  {
    accession: "ARO:3001027",
    name: "TEM-161",
    description:
      "TEM-161 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001028",
    name: "TEM-162",
    description:
      "TEM-162 is a beta-lactamase found in Acinetobacter haemolyticus.",
  },
  {
    accession: "ARO:3001029",
    name: "TEM-163",
    description:
      "TEM-163 is an inhibitor resistant beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3001030",
    name: "TEM-164",
    description:
      "TEM-164 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001031",
    name: "TEM-165",
    description:
      "TEM-165 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001032",
    name: "TEM-166",
    description: "TEM-166 is a beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3001033",
    name: "TEM-167",
    description:
      "TEM-167 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3001034",
    name: "TEM-168",
    description: "TEM-168 is an extended-spectrum beta-lactamase.",
  },
  {
    accession: "ARO:3001035",
    name: "TEM-169",
    description: "TEM-169 is an extended spectrum beta-lactamase.",
  },
  {
    accession: "ARO:3001036",
    name: "TEM-170",
    description: "TEM-170 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001037",
    name: "TEM-171",
    description: "TEM-171 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001038",
    name: "TEM-172",
    description: "TEM-172 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001039",
    name: "TEM-173",
    description: "TEM-173 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001040",
    name: "TEM-174",
    description: "TEM-174 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001041",
    name: "TEM-176",
    description: "TEM-176 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001042",
    name: "TEM-177",
    description: "TEM-177 is an extended-spectrum beta-lactamase.",
  },
  {
    accession: "ARO:3001043",
    name: "TEM-178",
    description:
      "TEM-178 is an inhibitor-resistant, extended-spectrum beta-lactamase.",
  },
  {
    accession: "ARO:3001044",
    name: "TEM-181",
    description: "TEM-181 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001045",
    name: "TEM-183",
    description: "TEM-183 is a broad-spectrum beta-lactamase.",
  },
  {
    accession: "ARO:3001046",
    name: "TEM-186",
    description: "TEM-186 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001047",
    name: "TEM-187",
    description:
      "TEM-187 is an extended-spectrum beta-lactamase found in P. mirabilis",
  },
  {
    accession: "ARO:3001048",
    name: "TEM-188",
    description: "TEM-188 is an extended-spectrum beta-lactamase.",
  },
  {
    accession: "ARO:3001049",
    name: "TEM-189",
    description: "TEM-189 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001050",
    name: "TEM-190",
    description: "TEM-190 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001051",
    name: "TEM-191",
    description: "TEM-191 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001052",
    name: "TEM-192",
    description: "TEM-192 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001053",
    name: "TEM-193",
    description: "TEM-193 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001054",
    name: "TEM-194",
    description: "TEM-194 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001055",
    name: "TEM-195",
    description: "TEM-195 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001056",
    name: "TEM-197",
    description: "TEM-197 is an extended-spectrum beta-lactamase.",
  },
  {
    accession: "ARO:3001057",
    name: "TEM-198",
    description: "TEM-198 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001058",
    name: "TEM-199",
    description: "TEM-199 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001059",
    name: "SHV-1",
    description:
      "SHV-1 is a broad-spectrum beta-lactamase found in Klebsiella spp., as well as Acinetobacter spp., E. coli., Raoultella terrigena, and Yersinia pestis.",
  },
  {
    accession: "ARO:3001060",
    name: "SHV-2",
    description:
      "SHV-2 is an extended-spectrum beta-lactamase found in E. coli., Klebsiella pneumoniae, and Shigella flexneri.",
  },
  {
    accession: "ARO:3001061",
    name: "SHV-2A",
    description:
      "SHV-2A is an extended-spectrum beta-lactamase found in Enterobacter cloacae, Klebsiella pneumoniae, Pseudomonas aeruginosa, and Salmonella enterica.",
  },
  {
    accession: "ARO:3001062",
    name: "SHV-3",
    description:
      "SHV-3 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001063",
    name: "SHV-4",
    description:
      "SHV-4 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001064",
    name: "SHV-5",
    description:
      "SHV-5 is an extended-spectrum beta-lactamase found in Acinetobacter baumannii, Enterobacter cloacae, E. coli, Klebsiella pneumoniae, Pseudomonas aeruginosa, and Salmonella enterica.",
  },
  {
    accession: "ARO:3001065",
    name: "SHV-6",
    description:
      "SHV-6 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001066",
    name: "SHV-7",
    description:
      "SHV-7 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3001067",
    name: "SHV-8",
    description:
      "SHV-8 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3001068",
    name: "SHV-9",
    description:
      "SHV-9 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001069",
    name: "SHV-10",
    description:
      "SHV-10 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001070",
    name: "SHV-11",
    description:
      "SHV-11 is a broad-spectrum beta-lactamase found in E. coli, Klebsiella pneumoniae, Proteus mirabilis, and Shigella dysenteriae.",
  },
  {
    accession: "ARO:3001071",
    name: "SHV-12",
    description:
      "SHV-12 is an extended-spectrum beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3001072",
    name: "SHV-13",
    description:
      "SHV-13 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001073",
    name: "SHV-14",
    description:
      "SHV-14 is an broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001074",
    name: "SHV-15",
    description:
      "SHV-15 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3001075",
    name: "SHV-16",
    description:
      "SHV-16 is an extended-spectrum beta-lactamase that has been found in clinical isolates. It differs from SHV-1 by a 163DRWET167 insertion.",
  },
  {
    accession: "ARO:3001076",
    name: "SHV-18",
    description:
      "SHV-18 is an extended-spectrum beta-lactamase found in Acinetobacter baumannii and Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001077",
    name: "SHV-19",
    description:
      "SHV-19 is a broad-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001078",
    name: "SHV-20",
    description:
      "SHV-20 is a broad-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001079",
    name: "SHV-21",
    description:
      "SHV-21 is a broad-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001080",
    name: "SHV-22",
    description:
      "SHV-22 is a broad-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001081",
    name: "SHV-23",
    description:
      "SHV-23 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001082",
    name: "SHV-24",
    description:
      "SHV-24 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3001083",
    name: "SHV-25",
    description:
      "SHV-25 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001084",
    name: "SHV-26",
    description:
      "SHV-26 is an inhibitor-resistant beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001085",
    name: "SHV-27",
    description:
      "SHV-27 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001086",
    name: "SHV-28",
    description: "SHV-28 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001087",
    name: "SHV-29",
    description:
      "SHV-29 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001088",
    name: "SHV-30",
    description:
      "SHV-30 is an extended-spectrum beta-lactamase found in Enterobacter cloacae.",
  },
  {
    accession: "ARO:3001089",
    name: "SHV-31",
    description:
      "SHV-31 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001090",
    name: "SHV-32",
    description:
      "SHV-32 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001091",
    name: "SHV-33",
    description:
      "SHV-33 is an broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001092",
    name: "SHV-34",
    description:
      "SHV-34 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3001093",
    name: "SHV-35",
    description: "SHV-35 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001094",
    name: "SHV-36",
    description: "SHV-36 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001095",
    name: "SHV-37",
    description: "SHV-37 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001096",
    name: "SHV-38",
    description:
      "SHV-38 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001097",
    name: "SHV-39",
    description:
      "SHV-39 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001098",
    name: "SHV-40",
    description:
      "SHV-40 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001099",
    name: "SHV-41",
    description:
      "SHV-41 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001100",
    name: "SHV-42",
    description:
      "SHV-42 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001101",
    name: "SHV-43",
    description:
      "SHV-43 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001102",
    name: "SHV-44",
    description:
      "SHV-44 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001103",
    name: "SHV-45",
    description:
      "SHV-45 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001104",
    name: "SHV-46",
    description:
      "SHV-46 is an extended-spectrum beta-lactamase found in Klebsiella oxytoca.",
  },
  {
    accession: "ARO:3001105",
    name: "SHV-48",
    description:
      "SHV-48 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001106",
    name: "SHV-49",
    description:
      "SHV-49 is an inhibitor-resistant beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001107",
    name: "SHV-50",
    description: "SHV-50 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001108",
    name: "SHV-51",
    description: "SHV-51 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001109",
    name: "SHV-52",
    description:
      "SHV-52 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001110",
    name: "SHV-53",
    description:
      "SHV-53 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001111",
    name: "SHV-55",
    description:
      "SHV-55 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001112",
    name: "SHV-56",
    description:
      "SHV-56 is an inhibitor-resistant beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001113",
    name: "SHV-57",
    description:
      "SHV-57 is an extended-spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3001114",
    name: "SHV-59",
    description: "SHV-59 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001115",
    name: "SHV-60",
    description:
      "SHV-60 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001116",
    name: "SHV-61",
    description:
      "SHV-61 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001117",
    name: "SHV-62",
    description:
      "SHV-62 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001118",
    name: "SHV-63",
    description: "SHV-63 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001119",
    name: "SHV-64",
    description:
      "SHV-64 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001120",
    name: "SHV-65",
    description: "SHV-65 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001121",
    name: "SHV-66",
    description:
      "SHV-66 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001122",
    name: "SHV-67",
    description: "SHV-67 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001123",
    name: "SHV-69",
    description: "SHV-69 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001124",
    name: "SHV-70",
    description:
      "SHV-70 is an extended-spectrum beta-lactamase found in Enterobacter cloacae.",
  },
  {
    accession: "ARO:3001125",
    name: "SHV-71",
    description:
      "SHV-71 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae and Shigella.",
  },
  {
    accession: "ARO:3001126",
    name: "SHV-72",
    description:
      "SHV-72 is an inhibitor-resistant beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001127",
    name: "SHV-73",
    description:
      "SHV-73 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001128",
    name: "SHV-74",
    description:
      "SHV-74 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001129",
    name: "SHV-75",
    description:
      "SHV-75 is a broad-spectrum beta-lactamase that is found in clinical isolates.",
  },
  {
    accession: "ARO:3001130",
    name: "SHV-76",
    description:
      "SHV-76 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001131",
    name: "SHV-77",
    description:
      "SHV-77 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae and E. coli.",
  },
  {
    accession: "ARO:3001132",
    name: "SHV-78",
    description:
      "SHV-78 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001133",
    name: "SHV-79",
    description:
      "SHV-79 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001134",
    name: "SHV-80",
    description:
      "SHV-80 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001135",
    name: "SHV-81",
    description:
      "SHV-81 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001136",
    name: "SHV-82",
    description:
      "SHV-82 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001137",
    name: "SHV-83",
    description:
      "SHV-83 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001138",
    name: "SHV-84",
    description:
      "SHV-84 is an inhibitor-resistant beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001139",
    name: "SHV-85",
    description:
      "SHV-85 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001140",
    name: "SHV-86",
    description:
      "SHV-86 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001141",
    name: "SHV-89",
    description:
      "SHV-89 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001142",
    name: "SHV-90",
    description:
      "SHV-90 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001143",
    name: "SHV-91",
    description:
      "SHV-91 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001144",
    name: "SHV-92",
    description: "SHV-92 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001145",
    name: "SHV-93",
    description: "SHV-93 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001146",
    name: "SHV-94",
    description: "SHV-94 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001147",
    name: "SHV-95",
    description: "SHV-95 is a beta-lactamase found in Citrobacter freundii.",
  },
  {
    accession: "ARO:3001148",
    name: "SHV-96",
    description: "SHV-96 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3001149",
    name: "SHV-97",
    description: "SHV-97 is a beta-lactamase found in Enterococcus faecalis.",
  },
  {
    accession: "ARO:3001150",
    name: "SHV-101",
    description: "SHV-101 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001151",
    name: "SHV-102",
    description:
      "SHV-102 is an extended spectrum beta-lactamase found in E. coli.",
  },
  {
    accession: "ARO:3001152",
    name: "SHV-103",
    description:
      "SHV-103 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001153",
    name: "SHV-104",
    description:
      "SHV-104 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001154",
    name: "SHV-105",
    description:
      "SHV-105 is an extended-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001155",
    name: "SHV-106",
    description:
      "SHV-106 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001156",
    name: "SHV-107",
    description:
      "SHV-107 is an inhibitor-resistant beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001157",
    name: "SHV-108",
    description: "SHV-108 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001158",
    name: "SHV-109",
    description:
      "SHV-109 is a broad-spectrum beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001159",
    name: "SHV-110",
    description: "SHV-110 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001160",
    name: "SHV-111",
    description: "SHV-111 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001161",
    name: "SHV-112",
    description:
      "SHV-112 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001162",
    name: "SHV-113",
    description:
      "SHV-113 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001163",
    name: "SHV-114",
    description:
      "SHV-114 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001164",
    name: "SHV-115",
    description:
      "SHV-115 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001165",
    name: "SHV-116",
    description:
      "SHV-116 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001166",
    name: "SHV-117",
    description:
      "SHV-117 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001167",
    name: "SHV-120",
    description:
      "SHV-120 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001168",
    name: "SHV-121",
    description:
      "SHV-121 is a beta-lactamase that has been found in clinical isolates. Identical to SHV-136.",
  },
  {
    accession: "ARO:3001169",
    name: "SHV-122",
    description:
      "SHV-122 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001170",
    name: "SHV-123",
    description:
      "SHV-123 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001171",
    name: "SHV-124",
    description:
      "SHV-124 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001172",
    name: "SHV-125",
    description:
      "SHV-125 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001173",
    name: "SHV-126",
    description:
      "SHV-126 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001174",
    name: "SHV-127",
    description:
      "SHV-127 is a beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001175",
    name: "SHV-128",
    description:
      "SHV-128 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001176",
    name: "SHV-129",
    description:
      "SHV-129 is an extended-spectrum beta-lactamase that has been found in clinical isolates.",
  },
  {
    accession: "ARO:3001177",
    name: "SHV-133",
    description: "SHV-133 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001178",
    name: "SHV-134",
    description: "SHV-134 is an extended-spectrum beta-lactamase.",
  },
  {
    accession: "ARO:3001179",
    name: "SHV-135",
    description: "SHV-135 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001181",
    name: "SHV-137",
    description: "SHV-137 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001182",
    name: "SHV-140",
    description: "SHV-140 is a broad-spectrum beta-lactamase.",
  },
  {
    accession: "ARO:3001183",
    name: "SHV-141",
    description: "SHV-141 is a broad-spectrum beta-lactamase.",
  },
  {
    accession: "ARO:3001184",
    name: "SHV-142",
    description: "SHV-142 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001185",
    name: "SHV-145",
    description: "SHV-145 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001186",
    name: "SHV-147",
    description: "SHV-147 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001187",
    name: "SHV-148",
    description: "SHV-148 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001188",
    name: "SHV-149",
    description: "SHV-149 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001189",
    name: "SHV-150",
    description: "SHV-150 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001190",
    name: "SHV-151",
    description: "SHV-151 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001191",
    name: "SHV-152",
    description: "SHV-152 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001192",
    name: "SHV-153",
    description: "SHV-153 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001193",
    name: "SHV-154",
    description: "SHV-154 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001194",
    name: "SHV-155",
    description: "SHV-155 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001195",
    name: "SHV-156",
    description: "SHV-156 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001196",
    name: "SHV-157",
    description: "SHV-157 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001197",
    name: "SHV-158",
    description: "SHV-158 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001198",
    name: "SHV-159",
    description: "SHV-159 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001199",
    name: "SHV-160",
    description: "SHV-160 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001200",
    name: "SHV-161",
    description: "SHV-161 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001201",
    name: "SHV-162",
    description: "SHV-162 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001202",
    name: "SHV-163",
    description: "SHV-163 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001203",
    name: "SHV-165",
    description: "SHV-165 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001204",
    name: "SHV-167",
    description: "SHV-167 is a beta-lactamase.",
  },
  {
    accession: "ARO:3001205",
    name: "BRP(MBL)",
    description:
      "A novel bleomycin resistance protein encoded by a metallo-beta-lactamase-associated ble gene. Expression of BRP(MBL) confers resistance to bleomycin and bleomycin-like antibiotics in Enterobacteriaceae and Acinetobacter, where it is co-expressed with an MBL and controlled by the same promoter region.",
  },
  {
    accession: "ARO:3001206",
    name: "antibiotic inactivation by sequestration",
    description:
      "Inactivation of an antibiotic by formation of a complex, preventing interaction of the antibiotic with its target.",
  },
  {
    accession: "ARO:3001207",
    name: "gene involved in antibiotic sequestration",
    description:
      "A gene whose product inactivates an antibiotic by formation of a complex, preventing interaction of the antibiotic with its target.",
  },
  {
    accession: "ARO:3001208",
    name: "methicillin resistant PBP2",
    description:
      "In methicillin sensitive S. aureus (MSSA), beta-lactams bind to native penicillin-binding proteins (PBPs) and disrupt synthesis of the cell membrane's peptidoglycan layer. In methicillin resistant S. aureus (MRSA), foreign PBP2a acquired by lateral gene transfer is able to perform peptidoglycan synthesis in the presence of beta-lactams.",
  },
  {
    accession: "ARO:3001209",
    name: "mecC",
    description:
      "A foreign PBP2a acquired by lateral gene transfer that able to perform peptidoglycan synthesis in the presence of beta-lactams.",
  },
  {
    accession: "ARO:3001214",
    name: "mdtM",
    description: "Multidrug resistance protein MdtM",
  },
  {
    accession: "ARO:3001215",
    name: "mdtL",
    description: "Putative/predicted multidrug resistance protein MdtL (YidY).",
  },
  {
    accession: "ARO:3001216",
    name: "mdtH",
    description: "Multidrug resistance protein MdtH",
  },
  {
    accession: "ARO:3001218",
    name: "trimethoprim resistant dihydrofolate reductase dfr",
    description:
      "Alternative dihydropteroate synthase dfr present on plasmids produces alternate proteins that are less sensitive to trimethoprim from inhibiting its role in folate synthesis, thus conferring trimethoprim resistance.",
  },
  {
    accession: "ARO:3001219",
    name: "elfamycin antibiotic",
    description:
      "Elfamycins are molecules that inhibit bacterial elongation factor Tu (EF-Tu), a key protein which brings aminoacyl-tRNA (aa-tRNA) to the ribosome during protein synthesis. Elfamycins defined by their target (EF-Tu), rather than a conserved chemical backbone. Elfamycins follow two mechanisms to disrupt protein synthesis: 1. kirromycins and enacyloxin fix EF-Tu in the GTP bound conformation and lock EF-Tu onto the ribosome, and 2. pulvomycin and GE2270 cover the binding site of aa-tRNA disallowing EF-Tu from being charged with aa-tRNA. All elfamycins cause increased the affinity of EF-Tu for GTP.",
  },
  {
    accession: "ARO:3001220",
    name: "factumycin",
    description:
      "Factumycin is a kirromycin-like antibiotic produced by Kitasatospora setae and Streptomyces strains. Its biosynthetic cluster has been characterized which has interesting acetyl transferase domains in trans, or outside of the polyketide synthase domains. Factumycin has specific, rather than broad spectrum,  antibacterial properties, especially targeting various Acinetobacter baumanii strains.",
  },
  {
    accession: "ARO:3001221",
    name: "josamycin",
    description:
      "A macrolide antibiotic from Streptomyces narbonensis\u00a0subsp.\u00a0josamyceticus. The drug has antimicrobial activity against a wide spectrum of pathogens.",
  },
  {
    accession: "ARO:3001222",
    name: "chalcomycin",
    description: "Produced by Streptomyces bikiniensis",
  },
  {
    accession: "ARO:3001223",
    name: "midecamycin",
    description: "Produced by Streptomyces mycarofaciens.",
  },
  {
    accession: "ARO:3001224",
    name: "mycinamicin",
    description: "Produced by Micromonospora\u00a0griseorubida.",
  },
  {
    accession: "ARO:3001225",
    name: "megalomycin",
    description: "Produced by Micromonospora megalomicia.",
  },
  {
    accession: "ARO:3001226",
    name: "narbomycin",
    description: "Produced by Streptomyces narbonensis.",
  },
  {
    accession: "ARO:3001227",
    name: "kitasamycin",
    description:
      "Kitasamycin is a macrolide antibiotic and is produced by Streptoverticillium kitasatoense. The drug has antimicrobial activity against a wide spectrum of pathogens.",
  },
  {
    accession: "ARO:3001228",
    name: "carbomycin",
    description:
      "Produced by Streptomyces halstedii and Streptomyces thermotolerans.",
  },
  {
    accession: "ARO:3001229",
    name: "rosaramicin",
    description: "Produced by Micromonospora rosaria.",
  },
  {
    accession: "ARO:3001230",
    name: "niddamycin",
    description: "Produced by Streptomyces caelestis.",
  },
  {
    accession: "ARO:3001231",
    name: "celesticetin",
    description:
      "Produced by Streptomyces caelestis, which also produces the macrolide niddamycin.",
  },
  {
    accession: "ARO:3001232",
    name: "methymycin",
    description: "Produced by Streptomyces venezuelae ATCC 15439.",
  },
  {
    accession: "ARO:3001233",
    name: "pikromycin",
    description: "Produced by Streptomyces venezuelae ATCC 15439.",
  },
  {
    accession: "ARO:3001234",
    name: "kirromycin",
    description:
      "Kirromycin, also known as mocimycin, is the representative molecule of its own class of elfamycins which is composed of more than 10 analogs. Kirromycin binds to the domain 1,2 interface of elongation factor Tu. This interaction maintains the EF-Tu*GTP conformation even after GTP is hydrolyzed to GDP. EF-Tu*GDP normally releases aa-tRNA and then exits the ribosome; however, kirromycin*EF-Tu*GDP*aa-tRNA forms a strong complex and remains bound to the ribosome, prohibits translocation of the peptide chain and translation is halted.",
  },
  {
    accession: "ARO:3001235",
    name: "aurodox",
    description:
      "Aurodox is an elongation factor Tu inhibiting antibiotic. It was first referred to as X-5108. It is produced by Streptomyces goldiniensis var. goldiniensis and is active primarily against gram-positive bacteria.",
  },
  {
    accession: "ARO:3001236",
    name: "efrotomycin",
    description:
      "Efrotomycin is produced by Streptomyces lactamodurans NRRL 3802. This kirromycin-like antibiotic is used in animal agriculture, especially in the production of swine. Generally gram negative bacteria are susceptible to efrotomycin, while gram positive bacteria are resistant.",
  },
  {
    accession: "ARO:3001237",
    name: "GE2270A",
    description:
      "GE2270A is the model molecule of cyclic thiazolyl peptide elfamycins. \nGE2270A is produced by Planobispora rosea. Biosynthesis of the molecule has been shown to originate as a ribosomally synthesized peptide that undergoes significant post-translational modification. Clinical use of cyclic thiazolyl peptides is hindered by their low water solubility and bioavailability.",
  },
  {
    accession: "ARO:3001242",
    name: "enacyloxin IIa",
    description:
      "Enacyloxin IIa is structurally distinct but acts in a similar mechanism to kirromycin-like elfamycins. It prohibits the transfer of the amino acid at the A site to the elongating peptide chain. It is most likely that the mechanism of action is that EF-Tu*GDP is locked in the EF-Tu*GTP form, and EF-Tu*GDP*aa-tRNA is immobilized on the ribosome. It is an open question whether enacyloxin IIa actually belongs to the kirromycin-like group of elfamycins due to their high similarity.",
  },
  {
    accession: "ARO:3001246",
    name: "kirrothricin",
    description:
      "Kirrothricin is produced by Streptomyces cinnamoneus and is structurally related to kirromycin, aurodox, heneicomycin and shares a similar spectrum of antibacterial activity.",
  },
  {
    accession: "ARO:3001247",
    name: "azdimycin",
    description:
      "Azdimycin is produced by Streptomyces diastatochromogenes ATCC 31013.",
  },
  {
    accession: "ARO:3001248",
    name: "dihydromocimycin",
    description:
      "Dihydromocimycin is produced by Streptomyces ramocissimus CBS190.69.",
  },
  {
    accession: "ARO:3001249",
    name: "phenelfamycin A",
    description:
      "Phenelfamycin A was isolated from two environmental isolates called AB999F-80 and AB1047T-33. They were both identified as Streptomyces violaceoniger strains. Both these strains were found to produce phenelfamycins A-F and unphenelfamycin. Phenelfamycin A has a monosaccaride moeity and is isomeric with phenelfamycin B. All phenelfamycins were selected due to their activity against anaerobic bacteria, especially Clostridium difficile.",
  },
  {
    accession: "ARO:3001250",
    name: "phenelfamycin B",
    description:
      "Phenelfamycin B was isolated from two environmental isolates called AB999F-80 and AB1047T-33. They were both identified as Streptomyces violaceoniger strains. Both these strains were found to produce phenelfamycins A-F and unphenelfamycin. Phenelfamycin B has a monosaccaride moeity and is isomeric with phenelfamycin A. All phenelfamycins were selected due to their activity against anaerobic bacteria, especially Clostridium difficile.",
  },
  {
    accession: "ARO:3001251",
    name: "phenelfamycin C",
    description:
      "Phenelfamycin C was isolated from two environmental isolates called AB999F-80 and AB1047T-33. They were both identified as Streptomyces violaceoniger strains. Both these strains were found to produce phenelfamycins A-F and unphenelfamycin. Phenelfamycin C has a disaccaride moeity and is isomeric with phenelfamycin D. All phenelfamycins were selected due to their activity against anaerobic bacteria, especially Clostridium difficile.",
  },
  {
    accession: "ARO:3001252",
    name: "phenelfamycin D",
    description:
      "Phenelfamycin D was isolated from two environmental isolates called AB999F-80 and AB1047T-33. They were both identified as Streptomyces violaceoniger strains. Both these strains were found to produce phenelfamycins A-F and unphenelfamycin. Phenelfamycin D has a disaccaride moeity and is isomeric with phenelfamycin C. All phenelfamycins were selected due to their activity against anaerobic bacteria, especially Clostridium difficile.",
  },
  {
    accession: "ARO:3001253",
    name: "phenelfamycin E",
    description:
      "Phenelfamycin E was isolated from two environmental isolates called AB999F-80 and AB1047T-33. They were both identified as Streptomyces violaceoniger strains. Both these strains were found to produce phenelfamycins A-F and unphenelfamycin. Phenelfamycin E has a trisaccaride moeity and is isomeric with phenelfamycin F. All phenelfamycins were selected due to their activity against anaerobic bacteria, especially Clostridium difficile. Phenelfamycin E and F are the most potent of the phenelfamycins.",
  },
  {
    accession: "ARO:3001256",
    name: "phenelfamycin F",
    description:
      "Phenelfamycin F was isolated from two environmental isolates called AB999F-80 and AB1047T-33. They were both identified as Streptomyces violaceoniger strains. Both these strains were found to produce phenelfamycins A-F and unphenelfamycin. Phenelfamycin F has a trisaccaride moeity and is isomeric with phenelfamycin E. All phenelfamycins were selected due to their activity against anaerobic bacteria, especially Clostridium difficile. Phenelfamycin E and F are the most potent of the phenelfamycins.",
  },
  {
    accession: "ARO:3001257",
    name: "unphenelfamycin",
    description:
      "Unphenelfamycin was isolated from two environmental isolates called AB999F-80 and AB1047T-33. They were both identified as Streptomyces violaceoniger strains. Both these strains were found to produce phenelfamycins A-F and unphenelfamycin. Unphenelfamycin has a monosaccaride moeity and lacks the  phenacetyl moeity present in the other phenelfamycins. All phenelfamycins were selected due to their activity against anaerobic bacteria, especially Clostridium difficile.",
  },
  {
    accession: "ARO:3001258",
    name: "L-681,217",
    description:
      "L-681217 is a kirromycin-like antibiotic produced by Streptomyces cattleya. Its biosynthetic gene cluster is characterized.",
  },
  {
    accession: "ARO:3001259",
    name: "GE37468",
    description:
      "GE37468 is a thiazolyl peptide produced by Streptomyces ATCC 55365. GE37468 is produced by post-transcription modification of a ribosomally synthesized preprotein, of which the biosynthetic gene cluster has been elucidated.",
  },
  {
    accession: "ARO:3001260",
    name: "amythiamicin A",
    description:
      "Amythiamicin A is a cyclic thiazolyl peptide antibiotic with a structure similar to GE2270A. Amythiamicin B, C, and D also exist, but are not discussed further here. Amythiamicin A was shown to have the highest activity in cell free protein synthesis assays.",
  },
  {
    accession: "ARO:3001264",
    name: "heneicomycin",
    description:
      "Heneicomycin is a kirromycin-like antibiotic, isolated from Streptomyces filipinensis. It is structurally related to kirromycin, aurodox and efrotomycin.",
  },
  {
    accession: "ARO:3001265",
    name: "Erm(30)",
    description:
      "Erm(30) confers a MLSb resistant phenotype. Along with erm(31), these genes are responsible for self-resistance in the pikromycin/narbomycin/methymycin/neomethymycin producer, Streptomyces venezuelae.",
  },
  {
    accession: "ARO:3001294",
    name: "verdamicin",
    description:
      "A gentamicin-group aminoglycoside isolated from Micromonospora.",
  },
  {
    accession: "ARO:3001295",
    name: "vertilimicin",
    description:
      "A semisynthetic genamicin-group aminoglycoside, derived from verdamicin.",
  },
  {
    accession: "ARO:3001296",
    name: "kasugamycin",
    description:
      "An unusual aminoglycoside because the cyclitol ring is not amino substituted; it was discovered as a fermentation product of Streptomyces kasugaensis.",
  },
  {
    accession: "ARO:3001298",
    name: "non-erm 23S ribosomal RNA methyltransferase (G748)",
    description:
      "Non-erm 23S ribosomal RNA methyltransferases modify guanosine 748 (E. coli numbering) to confer resistance to some macrolides and lincosamides",
  },
  {
    accession: "ARO:3001299",
    name: "tlrB conferring tylosin resistance",
    description:
      "TlrB is a methyltransferase found in Streptomyces fradiae and confers resistance to mycinamicin, tylosin and lincosamides. Specifically, this enzyme adds a methyl group to guanosine 748 (E. coli numbering). TlrB is found in the tylosin biosynthetic cluster and is one mechanism by which S. fradiae protects itself from self-destruction when producing this macrolide.",
  },
  {
    accession: "ARO:3001300",
    name: "myrA",
    description:
      "myrA is a methyltransferase found in Micromonospora griseorubida and confers resistance to mycinamicin, tylosin and lincosamides. Specifically, this enzyme adds a methyl group to guanosine 748 (E. coli numbering) of 23S ribosomal RNA. MyrA is found in the mycinamicin biosynthetic cluster and is one mechanism by which M. griseorubida protects itself from self-destruction when producing this macrolide.",
  },
  {
    accession: "ARO:3001301",
    name: "RlmA(II)",
    description:
      "RlmA(II) is a methyltransferase found in Streptococcus pneumoniae and confers resistance to mycinamicin, tylosin and lincosamides. Specifically, this enzyme adds a methyl group to guanosine 748 (E. coli numbering) of 23S ribosomal RNA.",
  },
  {
    accession: "ARO:3001302",
    name: "chrB",
    description:
      "ChrB is a methyltransferase found in Streptomyces bikiniensis and confers resistance to chalcomycin, mycinamicin, tylosin and lincosamides. Specifically, this enzyme adds a methyl group to guanosine 748 (E. coli numbering). chrB is found in the chalcomycin biosynthetic cluster and is responsible for self-resistance in S. bikiniensis.",
  },
  {
    accession: "ARO:3001303",
    name: "ErmO",
    description:
      "ErmO is a methyltransferase found in the spiramycin producer Streptomyces ambofaciens. Like other Erm enzymes, it catalyzes the methylation of A2058 of the 23S ribosomal RNA. Specifically, this enzyme transfers only one methyl group. The gene is responsible for self-resistance to spiramycin.",
  },
  {
    accession: "ARO:3001304",
    name: "ErmS",
    description:
      "ErmS is a methyltransferase found in the tylosin producer Streptomyces fradiae. Like other Erm enzymes, it catalyzes the methylation of A2058 of the 23S ribosomal RNA. Specifically, this enzyme transfers two methyl groups. The gene is found within the tylosin biosynthetic cluster and is responsible for self-resistance",
  },
  {
    accession: "ARO:3001305",
    name: "ErmU",
    description:
      "ErmU is a methyltransferase found in the lincomycin producer Streptomyces lincolnensis. Like other Erm enzymes, it catalyzes the methylation of A2058 of the 23S ribosomal RNA. Specifically, this enzyme transfers only one methyl group. The gene is found in the lincomycin biosynthetic cluster and is responsible for self-resistance.",
  },
  {
    accession: "ARO:3001306",
    name: "ErmW",
    description:
      "ErmW is a methyltransferase found in the mycinamicin producer Micromonospora griseorubida. Like other Erm enzymes, it catalyzes the methylation of A2058 of the 23S ribosomal RNA. The gene is found within the mycinamicin biosynthetic cluster and is responsible for self-resistance.",
  },
  {
    accession: "ARO:3001307",
    name: "vgbA",
    description:
      "vgbA inactivates streptogramin B-type antibiotics by linearizing the lactone ring on the ester bond using an elimination mechanism, thus conferring resistance to these compounds.",
  },
  {
    accession: "ARO:3001308",
    name: "vgbB",
    description:
      "VgbB inactivates streptogramin B-type antibiotics by linearizing the lactone ring on the ester bond through an elimination mechanism, thus conferring resistance.",
  },
  {
    accession: "ARO:3001309",
    name: "cyclic thiazolyl peptide elfamycin",
    description:
      "Cyclic thiazolyl peptide elfamycins inhibit protein synthesis by preventing the formation of the EF-Tu*GTP-aa-tRNA ternary complex, a mechanism similar to that of pulvomycin. There are non-elfamycin cyclic thiazolyl peptide antibiotics which inhibit protein synthesis by by targeting the 50S ribosomal subunit and are not discussed here further.",
  },
  {
    accession: "ARO:3001310",
    name: "kirromycin-like antibiotics",
    description:
      "Kirromycin-like antibiotics bind to the domain 1,2 interface of elongation factor Tu. This interaction maintains the EF-Tu*GTP conformation even after GTP is hydrolyzed to GDP. EF-Tu*GDP normally releases aa-tRNA and then exits the ribosome; however, kirromycin*EF-Tu*GDP*aa-tRNA forms a strong complex and remains bound to the ribosome, prohibits translocation of the peptide chain and translation is halted.",
  },
  {
    accession: "ARO:3001311",
    name: "determinant of elfamycin resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to elfamycin antibiotics.",
  },
  {
    accession: "ARO:3001312",
    name: "elfamycin resistant EF-Tu",
    description:
      "Sequence variants of elongation factor Tu that confer resistance to elfamycin antibiotics.",
  },
  {
    accession: "ARO:3001313",
    name: "facT",
    description:
      "Efflux protein facT confers resistance to factumycin. The gene has been heterologously expressed in S. coelicolor and its function was confirmed (Thaker et al. Med. Chem. Commun. 2012, 3: 1020). It is possible that this gene can efflux other kirromycin-like elfamycins, but this has not been tested.",
  },
  {
    accession: "ARO:3001314",
    name: "retapamulin",
    description:
      "Retapamulin is a semi-synthetic pleuromutilin antibiotic approved for the treatment of skin infections.",
  },
  {
    accession: "ARO:3001315",
    name: "valnemulin",
    description:
      "Two semi-synthetic pleuromutilin derivatives, tiamulin and valnemulin, have been developed as antibiotics for veterinary use.",
  },
  {
    accession: "ARO:3001316",
    name: "azamulin",
    description:
      "Azamulin is a semi-synthetic pleuromutilin derivative that failed Stage 1 clinical trials due to poor bioavailability.",
  },
  {
    accession: "ARO:3001317",
    name: "pleuromutilin",
    description:
      "Pleuromutilin is a natural product antibiotic produced by Clitopilus passeckerianus. Related antibiotics of clinical significance, such as tiamulin and retapamulin, are semi-synthetic derivatives of this compound.",
  },
  {
    accession: "ARO:3001318",
    name: "macrocyclic antibiotic",
    description:
      "Narrow spectrum macrocyclic antibiotics, in the general sense, is a cyclic molecule containing more than 12 atoms in the macrocyclic lactone ring.",
  },
  {
    accession: "ARO:3001319",
    name: "fidaxomicin",
    description:
      "A narrow spectrum antibiotic effective against Clostridium difficile infection (CDI). It is active against gram positive and gram negative bacteria by inhibiting RNA polymerase's switch region leading to cell death.",
  },
  {
    accession: "ARO:3001320",
    name: "SB22484",
    description:
      "SB22484 antibiotics are of the same chemical group as kirromycins. SB22484 antibiotics are produced by Streptomyces sp. NRRL 15496. There are two SB22484 isomers with the same UV absorption spectra, but differing molecular masses. These antibiotics are strongly active against Neisseria species, but also show activity against Streptococci, and Haemophilus influenzae.",
  },
  {
    accession: "ARO:3001321",
    name: "UK-69,753",
    description:
      "UK-69,753 is a glycosylated form of factumycin. The disaccharide moiety of UK-69,753 is virtually identical to that of efrotomycin, another member of this class of antibiotics. UK-69,753 has been shown to have strong in vitro and in vivo activity toward a swine pathogen, Treponema hyodysenteriae.",
  },
  {
    accession: "ARO:3001322",
    name: "phenelfamycin G",
    description:
      "Phenelfamycin G is produced by Streptomyces albospinus strain Acta 3619. Phenelfamycin G and H are isomeric and similar to phenelfamycin E and F only with the addition of a hydroxyl group at position C-30. Phenelfamycin G and H have a narrow antibacterial spectrum with potent activity against Propionibacterium acnes.",
  },
  {
    accession: "ARO:3001323",
    name: "phenelfamycin H",
    description:
      "Phenelfamycin H is produced by Streptomyces albospinus strain Acta 3619. Phenelfamycin G and H are isomeric and similar to phenelfamycin E and F only with the addition of a hydroxyl group at position C-30. Phenelfamycin G and H have a narrow antibacterial spectrum with potent activity against Propionibacterium acnes.",
  },
  {
    accession: "ARO:3001326",
    name: "ganefromycin",
    description:
      "One of a small group of kirromycin-like elfamycins which lack the pyridone moiety. Ganefromycin is produced by Streptomyces lydicus spp. tanzanius, and was considered for use as a growth promoting antibiotic in agriculture.",
  },
  {
    accession: "ARO:3001327",
    name: "MdtK",
    description:
      "A multidrug and toxic compound extrusions (MATE) transporter conferring resistance to norfloxacin, doxorubicin and acriflavine.",
  },
  {
    accession: "ARO:3001328",
    name: "Escherichia coli mdfA",
    description:
      "Multidrug efflux pump in E. coli. This multidrug efflux system was originally identified as the Cmr/CmlA chloramphenicol exporter.",
  },
  {
    accession: "ARO:3001329",
    name: "mdtG",
    description:
      "The MdtG protein, also named YceE, appears to be a member of the major facilitator superfamily of transporters, and it has been reported, when overexpressed, to increase fosfomycin and deoxycholate resistances. mdtG is a member of the marA-soxS-rob regulon.",
  },
  {
    accession: "ARO:3001331",
    name: "SHV-54",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001332",
    name: "SHV-58",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001333",
    name: "SHV-68",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001334",
    name: "SHV-87",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001335",
    name: "SHV-88",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001336",
    name: "SHV-98",
    description: "SHV-98 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001337",
    name: "SHV-99",
    description: "SHV-99 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001338",
    name: "SHV-100",
    description:
      "SHV-100 is a beta-lactamase found in Klebsiella pneumoniae. It differs from SHV-1 by a 35SESQLSGRVGMIE36 insertion.",
  },
  {
    accession: "ARO:3001339",
    name: "SHV-118",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001340",
    name: "SHV-119",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001341",
    name: "SHV-132",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001342",
    name: "SHV-138",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001343",
    name: "SHV-139",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001344",
    name: "SHV-143",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3001345",
    name: "SHV-144",
    description: "SHV-144 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001346",
    name: "SHV-146",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001347",
    name: "SHV-164",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3001348",
    name: "SHV-166",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001349",
    name: "SHV-174",
    description: "SHV-174 is a beta-lactamase found in Enterobacteriaceae",
  },
  {
    accession: "ARO:3001350",
    name: "SHV-180",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001351",
    name: "SHV-183",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001352",
    name: "SHV-168",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3001353",
    name: "SHV-169",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3001354",
    name: "SHV-170",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3001355",
    name: "SHV-171",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3001356",
    name: "SHV-172",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3001357",
    name: "SHV-173",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3001358",
    name: "SHV-175",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3001359",
    name: "SHV-176",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3001360",
    name: "SHV-177",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3001361",
    name: "SHV-178",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3001362",
    name: "SHV-179",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3001363",
    name: "SHV-181",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3001364",
    name: "SHV-182",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3001365",
    name: "SHV-184",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3001366",
    name: "TEM-100",
    description: "From the Lahey list of beta-lactamases. Not yet released",
  },
  {
    accession: "ARO:3001367",
    name: "TEM-119",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001368",
    name: "TEM-140",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001369",
    name: "TEM-153",
    description: "TEM-153 is a beta-lactamase found in Enterobacteriaceae.",
  },
  {
    accession: "ARO:3001370",
    name: "TEM-175",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001371",
    name: "TEM-179",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001372",
    name: "TEM-180",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001373",
    name: "TEM-182",
    description:
      "TEM-182 is a beta-lactamase found in clinical isolates of H. parainfluenzae",
  },
  {
    accession: "ARO:3001374",
    name: "TEM-185",
    description: "TEM-185 is a beta-lactamase found in E. coli",
  },
  {
    accession: "ARO:3001375",
    name: "TEM-184",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001376",
    name: "TEM-196",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001377",
    name: "TEM-200",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001378",
    name: "TEM-201",
    description: "TEM-201 is a beta-lactamase found in Enterobacteriaceae.",
  },
  {
    accession: "ARO:3001379",
    name: "TEM-202",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001380",
    name: "TEM-203",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001381",
    name: "TEM-204",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001382",
    name: "TEM-205",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001383",
    name: "TEM-206",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001384",
    name: "TEM-207",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001385",
    name: "TEM-208",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001386",
    name: "TEM-209",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001387",
    name: "TEM-210",
    description: "TEM-210",
  },
  {
    accession: "ARO:3001388",
    name: "TEM-211",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001389",
    name: "TEM-212",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001390",
    name: "TEM-213",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001391",
    name: "TEM-214",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001392",
    name: "TEM-215",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001393",
    name: "TEM-216",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001394",
    name: "TEM-217",
    description:
      "From the Lahey list of beta-lactamases. Not yet released.\nIsolated from Enterobacter cloacae",
  },
  {
    accession: "ARO:3001395",
    name: "TEM-218",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3001396",
    name: "OXA-1",
    description: "OXA-1 is a beta-lactamase found in E. coli",
  },
  {
    accession: "ARO:3001397",
    name: "OXA-2",
    description:
      "OXA-2 is a beta-lactamase found in the Enterobacteriaceae family.",
  },
  {
    accession: "ARO:3001398",
    name: "OXA-3",
    description: "OXA-3 is a beta-lactamase found in P. aeruginosa",
  },
  {
    accession: "ARO:3001399",
    name: "OXA-4",
    description:
      "OXA-4 is a beta-lactamase found in Enterobacteriaceae and P. aeruginosa",
  },
  {
    accession: "ARO:3001400",
    name: "OXA-5",
    description:
      "OXA-5 is a beta-lactamase found in P. aeruginosa and Enterobacteriaceae.",
  },
  {
    accession: "ARO:3001401",
    name: "OXA-6",
    description:
      "OXA-5 is a beta-lactamase found in P. aeruginosa and Enterobacteriaceae.",
  },
  {
    accession: "ARO:3001402",
    name: "OXA-7",
    description:
      "OXA-7 is a beta-lactamase found in P. aeruginosa and Enterobacteriaceae.",
  },
  {
    accession: "ARO:3001403",
    name: "OXA-8",
    description: "OXA-8 is an uncharacterized beta-lactamase.",
  },
  {
    accession: "ARO:3001404",
    name: "OXA-9",
    description: "OXA-9 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001405",
    name: "OXA-10",
    description:
      "OXA-10 is a beta-lactamase found in Acinetobacter baumannii and P. aeruginosa.",
  },
  {
    accession: "ARO:3001406",
    name: "OXA-11",
    description: "OXA-11 is a beta-lactamase found in P. aeruginosa.",
  },
  {
    accession: "ARO:3001407",
    name: "OXA-12",
    description: "OXA-12 is a beta-lactamase found in Aeromonas jandaei",
  },
  {
    accession: "ARO:3001408",
    name: "OXA-13",
    description: "OXA-13 is a beta-lactamase found in P. aeruginosa.",
  },
  {
    accession: "ARO:3001409",
    name: "OXA-14",
    description: "OXA-14 is a beta-lactamase found in P. aeruginosa.",
  },
  {
    accession: "ARO:3001410",
    name: "OXA-15",
    description: "OXA-15 is a beta-lactamase found in P. aeruginosa.",
  },
  {
    accession: "ARO:3001411",
    name: "OXA-16",
    description: "OXA-16 is a beta-lactamase found in P. aeruginosa.",
  },
  {
    accession: "ARO:3001412",
    name: "OXA-17",
    description:
      "OXA-17 is a beta-lactamase found in P. aeruginosa and Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3001413",
    name: "OXA-18",
    description:
      "OXA-18 is a beta-lactamase found in P. aeruginosa and Rickettsia.",
  },
  {
    accession: "ARO:3001414",
    name: "OXA-19",
    description: "OXA-19 is a beta-lactamase found in P. aeruginosa.",
  },
  {
    accession: "ARO:3001415",
    name: "OXA-20",
    description: "OXA-20 is a beta-lactamase found in P. aeruginosa",
  },
  {
    accession: "ARO:3001416",
    name: "OXA-21",
    description: "OXA-21 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001417",
    name: "OXA-22",
    description: "OXA-22 is a beta-lactamase found in Ralstonia pickettii",
  },
  {
    accession: "ARO:3001418",
    name: "OXA-23",
    description: "OXA-23 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001419",
    name: "OXA-24",
    description:
      "OXA-24 is a beta-lactamase found in A. baumannii and P. aeruginosa.",
  },
  {
    accession: "ARO:3001420",
    name: "OXA-25",
    description: "OXA-25 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001421",
    name: "OXA-26",
    description: "OXA-26 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001422",
    name: "OXA-27",
    description: "OXA-27 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001423",
    name: "OXA-28",
    description: "OXA-28 is a beta-lactamase found in P. aeruginosa",
  },
  {
    accession: "ARO:3001424",
    name: "OXA-29",
    description: "OXA-29 is a beta-lactamase found in Legionella gormanii",
  },
  {
    accession: "ARO:3001425",
    name: "OXA-31",
    description: "OXA-31 is a beta-lactamase found in P. aeruginosa",
  },
  {
    accession: "ARO:3001426",
    name: "OXA-32",
    description: "OXA-32 is a beta-lactamase found in P. aeruginosa",
  },
  {
    accession: "ARO:3001427",
    name: "OXA-33",
    description:
      "OXA-33 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001428",
    name: "OXA-34",
    description:
      "OXA-34 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001429",
    name: "OXA-35",
    description: "OXA-35 is a beta-lactamase found in P. aeruginosa",
  },
  {
    accession: "ARO:3001430",
    name: "OXA-36",
    description:
      "OXA-36 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001431",
    name: "OXA-37",
    description: "OXA-37 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001432",
    name: "OXA-38",
    description:
      "OXA-38 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001433",
    name: "OXA-39",
    description:
      "OXA-39 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001434",
    name: "OXA-41",
    description:
      "OXA-41 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001435",
    name: "OXA-44",
    description:
      "OXA-44 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001436",
    name: "OXA-52",
    description:
      "OXA-52 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001437",
    name: "OXA-67",
    description:
      "OXA-67 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001438",
    name: "OXA-100",
    description:
      "OXA-100 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001439",
    name: "OXA-101",
    description:
      "OXA-101 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001440",
    name: "OXA-115",
    description:
      "OXA-115 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001441",
    name: "OXA-116",
    description:
      "OXA-116 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001442",
    name: "OXA-117",
    description:
      "OXA-117 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001443",
    name: "OXA-121",
    description:
      "OXA-121 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001444",
    name: "OXA-122",
    description:
      "OXA-122 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001445",
    name: "OXA-123",
    description:
      "OXA-123 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001446",
    name: "OXA-124",
    description:
      "OXA-124 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001447",
    name: "OXA-125",
    description:
      "OXA-125 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001448",
    name: "OXA-126",
    description:
      "OXA-126 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001449",
    name: "OXA-127",
    description:
      "OXA-127 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001450",
    name: "OXA-135",
    description:
      "OXA-135 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001451",
    name: "OXA-139",
    description:
      "OXA-139 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001452",
    name: "OXA-140",
    description:
      "OXA-140 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001453",
    name: "OXA-148",
    description:
      "OXA-148 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001454",
    name: "OXA-149",
    description:
      "OXA-149 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001455",
    name: "OXA-150",
    description:
      "OXA-150 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001456",
    name: "OXA-151",
    description:
      "OXA-151 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001457",
    name: "OXA-152",
    description:
      "OXA-152 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001458",
    name: "OXA-153",
    description:
      "OXA-153 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001459",
    name: "OXA-154",
    description:
      "OXA-154 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001460",
    name: "OXA-155",
    description:
      "OXA-155 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001461",
    name: "OXA-156",
    description:
      "OXA-156 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001462",
    name: "OXA-157",
    description:
      "OXA-157 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001463",
    name: "OXA-158",
    description:
      "OXA-158 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001464",
    name: "OXA-159",
    description:
      "OXA-159 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001465",
    name: "OXA-165",
    description:
      "OXA-165 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001466",
    name: "OXA-166",
    description:
      "OXA-166 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001467",
    name: "OXA-167",
    description:
      "OXA-167 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001468",
    name: "OXA-168",
    description:
      "OXA-168 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001469",
    name: "OXA-169",
    description:
      "OXA-169 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001470",
    name: "OXA-170",
    description:
      "OXA-170 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001471",
    name: "OXA-171",
    description:
      "OXA-171 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001472",
    name: "OXA-178",
    description:
      "OXA-178 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001473",
    name: "OXA-179",
    description:
      "OXA-179 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001474",
    name: "OXA-180",
    description:
      "OXA-180 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001475",
    name: "OXA-183",
    description:
      "OXA-183 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001476",
    name: "OXA-184",
    description:
      "OXA-184 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001477",
    name: "OXA-185",
    description:
      "OXA-185 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001478",
    name: "OXA-193",
    description:
      "OXA-193 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001479",
    name: "OXA-194",
    description:
      "OXA-194 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001480",
    name: "OXA-195",
    description:
      "OXA-195 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001481",
    name: "OXA-196",
    description:
      "OXA-196 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001482",
    name: "OXA-197",
    description:
      "OXA-197 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001483",
    name: "OXA-205",
    description:
      "OXA-205 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001484",
    name: "OXA-206",
    description:
      "OXA-206 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001485",
    name: "OXA-207",
    description:
      "OXA-207 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001486",
    name: "OXA-208",
    description:
      "OXA-208 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001487",
    name: "OXA-210",
    description:
      "OXA-210 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001488",
    name: "OXA-216",
    description:
      "OXA-216 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001489",
    name: "OXA-217",
    description:
      "OXA-217 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001490",
    name: "OXA-220",
    description:
      "OXA-220 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001491",
    name: "OXA-221",
    description:
      "OXA-221 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001492",
    name: "OXA-222",
    description:
      "OXA-222 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001493",
    name: "OXA-225",
    description:
      "OXA-225 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001494",
    name: "OXA-227",
    description:
      "OXA-227 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001495",
    name: "OXA-233",
    description:
      "OXA-233 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001496",
    name: "OXA-234",
    description: "OXA-234 is a beta-lactamase found in Acinetobacter baumannii",
  },
  {
    accession: "ARO:3001497",
    name: "OXA-238",
    description:
      "OXA-238 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001498",
    name: "OXA-239",
    description:
      "OXA-239 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001499",
    name: "OXA-240",
    description:
      "OXA-240 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001500",
    name: "OXA-251",
    description:
      "OXA-251 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001501",
    name: "OXA-252",
    description: "OXA-252 is a beta-lactamase found in Shewanella spp.",
  },
  {
    accession: "ARO:3001502",
    name: "OXA-256",
    description:
      "OXA-256 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001503",
    name: "OXA-258",
    description:
      "OXA-258 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001504",
    name: "OXA-309",
    description:
      "OXA-309 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001505",
    name: "OXA-310",
    description:
      "OXA-310 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001506",
    name: "OXA-311",
    description:
      "OXA-311 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001507",
    name: "OXA-318",
    description:
      "OXA-318 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001508",
    name: "OXA-319",
    description:
      "OXA-319 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001509",
    name: "OXA-321",
    description:
      "OXA-321 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001510",
    name: "OXA-322",
    description:
      "OXA-322 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001511",
    name: "OXA-323",
    description:
      "OXA-323 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001512",
    name: "OXA-324",
    description:
      "OXA-324 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001513",
    name: "OXA-325",
    description:
      "OXA-325 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001514",
    name: "OXA-326",
    description:
      "OXA-326 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001515",
    name: "OXA-327",
    description:
      "OXA-327 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001516",
    name: "OXA-328",
    description:
      "OXA-328 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001517",
    name: "OXA-329",
    description:
      "OXA-329 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001518",
    name: "OXA-330",
    description:
      "OXA-330 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001519",
    name: "OXA-331",
    description:
      "OXA-331 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001520",
    name: "OXA-332",
    description:
      "OXA-332 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001521",
    name: "OXA-333",
    description:
      "OXA-333 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001522",
    name: "OXA-334",
    description:
      "OXA-334 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001523",
    name: "OXA-335",
    description:
      "OXA-335 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001524",
    name: "OXA-336",
    description:
      "OXA-336 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001525",
    name: "OXA-337",
    description:
      "OXA-337 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001526",
    name: "OXA-338",
    description:
      "OXA-338 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001527",
    name: "OXA-339",
    description:
      "OXA-339 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001528",
    name: "OXA-340",
    description:
      "OXA-340 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001529",
    name: "OXA-341",
    description:
      "OXA-341 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001530",
    name: "OXA-342",
    description:
      "OXA-342 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001531",
    name: "OXA-343",
    description:
      "OXA-343 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001532",
    name: "OXA-344",
    description:
      "OXA-344 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001533",
    name: "OXA-345",
    description:
      "OXA-345 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001534",
    name: "OXA-346",
    description:
      "OXA-346 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001535",
    name: "OXA-348",
    description:
      "OXA-348 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001536",
    name: "OXA-349",
    description:
      "OXA-349 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001537",
    name: "OXA-350",
    description:
      "OXA-350 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001538",
    name: "OXA-351",
    description:
      "OXA-351 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001539",
    name: "OXA-352",
    description:
      "OXA-352 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001540",
    name: "OXA-353",
    description:
      "OXA-353 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001541",
    name: "OXA-354",
    description:
      "OXA-354 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001542",
    name: "OXA-355",
    description:
      "OXA-355 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001543",
    name: "OXA-356",
    description:
      "OXA-356 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001544",
    name: "OXA-357",
    description:
      "OXA-357 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001545",
    name: "OXA-358",
    description:
      "OXA-358 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001546",
    name: "OXA-359",
    description:
      "OXA-359 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001547",
    name: "OXA-360",
    description:
      "OXA-360 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001548",
    name: "OXA-361",
    description:
      "OXA-361 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001549",
    name: "OXA-362",
    description:
      "OXA-362 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001550",
    name: "OXA-363",
    description:
      "OXA-363 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001551",
    name: "OXA-364",
    description:
      "OXA-364 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001552",
    name: "OXA-365",
    description:
      "OXA-365 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001553",
    name: "OXA-366",
    description:
      "OXA-366 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001554",
    name: "OXA-367",
    description:
      "OXA-367 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001555",
    name: "OXA-368",
    description:
      "OXA-368 is a beta-lactamase found in Aeromonas sobria. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001557",
    name: "OXA-371",
    description:
      "OXA-371 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001558",
    name: "OXA-372",
    description:
      "OXA-372 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001559",
    name: "OXA-373",
    description:
      "OXA-373 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001560",
    name: "OXA-374",
    description:
      "OXA-374 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001561",
    name: "OXA-375",
    description:
      "OXA-375 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001562",
    name: "OXA-376",
    description:
      "OXA-376 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001563",
    name: "OXA-377",
    description:
      "OXA-377 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001564",
    name: "OXA-378",
    description:
      "OXA-378 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001565",
    name: "OXA-379",
    description:
      "OXA-379 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001566",
    name: "OXA-380",
    description:
      "OXA-380 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001567",
    name: "OXA-381",
    description:
      "OXA-381 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001568",
    name: "OXA-382",
    description:
      "OXA-382 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001569",
    name: "OXA-383",
    description:
      "OXA-383 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001570",
    name: "OXA-384",
    description:
      "OXA-384 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001571",
    name: "OXA-385",
    description:
      "OXA-385 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001572",
    name: "OXA-386",
    description:
      "OXA-386 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001573",
    name: "OXA-387",
    description:
      "OXA-387 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001574",
    name: "OXA-388",
    description:
      "OXA-388 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001575",
    name: "OXA-389",
    description:
      "OXA-389 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001576",
    name: "OXA-390",
    description:
      "OXA-390 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001577",
    name: "OXA-391",
    description:
      "OXA-391 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001578",
    name: "OXA-392",
    description:
      "OXA-392 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001579",
    name: "OXA-393",
    description:
      "OXA-393 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001580",
    name: "OXA-394",
    description:
      "OXA-394 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001581",
    name: "OXA-395",
    description:
      "OXA-395 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001582",
    name: "OXA-396",
    description:
      "OXA-396 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001583",
    name: "OXA-397",
    description:
      "OXA-397 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001584",
    name: "OXA-398",
    description:
      "OXA-398 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001585",
    name: "OXA-399",
    description:
      "OXA-399 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001586",
    name: "OXA-400",
    description:
      "OXA-400 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001587",
    name: "OXA-401",
    description:
      "OXA-401 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001588",
    name: "OXA-402",
    description:
      "OXA-402 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001589",
    name: "OXA-403",
    description:
      "OXA-403 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001595",
    name: "OXA-404",
    description:
      "OXA-404 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001596",
    name: "OXA-405",
    description:
      "OXA-405 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001597",
    name: "OXA-406",
    description:
      "OXA-406 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001598",
    name: "OXA-407",
    description:
      "OXA-407 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001599",
    name: "OXA-408",
    description:
      "OXA-408 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001600",
    name: "OXA-409",
    description:
      "OXA-409 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001601",
    name: "OXA-410",
    description:
      "OXA-410 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001602",
    name: "OXA-411",
    description:
      "OXA-411 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001603",
    name: "OXA-412",
    description:
      "OXA-412 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001604",
    name: "OXA-413",
    description:
      "OXA-413 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001605",
    name: "OXA-414",
    description:
      "OXA-414 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001606",
    name: "OXA-415",
    description:
      "OXA-415 is a beta-lactamase. From the Lahey list of OXA beta-lactamases",
  },
  {
    accession: "ARO:3001607",
    name: "OXA-416",
    description:
      "OXA-416 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001608",
    name: "OXA-417",
    description:
      "OXA-417 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3001609",
    name: "OXA-114a",
    description:
      "OXA-114a is a beta-lactamase found in Achromobacter xylosoxidans.",
  },
  {
    accession: "ARO:3001610",
    name: "OXA-243",
    description:
      "OXA-243 is a beta-lactamase found in Achromobacter xylosoxidans.",
  },
  {
    accession: "ARO:3001611",
    name: "OXA-58",
    description: "OXA-58 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001612",
    name: "OXA-51",
    description: "OXA-51 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001613",
    name: "OXA-64",
    description: "OXA-64 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001614",
    name: "OXA-65",
    description: "OXA-65 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001615",
    name: "OXA-66",
    description: "OXA-66 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001616",
    name: "OXA-68",
    description: "OXA-68 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001617",
    name: "OXA-69",
    description: "OXA-69 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001618",
    name: "OXA-70",
    description: "OXA-70 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001619",
    name: "OXA-71",
    description: "OXA-71 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001620",
    name: "OXA-75",
    description: "OXA-75 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001621",
    name: "OXA-76",
    description: "OXA-76 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001622",
    name: "OXA-77",
    description: "OXA-77 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001623",
    name: "OXA-83",
    description: "OXA-83 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001624",
    name: "OXA-84",
    description: "OXA-84 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001625",
    name: "OXA-89",
    description: "OXA-89 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001626",
    name: "OXA-88",
    description: "OXA-88 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001627",
    name: "OXA-91",
    description: "OXA-91 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001628",
    name: "OXA-93",
    description: "OXA-93 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001629",
    name: "OXA-94",
    description: "OXA-94 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001630",
    name: "OXA-95",
    description: "OXA-95 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001631",
    name: "OXA-96",
    description: "OXA-96 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001632",
    name: "OXA-92",
    description: "OXA-92 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001633",
    name: "OXA-79",
    description: "OXA-79 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001634",
    name: "OXA-80",
    description: "OXA-80 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001635",
    name: "OXA-82",
    description: "OXA-82 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001636",
    name: "OXA-104",
    description: "OXA-104 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001637",
    name: "OXA-106",
    description: "OXA-106 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001638",
    name: "OXA-107",
    description: "OXA-107 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001639",
    name: "OXA-108",
    description: "OXA-108 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001640",
    name: "OXA-109",
    description: "OXA-109 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001641",
    name: "OXA-110",
    description: "OXA-110 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001642",
    name: "OXA-111",
    description: "OXA-111 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001643",
    name: "OXA-112",
    description: "OXA-112 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001644",
    name: "OXA-113",
    description: "OXA-113 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001645",
    name: "OXA-78",
    description: "OXA-78 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001646",
    name: "OXA-98",
    description: "OXA-98 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001647",
    name: "OXA-97",
    description: "OXA-97 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001648",
    name: "OXA-128",
    description: "OXA-128 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001649",
    name: "OXA-99",
    description: "OXA-99 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001650",
    name: "OXA-90",
    description: "OXA-90 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001651",
    name: "OXA-130",
    description: "OXA-130 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001652",
    name: "OXA-131",
    description: "OXA-131 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001653",
    name: "OXA-132",
    description: "OXA-132 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001654",
    name: "OXA-143",
    description: "OXA-143 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001655",
    name: "OXA-138",
    description: "OXA-138 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001656",
    name: "OXA-172",
    description: "OXA-172 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001657",
    name: "OXA-173",
    description: "OXA-173 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001658",
    name: "OXA-174",
    description: "OXA-174 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001659",
    name: "OXA-175",
    description: "OXA-175 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001660",
    name: "OXA-176",
    description: "OXA-176 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001661",
    name: "OXA-177",
    description: "OXA-177 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001662",
    name: "OXA-164",
    description: "OXA-164 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001663",
    name: "OXA-160",
    description: "OXA-160 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001664",
    name: "OXA-203",
    description: "OXA-203 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001665",
    name: "OXA-200",
    description: "OXA-200 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001666",
    name: "OXA-201",
    description: "OXA-201 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001667",
    name: "OXA-202",
    description: "OXA-202 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001668",
    name: "OXA-219",
    description: "OXA-219 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001669",
    name: "OXA-223",
    description: "OXA-223 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001670",
    name: "OXA-231",
    description: "OXA-231 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001671",
    name: "OXA-49",
    description: "OXA-49 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001672",
    name: "OXA-120",
    description: "OXA-120 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001673",
    name: "OXA-248",
    description: "OXA-248 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001674",
    name: "OXA-249",
    description: "OXA-249 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001675",
    name: "OXA-250",
    description: "OXA-250 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001676",
    name: "OXA-235",
    description: "OXA-235 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001677",
    name: "OXA-236",
    description: "OXA-236 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001678",
    name: "OXA-237",
    description: "OXA-237 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001679",
    name: "OXA-241",
    description: "OXA-241 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001680",
    name: "OXA-242",
    description: "OXA-242 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001681",
    name: "OXA-278",
    description: "OXA-278 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001682",
    name: "OXA-254",
    description: "OXA-254 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001683",
    name: "OXA-312",
    description: "OXA-312 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001684",
    name: "OXA-313",
    description: "OXA-313 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001685",
    name: "OXA-314",
    description: "OXA-314 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001686",
    name: "OXA-315",
    description: "OXA-315 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001687",
    name: "OXA-316",
    description: "OXA-316 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001688",
    name: "OXA-317",
    description: "OXA-317 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001689",
    name: "OXA-253",
    description: "OXA-253 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001690",
    name: "OXA-255",
    description: "OXA-255 is a beta-lactamase found in A. baumannii",
  },
  {
    accession: "ARO:3001691",
    name: "OXA-228",
    description: "OXA-228 is a beta-lactamase found in A. bereziniae",
  },
  {
    accession: "ARO:3001692",
    name: "OXA-229",
    description: "OXA-229 is a beta-lactamase found in A. bereziniae",
  },
  {
    accession: "ARO:3001693",
    name: "OXA-230",
    description: "OXA-230 is a beta-lactamase found in A. bereziniae",
  },
  {
    accession: "ARO:3001694",
    name: "OXA-257",
    description: "OXA-257 is a beta-lactamase found in A. bereziniae",
  },
  {
    accession: "ARO:3001695",
    name: "OXA-134",
    description: "OXA-134 is a beta-lactamase found in A. lwoffii",
  },
  {
    accession: "ARO:3001696",
    name: "OXA-186",
    description: "OXA-186 is a beta-lactamase found in A. lwoffii",
  },
  {
    accession: "ARO:3001697",
    name: "OXA-187",
    description: "OXA-187 is a beta-lactamase found in A. lwoffii",
  },
  {
    accession: "ARO:3001698",
    name: "OXA-188",
    description: "OXA-188 is a beta-lactamase found in A. lwoffii",
  },
  {
    accession: "ARO:3001699",
    name: "OXA-189",
    description: "OXA-189 is a beta-lactamase found in A. lwoffii",
  },
  {
    accession: "ARO:3001700",
    name: "OXA-190",
    description: "OXA-190 is a beta-lactamase found in A. lwoffii",
  },
  {
    accession: "ARO:3001701",
    name: "OXA-191",
    description: "OXA-191 is a beta-lactamase found in A. lwoffii",
  },
  {
    accession: "ARO:3001702",
    name: "OXA-133",
    description: "OXA-133 is a beta-lactamase found in A. radioresistens",
  },
  {
    accession: "ARO:3001703",
    name: "OXA-86",
    description: "OXA-86 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001704",
    name: "OXA-87",
    description: "OXA-87 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001705",
    name: "OXA-72",
    description: "OXA-72 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001706",
    name: "OXA-102",
    description: "OXA-102 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001707",
    name: "OXA-103",
    description: "OXA-103 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001708",
    name: "OXA-105",
    description: "OXA-105 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001709",
    name: "OXA-182",
    description: "OXA-182 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001710",
    name: "OXA-211",
    description: "OXA-211 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001711",
    name: "OXA-212",
    description: "OXA-212 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001712",
    name: "OXA-213",
    description: "OXA-213 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001713",
    name: "OXA-214",
    description: "OXA-214 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001714",
    name: "OXA-215",
    description: "OXA-215 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001715",
    name: "OXA-259",
    description: "OXA-259 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001716",
    name: "OXA-260",
    description: "OXA-260 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001717",
    name: "OXA-261",
    description: "OXA-261 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001718",
    name: "OXA-262",
    description: "OXA-262 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001719",
    name: "OXA-263",
    description: "OXA-263 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001720",
    name: "OXA-264",
    description: "OXA-264 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001721",
    name: "OXA-265",
    description: "OXA-265 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001722",
    name: "OXA-266",
    description: "OXA-266 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001723",
    name: "OXA-267",
    description: "OXA-267 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001724",
    name: "OXA-268",
    description: "OXA-268 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001725",
    name: "OXA-269",
    description: "OXA-269 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001726",
    name: "OXA-270",
    description: "OXA-270 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001727",
    name: "OXA-271",
    description: "OXA-271 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001728",
    name: "OXA-272",
    description: "OXA-272 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001729",
    name: "OXA-273",
    description: "OXA-273 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001730",
    name: "OXA-274",
    description: "OXA-274 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001731",
    name: "OXA-275",
    description: "OXA-275 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001732",
    name: "OXA-276",
    description: "OXA-276 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001733",
    name: "OXA-277",
    description: "OXA-277 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001734",
    name: "OXA-279",
    description: "OXA-279 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001735",
    name: "OXA-280",
    description: "OXA-280 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001736",
    name: "OXA-281",
    description: "OXA-281 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001737",
    name: "OXA-282",
    description: "OXA-282 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001738",
    name: "OXA-283",
    description: "OXA-283 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001739",
    name: "OXA-284",
    description: "OXA-284 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001740",
    name: "OXA-285",
    description: "OXA-285 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001741",
    name: "OXA-286",
    description: "OXA-286 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001742",
    name: "OXA-287",
    description: "OXA-287 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001743",
    name: "OXA-288",
    description: "OXA-288 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001744",
    name: "OXA-289",
    description: "OXA-289 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001745",
    name: "OXA-290",
    description: "OXA-290 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001746",
    name: "OXA-291",
    description: "OXA-291 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001747",
    name: "OXA-292",
    description: "OXA-292 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001748",
    name: "OXA-293",
    description: "OXA-293 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001749",
    name: "OXA-294",
    description: "OXA-294 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001750",
    name: "OXA-295",
    description: "OXA-295 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001751",
    name: "OXA-296",
    description: "OXA-296 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001752",
    name: "OXA-297",
    description: "OXA-297 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001753",
    name: "OXA-298",
    description: "OXA-298 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001754",
    name: "OXA-299",
    description: "OXA-299 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001755",
    name: "OXA-300",
    description: "OXA-300 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001756",
    name: "OXA-301",
    description: "OXA-301 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001757",
    name: "OXA-302",
    description: "OXA-302 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001758",
    name: "OXA-303",
    description: "OXA-303 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001759",
    name: "OXA-304",
    description: "OXA-304 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001760",
    name: "OXA-305",
    description: "OXA-305 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001761",
    name: "OXA-306",
    description: "OXA-306 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001762",
    name: "OXA-307",
    description: "OXA-307 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001763",
    name: "OXA-308",
    description: "OXA-308 is a beta-lactamase found in Acinetobacter spp.",
  },
  {
    accession: "ARO:3001764",
    name: "OXA-63",
    description: "OXA-63 is a beta-lactamase found in Brachyspira spp.",
  },
  {
    accession: "ARO:3001765",
    name: "OXA-136",
    description: "OXA-136 is a beta-lactamase found in Brachyspira spp.",
  },
  {
    accession: "ARO:3001766",
    name: "OXA-192",
    description: "OXA-192 is a beta-lactamase found in Brachyspira spp.",
  },
  {
    accession: "ARO:3001767",
    name: "OXA-137",
    description: "OXA-137 is a beta-lactamase found in Brachyspira spp.",
  },
  {
    accession: "ARO:3001768",
    name: "OXA-118",
    description: "OXA-118 is a beta-lactamase found in Burkholderia cepacia",
  },
  {
    accession: "ARO:3001769",
    name: "OXA-42",
    description:
      "OXA-42 is a beta-lactamase found in Burkholderia pseudomallei",
  },
  {
    accession: "ARO:3001770",
    name: "OXA-43",
    description:
      "OXA-43 is a beta-lactamase found in Burkholderia pseudomallei",
  },
  {
    accession: "ARO:3001771",
    name: "OXA-57",
    description:
      "OXA-57 is a beta-lactamase found in Burkholderia pseudomallei",
  },
  {
    accession: "ARO:3001772",
    name: "OXA-59",
    description:
      "OXA-59 is a beta-lactamase found in Burkholderia pseudomallei",
  },
  {
    accession: "ARO:3001773",
    name: "OXA-61",
    description: "OXA-61 is a beta-lactamase found in Campylobacter jejuni",
  },
  {
    accession: "ARO:3001774",
    name: "OXA-370",
    description: "OXA-370 is a beta-lactamase found in Enterobacter hormaechei",
  },
  {
    accession: "ARO:3001775",
    name: "OXA-119",
    description: "OXA-119 is a beta-lactamase found in Enterobacteriaceae",
  },
  {
    accession: "ARO:3001776",
    name: "OXA-162",
    description: "OXA-162 is a beta-lactamase found in Enterobacteriaceae",
  },
  {
    accession: "ARO:3001777",
    name: "OXA-347",
    description: "OXA-347 is a beta-lactamase found in Enterobacteriaceae",
  },
  {
    accession: "ARO:3001778",
    name: "OXA-232",
    description: "OXA-232 is a beta-lactamase found in Enterobacteriaceae",
  },
  {
    accession: "ARO:3001779",
    name: "OXA-146",
    description: "OXA-146 is a beta-lactamase found in Enterobacteriaceae",
  },
  {
    accession: "ARO:3001780",
    name: "OXA-85",
    description: "OXA-85 is a beta-lactamase found in Fusobacterium nucleatum",
  },
  {
    accession: "ARO:3001781",
    name: "OXA-47",
    description: "OXA-47 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001782",
    name: "OXA-48",
    description: "OXA-48 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001783",
    name: "OXA-163",
    description: "OXA-163 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001784",
    name: "OXA-181",
    description: "OXA-181 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001785",
    name: "OXA-73",
    description: "OXA-73 is a beta-lactamase found in Klebsiella pneumonia",
  },
  {
    accession: "ARO:3001786",
    name: "OXA-244",
    description: "OXA-244 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001787",
    name: "OXA-245",
    description: "OXA-245 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001788",
    name: "OXA-204",
    description: "OXA-204 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001791",
    name: "OXA-247",
    description: "OXA-247 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001792",
    name: "OXA-62",
    description: "OXA-62 is a beta-lactamase found in Pandoraea pnomenusa",
  },
  {
    accession: "ARO:3001793",
    name: "OXA-320",
    description: "OXA-320 is a beta-lactamase found in Proteus mirabilis",
  },
  {
    accession: "ARO:3001794",
    name: "OXA-45",
    description: "OXA-45 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3001795",
    name: "OXA-56",
    description: "OXA-56 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3001796",
    name: "OXA-50",
    description:
      "OXA-50 is a beta-lactamase found in Pseudomonas aeruginosa. It confers decreased susceptibility to ampicillin and ticarcillin and, interestingly, to moxalactam and meropenem in P. aeruginosa but not in E. coli. Also confers resistance to piperacillin-tazobactam and cephalotin.",
  },
  {
    accession: "ARO:3001797",
    name: "OXA-46",
    description: "OXA-46 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3001798",
    name: "OXA-74",
    description: "OXA-74 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3001799",
    name: "OXA-161",
    description: "OXA-161 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3001800",
    name: "OXA-226",
    description: "OXA-226 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3001801",
    name: "OXA-147",
    description: "OXA-147 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3001802",
    name: "OXA-141",
    description: "OXA-141 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3001803",
    name: "OXA-142",
    description: "OXA-142 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3001804",
    name: "OXA-145",
    description: "OXA-145 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3001805",
    name: "OXA-198",
    description: "OXA-198 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3001806",
    name: "OXA-224",
    description: "OXA-224 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3001807",
    name: "OXA-246",
    description: "OXA-246 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3001808",
    name: "OXA-60",
    description: "OXA-60 is a beta-lactamase found in Ralstonia pickettii",
  },
  {
    accession: "ARO:3001809",
    name: "OXA-209",
    description:
      "OXA-209 is a beta-lactamase found in Riemerella anatipestifer",
  },
  {
    accession: "ARO:3001810",
    name: "OXA-53",
    description: "OXA-53 is a beta-lactamase found in Salmonella enterica",
  },
  {
    accession: "ARO:3001811",
    name: "OXA-129",
    description: "OXA-129 is a beta-lactamase found in Salmonella enterica",
  },
  {
    accession: "ARO:3001812",
    name: "OXA-54",
    description: "OXA-54 is a beta-lactamase found in Shewanella spp.",
  },
  {
    accession: "ARO:3001813",
    name: "OXA-55",
    description: "OXA-55 is a beta-lactamase found in Shewanella spp.",
  },
  {
    accession: "ARO:3001814",
    name: "OXA-199",
    description: "OXA-199 is a beta-lactamase found in Shewanella spp.",
  },
  {
    accession: "ARO:3001815",
    name: "ACC-1",
    description: "ACC-1 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001816",
    name: "ACC-2",
    description: "ACC-2 is a beta-lactamase found in Hafnia alvei",
  },
  {
    accession: "ARO:3001817",
    name: "ACC-3",
    description: "ACC-3 is a beta-lactamase found in Hafnia alvei",
  },
  {
    accession: "ARO:3001818",
    name: "ACC-4",
    description: "ACC-4 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001819",
    name: "ACC-5",
    description:
      "ACC-5 is a beta-lactamase. From the Lahey list of ACC beta-lactamases. It is detected among Gram-negative bacteria isolates (Gamma- and Alpha-proteobacteria) from urban water sources.",
  },
  {
    accession: "ARO:3001821",
    name: "ACT-1",
    description: "ACT-1 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001822",
    name: "ACT-2",
    description: "ACT-2 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001823",
    name: "ACT-3",
    description: "ACT-3 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001824",
    name: "ACT-5",
    description: "ACT-5 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001825",
    name: "ACT-6",
    description: "ACT-6 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001826",
    name: "ACT-9",
    description: "ACT-9 is a beta-lactamase found in Pantoea agglomerans",
  },
  {
    accession: "ARO:3001827",
    name: "ACT-16",
    description: "ACT-16 is a beta-lactamase found in Enterobacteriaceae",
  },
  {
    accession: "ARO:3001828",
    name: "ACT-23",
    description: "ACT-23 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001829",
    name: "ACT-4",
    description:
      "ACT-4 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001830",
    name: "ACT-7",
    description:
      "ACT-7 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001831",
    name: "ACT-8",
    description:
      "ACT-8 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001832",
    name: "ACT-10",
    description:
      "ACT-10 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001833",
    name: "ACT-11",
    description:
      "ACT-11 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001834",
    name: "ACT-12",
    description:
      "ACT-12 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001835",
    name: "ACT-13",
    description:
      "ACT-13 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001836",
    name: "ACT-14",
    description:
      "ACT-14 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001837",
    name: "ACT-15",
    description:
      "ACT-15 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001838",
    name: "ACT-17",
    description:
      "ACT-17 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001839",
    name: "ACT-18",
    description:
      "ACT-18 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001840",
    name: "ACT-19",
    description:
      "ACT-19 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001841",
    name: "ACT-20",
    description:
      "ACT-20 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001842",
    name: "ACT-21",
    description:
      "ACT-21 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001843",
    name: "ACT-22",
    description:
      "ACT-22 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001844",
    name: "ACT-24",
    description:
      "ACT-24 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001845",
    name: "ACT-25",
    description:
      "ACT-25 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001847",
    name: "ACT-27",
    description:
      "ACT-27 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001848",
    name: "ACT-28",
    description:
      "ACT-28 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001849",
    name: "ACT-29",
    description:
      "ACT-29 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001850",
    name: "ACT-30",
    description:
      "ACT-30 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001851",
    name: "ACT-31",
    description:
      "ACT-31 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001852",
    name: "ACT-32",
    description:
      "ACT-32 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001853",
    name: "ACT-33",
    description:
      "ACT-33 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001854",
    name: "ACT-34",
    description:
      "ACT-34 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001855",
    name: "ACT-35",
    description:
      "ACT-35 is a beta-lactamase. From the Lahey list of ACT beta-lactamases.",
  },
  {
    accession: "ARO:3001856",
    name: "CFE-1",
    description: "CFE-1 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001857",
    name: "LAT-1",
    description: "LAT-1 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001858",
    name: "IMI-1",
    description: "IMI-1 is a beta-lactamase found in Enterobacter cloacae",
  },
  {
    accession: "ARO:3001859",
    name: "IMI-2",
    description: "IMI-2 is a beta-lactamase found in Enterobacteriaceae",
  },
  {
    accession: "ARO:3001860",
    name: "IMI-3",
    description: "IMI-3 is a beta-lactamase found in Enterobacteriaceae",
  },
  {
    accession: "ARO:3001861",
    name: "IMI-4",
    description:
      "IMI-4 is a beta-lactamase. From the Lahey list of IMI beta-lactamases.",
  },
  {
    accession: "ARO:3001862",
    name: "IMI-5",
    description:
      "IMI-5 is a beta-lactamase. From the Lahey list of IMI beta-lactamases.",
  },
  {
    accession: "ARO:3001863",
    name: "IMI-6",
    description:
      "IMI-6 is a beta-lactamase. From the Lahey list of IMI beta-lactamases.",
  },
  {
    accession: "ARO:3001864",
    name: "CTX-M-1",
    description:
      "CTX-M-1 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001865",
    name: "CTX-M-2",
    description:
      "CTX-M-2 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001866",
    name: "CTX-M-3",
    description: "CTX-M-3 is a beta-lactamase found in Citrobacter freundii",
  },
  {
    accession: "ARO:3001867",
    name: "CTX-M-4",
    description: "CTX-M-4 is a beta-lactamase found in Salmonella typhimurium",
  },
  {
    accession: "ARO:3001868",
    name: "CTX-M-5",
    description: "CTX-M-5 is a beta-lactamase found in Salmonella typhimurium",
  },
  {
    accession: "ARO:3001869",
    name: "CTX-M-6",
    description: "CTX-M-6 is a beta-lactamase found in Salmonella typhimurium",
  },
  {
    accession: "ARO:3001870",
    name: "CTX-M-7",
    description: "CTX-M-7 is a beta-lactamase found in Salmonella typhimurium",
  },
  {
    accession: "ARO:3001871",
    name: "CTX-M-8",
    description:
      "CTX-M-8 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001872",
    name: "CTX-M-9",
    description:
      "CTX-M-9 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001873",
    name: "CTX-M-10",
    description: "CTX-M-10 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001874",
    name: "CTX-M-11",
    description:
      "CTX-M-11 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001875",
    name: "CTX-M-12",
    description: "CTX-M-12 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001876",
    name: "CTX-M-13",
    description:
      "CTX-M-13 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001877",
    name: "CTX-M-14",
    description:
      "CTX-M-14 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001878",
    name: "CTX-M-15",
    description:
      "CTX-M-15 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001879",
    name: "CTX-M-16",
    description: "CTX-M-16 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001880",
    name: "CTX-M-17",
    description: "CTX-M-17 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001881",
    name: "CTX-M-19",
    description: "CTX-M-19 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001882",
    name: "CTX-M-20",
    description:
      "CTX-M-20 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001883",
    name: "CTX-M-21",
    description:
      "CTX-M-21 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001884",
    name: "CTX-M-22",
    description:
      "CTX-M-22 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001885",
    name: "CTX-M-23",
    description:
      "CTX-M-23 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001886",
    name: "CTX-M-24",
    description:
      "CTX-M-24 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001887",
    name: "CTX-M-25",
    description: "CTX-M-25 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001888",
    name: "CTX-M-26",
    description: "CTX-M-26 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001889",
    name: "CTX-M-27",
    description: "CTX-M-27 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001890",
    name: "CTX-M-28",
    description: "CTX-M-28 is a beta-lactamase found in Salmonella enterica",
  },
  {
    accession: "ARO:3001891",
    name: "CTX-M-29",
    description:
      "CTX-M-29 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001892",
    name: "CTX-M-30",
    description: "CTX-M-30 is a beta-lactamase found in Citrobacter freundii",
  },
  {
    accession: "ARO:3001893",
    name: "CTX-M-31",
    description:
      "CTX-M-31 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001894",
    name: "CTX-M-32",
    description: "CTX-M-32 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001895",
    name: "CTX-M-33",
    description: "CTX-M-33 is a beta-lactamase found in Salmonella enterica",
  },
  {
    accession: "ARO:3001896",
    name: "CTX-M-34",
    description:
      "CTX-M-34 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001897",
    name: "CTX-M-35",
    description: "CTX-M-35 is a beta-lactamase found in Citrobacter koseri",
  },
  {
    accession: "ARO:3001898",
    name: "CTX-M-36",
    description: "CTX-M-36 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001899",
    name: "CTX-M-37",
    description: "CTX-M-37 is a beta-lactamase found in Salmonella enterica",
  },
  {
    accession: "ARO:3001900",
    name: "CTX-M-38",
    description: "CTX-M-38 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001901",
    name: "CTX-M-39",
    description: "CTX-M-39 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001902",
    name: "CTX-M-40",
    description: "CTX-M-40 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001903",
    name: "CTX-M-41",
    description: "CTX-M-41 is a beta-lactamase found in Proteus mirabilis",
  },
  {
    accession: "ARO:3001904",
    name: "CTX-M-42",
    description: "CTX-M-42 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001905",
    name: "CTX-M-43",
    description:
      "CTX-M-43 is a beta-lactamase found in Pseudomonas aeruginosa and Acinetobacter spp.",
  },
  {
    accession: "ARO:3001906",
    name: "CTX-M-44",
    description:
      "CTX-M-44 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001907",
    name: "CTX-M-45",
    description: "CTX-M-45 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001908",
    name: "CTX-M-46",
    description:
      "CTX-M-46 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001909",
    name: "CTX-M-47",
    description:
      "CTX-M-47 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001910",
    name: "CTX-M-48",
    description:
      "CTX-M-48 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001911",
    name: "CTX-M-49",
    description:
      "CTX-M-49 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001912",
    name: "CTX-M-50",
    description:
      "CTX-M-50 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001913",
    name: "CTX-M-51",
    description:
      "CTX-M-51 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001914",
    name: "CTX-M-52",
    description: "CTX-M-52 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001915",
    name: "CTX-M-53",
    description: "CTX-M-53 is a beta-lactamase found in Salmonella enterica",
  },
  {
    accession: "ARO:3001916",
    name: "CTX-M-54",
    description: "CTX-M-54 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001917",
    name: "CTX-M-55",
    description:
      "CTX-M-55 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001918",
    name: "CTX-M-56",
    description: "CTX-M-56 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001919",
    name: "CTX-M-58",
    description: "CTX-M-58 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001920",
    name: "CTX-M-59",
    description: "CTX-M-59 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001921",
    name: "CTX-M-60",
    description:
      "CTX-M-60 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001922",
    name: "CTX-M-61",
    description:
      "CTX-M-61 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001923",
    name: "CTX-M-62",
    description: "CTX-M-62 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001924",
    name: "CTX-M-63",
    description: "CTX-M-63 is a beta-lactamase found in Salmonella enterica",
  },
  {
    accession: "ARO:3001925",
    name: "CTX-M-64",
    description: "CTX-M-64 is a beta-lactamase found in Shigella sonnei",
  },
  {
    accession: "ARO:3001926",
    name: "CTX-M-65",
    description: "CTX-M-65 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001927",
    name: "CTX-M-66",
    description: "CTX-M-66 is a beta-lactamase found in Proteus mirabilis",
  },
  {
    accession: "ARO:3001928",
    name: "CTX-M-67",
    description: "CTX-M-67 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001929",
    name: "CTX-M-68",
    description:
      "CTX-M-68 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001930",
    name: "CTX-M-69",
    description: "CTX-M-69 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001931",
    name: "CTX-M-70",
    description:
      "CTX-M-70 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001932",
    name: "CTX-M-71",
    description: "CTX-M-71 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001933",
    name: "CTX-M-72",
    description: "CTX-M-72 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001934",
    name: "CTX-M-73",
    description:
      "CTX-M-73 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001935",
    name: "CTX-M-74",
    description:
      "CTX-M-74 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001936",
    name: "CTX-M-75",
    description:
      "CTX-M-75 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001937",
    name: "CTX-M-76",
    description: "CTX-M-76 is a beta-lactamase found in Kluyvera spp.",
  },
  {
    accession: "ARO:3001938",
    name: "CTX-M-77",
    description: "CTX-M-77 is a beta-lactamase found in Kluyvera spp.",
  },
  {
    accession: "ARO:3001939",
    name: "CTX-M-78",
    description: "CTX-M-78 is a beta-lactamase found in Kluyvera georgiana",
  },
  {
    accession: "ARO:3001940",
    name: "CTX-M-79",
    description: "CTX-M-79 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001941",
    name: "CTX-M-80",
    description: "CTX-M-80 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001942",
    name: "CTX-M-81",
    description: "CTX-M-81 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001943",
    name: "CTX-M-82",
    description:
      "CTX-M-82 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001944",
    name: "CTX-M-83",
    description: "CTX-M-83 is a beta-lactamase found in Salmonella enterica",
  },
  {
    accession: "ARO:3001945",
    name: "CTX-M-84",
    description: "CTX-M-84 is a beta-lactamase found in Salmonella enterica",
  },
  {
    accession: "ARO:3001946",
    name: "CTX-M-86",
    description: "CTX-M-86 is a beta-lactamase found in Salmonella enterica",
  },
  {
    accession: "ARO:3001947",
    name: "CTX-M-87",
    description: "CTX-M-87 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001948",
    name: "CTX-M-88",
    description:
      "CTX-M-88 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001949",
    name: "CTX-M-89",
    description:
      "CTX-M-89 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001950",
    name: "CTX-M-90",
    description: "CTX-M-90 is a beta-lactamase found in Proteus mirabilis",
  },
  {
    accession: "ARO:3001951",
    name: "CTX-M-91",
    description: "CTX-M-91 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001952",
    name: "CTX-M-92",
    description: "CTX-M-92 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001953",
    name: "CTX-M-93",
    description: "CTX-M-93 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001954",
    name: "CTX-M-94",
    description: "CTX-M-94 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001955",
    name: "CTX-M-95",
    description: "CTX-M-95 is a beta-lactamase found in Kluyvera spp.",
  },
  {
    accession: "ARO:3001956",
    name: "CTX-M-96",
    description:
      "CTX-M-96 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001957",
    name: "CTX-M-98",
    description: "CTX-M-98 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001958",
    name: "CTX-M-99",
    description: "CTX-M-99 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3001959",
    name: "CTX-M-100",
    description: "CTX-M-100 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001960",
    name: "CTX-M-101",
    description:
      "CTX-M-101 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001961",
    name: "CTX-M-102",
    description: "CTX-M-102 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001962",
    name: "CTX-M-103",
    description: "CTX-M-103 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001963",
    name: "CTX-M-104",
    description: "CTX-M-104 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001965",
    name: "CTX-M-105",
    description: "CTX-M-105 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001966",
    name: "CTX-M-106",
    description:
      "CTX-M-106 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001967",
    name: "CTX-M-107",
    description: "CTX-M-107 is a beta-lactamase found in Shigella spp.",
  },
  {
    accession: "ARO:3001968",
    name: "CTX-M-108",
    description: "CTX-M-108 is a beta-lactamase found in Shigella spp.",
  },
  {
    accession: "ARO:3001969",
    name: "CTX-M-109",
    description: "CTX-M-109 is a beta-lactamase found in Shigella spp.",
  },
  {
    accession: "ARO:3001970",
    name: "CTX-M-110",
    description: "CTX-M-110 is a beta-lactamase found in Shigella spp.",
  },
  {
    accession: "ARO:3001971",
    name: "CTX-M-111",
    description: "CTX-M-111 is a beta-lactamase found in Shigella spp.",
  },
  {
    accession: "ARO:3001972",
    name: "CTX-M-112",
    description: "CTX-M-112 is a beta-lactamase found in Shigella spp.",
  },
  {
    accession: "ARO:3001973",
    name: "CTX-M-113",
    description: "CTX-M-113 is a beta-lactamase found in Shigella spp.",
  },
  {
    accession: "ARO:3001974",
    name: "CTX-M-114",
    description:
      "CTX-M-114 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001975",
    name: "CTX-M-115",
    description:
      "CTX-M-115 is a beta-lactamase found in Acinetobacter baumannii",
  },
  {
    accession: "ARO:3001976",
    name: "CTX-M-116",
    description: "CTX-M-116 is a beta-lactamase found in Proteus mirabilis",
  },
  {
    accession: "ARO:3001977",
    name: "CTX-M-117",
    description: "CTX-M-117 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001978",
    name: "CTX-M-119",
    description:
      "CTX-M-119 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001979",
    name: "CTX-M-120",
    description:
      "CTX-M-120 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001980",
    name: "CTX-M-121",
    description: "CTX-M-121 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001981",
    name: "CTX-M-122",
    description:
      "CTX-M-122 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001982",
    name: "CTX-M-123",
    description: "CTX-M-123 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001983",
    name: "CTX-M-124",
    description: "CTX-M-124 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001984",
    name: "CTX-M-125",
    description: "CTX-M-125 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001985",
    name: "CTX-M-126",
    description:
      "CTX-M-126 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001986",
    name: "CTX-M-127",
    description:
      "CTX-M-127 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001987",
    name: "CTX-M-128",
    description:
      "CTX-M-128 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001988",
    name: "CTX-M-129",
    description:
      "CTX-M-129 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001989",
    name: "CTX-M-130",
    description:
      "CTX-M-130 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001990",
    name: "CTX-M-131",
    description:
      "CTX-M-131 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001991",
    name: "CTX-M-132",
    description:
      "CTX-M-132 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001992",
    name: "CTX-M-134",
    description:
      "CTX-M-134 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001993",
    name: "CTX-M-135",
    description:
      "CTX-M-135 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3001994",
    name: "CTX-M-137",
    description: "CTX-M-137 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3001995",
    name: "CTX-M-136",
    description:
      "CTX-M-136 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001996",
    name: "CTX-M-138",
    description:
      "CTX-M-138 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001997",
    name: "CTX-M-139",
    description:
      "CTX-M-139 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001998",
    name: "CTX-M-140",
    description:
      "CTX-M-140 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3001999",
    name: "CTX-M-141",
    description:
      "CTX-M-141 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3002000",
    name: "CTX-M-142",
    description:
      "CTX-M-142 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3002001",
    name: "CTX-M-143",
    description:
      "CTX-M-143 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3002002",
    name: "CTX-M-144",
    description:
      "CTX-M-144 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3002003",
    name: "CTX-M-145",
    description:
      "CTX-M-145 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3002004",
    name: "CTX-M-146",
    description:
      "CTX-M-146 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3002005",
    name: "CTX-M-147",
    description:
      "CTX-M-147 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3002006",
    name: "CTX-M-148",
    description:
      "CTX-M-148 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3002007",
    name: "CTX-M-150",
    description:
      "CTX-M-150 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3002008",
    name: "CTX-M-151",
    description:
      "CTX-M-151 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3002009",
    name: "CTX-M-152",
    description:
      "CTX-M-152 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3002010",
    name: "CTX-M-153",
    description:
      "CTX-M-153 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3002011",
    name: "CTX-M-154",
    description:
      "CTX-M-154 is a beta-lactamase. From the Lahey list of CTX-M beta-lactamases.",
  },
  {
    accession: "ARO:3002012",
    name: "CMY-1",
    description:
      "CMY-1 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002013",
    name: "CMY-2",
    description: "CMY-2 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002014",
    name: "CMY-3",
    description: "CMY-3 is a beta-lactamase found in Proteus mirabilis",
  },
  {
    accession: "ARO:3002015",
    name: "CMY-4",
    description: "CMY-4 is a beta-lactamase found in Proteus mirabilis",
  },
  {
    accession: "ARO:3002016",
    name: "CMY-5",
    description: "CMY-5 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002017",
    name: "CMY-6",
    description: "CMY-6 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002018",
    name: "CMY-7",
    description: "CMY-7 is a beta-lactamase found in Salmonella typhimurium",
  },
  {
    accession: "ARO:3002019",
    name: "CMY-8",
    description: "CMY-8 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002020",
    name: "CMY-9",
    description: "CMY-9 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002021",
    name: "CMY-10",
    description: "CMY-10 is a beta-lactamase found in Enterobacter aerogenes",
  },
  {
    accession: "ARO:3002022",
    name: "CMY-11",
    description: "CMY-11 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002023",
    name: "CMY-12",
    description:
      "CMY-12 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002024",
    name: "CMY-13",
    description: "CMY-13 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002025",
    name: "CMY-14",
    description: "CMY-14 is a beta-lactamase found in Proteus mirabilis",
  },
  {
    accession: "ARO:3002026",
    name: "CMY-15",
    description: "CMY-15 is a beta-lactamase found in Proteus mirabilis",
  },
  {
    accession: "ARO:3002027",
    name: "CMY-16",
    description: "CMY-16 is a beta-lactamase found in Proteus mirabilis",
  },
  {
    accession: "ARO:3002028",
    name: "CMY-17",
    description:
      "CMY-17 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002029",
    name: "CMY-18",
    description:
      "CMY-18 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002030",
    name: "CMY-19",
    description: "CMY-19 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002031",
    name: "CMY-20",
    description: "CMY-20 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002032",
    name: "CMY-21",
    description: "CMY-21 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002033",
    name: "CMY-22",
    description:
      "CMY-22 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002034",
    name: "CMY-23",
    description:
      "CMY-23 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002035",
    name: "CMY-24",
    description:
      "CMY-24 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002036",
    name: "CMY-25",
    description:
      "CMY-25 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002037",
    name: "CMY-26",
    description:
      "CMY-26 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002038",
    name: "CMY-27",
    description:
      "CMY-27 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002039",
    name: "CMY-28",
    description:
      "CMY-28 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002040",
    name: "CMY-29",
    description:
      "CMY-29 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002041",
    name: "CMY-30",
    description:
      "CMY-30 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002042",
    name: "CMY-31",
    description:
      "CMY-31 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002043",
    name: "CMY-32",
    description: "CMY-32 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002044",
    name: "CMY-33",
    description: "CMY-33 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002045",
    name: "CMY-34",
    description: "CMY-34 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002046",
    name: "CMY-35",
    description:
      "CMY-35 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002047",
    name: "CMY-36",
    description: "CMY-36 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002048",
    name: "CMY-37",
    description: "CMY-37 is a beta-lactamase found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002049",
    name: "CMY-38",
    description:
      "CMY-38 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002050",
    name: "CMY-39",
    description:
      "CMY-39 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002051",
    name: "CMY-40",
    description:
      "CMY-40 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002052",
    name: "CMY-41",
    description:
      "CMY-41 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002053",
    name: "CMY-42",
    description:
      "CMY-42 is a plasmid-borne AmpC cephalosporinase gene found in Escherichia coli",
  },
  {
    accession: "ARO:3002054",
    name: "CMY-43",
    description:
      "CMY-43 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002055",
    name: "CMY-44",
    description: "CMY-44 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002056",
    name: "CMY-45",
    description: "CMY-45 is a beta-lactamase found in Proteus mirabilis",
  },
  {
    accession: "ARO:3002057",
    name: "CMY-46",
    description:
      "CMY-46 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002058",
    name: "CMY-47",
    description:
      "CMY-47 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002059",
    name: "CMY-48",
    description:
      "CMY-48 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002060",
    name: "CMY-49",
    description:
      "CMY-49 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002061",
    name: "CMY-50",
    description:
      "CMY-50 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002062",
    name: "CMY-51",
    description:
      "CMY-51 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002063",
    name: "CMY-52",
    description:
      "CMY-52 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002064",
    name: "CMY-53",
    description: "CMY-53 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002065",
    name: "CMY-54",
    description:
      "CMY-54 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002066",
    name: "CMY-55",
    description:
      "CMY-55 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002067",
    name: "CMY-56",
    description:
      "CMY-56 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002068",
    name: "CMY-57",
    description:
      "CMY-57 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002069",
    name: "CMY-59",
    description: "CMY-59 is a beta-lactamase found in Shigella spp.",
  },
  {
    accession: "ARO:3002070",
    name: "CMY-99",
    description: "CMY-99 is a beta-lactamase found in Proteus mirabilis",
  },
  {
    accession: "ARO:3002071",
    name: "CMY-110",
    description: "CMY-110 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002072",
    name: "CMY-58",
    description:
      "CMY-58 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002073",
    name: "CMY-60",
    description:
      "CMY-60 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002074",
    name: "CMY-61",
    description:
      "CMY-61 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002075",
    name: "CMY-62",
    description:
      "CMY-62 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002076",
    name: "CMY-63",
    description:
      "CMY-63 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002077",
    name: "CMY-64",
    description:
      "CMY-64 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002078",
    name: "CMY-65",
    description:
      "CMY-65 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002079",
    name: "CMY-66",
    description:
      "CMY-66 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002080",
    name: "CMY-67",
    description:
      "CMY-67 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002081",
    name: "CMY-68",
    description:
      "CMY-68 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002082",
    name: "CMY-69",
    description:
      "CMY-69 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002083",
    name: "CMY-70",
    description:
      "CMY-70 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002084",
    name: "CMY-71",
    description:
      "CMY-71 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002085",
    name: "CMY-72",
    description:
      "CMY-72 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002086",
    name: "CMY-73",
    description:
      "CMY-73 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002087",
    name: "CMY-74",
    description:
      "CMY-74 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002088",
    name: "CMY-75",
    description:
      "CMY-75 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002089",
    name: "CMY-76",
    description:
      "CMY-76 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002090",
    name: "CMY-77",
    description:
      "CMY-77 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002091",
    name: "CMY-78",
    description:
      "CMY-78 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002092",
    name: "CMY-79",
    description:
      "CMY-79 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002093",
    name: "CMY-80",
    description:
      "CMY-80 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002094",
    name: "CMY-81",
    description:
      "CMY-81 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002095",
    name: "CMY-82",
    description:
      "CMY-82 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002096",
    name: "CMY-83",
    description:
      "CMY-83 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002097",
    name: "CMY-84",
    description:
      "CMY-84 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002098",
    name: "CMY-85",
    description:
      "CMY-85 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002099",
    name: "CMY-86",
    description:
      "CMY-86 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002100",
    name: "CMY-87",
    description:
      "CMY-87 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002101",
    name: "CMY-88",
    description:
      "CMY-88 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002102",
    name: "CMY-89",
    description:
      "CMY-89 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002103",
    name: "CMY-90",
    description:
      "CMY-90 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002104",
    name: "CMY-91",
    description:
      "CMY-91 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002105",
    name: "CMY-92",
    description:
      "CMY-92 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002106",
    name: "CMY-93",
    description:
      "CMY-93 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002107",
    name: "CMY-94",
    description:
      "CMY-94 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002108",
    name: "CMY-95",
    description:
      "CMY-95 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002109",
    name: "CMY-96",
    description:
      "CMY-96 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002110",
    name: "CMY-97",
    description:
      "CMY-97 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002111",
    name: "CMY-98",
    description:
      "CMY-98 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002112",
    name: "CMY-100",
    description:
      "CMY-100 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002113",
    name: "CMY-101",
    description:
      "CMY-101 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002114",
    name: "CMY-102",
    description:
      "CMY-102 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002115",
    name: "CMY-103",
    description:
      "CMY-103 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002116",
    name: "CMY-104",
    description:
      "CMY-104 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002117",
    name: "CMY-105",
    description:
      "CMY-105 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002118",
    name: "CMY-106",
    description:
      "CMY-106 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002119",
    name: "CMY-107",
    description:
      "CMY-107 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002120",
    name: "CMY-108",
    description:
      "CMY-108 is a beta-lactamase present in plasmids of clinical Escherichia coli from humans and companion animals in the upper Midwestern USA . From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002121",
    name: "CMY-109",
    description:
      "CMY-109 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002122",
    name: "CMY-111",
    description:
      "CMY-111 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002123",
    name: "CMY-112",
    description:
      "CMY-112 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002124",
    name: "CMY-113",
    description:
      "CMY-113 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002125",
    name: "CMY-114",
    description:
      "CMY-114 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002126",
    name: "CMY-115",
    description:
      "CMY-115 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002127",
    name: "CMY-116",
    description:
      "CMY-116 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002128",
    name: "CMY-117",
    description:
      "CMY-117 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002129",
    name: "CMY-118",
    description:
      "CMY-118 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002130",
    name: "CMY-119",
    description:
      "CMY-119 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002131",
    name: "CMY-120",
    description:
      "CMY-120 is a beta-lactamase. From the Lahey list of CMY beta-lactamases.",
  },
  {
    accession: "ARO:3002132",
    name: "DHA-1",
    description:
      "DHA-1 is a class C beta-lactamase found in Morganella morganii and Salmonella enterica",
  },
  {
    accession: "ARO:3002133",
    name: "DHA-2",
    description: "DHA-2 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002134",
    name: "DHA-3",
    description: "DHA-3 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002135",
    name: "DHA-6",
    description:
      "DHA-6 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002136",
    name: "DHA-7",
    description:
      "DHA-7 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002137",
    name: "DHA-4",
    description:
      "DHA-4 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002138",
    name: "DHA-5",
    description:
      "DHA-5 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002139",
    name: "DHA-8",
    description:
      "DHA-8 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002140",
    name: "DHA-9",
    description:
      "DHA-9 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002141",
    name: "DHA-10",
    description:
      "DHA-10 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002142",
    name: "DHA-11",
    description:
      "DHA-11 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002143",
    name: "DHA-12",
    description:
      "DHA-12 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002144",
    name: "DHA-13",
    description:
      "DHA-13 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002145",
    name: "DHA-14",
    description:
      "DHA-14 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002146",
    name: "DHA-15",
    description:
      "DHA-15 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002147",
    name: "DHA-16",
    description:
      "DHA-16 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002148",
    name: "DHA-17",
    description:
      "DHA-17 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002149",
    name: "DHA-18",
    description:
      "DHA-18 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002150",
    name: "DHA-19",
    description:
      "DHA-19 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002151",
    name: "DHA-20",
    description:
      "DHA-20 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002152",
    name: "DHA-21",
    description:
      "DHA-21 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002153",
    name: "DHA-22",
    description:
      "DHA-22 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002154",
    name: "DHA-23",
    description:
      "DHA-23 is a beta-lactamase. From the Lahey list of DHA beta-lactamases.",
  },
  {
    accession: "ARO:3002155",
    name: "FOX-1",
    description: "FOX-1 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002156",
    name: "FOX-2",
    description: "FOX-2 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002157",
    name: "FOX-3",
    description: "FOX-3 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002158",
    name: "FOX-4",
    description: "FOX-4 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002159",
    name: "FOX-5",
    description: "FOX-5 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002160",
    name: "FOX-7",
    description: "FOX-7 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002161",
    name: "FOX-8",
    description:
      "FOX-8 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002162",
    name: "FOX-10",
    description:
      "FOX-10 is a beta-lactamase. From the Lahey list of FOX beta-lactamases.",
  },
  {
    accession: "ARO:3002163",
    name: "FOX-9",
    description:
      "FOX-9 is a beta-lactamase. From the Lahey list of FOX beta-lactamases.",
  },
  {
    accession: "ARO:3002164",
    name: "FOX-11",
    description:
      "FOX-11 is a beta-lactamase. From the Lahey list of FOX beta-lactamases.",
  },
  {
    accession: "ARO:3002165",
    name: "FOX-12",
    description:
      "FOX-12 is a beta-lactamase. From the Lahey list of FOX beta-lactamases.",
  },
  {
    accession: "ARO:3002166",
    name: "MIR-1",
    description: "MIR-1 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002167",
    name: "MIR-4",
    description: "MIR-4 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002168",
    name: "MIR-2",
    description:
      "MIR-2 is a beta-lactamase. From the Lahey list of MIR beta-lactamases.",
  },
  {
    accession: "ARO:3002169",
    name: "MIR-3",
    description:
      "MIR-3 is a beta-lactamase. From the Lahey list of MIR beta-lactamases.",
  },
  {
    accession: "ARO:3002170",
    name: "MIR-5",
    description:
      "MIR-5 is a beta-lactamase. From the Lahey list of MIR beta-lactamases.",
  },
  {
    accession: "ARO:3002171",
    name: "MIR-6",
    description:
      "MIR-6 is a beta-lactamase. From the Lahey list of MIR beta-lactamases.",
  },
  {
    accession: "ARO:3002172",
    name: "MIR-7",
    description:
      "MIR-7 is a beta-lactamase. From the Lahey list of MIR beta-lactamases.",
  },
  {
    accession: "ARO:3002173",
    name: "MIR-8",
    description:
      "MIR-8 is a beta-lactamase. From the Lahey list of MIR beta-lactamases.",
  },
  {
    accession: "ARO:3002174",
    name: "MIR-9",
    description:
      "MIR-9 is a beta-lactamase. From the Lahey list of MIR beta-lactamases.",
  },
  {
    accession: "ARO:3002175",
    name: "MIR-10",
    description:
      "MIR-10 is a beta-lactamase. From the Lahey list of MIR beta-lactamases.",
  },
  {
    accession: "ARO:3002176",
    name: "MIR-11",
    description:
      "MIR-11 is a beta-lactamase. From the Lahey list of MIR beta-lactamases.",
  },
  {
    accession: "ARO:3002177",
    name: "MIR-12",
    description:
      "MIR-12 is a beta-lactamase. From the Lahey list of MIR beta-lactamases.",
  },
  {
    accession: "ARO:3002178",
    name: "MIR-13",
    description:
      "MIR-13 is a beta-lactamase. From the Lahey list of MIR beta-lactamases.",
  },
  {
    accession: "ARO:3002179",
    name: "MIR-14",
    description:
      "MIR-14 is a beta-lactamase. From the Lahey list of MIR beta-lactamases.",
  },
  {
    accession: "ARO:3002180",
    name: "MIR-15",
    description:
      "MIR-15 is a beta-lactamase. From the Lahey list of MIR beta-lactamases.",
  },
  {
    accession: "ARO:3002181",
    name: "MIR-16",
    description:
      "MIR-16 is a beta-lactamase. From the Lahey list of MIR beta-lactamases.",
  },
  {
    accession: "ARO:3002182",
    name: "MOX-1",
    description: "MOX-1 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002183",
    name: "MOX-2",
    description: "MOX-2 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002184",
    name: "MOX-4",
    description: "MOX-4 is a beta-lactamase found in Aeromonas caviae",
  },
  {
    accession: "ARO:3002185",
    name: "MOX-6",
    description: "MOX-6 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002186",
    name: "MOX-3",
    description:
      "MOX-3 is a beta-lactamase. From the Lahey list of MOX beta-lactamases.",
  },
  {
    accession: "ARO:3002188",
    name: "MOX-5",
    description:
      "MOX-5 is a beta-lactamase. From the Lahey list of MOX beta-lactamases.",
  },
  {
    accession: "ARO:3002189",
    name: "MOX-7",
    description:
      "MOX-7 is a beta-lactamase. From the Lahey list of MOX beta-lactamases.",
  },
  {
    accession: "ARO:3002190",
    name: "MOX-8",
    description:
      "MOX-8 is a beta-lactamase. From the Lahey list of MOX beta-lactamases.",
  },
  {
    accession: "ARO:3002191",
    name: "MOX-9",
    description: "MOX-9 is a beta-lactamase found in Citrobacter freundii.",
  },
  {
    accession: "ARO:3002192",
    name: "IMP-1",
    description: "IMP-1 is a beta-lactamase found in Serratia marcescens",
  },
  {
    accession: "ARO:3002193",
    name: "IMP-2",
    description: "IMP-2 is a beta-lactamase found in Acinetobacter baumannii",
  },
  {
    accession: "ARO:3002194",
    name: "IMP-3",
    description: "IMP-3 is a beta-lactamase found in Acinetobacter baumannii",
  },
  {
    accession: "ARO:3002195",
    name: "IMP-4",
    description: "IMP-4 is a beta-lactamase found in Acinetobacter baumannii",
  },
  {
    accession: "ARO:3002196",
    name: "IMP-5",
    description: "IMP-5 is a beta-lactamase found in Acinetobacter baumannii",
  },
  {
    accession: "ARO:3002197",
    name: "IMP-6",
    description: "IMP-6 is a beta-lactamase found in Serratia marcescens",
  },
  {
    accession: "ARO:3002198",
    name: "IMP-7",
    description: "IMP-7 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002199",
    name: "IMP-8",
    description: "IMP-8 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002200",
    name: "IMP-9",
    description: "IMP-9 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002201",
    name: "IMP-10",
    description: "IMP-10 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002202",
    name: "IMP-11",
    description:
      "IMP-11 is a beta-lactamase found in Pseudomonas and Acinetobacter spp.",
  },
  {
    accession: "ARO:3002203",
    name: "IMP-12",
    description: "IMP-12 is a beta-lactamase found in Pseudomonas putida",
  },
  {
    accession: "ARO:3002204",
    name: "IMP-13",
    description: "IMP-13 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002205",
    name: "IMP-14",
    description:
      "IMP-14 is a beta-lactamase found in Pseudomonas and Acinetobacter spp.",
  },
  {
    accession: "ARO:3002206",
    name: "IMP-15",
    description: "IMP-15 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002207",
    name: "IMP-16",
    description: "IMP-16 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002208",
    name: "IMP-17",
    description:
      "IMP-17 is a beta-lactamase. From the Lahey list of IMP beta-lactamases.",
  },
  {
    accession: "ARO:3002209",
    name: "IMP-18",
    description: "IMP-18 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002210",
    name: "IMP-19",
    description: "IMP-19 is a beta-lactamase found in Aeromonas caviae",
  },
  {
    accession: "ARO:3002211",
    name: "IMP-20",
    description:
      "IMP-20 is a beta-lactamase found in Pseudomonas and Acinetobacter spp.",
  },
  {
    accession: "ARO:3002212",
    name: "IMP-21",
    description:
      "IMP-21 is a beta-lactamase found in Pseudomonas and Acinetobacter spp.",
  },
  {
    accession: "ARO:3002213",
    name: "IMP-22",
    description: "IMP-22 is a beta-lactamase found in Pseudomonas spp.",
  },
  {
    accession: "ARO:3002214",
    name: "IMP-23",
    description:
      "IMP-23 is a beta-lactamase. From the Lahey list of IMP beta-lactamases.",
  },
  {
    accession: "ARO:3002215",
    name: "IMP-24",
    description: "IMP-24 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002216",
    name: "IMP-25",
    description: "IMP-25 is a beta-lactamase found in Serratia marcescens",
  },
  {
    accession: "ARO:3002217",
    name: "IMP-26",
    description: "IMP-26 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002218",
    name: "IMP-27",
    description:
      "IMP-27 is a beta-lactamase. From the Lahey list of IMP beta-lactamases.",
  },
  {
    accession: "ARO:3002219",
    name: "IMP-28",
    description: "IMP-28 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002220",
    name: "IMP-29",
    description: "IMP-29 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002221",
    name: "IMP-30",
    description: "IMP-30 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002222",
    name: "IMP-31",
    description:
      "IMP-31 is a beta-lactamase. From the Lahey list of IMP beta-lactamases.",
  },
  {
    accession: "ARO:3002223",
    name: "IMP-32",
    description:
      "IMP-32 is a beta-lactamase. From the Lahey list of IMP beta-lactamases.",
  },
  {
    accession: "ARO:3002224",
    name: "IMP-33",
    description: "IMP-33 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002225",
    name: "IMP-34",
    description: "IMP-34 is a beta-lactamase found in Klebsiella spp.",
  },
  {
    accession: "ARO:3002226",
    name: "IMP-35",
    description: "IMP-35 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002227",
    name: "IMP-36",
    description:
      "IMP-36 is a beta-lactamase. From the Lahey list of IMP beta-lactamases.",
  },
  {
    accession: "ARO:3002228",
    name: "IMP-37",
    description:
      "IMP-37 is a beta-lactamase. From the Lahey list of IMP beta-lactamases.",
  },
  {
    accession: "ARO:3002229",
    name: "IMP-38",
    description:
      "IMP-38 is a beta-lactamase. From the Lahey list of IMP beta-lactamases.",
  },
  {
    accession: "ARO:3002230",
    name: "IMP-39",
    description:
      "IMP-39 is a beta-lactamase. From the Lahey list of IMP beta-lactamases.",
  },
  {
    accession: "ARO:3002231",
    name: "IMP-40",
    description:
      "IMP-40 is a beta-lactamase found in Pseudomonas and Acinetobacter spp.",
  },
  {
    accession: "ARO:3002232",
    name: "IMP-41",
    description:
      "IMP-41 is a beta-lactamase found in Pseudomonas and Acinetobacter spp.",
  },
  {
    accession: "ARO:3002233",
    name: "IMP-42",
    description:
      "IMP-42 is a beta-lactamase found in Pseudomonas and Acinetobacter spp.",
  },
  {
    accession: "ARO:3002234",
    name: "IMP-43",
    description: "IMP-43 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002235",
    name: "IMP-44",
    description: "IMP-44 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002236",
    name: "IMP-45",
    description: "IMP-45 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002237",
    name: "IMP-46",
    description:
      "IMP-46 is a beta-lactamase. From the Lahey list of IMP beta-lactamases.",
  },
  {
    accession: "ARO:3002239",
    name: "IMP-48",
    description:
      "IMP-48 is a beta-lactamase. From the Lahey list of IMP beta-lactamases.",
  },
  {
    accession: "ARO:3002240",
    name: "CARB-1",
    description: "CARB-1 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002241",
    name: "CARB-2",
    description: "CARB-2 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002242",
    name: "CARB-3",
    description: "CARB-3 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002243",
    name: "CARB-4",
    description: "CARB-4 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002244",
    name: "CARB-5",
    description:
      "CARB-5 is a beta-lactamase found in Acinetobacter calcoaceticus",
  },
  {
    accession: "ARO:3002245",
    name: "CARB-6",
    description: "CARB-6 is a beta-lactamase found in Vibrio cholerae",
  },
  {
    accession: "ARO:3002246",
    name: "CARB-7",
    description: "CARB-7 is a beta-lactamase found in Vibrio cholerae.",
  },
  {
    accession: "ARO:3002247",
    name: "CARB-8",
    description: "CARB-8 is a beta-lactamase found in Oligella urethralis",
  },
  {
    accession: "ARO:3002248",
    name: "CARB-9",
    description: "CARB-9 is a beta-lactamase found in Vibrio cholerae",
  },
  {
    accession: "ARO:3002249",
    name: "CARB-10",
    description: "CARB-10 is a beta-lactamase found in Acinetobacter baumannii",
  },
  {
    accession: "ARO:3002250",
    name: "CARB-12",
    description:
      "CARB-12 is a beta-lactamase. From the Lahey list of IMP beta-lactamases.",
  },
  {
    accession: "ARO:3002251",
    name: "CARB-13",
    description:
      "CARB-13 is a beta-lactamase. From the Lahey list of IMP beta-lactamases.",
  },
  {
    accession: "ARO:3002252",
    name: "CARB-14",
    description: "CARB-14 is a beta-lactamase found in Acinetobacter baumannii",
  },
  {
    accession: "ARO:3002253",
    name: "CARB-15",
    description:
      "CARB-15 is a beta-lactamase. From the Lahey list of IMP beta-lactamases.",
  },
  {
    accession: "ARO:3002254",
    name: "CARB-17",
    description:
      "CARB-17 is a beta-lactamase. From the Lahey list of IMP beta-lactamases.",
  },
  {
    accession: "ARO:3002255",
    name: "CARB-16",
    description:
      "CARB-16 is a beta-lactamase. From the Lahey list of IMP beta-lactamases.",
  },
  {
    accession: "ARO:3002256",
    name: "IND-1",
    description:
      "IND-1 is a beta-lactamase found in Chryseobacterium indologenes",
  },
  {
    accession: "ARO:3002257",
    name: "IND-2",
    description:
      "IND-2 is a beta-lactamase found in Chryseobacterium indologenes",
  },
  {
    accession: "ARO:3002258",
    name: "IND-2a",
    description:
      "IND-2a is a beta-lactamase found in Chryseobacterium indologenes",
  },
  {
    accession: "ARO:3002259",
    name: "IND-3",
    description:
      "IND-3 is a beta-lactamase found in Chryseobacterium indologenes",
  },
  {
    accession: "ARO:3002260",
    name: "IND-4",
    description:
      "IND-4 is a beta-lactamase found in Chryseobacterium indologenes",
  },
  {
    accession: "ARO:3002261",
    name: "IND-5",
    description:
      "IND-5 is a beta-lactamase found in Chryseobacterium indologenes",
  },
  {
    accession: "ARO:3002262",
    name: "IND-6",
    description:
      "IND-6 is a beta-lactamase found in Chryseobacterium indologenes",
  },
  {
    accession: "ARO:3002263",
    name: "IND-7",
    description:
      "IND-7 is a beta-lactamase found in Chryseobacterium indologenes",
  },
  {
    accession: "ARO:3002264",
    name: "IND-8",
    description: "IND-8 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002265",
    name: "IND-9",
    description: "IND-9 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002266",
    name: "IND-10",
    description: "IND-10 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002267",
    name: "IND-11",
    description: "IND-11 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002268",
    name: "IND-12",
    description: "IND-12 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002269",
    name: "IND-14",
    description: "IND-14 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002270",
    name: "IND-15",
    description: "IND-15 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002271",
    name: "VIM-1",
    description: "VIM-1 is a beta-lactamase found in Pseudomonas spp.",
  },
  {
    accession: "ARO:3002272",
    name: "VIM-2",
    description: "VIM-2 is a beta-lactamase found in Pseudomonas spp.",
  },
  {
    accession: "ARO:3002273",
    name: "VIM-3",
    description: "VIM-3 is a beta-lactamase found in Pseudomonas spp.",
  },
  {
    accession: "ARO:3002274",
    name: "VIM-4",
    description: "VIM-4 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002275",
    name: "VIM-5",
    description: "VIM-5 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002276",
    name: "VIM-6",
    description: "VIM-6 is a beta-lactamase found in Pseudomonas spp.",
  },
  {
    accession: "ARO:3002277",
    name: "VIM-7",
    description: "VIM-7 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002278",
    name: "VIM-8",
    description: "VIM-8 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002279",
    name: "VIM-9",
    description: "VIM-9 is a beta-lactamase found in Pseudomonas spp.",
  },
  {
    accession: "ARO:3002280",
    name: "VIM-10",
    description: "VIM-10 is a beta-lactamase found in Pseudomonas spp.",
  },
  {
    accession: "ARO:3002281",
    name: "VIM-11",
    description: "VIM-11 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002282",
    name: "VIM-12",
    description: "VIM-12 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002283",
    name: "VIM-13",
    description: "VIM-13 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002284",
    name: "VIM-14",
    description: "VIM-14 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002285",
    name: "VIM-15",
    description: "VIM-15 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002286",
    name: "VIM-16",
    description: "VIM-16 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002287",
    name: "VIM-17",
    description: "VIM-17 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002288",
    name: "VIM-18",
    description: "VIM-18 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002289",
    name: "VIM-19",
    description:
      "VIM-19 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002290",
    name: "VIM-20",
    description:
      "VIM-20 is a beta-lactamase. From the Lahey list of VIM beta-lactamases.",
  },
  {
    accession: "ARO:3002291",
    name: "VIM-21",
    description:
      "VIM-21 is a beta-lactamase. From the Lahey list of VIM beta-lactamases.",
  },
  {
    accession: "ARO:3002292",
    name: "VIM-22",
    description:
      "VIM-22 is a beta-lactamase. From the Lahey list of VIM beta-lactamases.",
  },
  {
    accession: "ARO:3002293",
    name: "VIM-23",
    description: "VIM-23 is a beta-lactamase found in Enterobacter cloacae",
  },
  {
    accession: "ARO:3002294",
    name: "VIM-24",
    description: "VIM-24 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002295",
    name: "VIM-25",
    description:
      "VIM-25 is a beta-lactamase. From the Lahey list of VIM beta-lactamases.",
  },
  {
    accession: "ARO:3002296",
    name: "VIM-26",
    description:
      "VIM-26 is a beta-lactamase. From the Lahey list of VIM beta-lactamases.",
  },
  {
    accession: "ARO:3002297",
    name: "VIM-27",
    description: "VIM-27 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002298",
    name: "VIM-28",
    description:
      "VIM-28 is a beta-lactamase. From the Lahey list of VIM beta-lactamases.",
  },
  {
    accession: "ARO:3002299",
    name: "VIM-29",
    description:
      "VIM-29 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002300",
    name: "VIM-30",
    description:
      "VIM-30 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002301",
    name: "VIM-31",
    description: "VIM-31 is a beta-lactamase found in Enterobacter cloacae",
  },
  {
    accession: "ARO:3002302",
    name: "VIM-32",
    description:
      "VIM-32 is a beta-lactamase. From the Lahey list of VIM beta-lactamases.",
  },
  {
    accession: "ARO:3002303",
    name: "VIM-33",
    description: "VIM-33 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002304",
    name: "VIM-34",
    description: "VIM-34 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002305",
    name: "VIM-35",
    description:
      "VIM-35 is a beta-lactamase. From the Lahey list of VIM beta-lactamases.",
  },
  {
    accession: "ARO:3002306",
    name: "VIM-36",
    description: "VIM-36 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002307",
    name: "VIM-37",
    description: "VIM-37 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002308",
    name: "VIM-38",
    description: "VIM-38 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002309",
    name: "VIM-39",
    description:
      "VIM-39 is a beta-lactamase. From the Lahey list of VIM beta-lactamases.",
  },
  {
    accession: "ARO:3002310",
    name: "VIM-40",
    description:
      "VIM-40 is a beta-lactamase. From the Lahey list of VIM beta-lactamases.",
  },
  {
    accession: "ARO:3002311",
    name: "VIM-41",
    description:
      "VIM-41 is a beta-lactamase. From the Lahey list of VIM beta-lactamases.",
  },
  {
    accession: "ARO:3002312",
    name: "KPC-2",
    description:
      "KPC-2 is a beta-lactamase found in Klebsiella pneumoniae and Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3002313",
    name: "KPC-3",
    description: "KPC-3 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002314",
    name: "KPC-4",
    description: "KPC-4 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002315",
    name: "KPC-5",
    description:
      "KPC-5 is a beta-lactamase found in Klebsiella pneumoniae and Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3002316",
    name: "KPC-6",
    description: "KPC-6 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002317",
    name: "KPC-7",
    description: "KPC-7 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002318",
    name: "KPC-8",
    description: "KPC-8 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002319",
    name: "KPC-9",
    description: "KPC-9 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002320",
    name: "KPC-10",
    description: "KPC-10 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002321",
    name: "KPC-11",
    description: "KPC-11 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002322",
    name: "KPC-12",
    description:
      "KPC-12 is a beta-lactamase. From the Lahey list of KPC beta-lactamases.",
  },
  {
    accession: "ARO:3002323",
    name: "KPC-13",
    description:
      "KPC-13 is a beta-lactamase. From the Lahey list of KPC beta-lactamases.",
  },
  {
    accession: "ARO:3002324",
    name: "KPC-14",
    description:
      "KPC-14 is a beta-lactamase. From the Lahey list of KPC beta-lactamases.",
  },
  {
    accession: "ARO:3002325",
    name: "KPC-15",
    description:
      "KPC-15 is a beta-lactamase. From the Lahey list of KPC beta-lactamases.",
  },
  {
    accession: "ARO:3002326",
    name: "KPC-16",
    description:
      "KPC-16 is a beta-lactamase. From the Lahey list of KPC beta-lactamases.",
  },
  {
    accession: "ARO:3002327",
    name: "KPC-17",
    description:
      "KPC-17 is a beta-lactamase. From the Lahey list of KPC beta-lactamases.",
  },
  {
    accession: "ARO:3002328",
    name: "KPC-18",
    description:
      "KPC-18 is a beta-lactamase. From the Lahey list of KPC beta-lactamases.",
  },
  {
    accession: "ARO:3002329",
    name: "KPC-19",
    description:
      "KPC-19 is a beta-lactamase. From the Lahey list of KPC beta-lactamases.",
  },
  {
    accession: "ARO:3002330",
    name: "GES-1",
    description: "GES-1 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002331",
    name: "GES-2",
    description: "GES-2 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002332",
    name: "GES-3",
    description: "GES-3 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002333",
    name: "GES-4",
    description: "GES-4 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002334",
    name: "GES-5",
    description:
      "GES-5 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002335",
    name: "GES-6",
    description:
      "GES-6 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002336",
    name: "GES-7",
    description: "GES-7 is a beta-lactamase found in Enterobacter cloacae",
  },
  {
    accession: "ARO:3002337",
    name: "GES-8",
    description: "GES-8 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002338",
    name: "GES-9",
    description: "GES-9 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002339",
    name: "GES-10",
    description:
      "GES-10 is a beta-lactamase. From the Lahey list of GES beta-lactamases.",
  },
  {
    accession: "ARO:3002340",
    name: "GES-11",
    description: "GES-11 is a beta-lactamase found in Acinetobacter baumannii",
  },
  {
    accession: "ARO:3002341",
    name: "GES-12",
    description: "GES-12 is a beta-lactamase found in Acinetobacter baumannii",
  },
  {
    accession: "ARO:3002342",
    name: "GES-13",
    description: "GES-13 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002343",
    name: "GES-14",
    description: "GES-14 is a beta-lactamase found in Acinetobacter baumannii",
  },
  {
    accession: "ARO:3002344",
    name: "GES-15",
    description:
      "GES-15 is a beta-lactamase. From the Lahey list of GES beta-lactamases.",
  },
  {
    accession: "ARO:3002345",
    name: "GES-16",
    description:
      "GES-16 is a beta-lactamase. From the Lahey list of GES beta-lactamases.",
  },
  {
    accession: "ARO:3002346",
    name: "GES-17",
    description:
      "GES-17 is a beta-lactamase. From the Lahey list of GES beta-lactamases.",
  },
  {
    accession: "ARO:3002347",
    name: "GES-18",
    description: "GES-18 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002348",
    name: "GES-19",
    description:
      "GES-19 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002349",
    name: "GES-20",
    description:
      "GES-20 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002350",
    name: "GES-21",
    description:
      "GES-21 is a beta-lactamase. From the Lahey list of GES beta-lactamases.",
  },
  {
    accession: "ARO:3002351",
    name: "GES-22",
    description:
      "GES-22 is a beta-lactamase. From the Lahey list of GES beta-lactamases.",
  },
  {
    accession: "ARO:3002352",
    name: "GES-23",
    description:
      "GES-23 is a beta-lactamase. From the Lahey list of GES beta-lactamases.",
  },
  {
    accession: "ARO:3002353",
    name: "GES-24",
    description:
      "GES-24 is a beta-lactamase. From the Lahey list of GES beta-lactamases.",
  },
  {
    accession: "ARO:3002354",
    name: "NDM-3",
    description:
      "NDM-3 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002355",
    name: "NDM-4",
    description: "NDM-4 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002356",
    name: "NDM-6",
    description:
      "NDM-6 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002357",
    name: "NDM-7",
    description: "NDM-7 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002358",
    name: "NDM-8",
    description: "NDM-8 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002359",
    name: "NDM-9",
    description: "NDM-9 is a beta-lactamase found in Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3002360",
    name: "NDM-10",
    description:
      "NDM-10 is a beta-lactamase. From the Lahey list of NDM beta-lactamases.",
  },
  {
    accession: "ARO:3002361",
    name: "NDM-11",
    description:
      "NDM-11 is a beta-lactamase. From the Lahey list of NDM beta-lactamases.",
  },
  {
    accession: "ARO:3002362",
    name: "NDM-12",
    description: "NDM-12 is a beta-lactamase found in Escherichia coli.",
  },
  {
    accession: "ARO:3002363",
    name: "PER-1",
    description: "PER-1 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002364",
    name: "PER-2",
    description: "PER-2 is a beta-lactamase found in Salmonella typhimurium",
  },
  {
    accession: "ARO:3002365",
    name: "PER-3",
    description: "PER-3 is a beta-lactamase found in Aeromonas spp.",
  },
  {
    accession: "ARO:3002366",
    name: "PER-4",
    description:
      "PER-4 is a beta-lactamase. From the Lahey list of PER beta-lactamases.",
  },
  {
    accession: "ARO:3002367",
    name: "PER-5",
    description:
      "PER-5 is a beta-lactamase found in the Enterobacteriaceae family",
  },
  {
    accession: "ARO:3002368",
    name: "PER-6",
    description:
      "PER-6 is a beta-lactamase found in Aeromonas allosaccharophila",
  },
  {
    accession: "ARO:3002369",
    name: "PER-7",
    description: "PER-7 is a beta-lactamase found in Acinetobacter baumannii",
  },
  {
    accession: "ARO:3002370",
    name: "VEB-1",
    description:
      "VEB-1 is a beta-lactamase found in Escherichia coli and Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3002371",
    name: "VEB-2",
    description: "VEB-2 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002372",
    name: "VEB-3",
    description: "VEB-3 is a beta-lactamase found in Enterobacter cloacae",
  },
  {
    accession: "ARO:3002373",
    name: "VEB-4",
    description: "VEB-4 is a beta-lactamase found in Proteus mirabilis",
  },
  {
    accession: "ARO:3002374",
    name: "VEB-6",
    description: "VEB-6 is a beta-lactamase found in Proteus mirabilis",
  },
  {
    accession: "ARO:3002375",
    name: "VEB-5",
    description:
      "VEB-5 is a beta-lactamase. From the Lahey list of VEB beta-lactamases.",
  },
  {
    accession: "ARO:3002376",
    name: "VEB-7",
    description:
      "VEB-7 is a beta-lactamase. From the Lahey list of VEB beta-lactamases.",
  },
  {
    accession: "ARO:3002377",
    name: "VEB-8",
    description:
      "VEB-8 is a beta-lactamase. From the Lahey list of VEB beta-lactamases.",
  },
  {
    accession: "ARO:3002378",
    name: "VEB-9",
    description:
      "VEB-9 is a beta-lactamase. From the Lahey list of VEB beta-lactamases.",
  },
  {
    accession: "ARO:3002379",
    name: "SME-1",
    description: "SME-1 is a beta-lactamase found in Serratia marcescens",
  },
  {
    accession: "ARO:3002380",
    name: "SME-2",
    description: "SME-2 is a beta-lactamase found in Serratia marcescens",
  },
  {
    accession: "ARO:3002381",
    name: "SME-3",
    description: "SME-3 is a beta-lactamase found in Serratia marcescens",
  },
  {
    accession: "ARO:3002382",
    name: "SME-4",
    description:
      "SME-4 is a beta-lactamase. From the Lahey list of SME beta-lactamases.",
  },
  {
    accession: "ARO:3002383",
    name: "SME-5",
    description:
      "SME-5 is a beta-lactamase. From the Lahey list of SME beta-lactamases.",
  },
  {
    accession: "ARO:3002384",
    name: "BEL beta-lactamase",
    description:
      "BEL beta-lactamases are class A expanded-spectrum beta-lactamases that are inhibited by clavulanic acid. They are chromosomally encoded and hydrolyze most cephalosporins and aztreonam.",
  },
  {
    accession: "ARO:3002385",
    name: "BEL-1",
    description: "BEL-1 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002386",
    name: "BEL-2",
    description: "BEL-2 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002387",
    name: "BEL-3",
    description: "BEL-3 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002388",
    name: "OXY beta-lactamase",
    description:
      "OXY beta-lactamases are chromosomal class A beta-lactamases that are found in Klebsiella oxytoca. At constitutive low levels, OXY beta-lactamases confer resistance to aminopenicillins and carboxypenicillins. At high induced levels,  OXY beta-lactamases confer resistance to penicillins, cephalosporins and aztreonam.",
  },
  {
    accession: "ARO:3002389",
    name: "OXY-1-1",
    description: "OXY-1-1 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002390",
    name: "OXY-1-2",
    description: "OXY-1-2 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002391",
    name: "OXY-1-3",
    description: "OXY-1-3 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002392",
    name: "OXY-1-4",
    description: "OXY-1-4 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002394",
    name: "OXY-1-6",
    description: "OXY-1-6 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002396",
    name: "OXY-2-1",
    description: "OXY-2-1 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002397",
    name: "OXY-2-2",
    description: "OXY-2-2 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002398",
    name: "OXY-2-3",
    description: "OXY-2-3 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002399",
    name: "OXY-2-4",
    description: "OXY-2-4 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002400",
    name: "OXY-2-5",
    description: "OXY-2-5 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002401",
    name: "OXY-2-6",
    description: "OXY-2-6 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002402",
    name: "OXY-2-7",
    description: "OXY-2-7 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002403",
    name: "OXY-2-8",
    description: "OXY-2-8 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002404",
    name: "OXY-2-9",
    description: "OXY-2-9 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002405",
    name: "OXY-2-10",
    description:
      "OXY-2-10 is a beta-lactamase. From the Pasteur Institute (Genopole) list of OXY beta-lactamases.",
  },
  {
    accession: "ARO:3002406",
    name: "OXY-2-11",
    description:
      "OXY-2-11 is a beta-lactamase. From the Pasteur Institute (Genopole) list of OXY beta-lactamases.",
  },
  {
    accession: "ARO:3002407",
    name: "OXY-2-12",
    description:
      "OXY-2-12 is a beta-lactamase. From the Pasteur Institute (Genopole) list of OXY beta-lactamases.",
  },
  {
    accession: "ARO:3002408",
    name: "OXY-2-13",
    description:
      "OXY-2-13 is a beta-lactamase. From the Pasteur Institute (Genopole) list of OXY beta-lactamases.",
  },
  {
    accession: "ARO:3002409",
    name: "OXY-3-1",
    description: "OXY-3-1 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002410",
    name: "OXY-4-1",
    description: "OXY-4-1 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002411",
    name: "OXY-5-1",
    description: "OXY-5-1 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002412",
    name: "OXY-5-2",
    description: "OXY-5-2 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002413",
    name: "OXY-6-1",
    description: "OXY-6-1 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002414",
    name: "OXY-6-2",
    description: "OXY-6-2 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002415",
    name: "OXY-6-3",
    description: "OXY-6-3 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002416",
    name: "OXY-6-4",
    description: "OXY-6-4 is a beta-lactamase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002417",
    name: "OKP beta-lactamase",
    description:
      "OKP beta-lactamases are chromosomal class A beta-lactamase that confer resistance to penicillins and early cephalosporins in Klebsiella pneumoniae. OKP beta-lactamases can be subdivided into two groups: OKP-A and OKP-B which diverge by about 4.2%",
  },
  {
    accession: "ARO:3002418",
    name: "OKP-A-1",
    description: "OKP-A-1 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002419",
    name: "OKP-A-2",
    description: "OKP-A-2 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002420",
    name: "OKP-A-3",
    description: "OKP-A-3 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002421",
    name: "OKP-A-4",
    description: "OKP-A-4 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002422",
    name: "OKP-A-5",
    description: "OKP-A-5 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002423",
    name: "OKP-A-6",
    description: "OKP-A-6 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002424",
    name: "OKP-A-7",
    description: "OKP-A-7 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002425",
    name: "OKP-A-8",
    description: "OKP-A-8 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002426",
    name: "OKP-A-9",
    description: "OKP-A-9 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002427",
    name: "OKP-A-10",
    description: "OKP-A-10 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002428",
    name: "OKP-A-11",
    description: "OKP-A-11 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002429",
    name: "OKP-A-12",
    description: "OKP-A-12 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002430",
    name: "OKP-A-13",
    description: "OKP-A-13 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002431",
    name: "OKP-A-14",
    description: "OKP-A-14 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002432",
    name: "OKP-A-15",
    description: "OKP-A-15 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002433",
    name: "OKP-A-16",
    description: "OKP-A-16 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002434",
    name: "OKP-B-1",
    description: "OKP-B-1 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002435",
    name: "OKP-B-2",
    description: "OKP-B-2 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002436",
    name: "OKP-B-3",
    description: "OKP-B-3 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002437",
    name: "OKP-B-4",
    description: "OKP-B-4 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002438",
    name: "OKP-B-5",
    description: "OKP-B-5 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002439",
    name: "OKP-B-6",
    description: "OKP-B-6 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002440",
    name: "OKP-B-7",
    description: "OKP-B-7 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002441",
    name: "OKP-B-8",
    description: "OKP-B-8 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002442",
    name: "OKP-B-9",
    description: "OKP-B-9 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002443",
    name: "OKP-B-10",
    description: "OKP-B-10 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002444",
    name: "OKP-B-11",
    description: "OKP-B-11 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002445",
    name: "OKP-B-12",
    description: "OKP-B-12 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002446",
    name: "OKP-B-13",
    description: "OKP-B-13 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002447",
    name: "OKP-B-14",
    description: "OKP-B-14 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002450",
    name: "OKP-B-17",
    description: "OKP-B-17 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002451",
    name: "OKP-B-18",
    description: "OKP-B-18 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002452",
    name: "OKP-B-19",
    description: "OKP-B-19 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002453",
    name: "OKP-B-20",
    description: "OKP-B-20 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002454",
    name: "LEN-1",
    description: "LEN-1 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002455",
    name: "LEN-2",
    description: "LEN-2 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002456",
    name: "LEN-5",
    description: "LEN-5 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002457",
    name: "LEN-7",
    description: "LEN-7 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002458",
    name: "LEN-8",
    description: "LEN-8 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002459",
    name: "LEN-9",
    description: "LEN-9 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002460",
    name: "LEN-10",
    description: "LEN-10 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002461",
    name: "LEN-11",
    description: "LEN-11 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002462",
    name: "LEN-12",
    description: "LEN-12 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002463",
    name: "LEN-13",
    description:
      "LEN-13 is a beta-lactamase found in Klebsiella pneumoniae and Klebsiella variicola.",
  },
  {
    accession: "ARO:3002464",
    name: "LEN-16",
    description: "LEN-16 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002465",
    name: "LEN-18",
    description: "LEN-18 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002466",
    name: "LEN-19",
    description: "LEN-19 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002467",
    name: "LEN-20",
    description: "LEN-20 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002468",
    name: "LEN-21",
    description: "LEN-21 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002469",
    name: "LEN-22",
    description: "LEN-22 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002470",
    name: "LEN-23",
    description: "LEN-23 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002471",
    name: "LEN-24",
    description: "LEN-24 is a beta-lactamase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002472",
    name: "LEN-4",
    description: "LEN-4 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002473",
    name: "LEN-17",
    description: "LEN-17 is a beta-lactamase found in Escherichia coli",
  },
  {
    accession: "ARO:3002474",
    name: "LEN-26",
    description:
      "LEN-26 is a beta-lactamase found in Escherichia coli and Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3002475",
    name: "LEN-3",
    description:
      "LEN-3 is a beta-lactamase. From the Pasteur Institute list of LEN beta-lactamases.",
  },
  {
    accession: "ARO:3002476",
    name: "LEN-6",
    description:
      "LEN-6 is a beta-lactamase. From the Pasteur Institute list of LEN beta-lactamases.",
  },
  {
    accession: "ARO:3002477",
    name: "LEN-14",
    description:
      "LEN-14 is a beta-lactamase. From the Pasteur Institute list of LEN beta-lactamases.",
  },
  {
    accession: "ARO:3002478",
    name: "LEN-15",
    description:
      "LEN-15 is a beta-lactamase. From the Pasteur Institute list of LEN beta-lactamases.",
  },
  {
    accession: "ARO:3002479",
    name: "LEN-25",
    description:
      "LEN-25 is a beta-lactamase. From the Pasteur Institute list of LEN beta-lactamases.",
  },
  {
    accession: "ARO:3002480",
    name: "OXA-144",
    description: "OXA-144 is a beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002481",
    name: "AER-1",
    description: "AER-1 is a beta-lactamase found in Aeromonas hydrophila",
  },
  {
    accession: "ARO:3002482",
    name: "LRA-1",
    description:
      "LRA-1 is a beta-lactamase isolated from soil samples in Alaska",
  },
  {
    accession: "ARO:3002483",
    name: "LRA-5",
    description:
      "LRA-5 is a beta-lactamase isolated from soil samples in Alaska",
  },
  {
    accession: "ARO:3002484",
    name: "LRA-13",
    description:
      "LRA-13 is a class D/class C fusion bifunctional beta-lactamase isolated from soil samples in Alaska.",
  },
  {
    accession: "ARO:3002485",
    name: "LRA-2",
    description:
      "LRA-2 is a beta-lactamase isolated from soil samples in Alaska",
  },
  {
    accession: "ARO:3002486",
    name: "LRA-7",
    description:
      "LRA-7 is a beta-lactamase isolated from soil samples in Alaska",
  },
  {
    accession: "ARO:3002487",
    name: "LRA-8",
    description:
      "LRA-8 is a beta-lactamase isolated from soil samples in Alaska",
  },
  {
    accession: "ARO:3002488",
    name: "LRA-9",
    description:
      "LRA-9 is a beta-lactamase isolated from soil samples in Alaska",
  },
  {
    accession: "ARO:3002489",
    name: "LRA-10",
    description:
      "LRA-10 is a beta-lactamase isolated from soil samples in Alaska",
  },
  {
    accession: "ARO:3002492",
    name: "LRA-18",
    description:
      "LRA-18 is a beta-lactamase isolated from soil samples in Alaska",
  },
  {
    accession: "ARO:3002493",
    name: "SRT-1",
    description:
      "SRT-1, isolated from Serratia marcescens, confers resistance to cephalosporins but not carbapenems, penems and monobactams.",
  },
  {
    accession: "ARO:3002494",
    name: "SRT-2",
    description:
      "SRT-2, isolated from Serratia marcescens, is a chromosomal beta-lactamase that confers resistance to cefotaxime",
  },
  {
    accession: "ARO:3002495",
    name: "CTX-M-85",
    description: "CTX-M-85 is a beta-lactamase found in Salmonella enterica",
  },
  {
    accession: "ARO:3002496",
    name: "OXA-418",
    description:
      "OXA-418 is a beta-lactamase. From the Lahey list of OXA beta-lactamases.",
  },
  {
    accession: "ARO:3002497",
    name: "PDC-1",
    description: "PDC-1 is a beta-lactamase found in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3002498",
    name: "PDC-2",
    description:
      "PDC-2 is a extended-spectrum beta-lactamase found in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3002500",
    name: "PDC-3",
    description:
      "PDC-3 is a extended-spectrum beta-lactamase found in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3002501",
    name: "PDC-4",
    description:
      "PDC-4 is a extended-spectrum beta-lactamase found in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3002502",
    name: "PDC-5",
    description:
      "PDC-5 is a extended-spectrum beta-lactamase found in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3002505",
    name: "PDC-6",
    description: "PDC-6 is a beta-lactamase found in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3002506",
    name: "PDC-7",
    description:
      "PDC-7 is a extended-spectrum beta-lactamase found in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3002507",
    name: "PDC-8",
    description:
      "PDC-8 is a extended-spectrum beta-lactamase found in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3002508",
    name: "PDC-9",
    description:
      "PDC-9 is a extended-spectrum beta-lactamase found in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3002509",
    name: "PDC-10",
    description:
      "PDC-10 is a extended-spectrum beta-lactamase found in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3002510",
    name: "LRA-3",
    description:
      "LRA-3 is a beta-lactamase isolated from soil samples in Alaska",
  },
  {
    accession: "ARO:3002511",
    name: "LRA-12",
    description:
      "LRA-12 is a beta-lactamase isolated from soil samples in Alaska",
  },
  {
    accession: "ARO:3002512",
    name: "LRA-17",
    description:
      "LRA-17 is a beta-lactamase isolated from soil samples in Alaska",
  },
  {
    accession: "ARO:3002513",
    name: "LRA-19",
    description:
      "LRA-19 is a beta-lactamase isolated from soil samples in Alaska",
  },
  {
    accession: "ARO:3002514",
    name: "OCH-1",
    description:
      "OCH-1 beta-lactamase is an Ambler class C chromosomal-encoded beta-lactamases in Ochrobactrum anthropi",
  },
  {
    accession: "ARO:3002515",
    name: "OCH-2",
    description:
      "OCH-2 beta-lactamase is an Ambler class C chromosomal-encoded beta-lactamases in Ochrobactrum anthropi",
  },
  {
    accession: "ARO:3002516",
    name: "OCH-3",
    description:
      "OCH-3 beta-lactamase is an Ambler class C chromosomal-encoded beta-lactamases in Ochrobactrum anthropi",
  },
  {
    accession: "ARO:3002517",
    name: "OCH-4",
    description:
      "OCH-4 beta-lactamase is an Ambler class C chromosomal-encoded beta-lactamases in Ochrobactrum anthropi",
  },
  {
    accession: "ARO:3002518",
    name: "OCH-5",
    description:
      "OCH-5 beta-lactamase is an Ambler class C chromosomal-encoded beta-lactamases in Ochrobactrum anthropi",
  },
  {
    accession: "ARO:3002519",
    name: "OCH-6",
    description:
      "OCH-6 beta-lactamase is an Ambler class C chromosomal-encoded beta-lactamases in Ochrobactrum anthropi",
  },
  {
    accession: "ARO:3002520",
    name: "OCH-7",
    description:
      "OCH-7 beta-lactamase is an Ambler class C chromosomal-encoded beta-lactamases in Ochrobactrum anthropi",
  },
  {
    accession: "ARO:3002521",
    name: "OCH-8",
    description:
      "OCH-8 beta-lactamase is an Ambler class C chromosomal-encoded beta-lactamases in Ochrobactrum anthropi",
  },
  {
    accession: "ARO:3002522",
    name: "novA",
    description:
      "A type III ABC transporter, identified on the novobiocin biosynthetic gene cluster, involved in the transport and resistance of novobiocin.",
  },
  {
    accession: "ARO:3002523",
    name: "AAC(2')-Ia",
    description:
      "AAC(2')-Ia is a chromosomal-encoded aminoglycoside acetyltransferase in P. stuartii",
  },
  {
    accession: "ARO:3002524",
    name: "AAC(2')-Ib",
    description:
      "AAC(2')-Ib is a chromosomal-encoded aminoglycoside acetyltransferase in M. fortuitum and A. baumannii",
  },
  {
    accession: "ARO:3002525",
    name: "AAC(2')-Ic",
    description:
      "AAC(2')-Ic is a chromosomal-encoded aminoglycoside acetyltransferase in M. tuberculosis and M. bovis",
  },
  {
    accession: "ARO:3002526",
    name: "AAC(2')-Id",
    description:
      "AAC(2')-Id is a chromosomal-encoded aminoglycoside acetyltransferase in M. smegmatis",
  },
  {
    accession: "ARO:3002527",
    name: "AAC(2')-Ie",
    description:
      "AAC(2')-Ie is a chromosomal-encoded aminoglycoside acetyltransferase in M. leprae",
  },
  {
    accession: "ARO:3002528",
    name: "AAC(3)-Ia",
    description:
      "AAC(3)-Ia is an aminoglycoside acetyltransferase encoded by plasmids, transposons, integrons in S. marcescens, E. coli, Acinetobacter baumannii, Klebsiella pneumoniae, Klebsiella oxytoca, P. aeruginosa, Salmonella typhimurium and Proteus mirabilis",
  },
  {
    accession: "ARO:3002529",
    name: "AAC(3)-Id",
    description:
      "AAC(3)-Id is an aminoglycoside acetyltransferase encoded by genomic islands and integrons in S. enterica, P. mirabilis and Vibrio fluvialis",
  },
  {
    accession: "ARO:3002530",
    name: "AAC(3)-Ib",
    description:
      "AAC(3)-Ib is an integron-encoded aminoglycoside acetyltransferase in P. aeruginosa",
  },
  {
    accession: "ARO:3002531",
    name: "AAC(3)-Ic",
    description:
      "AAC(3)-Ic is an integron-encoded aminoglycoside acetyltransferase in P. aeruginosa",
  },
  {
    accession: "ARO:3002533",
    name: "AAC(3)-IIa",
    description:
      "AAC(3)-IIa is a plasmid-encoded aminoglycoside acetyltransferase in K. pneumoniae, E. cloacae, Actinobacillus pleuropneumoniae, S. typhimurium, Citrobacter freundii, and P. aeruginosa.",
  },
  {
    accession: "ARO:3002534",
    name: "AAC(3)-IIb",
    description:
      "AAC(3)-IIb is an aminoglycoside acetyltransferase in E. coli, A. faecalis and S. marcescens",
  },
  {
    accession: "ARO:3002535",
    name: "AAC(3)-IIc",
    description:
      "AAC(3)-IIc is a plasmid-encoded aminoglycoside acetyltransferase in E. coli and P. aeruginosa",
  },
  {
    accession: "ARO:3002536",
    name: "AAC(3)-IIIa",
    description:
      "AAC(3)-IIIa is a chromosomal-encoded aminoglycoside acetyltransferase in P. aeruginosa",
  },
  {
    accession: "ARO:3002537",
    name: "AAC(3)-IIIb",
    description:
      "AAC(3)-IIIb is an aminoglycoside acetyltransferase in P. aeruginosa",
  },
  {
    accession: "ARO:3002538",
    name: "AAC(3)-IIIc",
    description:
      "AAC(3)-IIIc is an aminoglycoside acetyltransferase in P. aeruginosa",
  },
  {
    accession: "ARO:3002539",
    name: "AAC(3)-IV",
    description:
      "AAC(3)-IV is a plasmid-encoded aminoglycoside acetyltransferase in E. coli, C. jejuni and P. stutzeri",
  },
  {
    accession: "ARO:3002540",
    name: "AAC(3)-VIa",
    description:
      "AAC(3)-VIa is a plasmid-encoded aminoglycoside acetyltransferase in E. cloacae, S. enterica and E. coli",
  },
  {
    accession: "ARO:3002541",
    name: "AAC(3)-VIIa",
    description:
      "AAC(3)-VIIa is a chromosomal-encoded aminoglycoside acetyltransferase in Streptomyces rimosus",
  },
  {
    accession: "ARO:3002542",
    name: "AAC(3)-VIIIa",
    description:
      "AAC(3)-VIIIa is a chromosomal-encoded aminoglycoside acetyltransferase in Streptomyces fradiae",
  },
  {
    accession: "ARO:3002543",
    name: "AAC(3)-IXa",
    description:
      "AAC(3)-IXa is a chromosomal-encoded aminoglycoside acetyltransferase in Micromonospora chalcea",
  },
  {
    accession: "ARO:3002544",
    name: "AAC(3)-Xa",
    description:
      "AAC(3)-Xa is a chromosomal-encoded aminoglycoside acetyltransferase in Streptomyces griseus",
  },
  {
    accession: "ARO:3002545",
    name: "AAC(6')-Ia",
    description:
      "AAC(6')-Ia is an aminoglycoside acetyltransferase encoded by plasmids, transposons, integrons in Citrobacter diversus, E. coli, K. pneumoniae, Shigella sonnei, and P. aeruginosa.",
  },
  {
    accession: "ARO:3002546",
    name: "AAC(6')-Ib",
    description:
      "AAC(6')-Ib is an aminoglycoside acetyltransferase encoded by plasmids, transposons, integrons in K. pneumoniae, P. mirabilis, P. aeruginosa, S. enterica, K. oxytoca, S. maltophilia, E. cloacae and V. cholerae",
  },
  {
    accession: "ARO:3002547",
    name: "AAC(6')-Ib-cr",
    description:
      "AAC(6')-Ib-cr is an aminoglycoside acetyltransferase encoded by plasmids, transposons, integrons in Enterobacteriaceae. The aac(6')-Ib-cr variant gene can induce resistance against aminoglycoside and fluoroquinolone simultaneously",
  },
  {
    accession: "ARO:3002548",
    name: "AAC(6')-IId",
    description:
      "AAC(6')-IId is an integron-encoded aminoglycoside acetyltransferase in P. aeruginosa, E. cloacae, and C. freundii.",
  },
  {
    accession: "ARO:3002549",
    name: "AAC(6')-Ic",
    description:
      "AAC(6')-Ic is a chromosomal-encoded aminoglycoside acetyltransferase in S. marcescens",
  },
  {
    accession: "ARO:3002553",
    name: "AAC(6')-If",
    description:
      "AAC(6')-If is a plasmid-encoded aminoglycoside acetyltransferase in E. cloacae",
  },
  {
    accession: "ARO:3002554",
    name: "AAC(6')-Ig",
    description:
      "AAC(6')-Ig is a chromosomal-encoded aminoglycoside acetyltransferase in Acinetobacter haemolyticus",
  },
  {
    accession: "ARO:3002555",
    name: "AAC(6')-Ih",
    description:
      "AAC(6')-Ih is a plasmid-encoded aminoglycoside acetyltransferase in A. baumannii",
  },
  {
    accession: "ARO:3002556",
    name: "AAC(6')-Ii",
    description:
      "AAC(6')-Ii is a chromosomal-encoded aminoglycoside acetyltransferase in Enterococcus spp.",
  },
  {
    accession: "ARO:3002557",
    name: "AAC(6')-Ij",
    description:
      "AAC(6')-Ij is a chromosomal-encoded aminoglycoside acetyltransferase in Acinetobacter genomosp. 13",
  },
  {
    accession: "ARO:3002558",
    name: "AAC(6')-Ik",
    description:
      "AAC(6')-Ik is a chromosomal-encoded aminoglycoside acetyltransferase in Acinetobacter sp.",
  },
  {
    accession: "ARO:3002559",
    name: "AAC(6')-Ip",
    description:
      "AAC(6')-Ip is a aminoglycoside acetyltransferase encoded by plasmids and integrons in C. freundii, E. coli, E. faecium and Enterobacter aerogenes",
  },
  {
    accession: "ARO:3002560",
    name: "AAC(6')-Iq",
    description:
      "AAC(6')-Iq is a aminoglycoside acetyltransferase encoded by plasmids and integrons in K. pneumoniae",
  },
  {
    accession: "ARO:3002561",
    name: "AAC(6')-Ir",
    description:
      "AAC(6')-Ir is a chromosomal-encoded aminoglycoside acetyltransferase in Acinetobacter genomosp. 14",
  },
  {
    accession: "ARO:3002562",
    name: "AAC(6')-Is",
    description:
      "AAC(6')-Is is a chromosomal-encoded aminoglycoside acetyltransferase in Acinetobacter genomosp. 15",
  },
  {
    accession: "ARO:3002563",
    name: "AAC(6')-Isa",
    description:
      "AAC(6')-Isa is a plasmid-encoded aminoglycoside acetyltransferase in Streptomyces albulus",
  },
  {
    accession: "ARO:3002564",
    name: "AAC(6')-It",
    description:
      "AAC(6')-It is a chromosomal-encoded aminoglycoside acetyltransferase in Acinetobacter genomosp. 16",
  },
  {
    accession: "ARO:3002565",
    name: "AAC(6')-Iu",
    description:
      "AAC(6')-Iu is a chromosomal-encoded aminoglycoside acetyltransferase in Acinetobacter genomosp. 17",
  },
  {
    accession: "ARO:3002566",
    name: "AAC(6')-Iv",
    description:
      "AAC(6')-Iv is a chromosomal-encoded aminoglycoside acetyltransferase in Acinetobacter sp.",
  },
  {
    accession: "ARO:3002567",
    name: "AAC(6')-Iw",
    description:
      "AAC(6')-Iw is a chromosomal-encoded aminoglycoside acetyltransferase in Acinetobacter sp.",
  },
  {
    accession: "ARO:3002568",
    name: "AAC(6')-Ix",
    description:
      "AAC(6')-Ix is a chromosomal-encoded aminoglycoside acetyltransferase in Acinetobacter sp.",
  },
  {
    accession: "ARO:3002569",
    name: "AAC(6')-Iy",
    description:
      "AAC(6')-Iy is a chromosomal-encoded aminoglycoside acetyltransferase in S. enteritidis and S. enterica. Regulatory mutation required to increase expression of this chromosomally-encoded gene for resistance. In the specific system, aminoglycoside resistance was due to a transcriptional fusion secondary to a chromosomal deletion in which the downstream aac(6')-Iy gene was placed under the control of the upstream nmpC promoter.",
  },
  {
    accession: "ARO:3002570",
    name: "AAC(6')-Iz",
    description:
      "AAC(6')-Iz is a chromosomal-encoded aminoglycoside acetyltransferase in S. maltophilia",
  },
  {
    accession: "ARO:3002571",
    name: "AAC(6')-Iaa",
    description:
      "AAC(6')-Iaa is a chromosomal-encoded aminoglycoside acetyltransferase in S. typhimurium",
  },
  {
    accession: "ARO:3002572",
    name: "AAC(6')-Iad",
    description:
      "AAC(6')-Iad is a plasmid-encoded aminoglycoside acetyltransferase in Acinetobacter genomosp. 3",
  },
  {
    accession: "ARO:3002573",
    name: "AAC(6')-Iae",
    description:
      "AAC(6')-Iae is an integron-encoded aminoglycoside acetyltransferase in P. aeruginosa and S. enterica",
  },
  {
    accession: "ARO:3002574",
    name: "AAC(6')-Iaf",
    description:
      "AAC(6')-Iaf is an aminoglycoside acetyltransferase encoded by plasmids and integrons in P. aeruginosa",
  },
  {
    accession: "ARO:3002575",
    name: "AAC(6')-Iai",
    description:
      "AAC(6')-Iai is an aminoglycoside acetyltransferase encoded by plasmids and integrons in P. aeruginosa",
  },
  {
    accession: "ARO:3002576",
    name: "AAC(6')-Ib3",
    description:
      "AAC(6')-Ib3 is an integron-encoded aminoglycoside acetyltransferase in P. aeruginosa",
  },
  {
    accession: "ARO:3002577",
    name: "AAC(6')-Ib4",
    description:
      "AAC(6')-Ib4 is an aminoglycoside acetyltransferase in Serratia spp.",
  },
  {
    accession: "ARO:3002578",
    name: "AAC(6')-Ib7",
    description:
      "AAC(6')-Ib7 is a plasmid-encoded aminoglycoside acetyltransferase in E. cloacae and C. freundii",
  },
  {
    accession: "ARO:3002579",
    name: "AAC(6')-Ib8",
    description:
      "AAC(6')-Ib8 is a plasmid-encoded aminoglycoside acetyltransferase in E. cloacae",
  },
  {
    accession: "ARO:3002580",
    name: "AAC(6')-Ib9",
    description:
      "AAC(6')-Ib9 is an integron-encoded aminoglycoside acetyltransferase in P. aeruginosa",
  },
  {
    accession: "ARO:3002581",
    name: "AAC(6')-Ib10",
    description:
      "AAC(6')-Ib10 is an integron-encoded aminoglycoside acetyltransferase in P. aeruginosa",
  },
  {
    accession: "ARO:3002582",
    name: "AAC(6')-Ib11",
    description:
      "AAC(6')-Ib11 is an integron-encoded aminoglycoside acetyltransferase in S. enterica",
  },
  {
    accession: "ARO:3002583",
    name: "AAC(6')-29a",
    description:
      "AAC(6')-29a is an integron-encoded aminoglycoside acetyltransferase in P. aeruginosa",
  },
  {
    accession: "ARO:3002584",
    name: "AAC(6')-29b",
    description:
      "AAC(6')-29b is an integron-encoded aminoglycoside acetyltransferase in P. aeruginosa",
  },
  {
    accession: "ARO:3002585",
    name: "AAC(6')-31",
    description:
      "AAC(6')-31 is an integron-encoded aminoglycoside acetyltransferase in Pseudomonas putida, A. baumannii and K. pneumoniae",
  },
  {
    accession: "ARO:3002586",
    name: "AAC(6')-32",
    description:
      "AAC(6')-32 is an aminoglycoside acetyltransferase encoded by plasmids and integrons in P. aeruginosa",
  },
  {
    accession: "ARO:3002587",
    name: "AAC(6')-33",
    description:
      "AAC(6')-33 is an integron-encoded aminoglycoside acetyltransferase in P. aeruginosa",
  },
  {
    accession: "ARO:3002588",
    name: "AAC(6')-I30",
    description:
      "AAC(6')-I30 is an integron-encoded aminoglycoside acetyltransferase in S. enterica",
  },
  {
    accession: "ARO:3002589",
    name: "AAC(6')-Iid",
    description:
      "AAC(6')-Iid is a chromosomal-encoded aminoglycoside acetyltransferase in Enterococcus durans",
  },
  {
    accession: "ARO:3002590",
    name: "AAC(6')-Iih",
    description:
      "AAC(6')-Iih is a chromosomal-encoded aminoglycoside acetyltransferase in Enterococcus hirae",
  },
  {
    accession: "ARO:3002591",
    name: "AAC(6')-Ib-Suzhou",
    description:
      "AAC(6')-Ib-Suzhou is an aminoglycoside acetyltransferase in E. cloacae and K. pneumoniae",
  },
  {
    accession: "ARO:3002592",
    name: "AAC(6')-Ib-Hangzhou",
    description:
      "AAC(6')-Ib-Hangzhou is an aminoglycoside acetyltransferase in A. baumannii",
  },
  {
    accession: "ARO:3002593",
    name: "AAC(6')-Ib-SK",
    description:
      "AAC(6')-Ib-SK is a chromosomal-encoded aminoglycoside acetyltransferase in Streptomyces kanamyceticus",
  },
  {
    accession: "ARO:3002594",
    name: "AAC(6')-IIa",
    description:
      "AAC(6')-IIa is an aminoglycoside acetyltransferase encoded by plasmids and integrons in P. aeruginosa and S. enterica",
  },
  {
    accession: "ARO:3002595",
    name: "AAC(6')-IIb",
    description:
      "AAC(6')-IIb is an integron-encoded aminoglycoside acetyltransferase in P. fluorescens",
  },
  {
    accession: "ARO:3002596",
    name: "AAC(6')-IIc",
    description:
      "AAC(6')-IIc is an aminoglycoside acetyltransferase encoded by plasmids and integrons in E. cloacae",
  },
  {
    accession: "ARO:3002597",
    name: "AAC(6')-Ie-APH(2'')-Ia",
    description:
      "AAC(6')-Ie-APH(2'')-Ia is an aminoglycoside acetyltransferase encoded by plasmids and transposons in S. aureus, E. faecalis, E. faecium and Staphylococcus warneri",
  },
  {
    accession: "ARO:3002598",
    name: "ANT(3'')-Ii-AAC(6')-IId fusion protein",
    description:
      "ANT(3'')-Ii-AAC(6')-Iid is an integron-encoded aminoglycoside acetyltransferase in S. marcescens.",
  },
  {
    accession: "ARO:3002599",
    name: "AAC(6')-30/AAC(6')-Ib' fusion protein",
    description:
      "AAC(6')-30/AAC(6')-Ib' is an integron-encoded aminoglycoside acetyltransferase in P. aeruginosa",
  },
  {
    accession: "ARO:3002600",
    name: "AAC(3)-Ib/AAC(6')-Ib''",
    description:
      "AAC(3)-Ib/AAC(6')-Ib'' is an integron-encoded aminoglycoside acetyltransferase in P. aeruginosa",
  },
  {
    accession: "ARO:3002601",
    name: "aadA",
    description:
      "ANT(3'')-Ia is an aminoglycoside nucleotidyltransferase gene encoded by plasmids, transposons, integrons in Enterobacteriaceae, A. baumannii, P. aeruginosa and Vibrio cholerae",
  },
  {
    accession: "ARO:3002602",
    name: "aadA2",
    description:
      "aadA2 is an aminoglycoside nucleotidyltransferase gene encoded by plasmids and integrons in K. pneumoniae, Salmonella spp., Corynebacterium glutamicum, C. freundii and Aeromonas spp.",
  },
  {
    accession: "ARO:3002603",
    name: "aadA3",
    description:
      "aadA3 is an aminoglycoside nucleotidyltransferase gene encoded by plasmids, transposons and integrons in E. coli",
  },
  {
    accession: "ARO:3002604",
    name: "aadA4",
    description:
      "aadA4 is an aminoglycoside nucleotidyltransferase gene encoded by plasmids and chromosomes in Bordetella parapertussis and E. coli",
  },
  {
    accession: "ARO:3002605",
    name: "aadA5",
    description:
      "aadA5 is an aminoglycoside nucleotidyltransferase gene encoded by plasmids, transposons and integrons in E. coli, K. pneumoniae, Kluyvera georgiana, P. aeruginosa and E. cloacae",
  },
  {
    accession: "ARO:3002606",
    name: "aadA6",
    description:
      "aadA6 is an integron-encoded aminoglycoside nucleotidyltransferase gene in P. aeruginosa",
  },
  {
    accession: "ARO:3002607",
    name: "aadA7",
    description:
      "aadA7 is an integron-encoded aminoglycoside nucleotidyltransferase gene in V. fluvialis, P. aeruginosa, E. coli, V. cholerae and S. enterica",
  },
  {
    accession: "ARO:3002608",
    name: "aadA8",
    description:
      "aadA8 is an aminoglycoside nucleotidyltransferase gene encoded by plasmids and integrons in V. cholerae, K. pneumoniae and Bacillus endophyticus",
  },
  {
    accession: "ARO:3002609",
    name: "aadA9",
    description:
      "aadA9 is a plasmid-encoded aminoglycoside nucleotidyltransferase gene in C. glutamicum",
  },
  {
    accession: "ARO:3002611",
    name: "aadA11",
    description:
      "aadA11 is an integron-encoded aminoglycoside nucleotidyltransferase gene in E. coli and P. aeruginosa",
  },
  {
    accession: "ARO:3002612",
    name: "aadA12",
    description:
      "aadA12 is an integron-encoded aminoglycoside nucleotidyltransferase gene in E. coli, Yersinia enterocolitica and S. enterica",
  },
  {
    accession: "ARO:3002613",
    name: "aadA13",
    description:
      "aadA13 is an aminoglycoside nucleotidyltransferase gene encoded by plasmids and integrons in Pseudomonas rettgeri, P. aeruginosa, Y. enterocolitica and E. coli",
  },
  {
    accession: "ARO:3002614",
    name: "aadA14",
    description:
      "aadA14 is a plasmid-encoded aminoglycoside nucleotidyltransferase gene in Pasteurella multocida",
  },
  {
    accession: "ARO:3002615",
    name: "aadA15",
    description:
      "aadA15 is an integron-encoded aminoglycoside nucleotidyltransferase gene in P. aeruginosa",
  },
  {
    accession: "ARO:3002616",
    name: "aadA16",
    description:
      "aadA16 is an aminoglycoside nucleotidyltransferase gene encoded by plasmids and integrons in E. coli, V. cholerae and K. pneumoniae",
  },
  {
    accession: "ARO:3002617",
    name: "aadA17",
    description:
      "aadA17 is an integron-encoded aminoglycoside nucleotidyltransferase gene in Aeromonas media",
  },
  {
    accession: "ARO:3002618",
    name: "aadA21",
    description:
      "aadA21 is an integron-encoded aminoglycoside nucleotidyltransferase gene in Salmonella spp.",
  },
  {
    accession: "ARO:3002619",
    name: "aadA22",
    description:
      "aadA22 is an aminoglycoside nucleotidyltransferase gene encoded by plasmids and integrons in S. enterica and E. coli",
  },
  {
    accession: "ARO:3002620",
    name: "aadA23",
    description:
      "aadA23 is an integron-encoded aminoglycoside nucleotidyltransferase gene in S. enterica",
  },
  {
    accession: "ARO:3002621",
    name: "aadA24",
    description:
      "aadA24 is an integron-encoded aminoglycoside nucleotidyltransferase gene in Salmonella spp.",
  },
  {
    accession: "ARO:3002622",
    name: "aadA6/aadA10",
    description:
      "aadA6/aadA10 is an integron-encoded aminoglycoside nucleotidyltransferase gene cassette in P. aeruginosa",
  },
  {
    accession: "ARO:3002623",
    name: "ANT(4')-Ia",
    description:
      "ANT(4')-Ia is a plasmid-encoded aminoglycoside nucleotidyltransferase in S. epidermidis, S. aureus, Enterococcus spp. and Bacillus spp.",
  },
  {
    accession: "ARO:3002624",
    name: "ANT(4')-IIa",
    description:
      "ANT(4')-IIa is a plasmid-encoded aminoglycoside nucleotidyltransferase in P. aeruginosa and Enterobacteriaceae",
  },
  {
    accession: "ARO:3002625",
    name: "ANT(4')-IIb",
    description:
      "ANT(4')-IIb is a transposon-encoded aminoglycoside nucleotidyltransferase in P. aeruginosa",
  },
  {
    accession: "ARO:3002626",
    name: "ANT(6)-Ia",
    description:
      "ANT(6)-Ia is an aminoglycoside nucleotidyltransferase gene encoded by plasmids and chromosomes in Staphylococcus epidermidis, E. faecium, Streptococcus suis, S. aureus, E. faecalis and Streptococcus mitis",
  },
  {
    accession: "ARO:3002627",
    name: "aadK",
    description:
      "aadK is a chromosomal-encoded aminoglycoside nucleotidyltransferase gene in B. subtilis and Bacillus spp.",
  },
  {
    accession: "ARO:3002628",
    name: "aad(6)",
    description:
      "aad(6) is a plasmid-encoded aminoglycoside nucleotidyltransferase gene in E. faecalis and Streptococcus oralis",
  },
  {
    accession: "ARO:3002629",
    name: "ANT(6)-Ib",
    description:
      "ANT(6)-Ib is an aminoglycoside nucleotidyltransferase gene encoded by transferable pathogenicity islands in C. fetus subsp. fetus and B. subtilis",
  },
  {
    accession: "ARO:3002630",
    name: "ANT(9)-Ia",
    description:
      "ANT(9)-Ia is an aminoglycoside nucleotidyltransferase encoded by plasmids and transposons in S. aureus, Enterococcus spp., Staphylococcus sciuri and E. faecalis",
  },
  {
    accession: "ARO:3002631",
    name: "spd",
    description:
      "spd is a plasmid-encoded aminoglycoside nucleotidyltransferase gene in S. aureus",
  },
  {
    accession: "ARO:3002634",
    name: "APH(2'')-Ie",
    description:
      "APH(2'')-Ie is a plasmid or transposon-encoded aminoglycoside phosphotransferase in E. faecium and E. casseliflavus",
  },
  {
    accession: "ARO:3002635",
    name: "APH(2'')-IIa",
    description:
      "APH(2'')-IIa is a chromosomal-encoded aminoglycoside phosphotransferase in E. faecium and E. coli",
  },
  {
    accession: "ARO:3002636",
    name: "APH(2'')-IIIa",
    description:
      "APH(2'')-IIIa is a plasmid-encoded aminoglycoside phosphotransferase in Enterococcus gallinarum",
  },
  {
    accession: "ARO:3002637",
    name: "APH(2'')-IVa",
    description:
      "APH(2'')-IVa is a chromosomal-encoded aminoglycoside phosphotransferase in E. casseliflavus",
  },
  {
    accession: "ARO:3002638",
    name: "APH(3'')-Ia",
    description:
      "APH(3'')-Ia is a chromosomal-encoded aminoglycoside phosphotransferase in S. griseus",
  },
  {
    accession: "ARO:3002639",
    name: "APH(3'')-Ib",
    description:
      "APH(3'')-Ib is an aminoglycoside phosphotransferase encoded by plasmids, transposons, integrative conjugative elements and chromosomes in Enterobacteriaceae and Pseudomonas spp.",
  },
  {
    accession: "ARO:3002640",
    name: "APH(3'')-Ic",
    description:
      "APH(3'')-Ic is a chromosomal-encoded aminoglycoside phosphotransferase in M. fortuitum",
  },
  {
    accession: "ARO:3002641",
    name: "APH(3')-Ia",
    description:
      "APH(3')-Ia is a transposon-encoded aminoglycoside phosphotransferase in E. coli and S. enterica. It is identical at the protein sequence to APH(3')-Ic, an aminoglycoside phosphotransferase encoded by plasmids, transposons and genomic islands in K. pneumoniae, A. baumannii, S. marcescens, Corynebacterium spp., Photobacterium spp. and Citrobacter spp.",
  },
  {
    accession: "ARO:3002642",
    name: "APH(3')-Ib",
    description:
      "APH(3')-Ib is a plasmid-encoded aminoglycoside phosphotransferase in E. coli",
  },
  {
    accession: "ARO:3002644",
    name: "APH(3')-IIa",
    description:
      "APH(3')-IIa is a transposon-encoded aminoglycoside phosphotransferase in E. coli",
  },
  {
    accession: "ARO:3002645",
    name: "APH(3')-IIb",
    description:
      "APH(3')-IIb is a chromosomal-encoded aminoglycoside phosphotransferase in P. aeruginosa",
  },
  {
    accession: "ARO:3002646",
    name: "APH(3')-IIc",
    description:
      "APH(3')-IIc is a chromosomal-encoded aminoglycoside phosphotransferase in S. maltophilia",
  },
  {
    accession: "ARO:3002647",
    name: "APH(3')-IIIa",
    description:
      "APH(3')-IIIa is a plasmid-encoded aminoglycoside phosphotransferase in S. aureus and Enterococcus spp.",
  },
  {
    accession: "ARO:3002648",
    name: "APH(3')-IVa",
    description:
      "APH(3')-IVa is a chromosomal-encoded aminoglycoside phosphotransferase in B. circulans",
  },
  {
    accession: "ARO:3002649",
    name: "APH(3')-Va",
    description:
      "APH(3')-Va is a chromosomal-encoded aminoglycoside phosphotransferase in Streptomyces fradiae",
  },
  {
    accession: "ARO:3002650",
    name: "APH(3')-Vb",
    description:
      "APH(3')-Vb is a chromosomal-encoded aminoglycoside phosphotransferase in Streptomyces ribosidificus",
  },
  {
    accession: "ARO:3002651",
    name: "APH(3')-Vc",
    description:
      "APH(3')-Vc is a chromosomal-encoded aminoglycoside phosphotransferase in M. chalcea",
  },
  {
    accession: "ARO:3002652",
    name: "APH(3')-VIa",
    description:
      "APH(3')-VIa is a plasmid-encoded aminoglycoside phosphotransferase in A. baumannii",
  },
  {
    accession: "ARO:3002653",
    name: "APH(3')-VIb",
    description:
      "APH(3')-VIb is a plasmid-encoded aminoglycoside phosphotransferase in K. pneumoniae and S. marcescens",
  },
  {
    accession: "ARO:3002654",
    name: "APH(3')-VIIa",
    description:
      "APH(3')-VIIa is a plasmid-encoded aminoglycoside phosphotransferase in C. jejuni",
  },
  {
    accession: "ARO:3002655",
    name: "APH(4)-Ia",
    description:
      "APH(4)-Ia is a plasmid-encoded aminoglycoside phosphotransferase in E. coli",
  },
  {
    accession: "ARO:3002656",
    name: "APH(4)-Ib",
    description:
      "APH(4)-Ib is a chromosomal-encoded aminoglycoside phosphotransferase in Pseudomonas pseudomallei",
  },
  {
    accession: "ARO:3002657",
    name: "APH(6)-Ia",
    description:
      "APH(6)-Ia is a chromosomal-encoded aminoglycoside phosphotransferase in S. griseus",
  },
  {
    accession: "ARO:3002658",
    name: "APH(6)-Ib",
    description:
      "APH(6)-Ib is a chromosomal-encoded aminoglycoside phosphotransferase in S. glaucescens",
  },
  {
    accession: "ARO:3002659",
    name: "APH(6)-Ic",
    description:
      "APH(6)-Ic is a transposon-encoded aminoglycoside phosphotransferase in S. enterica, P. aeruginosa and E. coli",
  },
  {
    accession: "ARO:3002660",
    name: "APH(6)-Id",
    description:
      "APH(6)-Id is an aminoglycoside phosphotransferase encoded by plasmids, integrative conjugative elements and chromosomal genomic islands in K. pneumoniae, Salmonella spp., E. coli, Shigella flexneri, Providencia alcalifaciens, Pseudomonas spp., V. cholerae, Edwardsiella tarda, Pasteurella multocida and Aeromonas bestiarum",
  },
  {
    accession: "ARO:3002661",
    name: "APH(7'')-Ia",
    description:
      "APH(7'')-Ia is a chromosomal-encoded aminoglycoside phosphotransferase in S. hygroscopicus",
  },
  {
    accession: "ARO:3002662",
    name: "APH(9)-Ia",
    description:
      "APH(9)-Ia is a chromosomal-encoded aminoglycoside phosphotransferase in L. pneumophila",
  },
  {
    accession: "ARO:3002663",
    name: "APH(9)-Ib",
    description:
      "APH(9)-Ib is a chromosomal-encoded aminoglycoside phosphotransferase in S. flavopersicus",
  },
  {
    accession: "ARO:3002665",
    name: "npmA",
    description:
      "NpmA is a plasmid-mediated aminoglycoside-resistance 16S rRNA methyltransferase by interfering with aminoglycoside binding with the A site of 16S rRNA through N-1 methylation at position A1408.",
  },
  {
    accession: "ARO:3002666",
    name: "rmtF",
    description:
      "RmtF is a 16S rRNA methyltransferase found in Pseudomonas aeruginosa which methylates G1405 of the 16S rRNA. It confers high level resistance to many aminoglycosides.",
  },
  {
    accession: "ARO:3002667",
    name: "rmtD",
    description:
      "RmtD is a 16S rRNA methyltransferase found in Pseudomonas aeruginosa which methylates G1405 of the 16S rRNA. It confers high level resistance to many aminoglycosides.",
  },
  {
    accession: "ARO:3002668",
    name: "rmtG",
    description:
      "RmtG is a 16S rRNA methyltransferase found in Pseudomonas aeruginosa which methylates G1405 of the 16S rRNA. It confers high level resistance to many aminoglycosides.",
  },
  {
    accession: "ARO:3002669",
    name: "APH(2'')-Ig",
    description:
      "APH('')-Ig is a plasmid-encoded aminoglycoside phosphotransferase in Campylobacter coli",
  },
  {
    accession: "ARO:3002670",
    name: "cat",
    description:
      "cat is used to describe many variants of the chloramphenicol acetyltransferase gene in a range of organisms including Acinetobacter calcoaceticus, Agrobacterium tumefaciens, Bacillus clausii, Bacillus subtilis, Campylobacter coli, Enterococcus faecalis, Enterococcus faecium, Lactococcus lactis, Listeria monocytogenes, Listonella anguillarum\nMorganella morganii, Photobacterium damselae subsp. piscicida, Proteus mirabilis, Salmonella typhi, Serratia marcescens, Shigella flexneri, Staphylococcus aureus, Staphylococcus haemolyticus, Staphylococcus intermedius, Streptococcus agalactiae, Streptococcus suis and Streptomyces acrimycini",
  },
  {
    accession: "ARO:3002671",
    name: "Lactobacillus reuteri cat-TC",
    description:
      "cat-TC is a plasmid-encoded variant of the cat gene found in Lactobacillus reuteri",
  },
  {
    accession: "ARO:3002672",
    name: "Bacillus pumilus cat86",
    description:
      "cat86 is a chromosome-encoded variant of the cat gene found in Bacillus pumilus",
  },
  {
    accession: "ARO:3002674",
    name: "Clostridium butyricum catB",
    description:
      "catB is a chromosome-encoded variant of the cat gene found in Clostridium butyricum",
  },
  {
    accession: "ARO:3002675",
    name: "catB2",
    description:
      "catB2 is a plasmid-encoded variant of the cat gene found in Escherichia coli, Salmonella enteritidis and Pasteurella multocida",
  },
  {
    accession: "ARO:3002676",
    name: "catB3",
    description:
      "catB3 is a plasmid or chromosome-encoded variant of the cat gene found in Salmonella typhimurium, Acinetobacter baumannii and Escherichia coli",
  },
  {
    accession: "ARO:3002678",
    name: "Pseudomonas aeruginosa catB6",
    description:
      "catB6 is a plasmid-encoded variant of the cat gene found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002679",
    name: "Pseudomonas aeruginosa catB7",
    description:
      "catB7 is a chromosome-encoded variant of the cat gene found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002680",
    name: "catB8",
    description:
      "catB8 is a plasmid or integron-encoded variant of the cat gene found in Klebsiella pneumoniae, Salmonella typhi and Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002681",
    name: "catB9",
    description:
      "catB9 is a chromosome-encoded variant of the cat gene found in Vibrio cholerae",
  },
  {
    accession: "ARO:3002682",
    name: "catD",
    description:
      "catD is a chromosome and transposon-encoded variant of the cat gene found in Clostridium difficile",
  },
  {
    accession: "ARO:3002683",
    name: "catI",
    description:
      "catI is a chromosome and transposon-encoded variant of the cat gene found in Escherichia coli and Acinetobacter baumannii",
  },
  {
    accession: "ARO:3002684",
    name: "catII",
    description:
      "catII is a plasmid-encoded variant of the cat gene found in Haemophilus influenzae, Agrobacterium tumefaciens and Escherichia coli",
  },
  {
    accession: "ARO:3002685",
    name: "catIII",
    description:
      "catIII is a plasmid-encoded variant of the cat gene found in Shigella flexneri",
  },
  {
    accession: "ARO:3002686",
    name: "catP",
    description:
      "catP is a transposon and chromosome-encoded variant of the cat gene found in Clostridium perfringens and Neisseria meningitidis",
  },
  {
    accession: "ARO:3002687",
    name: "catQ",
    description:
      "catQ is a chromosome-encoded variant of the cat gene found in Clostridium perfringens",
  },
  {
    accession: "ARO:3002688",
    name: "catS",
    description:
      "catS is a chromosome-encoded variant of the cat gene found in Streptococcus pyogenes",
  },
  {
    accession: "ARO:3002689",
    name: "plasmid-encoded cat (pp-cat)",
    description:
      "pp-cat is a plasmid-encoded variant of the cat gene found in Photobacterium damselae subsp. piscicida",
  },
  {
    accession: "ARO:3002690",
    name: "Streptomyces lividans cmlR",
    description:
      "cmlR is a plasmid or chromosome-encoded chloramphenicol resistance determinant (putative transmembrane protein) that is found in Escherichia coli and Streptomyces lividans",
  },
  {
    accession: "ARO:3002691",
    name: "Salmonella enterica cmlA",
    description:
      "cmlA is a plasmid-encoded chloramphenicol exporter that is found in Salmonella typhimurium",
  },
  {
    accession: "ARO:3002693",
    name: "cmlA1",
    description:
      "cmlA1 is a plasmid or transposon-encoded chloramphenicol exporter that is found in Pseudomonas aeruginosa and Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002694",
    name: "cmlA4",
    description:
      "cmlA4 is a plasmid-encoded chloramphenicol exporter that is found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002695",
    name: "cmlA5",
    description:
      "cmlA5 is a plasmid or transposon-encoded chloramphenicol exporter that is found in Escherichia coli",
  },
  {
    accession: "ARO:3002696",
    name: "cmlA6",
    description:
      "cmlA6 is a plasmid-encoded chloramphenicol exporter that is found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002698",
    name: "cmlB",
    description:
      "cmlB is a plasmid-encoded chloramphenicol exporter that is found in Enterobacter aerogenes",
  },
  {
    accession: "ARO:3002699",
    name: "cmlB1",
    description:
      "cmlB1 is a plasmid-encoded chloramphenicol exporter that is found in Bordetella bronchiseptica",
  },
  {
    accession: "ARO:3002700",
    name: "cmlv",
    description:
      "cmlv is a chromosome-encoded chloramphenicol phoshotransferase that is found in Streptomyces venezuelae",
  },
  {
    accession: "ARO:3002701",
    name: "Rhodococcus fascians cmr",
    description:
      "cmr is a plasmid-encoded chloramphenicol exporter that is found in Rhodococcus fascians and Corynebacterium glutamicum",
  },
  {
    accession: "ARO:3002702",
    name: "cmrA",
    description:
      "cmrA is a transposon-encoded chloramphenicol exporter that is found in Rhodococcus rhodochrous",
  },
  {
    accession: "ARO:3002703",
    name: "cmx",
    description:
      "cmx is a plasmid or transposon-encoded chloramphenicol exporter that is found in Corynebacterium striatum and Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3002704",
    name: "fexA",
    description:
      "fexA is a plasmid-encoded chloramphenicol exporter that is found in Staphylococcus lentus",
  },
  {
    accession: "ARO:3002705",
    name: "floR",
    description:
      "floR is a plasmid or chromosome-encoded chloramphenicol exporter that is found in Bordetella bronchiseptica, Escherichia coli, Klebsiella pneumoniae,\nSalmonella typhimurium DT104 and Vibrio cholerae",
  },
  {
    accession: "ARO:3002706",
    name: "ethambutol resistant embC",
    description:
      "Mycobacterium tuberculosis embC mutations at codon 270 confer resistance to ethambutol.",
  },
  {
    accession: "ARO:3002707",
    name: "QnrA1",
    description:
      "QnrA1 is a plasmid-mediated quinolone resistance protein found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002708",
    name: "QnrA2",
    description:
      "QnrA2 is a plasmid-mediated quinolone resistance protein found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002709",
    name: "QnrA3",
    description:
      "QnrA3 is a plasmid-mediated quinolone resistance protein found in Shewanella algae",
  },
  {
    accession: "ARO:3002710",
    name: "QnrA4",
    description:
      "QnrA4 is a plasmid-mediated quinolone resistance protein found in Shewanella algae",
  },
  {
    accession: "ARO:3002711",
    name: "QnrA5",
    description:
      "QnrA5 is a plasmid-mediated quinolone resistance protein found in Shewanella algae",
  },
  {
    accession: "ARO:3002712",
    name: "QnrA6",
    description:
      "QnrA6 is a plasmid-mediated quinolone resistance protein found in Proteus mirabilis",
  },
  {
    accession: "ARO:3002713",
    name: "QnrA7",
    description:
      "QnrA7 is a plasmid-mediated quinolone resistance protein found in Shewanella algae",
  },
  {
    accession: "ARO:3002714",
    name: "QnrB1",
    description:
      "QnrB1 is a plasmid-mediated quinolone resistance protein found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002715",
    name: "QnrB2",
    description:
      "QnrB2 is a plasmid-mediated quinolone resistance protein found in Citrobacter koseri",
  },
  {
    accession: "ARO:3002716",
    name: "QnrB3",
    description:
      "QnrB3 is a plasmid-mediated quinolone resistance protein found in Escherichia coli",
  },
  {
    accession: "ARO:3002718",
    name: "QnrB4",
    description:
      "QnrB4 is a plasmid-mediated quinolone resistance protein found in Escherichia coli",
  },
  {
    accession: "ARO:3002719",
    name: "QnrB5",
    description:
      "QnrB5 is a plasmid-mediated quinolone resistance protein found in Salmonella enterica",
  },
  {
    accession: "ARO:3002720",
    name: "QnrB6",
    description:
      "QnrB6 is a plasmid-mediated quinolone resistance protein found in Pantoea agglomerans",
  },
  {
    accession: "ARO:3002721",
    name: "QnrB7",
    description:
      "QnrB7 is a plasmid-mediated quinolone resistance protein found in Enterobacter cloacae",
  },
  {
    accession: "ARO:3002722",
    name: "QnrB8",
    description:
      "QnrB8 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002723",
    name: "QnrB9",
    description:
      "QnrB9 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002724",
    name: "QnrB10",
    description:
      "QnrB10 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002725",
    name: "QnrB11",
    description:
      "QnrB11 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002726",
    name: "QnrB12",
    description:
      "QnrB12  is a plasmid-mediated quinolone resistance protein found in Citrobacter werkmanii",
  },
  {
    accession: "ARO:3002727",
    name: "QnrB13",
    description:
      "QnrB13 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002728",
    name: "QnrB14",
    description:
      "QnrB14 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002730",
    name: "QnrB15",
    description:
      "QnrB15 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002731",
    name: "QnrB16",
    description:
      "QnrB16 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002732",
    name: "QnrB17",
    description:
      "QnrB17 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002733",
    name: "QnrB18",
    description:
      "QnrB18 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002734",
    name: "QnrB19",
    description:
      "QnrB19 is a plasmid-mediated quinolone resistance protein found in Escherichia coli",
  },
  {
    accession: "ARO:3002735",
    name: "QnrB20",
    description:
      "QnrB20 is a plasmid-mediated quinolone resistance protein found in Escherichia coli",
  },
  {
    accession: "ARO:3002736",
    name: "QnrB21",
    description:
      "QnrB21 is a plasmid-mediated quinolone resistance protein found in Escherichia coli",
  },
  {
    accession: "ARO:3002737",
    name: "QnrB22",
    description:
      "QnrB22 is a plasmid-mediated quinolone resistance protein found in Citrobacter werkmanii",
  },
  {
    accession: "ARO:3002738",
    name: "QnrB23",
    description:
      "QnrB23 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002739",
    name: "QnrB24",
    description:
      "QnrB24 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002740",
    name: "QnrB25",
    description:
      "QnrB25 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002741",
    name: "QnrB26",
    description:
      "QnrB26 is a plasmid-mediated quinolone resistance protein found in Proteus vulgaris",
  },
  {
    accession: "ARO:3002742",
    name: "QnrB27",
    description:
      "QnrB27 is a plasmid-mediated quinolone resistance protein found in Citrobacter braakii",
  },
  {
    accession: "ARO:3002743",
    name: "QnrB28",
    description:
      "QnrB28 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002744",
    name: "QnrB29",
    description:
      "QnrB29 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002745",
    name: "QnrB30",
    description:
      "QnrB30 is a plasmid-mediated quinolone resistance protein found in Citrobacter braakii",
  },
  {
    accession: "ARO:3002746",
    name: "QnrB31",
    description:
      "QnrB31 is a plasmid-mediated quinolone resistance protein found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002747",
    name: "QnrB32",
    description:
      "QnrB32 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002748",
    name: "QnrB33",
    description:
      "QnrB33 is a plasmid-mediated quinolone resistance protein found in Citrobacter braakii",
  },
  {
    accession: "ARO:3002749",
    name: "QnrB34",
    description:
      "QnrB34 is a plasmid-mediated quinolone resistance protein found in Citrobacter werkmanii",
  },
  {
    accession: "ARO:3002750",
    name: "QnrB35",
    description:
      "QnrB35 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002751",
    name: "QnrB36",
    description:
      "QnrB36 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002752",
    name: "QnrB37",
    description:
      "QnrB37 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002753",
    name: "QnrB38",
    description:
      "QnrB38 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002754",
    name: "QnrB39",
    description:
      "QnrB39 is a chromosome-mediated quinolone resistance protein found in Citrobacter youngae",
  },
  {
    accession: "ARO:3002755",
    name: "QnrB40",
    description:
      "QnrB40 is a plasmid-mediated quinolone resistance protein found in Citrobacter braakii",
  },
  {
    accession: "ARO:3002756",
    name: "QnrB41",
    description:
      "QnrB41 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002757",
    name: "QnrB42",
    description:
      "QnrB42 is a plasmid-mediated quinolone resistance protein found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002758",
    name: "QnrB43",
    description:
      "QnrB43 is a plasmid-mediated quinolone resistance protein found in Escherichia coli",
  },
  {
    accession: "ARO:3002759",
    name: "QnrB44",
    description:
      "QnrB44 is a plasmid-mediated quinolone resistance protein found in Escherichia coli",
  },
  {
    accession: "ARO:3002760",
    name: "QnrB45",
    description:
      "QnrB45 is a plasmid-mediated quinolone resistance protein found in Escherichia coli",
  },
  {
    accession: "ARO:3002761",
    name: "QnrB46",
    description:
      "QnrB46 is a plasmid-mediated quinolone resistance protein found in Escherichia coli",
  },
  {
    accession: "ARO:3002762",
    name: "QnrB47",
    description:
      "QnrB47 is a plasmid-mediated quinolone resistance protein found in Escherichia coli",
  },
  {
    accession: "ARO:3002763",
    name: "QnrB48",
    description:
      "QnrB48 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002764",
    name: "QnrB49",
    description:
      "QnrB49 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002765",
    name: "QnrB50",
    description:
      "QnrB50 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002767",
    name: "QnrB54",
    description:
      "QnrB54 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002768",
    name: "QnrB55",
    description:
      "QnrB55 is a plasmid-mediated quinolone resistance protein found in Raoultella terrigena",
  },
  {
    accession: "ARO:3002769",
    name: "QnrB56",
    description:
      "QnrB56 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002770",
    name: "QnrB57",
    description:
      "QnrB57 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002771",
    name: "QnrB58",
    description:
      "QnrB58 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002772",
    name: "QnrB59",
    description:
      "QnrB59 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002773",
    name: "QnrB60",
    description:
      "QnrB60 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002774",
    name: "QnrB61",
    description:
      "QnrB61 is a plasmid-mediated quinolone resistance protein found in Citrobacter braakii",
  },
  {
    accession: "ARO:3002775",
    name: "QnrB62",
    description:
      "QnrB62 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002776",
    name: "QnrB64",
    description:
      "QnrB64 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002777",
    name: "QnrB65",
    description:
      "QnrB65 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002778",
    name: "QnrB66",
    description:
      "QnrB66 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002779",
    name: "QnrB67",
    description:
      "QnrB67 is a plasmid-mediated quinolone resistance protein found in Citrobacter braakii",
  },
  {
    accession: "ARO:3002780",
    name: "QnrB68",
    description:
      "QnrB68 is a plasmid-mediated quinolone resistance protein found in Citrobacter braakii",
  },
  {
    accession: "ARO:3002781",
    name: "QnrB69",
    description:
      "QnrB69 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002782",
    name: "QnrB70",
    description:
      "QnrB70 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002783",
    name: "QnrB71",
    description:
      "QnrB71 is a plasmid-mediated quinolone resistance protein found in Citrobacter braakii",
  },
  {
    accession: "ARO:3002784",
    name: "QnrB72",
    description:
      "QnrB72 is a plasmid-mediated quinolone resistance protein found in Citrobacter sp. TR21_24",
  },
  {
    accession: "ARO:3002785",
    name: "QnrB73",
    description:
      "QnrB73 is a plasmid-mediated quinolone resistance protein found in Citrobacter freundii",
  },
  {
    accession: "ARO:3002786",
    name: "QnrB74",
    description:
      "QnrB74 is a plasmid-mediated quinolone resistance protein found in Escherichia coli",
  },
  {
    accession: "ARO:3002787",
    name: "QnrC",
    description:
      "QnrC is a plasmid-mediated quinolone resistance protein found in Proteus mirabilis",
  },
  {
    accession: "ARO:3002788",
    name: "QnrD1",
    description:
      "QnrD1 is a plasmid-mediated quinolone resistance protein found in Salmonella enterica",
  },
  {
    accession: "ARO:3002789",
    name: "QnrD2",
    description:
      "QnrD2 is a plasmid-mediated quinolone resistance protein found in Salmonella enterica",
  },
  {
    accession: "ARO:3002790",
    name: "QnrS1",
    description:
      "QnrS1 is a plasmid-mediated quinolone resistance protein found in Shigella flexneri",
  },
  {
    accession: "ARO:3002791",
    name: "QnrS2",
    description:
      "QnrS2 is a plasmid-mediated quinolone resistance protein found in Salmonella enterica",
  },
  {
    accession: "ARO:3002792",
    name: "QnrS3",
    description:
      "QnrS3 is a plasmid-mediated quinolone resistance protein found in Escherichia coli",
  },
  {
    accession: "ARO:3002793",
    name: "QnrS4",
    description:
      "QnrS4 is a plasmid-mediated quinolone resistance protein found in Salmonella enterica",
  },
  {
    accession: "ARO:3002794",
    name: "QnrS5",
    description:
      "QnrS5 is a plasmid-mediated quinolone resistance protein found in Aeromonas sobria",
  },
  {
    accession: "ARO:3002795",
    name: "QnrS6",
    description:
      "QnrS6 is a plasmid-mediated quinolone resistance protein found in Aeromonas hydrophila",
  },
  {
    accession: "ARO:3002796",
    name: "QnrS7",
    description:
      "QnrS7 is a plasmid-mediated quinolone resistance protein found in Escherichia coli",
  },
  {
    accession: "ARO:3002797",
    name: "QnrS8",
    description:
      "QnrS8 is a plasmid-mediated quinolone resistance protein found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002798",
    name: "QnrS9",
    description:
      "QnrS9 is a plasmid-mediated quinolone resistance protein found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3002799",
    name: "QnrVC1",
    description:
      "QnrVC1 is an integron-mediated quinolone resistance protein found in Vibrio cholerae",
  },
  {
    accession: "ARO:3002800",
    name: "QnrVC3",
    description:
      "QnrVC3 is an integron-mediated quinolone resistance protein found in Vibrio cholerae",
  },
  {
    accession: "ARO:3002801",
    name: "QnrVC4",
    description:
      "QnrVC4 is an integron-mediated quinolone resistance protein found in Aeromonas punctata",
  },
  {
    accession: "ARO:3002802",
    name: "QnrVC5",
    description:
      "QnrVC5 is an integron-mediated quinolone resistance protein found in Vibrio fluvialis",
  },
  {
    accession: "ARO:3002803",
    name: "QnrVC6",
    description:
      "QnrVC6 is an integron-mediated quinolone resistance protein found in Acinetobacter baumannii",
  },
  {
    accession: "ARO:3002804",
    name: "FosA2",
    description:
      "An enzyme that confers resistance to fosfomycin in Enterobacter cloacae by breaking the epoxide ring of the molecule. It depends on the cofactors Manganese (II) and Potassium and uses Glutathione (GSH) as the nucleophilic molecule.",
  },
  {
    accession: "ARO:3002811",
    name: "murA transferase",
    description:
      "murA or UDP-N-acetylglucosamine enolpyruvyl transferase catalyses the initial step in peptidoglycan biosynthesis and is inhibited by fosfomycin. Overexpression of murA through mutations confers fosfomycin resistance.",
  },
  {
    accession: "ARO:3002812",
    name: "pp-flo",
    description:
      "pp-flo is a plasmid chloramphenicol exporter that is found in Photobacterium damselae subsp. piscicida",
  },
  {
    accession: "ARO:3002813",
    name: "lmrB",
    description:
      "lmrB is a chromosomally-encoded efflux pump that confers resistance to lincosamides in Bacillus subtilis",
  },
  {
    accession: "ARO:3002814",
    name: "clbA",
    description:
      "clbA is a plasmid-encoded cfr gene found in Bacillus velezensis (Bacillus amyloliquefaciens subsp. plantarum).",
  },
  {
    accession: "ARO:3002815",
    name: "clbB",
    description: "clbB is a plasmid-encoded cfr gene found in Bacillus brevis",
  },
  {
    accession: "ARO:3002816",
    name: "clbC",
    description: "clbC is a plasmid-encoded cfr gene found in Bacillus clausii",
  },
  {
    accession: "ARO:3002817",
    name: "carA",
    description:
      "carA is an ABC-F subfamily protein involved in macrolide resistance. It is found in Streptomyces thermotolerans",
  },
  {
    accession: "ARO:3002818",
    name: "msrB",
    description:
      "MsrB is an ABC-F subfamily protein expressed to Staphylococcus species that confers resistance to erythromycin and streptogramin B antibiotics. It is associated with plasmid DNA.",
  },
  {
    accession: "ARO:3002819",
    name: "msrC",
    description:
      "msrC is a chromosomal-encoded ABC-F subfamily protein expressed in Enterococcus faecium that confers resistance to erythromycin and other macrolide and streptogramin B antibiotics.",
  },
  {
    accession: "ARO:3002823",
    name: "ErmH",
    description:
      "ErmH is a plasmid-mediated methyltransferase found in Streptomyces thermotolerans",
  },
  {
    accession: "ARO:3002824",
    name: "ErmV",
    description:
      "ErmV is a plasmid-mediated methyltransferase found in Streptomyces viridochromogenes",
  },
  {
    accession: "ARO:3002825",
    name: "ErmY",
    description:
      "ErmY is a plasmid-mediated methyltransferase found in Staphylococcus aureus",
  },
  {
    accession: "ARO:3002826",
    name: "EreA2",
    description:
      "EreA2 is an integron-encoded erythromycin esterase that hydrolyses the drug's lactone ring. EreA2 is found in Providencia stuartii",
  },
  {
    accession: "ARO:3002827",
    name: "tlrC",
    description:
      "tlrC is an ABC-F subfamily protein found in Streptomyces fradiae and confers resistance to mycinamicin, tylosin and lincosamides. tlrC is found in the tylosin biosynthetic cluster and is one mechanism by which S. fradiae protects itself from self-destruction when producing this macrolide.",
  },
  {
    accession: "ARO:3002828",
    name: "srmB",
    description:
      "srmB is an ABC-F subfamily protein found in Streptomyces ambofaciens that confers resistance to spiramycin",
  },
  {
    accession: "ARO:3002829",
    name: "vgaA",
    description:
      "vgaA is an ABC-F subfamily protein expressed in staphylococci that confers resistance to streptogramin A antibiotics and related compounds. It is associated with plasmid DNA.",
  },
  {
    accession: "ARO:3002830",
    name: "vgaALC",
    description:
      "vgaALC is an ABC-F subfamily protein expressed in staphylococci that confers resistance to streptogramin A antibiotics and related compounds. It is associated with plasmid DNA.",
  },
  {
    accession: "ARO:3002831",
    name: "vgaC",
    description:
      "vgaC is an ABC-F subfamily protein expressed in staphylococci that confers resistance to streptogramin A antibiotics and related compounds. It is associated with plasmid DNA.",
  },
  {
    accession: "ARO:3002832",
    name: "vgaD",
    description:
      "vgaD is an ABC-F subfamily protein expressed in Enterococcus faecium that confers resistance to streptogramin A antibiotics and related compounds. It is associated with plasmid DNA.",
  },
  {
    accession: "ARO:3002833",
    name: "vgaE",
    description:
      "vgaE is an ABC-F subfamily protein expressed in staphylococci that confers resistance to streptogramin A antibiotics and related compounds. It is associated with transposon DNA.",
  },
  {
    accession: "ARO:3002835",
    name: "lnuA",
    description:
      "lnuA is a plasmid-mediated nucleotidyltransferase found in Staphylococcus chromogenes",
  },
  {
    accession: "ARO:3002836",
    name: "lnuB",
    description:
      "lnuB is a plasmid-mediated nucleotidyltransferase found in Streptococcus lutetiensis",
  },
  {
    accession: "ARO:3002837",
    name: "lnuC",
    description:
      "lnuC is a transposon-mediated nucleotidyltransferase found in Streptococcus agalactiae",
  },
  {
    accession: "ARO:3002838",
    name: "lnuD",
    description:
      "lnuD is a plasmid-mediated nucleotidyltransferase found in Streptococcus uberis",
  },
  {
    accession: "ARO:3002839",
    name: "lnuF",
    description:
      "lnuF is an integron-mediated nucleotidyltransferase found in Escherichia coli",
  },
  {
    accession: "ARO:3002840",
    name: "vatA",
    description:
      "vatA is a plasmid-mediated acetyltransferase found in Staphylococcus aureus",
  },
  {
    accession: "ARO:3002841",
    name: "vatB",
    description:
      "vatB is a plasmid-mediated acetyltransferase found in Staphylococcus aureus",
  },
  {
    accession: "ARO:3002842",
    name: "vatC",
    description:
      "vatC is a plasmid-mediated acetyltransferase found in Staphylococcus cohnii",
  },
  {
    accession: "ARO:3002843",
    name: "vatD",
    description:
      "vatD is a transposon-mediated acetyltransferase found in Enterococcus faecium",
  },
  {
    accession: "ARO:3002844",
    name: "vatE",
    description:
      "vatE is a transposon-mediated acetyltransferase found in Enterococcus faecium",
  },
  {
    accession: "ARO:3002845",
    name: "vatH",
    description:
      "vatH is a plasmid-mediated acetyltransferase found in Enterococcus faecium",
  },
  {
    accession: "ARO:3002846",
    name: "arr-1",
    description:
      "arr-1 is a chromosome-encoded ribosyltransferase found in Mycobacterium smegmatis",
  },
  {
    accession: "ARO:3002847",
    name: "arr-2",
    description:
      "arr-2 is an integron-encoded ribosyltransferase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002848",
    name: "arr-3",
    description:
      "arr-3 is a plasmid-encoded ribosyltransferase found in Vibrio fluvialis",
  },
  {
    accession: "ARO:3002849",
    name: "arr-4",
    description:
      "arr-4 is an integron-encoded ribosyltransferase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002850",
    name: "arr-5",
    description:
      "arr-5 is an integron-encoded ribosyltransferase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002852",
    name: "arr-7",
    description:
      "arr-7 is an integron-encoded ribosyltransferase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002853",
    name: "arr-8",
    description:
      "arr-8 is an integron-encoded ribosyltransferase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3002854",
    name: "dfrA1",
    description: "dfrA1 is an integron-encoded dihydrofolate reductase",
  },
  {
    accession: "ARO:3002856",
    name: "dfrA24",
    description:
      "dfrA24 is an integron-encoded dihydrofolate reductase found in Escherichia coli",
  },
  {
    accession: "ARO:3002857",
    name: "dfrA26",
    description:
      "dfrA26 is an integron-encoded dihydrofolate reductase found in Escherichia coli",
  },
  {
    accession: "ARO:3002858",
    name: "dfrA12",
    description:
      "dfrA12 is an integron-encoded dihydrofolate reductase found in Vibrio cholerae",
  },
  {
    accession: "ARO:3002859",
    name: "dfrA14",
    description:
      "dfrA14 is an integron-encoded dihydrofolate reductase found in Escherichia coli",
  },
  {
    accession: "ARO:3002860",
    name: "dfrA17",
    description:
      "dfrA17 is an integron-encoded dihydrofolate reductase found in Escherichia coli",
  },
  {
    accession: "ARO:3002861",
    name: "dfrA5",
    description:
      "dfrA5 is an integron-encoded dihydrofolate reductase found in Vibrio cholerae",
  },
  {
    accession: "ARO:3002862",
    name: "dfrA7",
    description:
      "dfrA7 is an integron-encoded dihydrofolate reductase found in Escherichia coli",
  },
  {
    accession: "ARO:3002863",
    name: "dfrA8",
    description:
      "dfrA8 is a transposon-encoded dihydrofolate reductase found in Salmonella enterica",
  },
  {
    accession: "ARO:3002864",
    name: "dfrB1",
    description:
      "A plasmid-associated trimethoprim-resistant dihydrofolate reductase detected in Bordetella bronchispetica on pKBB958.",
  },
  {
    accession: "ARO:3002865",
    name: "dfrC",
    description:
      "dfrC is a chromosome-encoded dihydrofolate reductase found in Staphylococcus aureus",
  },
  {
    accession: "ARO:3002866",
    name: "dfrD",
    description:
      "dfrD is a plasmid-encoded dihydrofolate reductase found in Listeria monocytogenes",
  },
  {
    accession: "ARO:3002867",
    name: "dfrF",
    description:
      "dfrF is a chromosome-encoded dihydrofolate reductase found in Streptococcus pyogenes",
  },
  {
    accession: "ARO:3002868",
    name: "dfrG",
    description:
      "dfrG is a plasmid-encoded dihydrofolate reductase found in Staphylococcus aureus",
  },
  {
    accession: "ARO:3002869",
    name: "dfrK",
    description:
      "dfrK is a plasmid-encoded dihydrofolate reductase found in Staphylococcus aureus",
  },
  {
    accession: "ARO:3002870",
    name: "tet34",
    description:
      "tet34 causes the activation of Mg2+-dependent purine nucleotide synthesis, which protects the protein synthesis pathway. It is found in Gram-negative Vibrio",
  },
  {
    accession: "ARO:3002871",
    name: "tet37",
    description:
      "tet37 is a chromosome-encoded oxidoreductase isolated from an uncultured bacterium that confers resistance to tetracycline",
  },
  {
    accession: "ARO:3002872",
    name: "FosA3",
    description:
      "An enzyme that confers resistance to fosfomycin in Escherichia coli by breaking the epoxide ring of the molecule. It depends on the cofactors Manganese (II) and Potassium and uses Glutathione (GSH) as the nucleophilic molecule.",
  },
  {
    accession: "ARO:3002873",
    name: "FosB3",
    description:
      "A thiol transferase that leads to the resistance of fosfomycin in Enterococcus faecium. Contrasting FosA, FosB is dependent on the cofactor Magnesium (II) and uses either bacillithiol to open up the epoxide ring of fosfomycin.",
  },
  {
    accession: "ARO:3002874",
    name: "FosC2",
    description:
      "FosC2 is an enzyme that phosphorylates fosfomycin to confer resistance in Escherichia coli",
  },
  {
    accession: "ARO:3002875",
    name: "dfrE",
    description:
      "dfrE is a chromosome-encoded dihydrofolate reductase found in Enterococcus faecalis",
  },
  {
    accession: "ARO:3002876",
    name: "ethambutol resistant arabinosyltransferase",
    description:
      "Arabinosyl transferases allow for the polymerization of arabinose to form arabinan. Arabinan is required for formation of mycobacterial cell walls and arabinosyltransferases are targets of the drug ethambutol. Mutations in these genes can confer resistance to ethambutol.",
  },
  {
    accession: "ARO:3002877",
    name: "BcI",
    description:
      "Bacillus cereus beta-lactamase I is a zinc metallo-beta-lactamase that hydrolyzes a large number of penicillins and cephalosporins in the Bacillus cereus strain 569/H/9",
  },
  {
    accession: "ARO:3002878",
    name: "BcII",
    description:
      "Bacillus cereus beta-lactamase II is a zinc metallo-beta-lactamase that hydrolyzes a large number of penicillins and cephalosporins in the Bacillus cereus strain 5/B/6",
  },
  {
    accession: "ARO:3002879",
    name: "linG",
    description:
      "linG is an integron-associated gene cassette, with aadA2, encoding a lincosomide nucleotidyltransferase found in Salmonella enterica. linG confers resistance to lincosomide antibiotics.",
  },
  {
    accession: "ARO:3002881",
    name: "lmrC",
    description:
      "lmrC is an ABC-F subfamily protein that confers resistance to lincosamides in Streptomyces lincolnensis and Lactococcus lactis. It can dimerize with lmrD",
  },
  {
    accession: "ARO:3002882",
    name: "lmrD",
    description:
      "lmrD is a chromosomally-encoded efflux pump that confers resistance to lincosamides in Streptomyces lincolnensis and Lactococcus lactis. It can dimerize with lmrC",
  },
  {
    accession: "ARO:3002883",
    name: "rgt1438",
    description:
      "rgt1438 is a glycosyltransferase that confers rifampin resistance in Streptomyces",
  },
  {
    accession: "ARO:3002884",
    name: "iri",
    description:
      "iri is a monooxygenase that confers resistance to rifampin found in Rhodococcus equi",
  },
  {
    accession: "ARO:3002891",
    name: "otr(A)",
    description:
      "otr(A) is an oxytetracycline resistance ribosomal protection protein found in Streptomyces rimosus",
  },
  {
    accession: "ARO:3002892",
    name: "otr(B)",
    description:
      "otr(B) is a tetracycline resistance efflux pump found in Streptomyces rimosus",
  },
  {
    accession: "ARO:3002893",
    name: "tcr3",
    description:
      "tcr3 is a tetracycline efflux pump that confers self-resistance to Streptomyces aureofaciens",
  },
  {
    accession: "ARO:3002894",
    name: "otrC",
    description:
      "otrC is a tetracycline resistance efflux pump found in Streptomyces rimosus",
  },
  {
    accession: "ARO:3002895",
    name: "SAT-2",
    description:
      "SAT-2 is a plasmid-mediated streptothricin acetyltransferase, which confers resistance to streptothricin, a nucleoside antibiotic. Originally described from an E. coli plasmid sequence by Heim et al., 1989.",
  },
  {
    accession: "ARO:3002897",
    name: "SAT-4",
    description:
      "SAT-4 is a plasmid-mediated streptothricin acetyltransferase and streptothricin (a nucleoside antibiotic) resistant determinant. Originally described from a Campylobacter coli BE/G4 plasmid gene sequence by Jacob et al, 1994.",
  },
  {
    accession: "ARO:3002898",
    name: "SAT-3",
    description:
      "SAT-3 is a plasmid-mediated streptothricin acetyltransferase and streptothricin resistance determinant. Originally described from an E. coli plasmid gene by Tietze and Brevet, 1995.",
  },
  {
    accession: "ARO:3002905",
    name: "VanS-VanR two-component regulatory system",
    description:
      "van genes that are involved in activating and regulating transcription of gene clusters",
  },
  {
    accession: "ARO:3002906",
    name: "van ligase",
    description:
      "van ligases synthesize alternative substrates for peptidoglycan synthesis that reduce vancomycin binding affinity.",
  },
  {
    accession: "ARO:3002907",
    name: "vanE",
    description:
      "VanE is a D-Ala-D-Ala ligase homolog that can synthesize D-Ala-D-Ser, an alternative substrate for peptidoglycan synthesis that reduces vancomycin binding affinity in Enterococcus faecalis",
  },
  {
    accession: "ARO:3002908",
    name: "vanF",
    description:
      "VanF is a D-Ala-D-Ala ligase homolog that can synthesize D-Ala-D-Lac, an alternative substrate for peptidoglycan synthesis that reduces vancomycin binding affinity. It is associated with both vancomycin and teicoplanin resistance in Paenibacillus popilliae",
  },
  {
    accession: "ARO:3002909",
    name: "vanG",
    description:
      "VanG is a D-Ala-D-Ala ligase homolog that can synthesize D-Ala-D-Ser, an alternative substrate for peptidoglycan synthesis that reduces vancomycin binding affinity in Enterococcus faecalis",
  },
  {
    accession: "ARO:3002910",
    name: "vanL",
    description:
      "VanL is a D-Ala-D-Ala ligase homolog that can synthesize D-Ala-D-Ser, an alternative substrate for peptidoglycan synthesis that reduces vancomycin binding affinity in Enterococcus faecalis",
  },
  {
    accession: "ARO:3002911",
    name: "vanM",
    description:
      "VanM is a D-Ala-D-Ala ligase homolog that can synthesize D-Ala-D-Lac, an alternative substrate for peptidoglycan synthesis that reduces vancomycin binding affinity. It is associated with both vancomycin and teicoplanin resistance.",
  },
  {
    accession: "ARO:3002912",
    name: "vanN",
    description:
      "VanN is a D-Ala-D-Ala ligase homolog that can synthesize D-Ala-D-Ser, an alternative substrate for peptidoglycan synthesis that reduces vancomycin binding affinity in Enterococcus faecium",
  },
  {
    accession: "ARO:3002913",
    name: "vanO",
    description:
      "VanO is a D-Ala-D-Ala ligase homolog that can synthesize D-Ala-D-Lac, an alternative substrate for peptidoglycan synthesis that reduces vancomycin binding affinity. It is associated with both vancomycin and teicoplanin resistance.",
  },
  {
    accession: "ARO:3002914",
    name: "vanJ",
    description:
      "vanJ is a novel membrane protein that confers resistance to teicoplanin and its derivatives in Streptomyces coelicolor by recycling undecaprenol pyrophosphate during cell wall biosynthesis.",
  },
  {
    accession: "ARO:3002915",
    name: "vanK",
    description:
      "VanK is a member of the Fem family of enzymes that add the cross-bridge amino acids to the stem pentapeptide of cell wall precursors in Streptomyces coelicolor that confers inducible, high-level vancomycin resistance",
  },
  {
    accession: "ARO:3002916",
    name: "vanV",
    description:
      "vanV is an accessory protein of van operons, first identified in the vanB operon. It is not required for vancomycin resistance.",
  },
  {
    accession: "ARO:3002917",
    name: "glycopeptide resistance gene cluster VanN",
    description:
      "Homologous to vanC, contains a D-Ala-D-Ser ligase. The plasmid-located vanM gene cluster is inducible and confers low resistance to vancomycin. vanN organisms remain susceptible to teicoplanin. Gene orientation: N(XY)TRS",
  },
  {
    accession: "ARO:3002918",
    name: "glycopeptide resistance gene cluster VanO",
    description:
      "Homologous to vanA, contains a D-Ala-D-Lac ligase. The chromosome-located vanO gene cluster is inducible. Not much is known about the biochemistry about the vanO gene cluster. Gene orientation: orf1 RS orf2 HOX",
  },
  {
    accession: "ARO:3002919",
    name: "vanRA",
    description:
      "vanRA, also known as vanR, is a vanR variant found in the vanA gene cluster",
  },
  {
    accession: "ARO:3002921",
    name: "vanRB",
    description: "vanRB is a vanR variant found in the vanB gene cluster",
  },
  {
    accession: "ARO:3002922",
    name: "vanRC",
    description: "vanRC is a vanR variant found in the vanC gene cluster",
  },
  {
    accession: "ARO:3002923",
    name: "vanRD",
    description:
      "vanRD is a mutated vanR variant found in the vanD gene cluster that caused constitutive expression of vanD peptidoglycan synthesis",
  },
  {
    accession: "ARO:3002924",
    name: "vanRE",
    description: "vanRE is a vanR variant found in the vanE gene cluster",
  },
  {
    accession: "ARO:3002925",
    name: "vanRF",
    description: "vanRF is a vanR variant found in the vanF gene cluster",
  },
  {
    accession: "ARO:3002926",
    name: "vanRG",
    description: "vanRG is a vanR variant found in the vanG gene cluster",
  },
  {
    accession: "ARO:3002927",
    name: "vanRL",
    description: "vanRL is a vanR variant found in the vanL gene cluster",
  },
  {
    accession: "ARO:3002928",
    name: "vanRM",
    description: "vanRM is a vanR variant found in the vanM gene cluster",
  },
  {
    accession: "ARO:3002929",
    name: "vanRN",
    description: "vanRN is a vanR variant found in the vanN gene cluster",
  },
  {
    accession: "ARO:3002930",
    name: "vanRO",
    description: "vanRO is a vanR variant found in the vanO gene cluster",
  },
  {
    accession: "ARO:3002931",
    name: "vanSA",
    description:
      "vanSA, also known as vanS, is a vanS variant found in the vanA gene cluster",
  },
  {
    accession: "ARO:3002932",
    name: "vanSB",
    description: "vanSB is a vanS variant found in the vanB gene cluster",
  },
  {
    accession: "ARO:3002933",
    name: "vanSC",
    description: "vanSC is a vanS variant found in the vanC gene cluster",
  },
  {
    accession: "ARO:3002934",
    name: "vanSD",
    description:
      "vanSD is a mutated vanS variant found in the vanD gene cluster that caused constitutive expression of vanD peptidoglycan synthesis",
  },
  {
    accession: "ARO:3002935",
    name: "vanSE",
    description: "vanSE is a vanS variant found in the vanE gene cluster",
  },
  {
    accession: "ARO:3002936",
    name: "vanSF",
    description: "vanSF is a vanS variant found in the vanF gene cluster",
  },
  {
    accession: "ARO:3002937",
    name: "vanSG",
    description: "vanSG is a vanS variant found in the vanG gene cluster",
  },
  {
    accession: "ARO:3002938",
    name: "vanSL",
    description: "vanSL is a vanS variant found in the vanL gene cluster",
  },
  {
    accession: "ARO:3002939",
    name: "vanSM",
    description: "vanSM is a vanS variant found in the vanM gene cluster",
  },
  {
    accession: "ARO:3002940",
    name: "vanSN",
    description: "vanSN is a vanS variant found in the vanN gene cluster",
  },
  {
    accession: "ARO:3002941",
    name: "vanSO",
    description: "vanSO is a vanS variant found in the vanO gene cluster",
  },
  {
    accession: "ARO:3002942",
    name: "vanHA",
    description:
      "vanHA, also known as vanH, is a vanH variant in the vanA gene cluster",
  },
  {
    accession: "ARO:3002943",
    name: "vanHB",
    description: "vanHB is a vanH variant in the vanB gene cluster",
  },
  {
    accession: "ARO:3002944",
    name: "vanHD",
    description: "vanHD is a vanH variant in the vanD gene cluster",
  },
  {
    accession: "ARO:3002945",
    name: "vanHF",
    description: "vanHF is a vanH variant in the vanF gene cluster",
  },
  {
    accession: "ARO:3002947",
    name: "vanHM",
    description: "vanHM is a vanH variant in the vanM gene cluster",
  },
  {
    accession: "ARO:3002948",
    name: "vanHO",
    description: "vanHO is a vanH variant in the vanO gene cluster",
  },
  {
    accession: "ARO:3002949",
    name: "vanXA",
    description:
      "vanXA, also known as vanX, is a vanX variant found in the vanA gene cluster",
  },
  {
    accession: "ARO:3002950",
    name: "vanXB",
    description: "vanXB is a vanX variant found in the vanB gene cluster",
  },
  {
    accession: "ARO:3002952",
    name: "vanXF",
    description: "vanXF is a vanX variant found in the vanF gene cluster",
  },
  {
    accession: "ARO:3002953",
    name: "vanXM",
    description: "vanXM is a vanX variant found in the vanM gene cluster",
  },
  {
    accession: "ARO:3002954",
    name: "vanXO",
    description: "vanXO is a vanX variant found in the vanO gene cluster",
  },
  {
    accession: "ARO:3002955",
    name: "vanYA",
    description:
      "vanYA, also known as vanY, is a vanY variant found in the vanA gene cluster",
  },
  {
    accession: "ARO:3002956",
    name: "vanYB",
    description: "vanYB is a vanY variant found in the vanB gene cluster",
  },
  {
    accession: "ARO:3002957",
    name: "vanYD",
    description: "vanYD is a vanY variant found in the vanD gene cluster",
  },
  {
    accession: "ARO:3002958",
    name: "vanYF",
    description: "vanYF is a vanY variant found in the vanF gene cluster",
  },
  {
    accession: "ARO:3002959",
    name: "vanYG1",
    description: "vanYG1 is a vanY variant found in the vanG gene cluster",
  },
  {
    accession: "ARO:3002961",
    name: "vanYM",
    description: "vanYM is a vanY variant found in the vanM gene cluster",
  },
  {
    accession: "ARO:3002962",
    name: "vanZA",
    description:
      "vanZA, also known as vanZ, is a vanZ variant found in the vanA gene cluster",
  },
  {
    accession: "ARO:3002963",
    name: "vanZF",
    description: "vanZF is a vanZ variant found in the vanF gene cluster",
  },
  {
    accession: "ARO:3002964",
    name: "vanWB",
    description:
      "vanWB, also known as vanW, is a vanW variant found in the vanB gene cluster",
  },
  {
    accession: "ARO:3002965",
    name: "vanWG",
    description: "vanWG is a vanW variant found in the vanG gene cluster",
  },
  {
    accession: "ARO:3002966",
    name: "vanXYC",
    description: "vanXYC is a vanXY variant found in the vanC gene cluster",
  },
  {
    accession: "ARO:3002967",
    name: "vanXYE",
    description: "vanXYE is a vanXY variant found in the vanE gene cluster",
  },
  {
    accession: "ARO:3002968",
    name: "vanXYL",
    description: "vanXYL is a vanXY variant found in the vanL gene cluster",
  },
  {
    accession: "ARO:3002969",
    name: "vanXYN",
    description: "vanXYN is a vanXY variant found in the vanN gene cluster",
  },
  {
    accession: "ARO:3002970",
    name: "vanTC",
    description: "vanTC is a vanT variant found in the vanC gene cluster",
  },
  {
    accession: "ARO:3002971",
    name: "vanTE",
    description: "vanTE is a vanT variant found in the vanE gene cluster",
  },
  {
    accession: "ARO:3002972",
    name: "vanTG",
    description: "vanTG is a vanT variant found in the vanG gene cluster",
  },
  {
    accession: "ARO:3002973",
    name: "vanTmL",
    description:
      "vanTmL is a vanT variant found in the vanL gene cluster. vanTmL codes for the membrane-binding domain of vanTL",
  },
  {
    accession: "ARO:3002974",
    name: "vanTrL",
    description:
      "vanTrL is a vanT variant found in the vanL gene cluster. vanTrL codes for the racemase component of vanT",
  },
  {
    accession: "ARO:3002975",
    name: "vanTN",
    description: "vanTN is a vanT variant found in the vanN gene cluster",
  },
  {
    accession: "ARO:3002976",
    name:
      "gene(s) or protein(s) associated with a glycopeptide resistance cluster",
    description:
      "Van genes encode for various proteins that are involved in conferring vancomycin and teicoplanin resistance.",
  },
  {
    accession: "ARO:3002978",
    name: "D-Ala-D-Lac ligase",
    description:
      "van ligases that synthesize D-Ala-D-Lac, an alternative peptidoglycan substrate that confers resistance to vancomycin",
  },
  {
    accession: "ARO:3002979",
    name: "D-Ala-D-Ser ligase",
    description:
      "van ligases that synthesize D-Ala-D-Ser, an alternative peptidoglycan substrate that confers resistance to vancomycin",
  },
  {
    accession: "ARO:3002980",
    name: "aspergillomarasmine A",
    description:
      "Aspergillomarasmine A, a fungal natural product, has been shown to inhibit the NDM-1 and VIM-2 enzymes, restoring of the activity of meropenem against carbapenem-resistant pathogens possessing either NDM-1 or VIM-2.",
  },
  {
    accession: "ARO:3002981",
    name: "AmrAB-OprM",
    description:
      "AmrAB is an RND efflux system that impairs aminoglycoside accumulation in clinical Burkholderia vietnamiensis strains",
  },
  {
    accession: "ARO:3002982",
    name: "amrA",
    description:
      "amrA is the efflux pump subunit of the AmrAB-OprM multidrug efflux complex. amrA corresponds to 1 locus in Pseudomonas aeruginosa PAO1 and 1 locus in Pseudomonas aeruginosa LESB58.",
  },
  {
    accession: "ARO:3002983",
    name: "amrB",
    description:
      "amrB is the membrane fusion protein of the AmrAB-OprM multidrug efflux complex.",
  },
  {
    accession: "ARO:3002984",
    name: "determinant of polymyxin resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to polymyxin antibiotics, i.e. colistin.",
  },
  {
    accession: "ARO:3002985",
    name: "arnA",
    description:
      "arnA modifies lipid A with 4-amino-4-deoxy-L-arabinose (Ara4N) which allows gram-negative bacteria to resist the antimicrobial activity of cationic antimicrobial peptides and antibiotics such as polymyxin. arnA is found in E. coli and P. aeruginosa.",
  },
  {
    accession: "ARO:3002986",
    name: "bacA",
    description:
      "The bacA gene product (BacA) recycles undecaprenyl pyrophosphate during cell wall biosynthesis which confers resistance to bacitracin.",
  },
  {
    accession: "ARO:3002987",
    name: "bcrA",
    description:
      "bcrA is an ABC transporter found in Bacillus licheniformis that confers bacitracin resistance",
  },
  {
    accession: "ARO:3002988",
    name: "bcrB",
    description:
      "bcrB is an ABC transporter found in Bacillus licheniformis that confers bacitracin resistance",
  },
  {
    accession: "ARO:3002992",
    name: "AQU beta-lactamase",
    description:
      "AQU beta-lactamases are chromosomal class C beta-lactamases that confer resistance to cephalosporins.",
  },
  {
    accession: "ARO:3002993",
    name: "AQU-1",
    description:
      "AQU-1 is a chromosomal class C beta-lactamase found in clinical Aeromonas dhakensis isolates",
  },
  {
    accession: "ARO:3002994",
    name: "ROB beta-lactamase",
    description: "ROB beta-lactamases are a class A beta-lactamases.",
  },
  {
    accession: "ARO:3002995",
    name: "ROB-1",
    description:
      "ROB-1 is a beta-lactamase found in Pasteurella and Haemophilus",
  },
  {
    accession: "ARO:3002996",
    name: "LCR beta-lactamase",
    description:
      "LCR beta-lactamases are a class D beta-lactamase that hydrolyze a variety of penams and some cephalosporins.",
  },
  {
    accession: "ARO:3002997",
    name: "LCR-1",
    description:
      "LCR-1 is a class D beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3002998",
    name: "CblA beta-lactamase",
    description:
      "CblA beta-lactamases are class A beta-lactamases that confer resistance to cephalosporins.",
  },
  {
    accession: "ARO:3002999",
    name: "CblA-1",
    description:
      "CblA-1 beta-lactamase is a class A beta-lactamase found in Bacteroides uniformis that is species-specific.",
  },
  {
    accession: "ARO:3003000",
    name: "CfxA beta-lactamase",
    description: "cfxA beta-lactamases are class A beta-lactamases",
  },
  {
    accession: "ARO:3003001",
    name: "CfxA",
    description:
      "cfxA beta-lactamase is a class A beta-lactamase found in Bacteroides vulgatus",
  },
  {
    accession: "ARO:3003002",
    name: "CfxA2",
    description:
      "cfxA2 beta-lactamase is a class A beta-lactamase found in Prevotella intermedia",
  },
  {
    accession: "ARO:3003003",
    name: "CfxA3",
    description:
      "cfxA3 beta-lactamase is a class A beta-lactamase found in Capnocytophaga ochracea",
  },
  {
    accession: "ARO:3003005",
    name: "CfxA4",
    description:
      "cfxA4 beta-lactamase is a class A beta-lactamase found in Bacteroides fragilis",
  },
  {
    accession: "ARO:3003006",
    name: "blt",
    description:
      "blt is an MFS efflux pump that confers resistance to multiple drugs such as rhodamine and acridine dyes, and fluoroquinolone antibiotics",
  },
  {
    accession: "ARO:3003007",
    name: "bmr",
    description:
      "bmr is an MFS antibiotic efflux pump that confers resistance to multiple drugs including acridine dyes, fluoroquinolone antibiotics, chloramphenicol, and puromycin",
  },
  {
    accession: "ARO:3003008",
    name: "CeoAB-OpcM",
    description:
      "CeoAB is an RND efflux system that confers multidrug resistance in Burkholderia cenocepacia",
  },
  {
    accession: "ARO:3003009",
    name: "ceoA",
    description:
      "ceoA is a periplasmic linker subunit of the CeoAB-OpcM efflux pump",
  },
  {
    accession: "ARO:3003010",
    name: "ceoB",
    description:
      "ceoB is a cytoplasmic membrane component of the CeoAB-OpcM efflux pump",
  },
  {
    accession: "ARO:3003011",
    name: "dfrA10",
    description:
      "dfrA10 is an integron-encoded dihydrofolate reductase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3003012",
    name: "dfrA13",
    description:
      "dfrA13 is an integron-encoded dihydrofolate reductase found in Escherichia coli",
  },
  {
    accession: "ARO:3003013",
    name: "dfrA15",
    description:
      "dfrA15 is an integron-encoded dihydrofolate reductase found in Vibrio cholerae",
  },
  {
    accession: "ARO:3003014",
    name: "dfrA16",
    description:
      "dfrA16 is an integron-encoded dihydrofolate reductase found in Salmonella enterica",
  },
  {
    accession: "ARO:3003015",
    name: "dfrA19",
    description:
      "dfrA19 is an integron-encoded dihydrofolate reductase found in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3003016",
    name: "dfrA20",
    description:
      "dfrA20 is a plasmid-encoded dihydrofolate reductase found in Pasteurella multocida",
  },
  {
    accession: "ARO:3003017",
    name: "dfrA21",
    description:
      "dfrA21 is an integron-encoded dihydrofolate reductase found in Salmonella enterica",
  },
  {
    accession: "ARO:3003018",
    name: "dfrA22",
    description:
      "dfrA22 is an integron-encoded dihydrofolate reductase found in Salmonella enterica",
  },
  {
    accession: "ARO:3003019",
    name: "dfrA23",
    description:
      "dfrA23 is an integron-encoded dihydrofolate reductase found in Salmonella enterica",
  },
  {
    accession: "ARO:3003020",
    name: "dfrA25",
    description:
      "dfrA25 is an integron-encoded dihydrofolate reductase found in Salmonella agona",
  },
  {
    accession: "ARO:3003021",
    name: "dfrB2",
    description:
      "dfrB2 is an integron-encoded dihydrofolate reductase found in an uncultured bacterium from a wastewater treatment plant",
  },
  {
    accession: "ARO:3003022",
    name: "dfrB3",
    description:
      "dfrB3 is an integron-encoded dihydrofolate reductase found in Klebsiella oxytoca",
  },
  {
    accession: "ARO:3003023",
    name: "dfrB6",
    description:
      "dfrB6 is an integron-encoded dihydrofolate reductase found in Salmonella enterica",
  },
  {
    accession: "ARO:3003024",
    name: "determinant of fusidic acid resistance",
    description:
      "Enzymes, other proteins and other gene products shown clinically to confer resistance to fusidic acid.",
  },
  {
    accession: "ARO:3003025",
    name: "fusidic acid inactivation enzyme",
    description:
      "Enzymes that confer resistance to fusidic acid by inactivation",
  },
  {
    accession: "ARO:3003026",
    name: "fusH",
    description:
      "fusH is a highly specific fusidic acid esterase found in Streptomyces lividans",
  },
  {
    accession: "ARO:3003027",
    name: "lmrAB Operon",
    description:
      "lmrAB operon is involved in lincosamide resistance in Bacillus subtilis",
  },
  {
    accession: "ARO:3003028",
    name: "lmrA",
    description:
      "lmrA is the repressor to the lmrAB operon in Bacillus subtilis. lmrA mutations result in lincomycin resistance.",
  },
  {
    accession: "ARO:3003029",
    name: "MexVW-OprM",
    description:
      "MexVW-OprM is a multidrug efflux protein expressed in Pseudomonas aeruginosa. MexV is the membrane fusion protein; MexW is the RND-type membrane protein; and OprM is the outer membrane channel. MexVW-OprM is associated with resistance to fluoroquinolones, tetracycline, chloramphenicol, erythromycin and acriflavine.",
  },
  {
    accession: "ARO:3003030",
    name: "MexV",
    description:
      "MexV is the membrane fusion protein of the MexVW-OprM multidrug efflux complex.",
  },
  {
    accession: "ARO:3003031",
    name: "MexW",
    description:
      "MexW is the RND-type membrane protein of the efflux complex MexVW-OprM.",
  },
  {
    accession: "ARO:3003032",
    name: "MexXY-OprM",
    description:
      "MexXY-OprM is a multidrug efflux protein expressed in Pseudomonas aeruginosa. MexY is the membrane fusion protein; MexX is the RND-type membrane protein; and OprM is the outer membrane channel. MexXY-OprM is associated with resistance to acriflavine, erythromycin, norfloxacin and ofloxacin. The efflux system may also be involved in acquisition of higher resistance, in particular when bacteria are repeatedly exposed to subinhibitory doses of AGs.",
  },
  {
    accession: "ARO:3003033",
    name: "mexY",
    description:
      "MexY is the RND-type membrane protein of the efflux complex MexXY-OprM.",
  },
  {
    accession: "ARO:3003034",
    name: "mexX",
    description:
      "MexX is the membrane fusion protein of the MexXY-OprM multidrug efflux complex.",
  },
  {
    accession: "ARO:3003035",
    name: "mfpA",
    description:
      "mfpA is a qnr homolog and a pentapeptide repeat protein that confers resistance to fluoroquinolones in Mycobacterium smegmatis",
  },
  {
    accession: "ARO:3003036",
    name: "oleB",
    description:
      "oleB is an ABC-F subfamily protein in Streptomyces antibioticus and is involved in oleandomycin secretion",
  },
  {
    accession: "ARO:3003037",
    name: "opcM",
    description:
      "OpcM is an outer membrane factor protein found in Burkholderia cepacia. It is part of the CeoAB-OpcM complex.",
  },
  {
    accession: "ARO:3003038",
    name: "MexXY-OprA",
    description:
      "MexXY-OprA is a multidrug efflux protein expressed in Pseudomonas aeruginosa. MexY is the membrane fusion protein; MexX is the RND-type membrane protein; and OprA is the outer membrane channel. MexXY-OprA is associated with resistance to aminoglycosides",
  },
  {
    accession: "ARO:3003039",
    name: "OprA",
    description:
      "OprA is an outer membrane factor protein found in Pseudomonas aeruginosa. It is part of the MexXY-OprA complex.",
  },
  {
    accession: "ARO:3003040",
    name: "beta-lactam resistant penicillin-binding proteins",
    description:
      "Pencillin-binding proteins are transpeptidase proteins essential for cell wall synthesis. PBPs bind to penicillin antibiotics, which inactivates the enzyme and therefore disrupts cell wall synthesis. Multiple mechanisms and determinants are implicated in antibiotic resistant PBP.",
  },
  {
    accession: "ARO:3003041",
    name: "Streptococcus pneumoniae PBP1a conferring resistance to amoxicillin",
    description:
      "PBP1a is a penicillin-binding protein found in Streptococcus pneumoniae",
  },
  {
    accession: "ARO:3003042",
    name: "Streptococcus pneumoniae PBP2b conferring resistance to amoxicillin",
    description:
      "PBP2b is a penicillin-binding protein found in Streptococcus pneumoniae",
  },
  {
    accession: "ARO:3003043",
    name: "Streptococcus pneumoniae PBP2x conferring resistance to amoxicillin",
    description:
      "PBP2x is a penicillin-binding protein found in Streptococcus pneumoniae",
  },
  {
    accession: "ARO:3003046",
    name: "qacA",
    description: "qacA is a subunit of the qac multidrug efflux pump",
  },
  {
    accession: "ARO:3003047",
    name: "qacB",
    description: "qacB is a subunit of the qac multidrug efflux pump",
  },
  {
    accession: "ARO:3003048",
    name: "rosA",
    description:
      "rosA is part of an efflux pump/potassium antiporter system (RosAB) in Yersinia that confers resistance to cationic antimicrobial peptides such as polymyxin B.",
  },
  {
    accession: "ARO:3003049",
    name: "rosB",
    description:
      "rosB is part of an efflux pump/potassium antiporter system (RosAB) in Yersinia that confers resistance to cationic antimicrobial peptides such as polymyxin B.",
  },
  {
    accession: "ARO:3003050",
    name: "SmeABC",
    description:
      "SmeABC is a multidrug resistance efflux pump that confers resistance to aminoglycosides, beta-lactams, and fluoroquinolones in Stenotrophomonas maltophilia",
  },
  {
    accession: "ARO:3003051",
    name: "smeA",
    description:
      "smeA is the membrane fusion protein of the smeABC multidrug efflux complex in Stenotrophomonas maltophilia",
  },
  {
    accession: "ARO:3003052",
    name: "smeB",
    description:
      "smeB is the inner membrane multidrug exporter of the efflux complex smeABC in Stenotrophomonas maltophilia",
  },
  {
    accession: "ARO:3003053",
    name: "smeC",
    description:
      "smeC is an outer membrane multidrug efflux protein of the smeABC complex in Stenotrophomonas maltophilia",
  },
  {
    accession: "ARO:3003054",
    name: "SmeDEF",
    description:
      "SmeDEF is a multidrug resistance efflux pump that confers resistance to quinolones, tetracyclines, macrolides, chloramphenicol, and novobiocin in Stenotrophomonas maltophilia",
  },
  {
    accession: "ARO:3003055",
    name: "smeD",
    description:
      "smeD is the membrane fusion protein of the smeDEF multidrug efflux complex in Stenotrophomonas maltophilia",
  },
  {
    accession: "ARO:3003056",
    name: "smeE",
    description:
      "smeE is the RND protein of the efflux complex smeDEF in Stenotrophomonas maltophilia",
  },
  {
    accession: "ARO:3003057",
    name: "smeF",
    description:
      "smeF is an outer membrane multidrug efflux protein of the smeDEF complex in Stenotrophomonas maltophilia",
  },
  {
    accession: "ARO:3003059",
    name: "tmrB",
    description:
      "tmrB is an ATP-binding tunicamycin resistance protein found in Bacillus subtilis",
  },
  {
    accession: "ARO:3003060",
    name: "tsnR",
    description:
      "tsnR is a 23S ribosomal RNA methyltransferase that methylates adenosine-1067 to confer resistance to thiostrepton",
  },
  {
    accession: "ARO:3003061",
    name: "vph",
    description:
      "vph is a phosphotransferase that confers resistance to viomycin in Streptomyces vinaceus",
  },
  {
    accession: "ARO:3003062",
    name: "ykkCD",
    description:
      "ykkCD is a multidrug efflux pump with two SMR-type protein subunits that confers resistance to cationic drugs, chloramphenicol, streptomycin and tetracycline.",
  },
  {
    accession: "ARO:3003063",
    name: "ykkC",
    description:
      "ykkC is an SMR-type protein that is a subunit of the ykkCD efflux pump",
  },
  {
    accession: "ARO:3003064",
    name: "ykkD",
    description:
      "ykkD is an SMR-type protein that is a subunit of the ykkCD efflux pump",
  },
  {
    accession: "ARO:3003065",
    name: "non-erm 23S ribosomal RNA methyltransferase (A1067)",
    description:
      "Non-erm 23S ribosomal RNA methyltransferases modify adenosine 1067 to confer resistance to peptide antibiotics",
  },
  {
    accession: "ARO:3003066",
    name: "smeR",
    description:
      "smeR is the responder component of a two component signal transduction system that includes smeS",
  },
  {
    accession: "ARO:3003067",
    name: "smeS",
    description:
      "smeS is the protein kinase sensor component of a two component signal transduction system that includes smeR",
  },
  {
    accession: "ARO:3003068",
    name: "smeRS",
    description:
      "smeRS is a two component regulatory system for smeABC where smeR is a response regulator, while smeS is a sensor kinase.",
  },
  {
    accession: "ARO:3003069",
    name: "vanXYG",
    description: "vanXYG is a vanXY variant found in the vanG gene cluster",
  },
  {
    accession: "ARO:3003070",
    name: "vanXD",
    description: "vanXD is a vanX variant found in the vanD gene cluster",
  },
  {
    accession: "ARO:3003071",
    name: "mphF",
    description:
      "mphF is a macrolide phosphotransferase and resistance gene identified on the IncP plasmid pRSB111",
  },
  {
    accession: "ARO:3003072",
    name: "mphL",
    description:
      "mphL is a chromosomally-encoded macrolide phosphotransferases that inactivate 14- and 15-membered macrolides such as erythromycin, clarithromycin, azithromycin.",
  },
  {
    accession: "ARO:3003073",
    name: "determinant of resistance to lipopeptide antibiotics",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to lipopeptide antibiotics.",
  },
  {
    accession: "ARO:3003074",
    name: "Staphylococcus aureus cls conferring resistance to daptomycin",
    description:
      "cls or cardiolipin synthetase is an inner membrane protein that is involved in membrane synthesis. Specific mutations in S. aureus can confer resistance to daptomycin.",
  },
  {
    accession: "ARO:3003076",
    name: "daptomycin resistant liaFSR",
    description:
      "liaFSR is a two-component signal transduction pathway in B. subtilis that responds to peptide and lipopeptide antibiotics. The liaFSR two-component system consists of a histidine kinase liaS and the response regulator liaR as well as the accessory protein liaF, which acts as a negative regulator of liaRS-dependent signal transduction. Mutations confer resistance to daptomycin.",
  },
  {
    accession: "ARO:3003077",
    name: "Enterococcus faecalis liaF mutant conferring daptomycin resistance",
    description:
      "liaF is an accessory protein that acts as a negative regulator of liaRS signal transduction pathway. Mutations confer daptomycin resistance.",
  },
  {
    accession: "ARO:3003078",
    name: "Enterococcus faecium liaR mutant conferring daptomycin resistance",
    description:
      "liaR is a response regulator found in the liaFSR signal transduction pathway. Mutations confer daptomycin resistance.",
  },
  {
    accession: "ARO:3003079",
    name: "Enterococcus faecium liaS mutant conferring daptomycin resistance",
    description:
      "liaS is a histidine kinase found in the liaFSR signal transduction pathway. Mutations confer daptomycin resistance.",
  },
  {
    accession: "ARO:3003080",
    name: "daptomycin resistant pgsA",
    description:
      "pgsA or phosphatidylglycerophosphate synthetase is an integral membrane protein involved in phospholipid biosynthesis. It is a CDP-diacylglycerol-glycerol-3-phosphate 3-phosphatidyltransferase. Laboratory experiments have detected mutations conferring daptomycin resistance in Entercoccus.",
  },
  {
    accession: "ARO:3003090",
    name: "daptomycin resistant beta-subunit of RNA polymerase (rpoB)",
    description:
      "Daptomycin resistant RNA polymerases include amino acids substitutions which alter expression of the dlt operon, which increases the cell surface positive charge. Known from S. aureus.",
  },
  {
    accession: "ARO:3003091",
    name: "daptomycin resistant mprF",
    description:
      "MprF is a integral membrane protein that modifies the negatively-charged phosphatidylglycerol on the membrane surface of both Gram-positive and Gram-negative bacteria. This confers resistance to cationic peptides that disrupt the cell membrane, including defensins. Mutations in mprF can additionally confer resistance to daptomycin in S. aureus.",
  },
  {
    accession: "ARO:3003092",
    name: "Enterococcus faecium cls conferring resistance to daptomycin",
    description:
      "cls or cardiolipin synthetase is an inner membrane protein that is involved in membrane synthesis. Specific mutations in Enterococcus can confer resistance to daptomycin.",
  },
  {
    accession: "ARO:3003093",
    name: "cphA3",
    description:
      "CEPH-A3 is an Ambler Class B MBL; subclass B2 originally isolated from Aeromonas veronii. This enzyme has specific activity against carbapenems and is active as a mono-zinc protein.",
  },
  {
    accession: "ARO:3003094",
    name: "imiH",
    description:
      "imiH is an Ambler Class B MBL; subclass B2 originally isolated from Aeromonas hydrophila. This enzyme has specific activity against carbapenems.",
  },
  {
    accession: "ARO:3003095",
    name: "imiS",
    description:
      "imiS is an Ambler Class B MBL; subclass B2 originally isolated from Aeromonas veronii. This enzyme has specific activity against carbapenems.",
  },
  {
    accession: "ARO:3003096",
    name: "CfxA5",
    description:
      "cfxA5 beta-lactamase is a class A beta-lactamase found in Bacteroides distasonis",
  },
  {
    accession: "ARO:3003097",
    name: "CfxA6",
    description:
      "cfxA6 beta-lactamase is a class A beta-lactamase found in an uncultured bacterium",
  },
  {
    accession: "ARO:3003099",
    name: "cphA2",
    description:
      "CphA2 is an Ambler Class B MBL; subclass B2 originally isolated from Aeromonas hydrophila. This enzyme has specific activity against carbapenems and is active as a mono-zinc protein",
  },
  {
    accession: "ARO:3003100",
    name: "cphA4",
    description:
      "CphA4 is an Ambler Class B MBL; subclass B2 originally isolated from Aeromonas allosaccharophila. This enzyme has specific activity against carbapenems and is active as a mono-zinc protein.",
  },
  {
    accession: "ARO:3003101",
    name: "cphA5",
    description:
      "cphA5 is an Ambler Class B MBL; subclass B2 originally isolated from Aeromonas salmonicida. This enzyme has specific activity against carbapenems and is active as a mono-zinc protein.",
  },
  {
    accession: "ARO:3003102",
    name: "cphA6",
    description:
      "cphA6 is an Ambler Class B MBL; subclass B2 originally isolated from Aeromonas veronii. This enzyme has specific activity against carbapenems and is active as a mono-zinc protein.",
  },
  {
    accession: "ARO:3003103",
    name: "cphA7",
    description:
      "cphA7 is an Ambler Class B MBL; subclass B2 originally isolated from Aeromonas jandaei. This enzyme has specific activity against carbapenems and is active as a mono-zinc protein.",
  },
  {
    accession: "ARO:3003104",
    name: "cphA8",
    description:
      "cphA8 is an Ambler Class B MBL; subclass B2 originally isolated from Aeromonas sobria. This enzyme has specific activity against carbapenems and is active as a mono-zinc protein.",
  },
  {
    accession: "ARO:3003105",
    name: "dfrA3",
    description:
      "dfrA3 is an integron-encoded dihydrofolate reductase found in Escherichia coli",
  },
  {
    accession: "ARO:3003106",
    name: "Erm(42)",
    description: "Erm42 confers MLSb phenotype in Pasteurella multocida",
  },
  {
    accession: "ARO:3003107",
    name: "mef(B)",
    description:
      "mef(B) is a macrolide efflux gene located in the vicinity of sul3 in Escherichia coli. There is also a mefB found in Streptococcus agalactiae that confers resistance to macrolides.",
  },
  {
    accession: "ARO:3003109",
    name: "msrE",
    description:
      "MsrE is an ABC-F subfamily protein expressed to Klebsiella pneumoniae that confers resistance to erythromycin and streptogramin B antibiotics. It is associated with plasmid DNA. It is also 100% identical to ABC-F type ribosomal protection protein Msr(E) which is in multiple species.",
  },
  {
    accession: "ARO:3003110",
    name: "catB10",
    description:
      "catB10 is an integron-encoded variant of the cat gene found in P. aeruginosa",
  },
  {
    accession: "ARO:3003111",
    name: "lsaB",
    description:
      "LsaB is an ABC-F subfamily protein expressed in Staphylococcus sciuri. It confers resistance to clindamycin.",
  },
  {
    accession: "ARO:3003112",
    name: "lsaC",
    description:
      "LsaC is an ABC-F subfamily protein expressed in Streptococcus agalactiae. It confers resistance to lincomycin, clindamycin, dalfopristin, and tiamulin",
  },
  {
    accession: "ARO:3003113",
    name: "TEM-221",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003114",
    name: "TEM-222",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003115",
    name: "OXA-427",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003116",
    name: "OXA-420",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003117",
    name: "OXA-421",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003119",
    name: "OXA-428",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003120",
    name: "OXA-429",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003121",
    name: "OXA-430",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003124",
    name: "OXA-431",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003125",
    name: "OXA-432",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003126",
    name: "OXA-433",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003127",
    name: "OXA-434",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003128",
    name: "OXA-436",
    description:
      "An OXA-48-like plasmid-encoded beta-lactamase shown to confer resistance to carbapenems through hydrolysis. Identified in multiple Enterobacteriaceae isolates from several patients in Denmark. Beta-lactamase activity against carbapenems and penicillins but little to no activity against cephalosporin antibiotics. Described by Samuelsen et al. 2017.",
  },
  {
    accession: "ARO:3003129",
    name: "CMY-121",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003130",
    name: "CMY-122",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003131",
    name: "CMY-123",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003132",
    name: "CMY-124",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003133",
    name: "CMY-125",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003134",
    name: "CMY-126",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003135",
    name: "CMY-127",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003136",
    name: "CMY-128",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003137",
    name: "CMY-129",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003138",
    name: "CMY-130",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003139",
    name: "MIR-18",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003140",
    name: "MOX-10",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003141",
    name: "MOX-11",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003142",
    name: "IMI-8",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003143",
    name: "IMI-9",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003144",
    name: "KPC-20",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003145",
    name: "KPC-21",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003146",
    name: "VEB-10",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003147",
    name: "OXA-424",
    description:
      "OXA-424 is a \u00df-lactamase found in clinical isolates of Acinetobacter baumannii found in China. It is carbapenem resistant.",
  },
  {
    accession: "ARO:3003148",
    name: "OXA-425",
    description:
      "OXA-425 is a \u00df-lactamase found in clinical isolates of Acinetobacter baumannii. It is carbapenem-resistant",
  },
  {
    accession: "ARO:3003149",
    name: "OXA-426",
    description:
      "OXA-426 is a beta-lactamase found in clinical isolates of Acinetobacter baumannii. It is carbapenem resistant",
  },
  {
    accession: "ARO:3003150",
    name: "CARB-20",
    description:
      "CARB-20 is a beta-lactamase found worldwide in Vibrio parahaemolyticus",
  },
  {
    accession: "ARO:3003151",
    name: "CARB-22",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003152",
    name: "SHV-185",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003153",
    name: "SHV-186",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003154",
    name: "SHV-187",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003155",
    name: "SHV-188",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003156",
    name: "SHV-189",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003157",
    name: "TEM-219",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003158",
    name: "TEM-220",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003159",
    name: "OXA-419",
    description: "From the Lahey list of beta-lactamases. Not yet released.",
  },
  {
    accession: "ARO:3003160",
    name: "OXA-422",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003161",
    name: "OXA-423",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003162",
    name: "OXA-435",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003163",
    name: "CTX-M-155",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003164",
    name: "CTX-M-156",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003165",
    name: "CTX-M-157",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003166",
    name: "CTX-M-158",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003167",
    name: "CTX-M-159",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003168",
    name: "CTX-M-160",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003169",
    name: "CTX-M-161",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003170",
    name: "CMY-131",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003171",
    name: "ACT-36",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003172",
    name: "ACT-37",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003173",
    name: "MIR-17",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003174",
    name: "CARB-18",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003175",
    name: "CARB-19",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003176",
    name: "CARB-21",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003177",
    name: "IMI-7",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003178",
    name: "VIM-42",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003179",
    name: "VIM-43",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003180",
    name: "KPC-22",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003181",
    name: "GES-26",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003182",
    name: "NDM-13",
    description:
      "Beta-lactamase found in Escherichia coli clinical isolates in Nepal.",
  },
  {
    accession: "ARO:3003183",
    name: "NDM-14",
    description: "NDM-14 is a beta-lactamase found in Acinetobacter lwoffii.",
  },
  {
    accession: "ARO:3003184",
    name: "PER-8",
    description: "From the Lahey list of beta-lactamases.",
  },
  {
    accession: "ARO:3003185",
    name: "GES-25",
    description:
      "Assigned by Lahey's list of beta-lactamases, not yet released",
  },
  {
    accession: "ARO:3003186",
    name: "CARB-23",
    description: "Present in Lahey's list of beta-lactamases, not yet released",
  },
  {
    accession: "ARO:3003187",
    name: "QnrB75",
    description:
      "QnrB75 is a fluoroquinolone resistance gene in Lahey's list of qnr genes",
  },
  {
    accession: "ARO:3003188",
    name: "QnrB76",
    description:
      "QnrB76 is a fluoroquinolone resistance gene in Lahey's list of qnr genes",
  },
  {
    accession: "ARO:3003189",
    name: "QnrB77",
    description:
      "QnrB77 is a fluoroquinolone resistance gene in Lahey's list of qnr genes",
  },
  {
    accession: "ARO:3003190",
    name: "QnrB78",
    description:
      "QnrB78 is a fluoroquinolone resistance gene in Lahey's list of qnr genes",
  },
  {
    accession: "ARO:3003191",
    name: "QnrB79",
    description:
      "QnrB79 is a fluoroquinolone resistance gene in Lahey's list of qnr genes",
  },
  {
    accession: "ARO:3003192",
    name: "QnrB80",
    description:
      "QnrB80 is a fluoroquinolone resistance gene in Lahey's list of qnr genes",
  },
  {
    accession: "ARO:3003193",
    name: "QnrVC7",
    description:
      "QnrVC7 is a fluoroquinolone resistance gene in Lahey's list of qnr genes.",
  },
  {
    accession: "ARO:3003194",
    name: "GIM-2",
    description:
      "GIM-2 is a metallo-beta-lactamase present on an integron found in clinical isolates of Enterobacter cloacae in Germany",
  },
  {
    accession: "ARO:3003195",
    name: "GIM beta-lactamase",
    description:
      "GIM beta-lactamase enzymes isolated from Pseudomonas aeruginosa, and found to confer broad-spectrum resistance to beta-lactam antibiotics.",
  },
  {
    accession: "ARO:3003196",
    name: "tet(45)",
    description:
      "tet45 is a tetracycline efflux pump found in Bhargavaea cecembensis strain previously isolated from a poultry-litter-impacted soil.",
  },
  {
    accession: "ARO:3003197",
    name: "aadA25",
    description:
      "streptomycin/spectinomycin resistance gene found in Pasteurella multocida isolated from bovine respiratory tract",
  },
  {
    accession: "ARO:3003198",
    name: "rmtH",
    description:
      "rmtH is a 16s ribosomal RNA methyltransferase found in Klebsiella pneumoniae strain MRSN2404 that was isolated from the chronic wound of a soldier wounded in Iraq in 2006. It confers high resistance to all aminoglycosides",
  },
  {
    accession: "ARO:3003199",
    name: "AAC(6')-Iak",
    description:
      "AAC(6')-Iak is a 6'-N-aminoglycoside acetyltransferase-encoding gene isolated from a multidrug-resistant clinical isolate of Stenotrophomonas maltophilia",
  },
  {
    accession: "ARO:3003200",
    name: "AAC(6')-Ian",
    description:
      "AAC(6')-Ian is an amikacin acetyltransferase gene found on a transferable plasmid of the Serratia marcescens strain NUBL-11663. IT catalyze the transfer of an acetyl group from acetyl coenzyme A onto an amine at the 6'-position of various aminoglycosides.",
  },
  {
    accession: "ARO:3003201",
    name: "TLA beta-lactamase",
    description:
      "The TLA beta-lactamases are resistant to expanded-spectrum cephalosporins, aztreonam, ciprofloxacin, and ofloxacin but was susceptible to amikacin, cefotetan, and imipenem.",
  },
  {
    accession: "ARO:3003202",
    name: "TLA-1",
    description:
      "TLA-1 is a beta-lactamase found in plasmids of clinical isolates of Escherichia coli strain R170 in Latin America. It preferentially hydrolyzed cephaloridine, cefotaxime, cephalothin, benzylpenicillin, and ceftazidime. The enzyme was markedly inhibited by sulbactam, tazobactam, and clavulanic acid.",
  },
  {
    accession: "ARO:3003203",
    name: "TLA-2",
    description:
      "TLA-2 is a beta-lactamase is present on a plasmid isolated from an unidentified bacterial strain from a waste water treatment plant. The enzyme mostly hydrolyzes cephalosporins.",
  },
  {
    accession: "ARO:3003204",
    name: "TLA-3",
    description:
      "TLA-3 is a beta-lactamase found in a transferable plasmid of Serratia marcescens clinical isolate. It confers resistance to ceftazidime, cefotaxime and cefepime, but not to cefmetazole and meropenem",
  },
  {
    accession: "ARO:3003205",
    name: "Erm(43)",
    description:
      "Erm(43) is a macrolide, lincosamide, and streptogramin B resistance gene found in Staphylococcus lentus chromosome isolated from human, dog and chicken.",
  },
  {
    accession: "ARO:3003206",
    name: "lsaE",
    description:
      "lsaE is an ABC-F subfamily protein found in porcine MRSA isolates. It confers resistance to pleuromutilin, lincosamide, and streptogramin A",
  },
  {
    accession: "ARO:3003207",
    name: "FosK",
    description:
      "FosK is a fosfomycin thiol transferase isolated from Acinetobacter soli. It is integrin-mediated. It confers a high level of resistance to fosfomycin.",
  },
  {
    accession: "ARO:3003208",
    name: "FosXCC",
    description:
      "fosXCC is a fosfomycin resistance gene that modifies MurA isolated from Campylobacter species. It is highly resistant to fosfomycin.",
  },
  {
    accession: "ARO:3003209",
    name: "FosA5",
    description:
      "fosA5 is a fosfomycin resistance gene isolated from clinical strain of Escherichia coli E265. It is susceptible to amikacin, tetracycline and imipenem, and resistant to sulphonamide, cephalosporins, gentamicin, ciprofloxacin, chloramphenicol and streptomycin",
  },
  {
    accession: "ARO:3003210",
    name: "FosA4",
    description:
      "fosA4 is an enzyme that confers resistance to fosfomycin in Escherichia coli by breaking the epoxide ring of the molecule.",
  },
  {
    accession: "ARO:3003211",
    name: "16S rRNA with mutation conferring antibiotic resistance",
    description:
      "Many antibiotics inhibit the growth of bacteria by blocking protein synthesis on the ribosome. The antibiotic-binding sites are located within functionally important structures in the ribosomal RNA. Antibiotic resistance is often conferred by base substitutions or methylations at these sites in the rRNA. Point mutations in the bacterial 16S ribosomal RNA in the small 30S subunit can confer resistance to antibiotics.",
  },
  {
    accession: "ARO:3003223",
    name: "Escherichia coli 16S rRNA mutation conferring resistance to edeine",
    description:
      "Point mutations in the 3' minor domain of the 16S rRNA in Escherichia coli can confer resistance to edeine",
  },
  {
    accession: "ARO:3003231",
    name: "Lipid II",
    description:
      "Precursor molecule during the synthesis of bacterial cell wall. Lipid II anchors itself onto the cell and it's disaccharide-peptide subunit is used in PG polymer assembly by peptidoglycan glycosyltransferases. Due to it's importance in cell wall synthesis, it is the target of antibiotics.",
  },
  {
    accession: "ARO:3003232",
    name: "Lipid III",
    description:
      "Lipid III is the precursor to the wall teichoic acid. It's inhibition leads to bacterial death as teichoic acid is required to anchor autolysins which control the hydrolysis of peptidoglycan cell wall.",
  },
  {
    accession: "ARO:3003236",
    name:
      "Mycobacterium abscessus 16S rRNA mutation conferring resistance to kanamycin",
    description:
      "Point mutations in the 16S rRNA of Mycobacterium abscessus conferring resistance to kanamycin",
  },
  {
    accession: "ARO:3003237",
    name:
      "Mycobacterium abscessus 16S rRNA mutation conferring resistance to tobramycin",
    description:
      "Point mutations in the 16S rRNA of Mycobacterium abscessus conferring resistance to tobramycin",
  },
  {
    accession: "ARO:3003238",
    name:
      "Mycobacterium abscessus 16S rRNA mutation conferring resistance to neomycin",
    description:
      "Point mutations in the 16S rRNA of Mycobacterium abscessus conferring resistance to neomycin",
  },
  {
    accession: "ARO:3003239",
    name:
      "Mycobacterium abscessus 16S rRNA mutation conferring resistance to amikacin",
    description:
      "Point mutations in the 16S rRNA of Mycobacterium abscessus conferring resistance to amikacin",
  },
  {
    accession: "ARO:3003240",
    name:
      "Mycobacterium abscessus 16S rRNA mutation conferring resistance to gentamicin",
    description:
      "Point mutations in the 16S rRNA of Mycobacterium abscessus conferring resistance to gentamicin",
  },
  {
    accession: "ARO:3003241",
    name: "teixobactin",
    description:
      "Teixobactin is a peptide antibiotic isolated from the Gram-negative bacteria, Eleftheria terrae. It inhibits the synthesis of peptidoglycan by interacting with peptidoglycan precursor, Lipid II and Lipid III.",
  },
  {
    accession: "ARO:3003250",
    name: "bcrC",
    description:
      "The bcrC gene product (BcrC) is an undecaprenyl pyrophosphate phosphatase originally isolated from Bacillus subtilis. When overexpressed it can confer resistance to bacitracin.",
  },
  {
    accession: "ARO:3003251",
    name: "undecaprenyl pyrophosphate",
    description:
      "Undecaprenyl pyrophosphate is a precursor to a cell wall polymer carrier. It's activation requires a phosphatase.",
  },
  {
    accession: "ARO:3003252",
    name: "bacitracin resistance gene",
    description: "Genes that provide a bacteria with resistance to bacitracin",
  },
  {
    accession: "ARO:3003253",
    name: "aminocoumarin sensitive parY",
    description: "Subunit of topoisomerase IV sensitive to aminocourmarins",
  },
  {
    accession: "ARO:3003254",
    name: "antibiotic sensitive DNA topoisomerase subunit gyrA",
    description:
      "DNA gyrase is responsible for DNA supercoiling and consists of two alpha and two beta subunits. Binding the alpha-subunit is sufficient to inhibit DNA supercoiling.",
  },
  {
    accession: "ARO:3003256",
    name: "antibiotic sensitive DNA topoisomerase subunit parC",
    description:
      "ParC is a subunit of topoisomerase IV, which decatenates and relaxes DNA to allow access to genes for transcription or translation. Targeting topoisomerase IV inhibits DNA synthesis, and may also lead to double-stranded cell breaks.",
  },
  {
    accession: "ARO:3003258",
    name: "antibiotic sensitive DNA topoisomerase subunit parY",
    description:
      "parY is a subunit of the topoisomerase IV. It decatenates and relaxes DNA to allow for transcription or translation.",
  },
  {
    accession: "ARO:3003259",
    name: "antibiotic sensitive DNA topoisomerase subunit parE",
    description:
      "ParE is a subunit of topoisomerase IV, which decatenates and relaxes DNA to allow access to genes for transcription or translation. Targeting topoisomerase IV inhibits DNA synthesis, and may also lead to double-stranded cell breaks.",
  },
  {
    accession: "ARO:3003260",
    name: "aminocoumarin sensitive gyrB",
    description:
      "Beta-subunit of DNA gyrase sensitive to inhibition by aminocoumarins",
  },
  {
    accession: "ARO:3003261",
    name: "fluoroquinolone sensitive gyrB",
    description:
      "Beta-subunit of DNA gyrase sensitive to inhibition by fluoroquinolones",
  },
  {
    accession: "ARO:3003262",
    name: "Antibiotic sensitive Escherichia coli 16S rRNA",
    description: "Sensitive form of 16s susceptable to antibiotics",
  },
  {
    accession: "ARO:3003263",
    name: "Antibiotic sensitive Mycobacterium abscessus 16S rRNA",
    description: "Sensitive form of 16s susceptable to antibiotics",
  },
  {
    accession: "ARO:3003264",
    name: "Aminoglycosides sensitive Mycobacterium abscessus 16S rRNA",
    description: "Sensitive form of 16s susceptable to antibiotics",
  },
  {
    accession: "ARO:3003265",
    name: "Amikacin sensitive Mycobacterium abscessus 16S rRNA mutation",
    description: "Sensitive form of 16s susceptable to antibiotics",
  },
  {
    accession: "ARO:3003272",
    name: "daptomycin resistant cls",
    description:
      "Cardiolipin synthetase catalyzes the formation of cardiolipin from two phosphatidylglycerol molecules. Cardiolipin is important in membrane translocation and permeabilization. Current known mutations on the enzyme confer resistance to daptomycin.",
  },
  {
    accession: "ARO:3003273",
    name: "cell membrane component enzymes targeted by antibiotic",
    description:
      "Interaction with the cell membrane phospholipids interferes with membrane permeability and integrity, leading to the leakage of cell contents. The outer membrane of Gram-negative bacteria is particularly sensitive.",
  },
  {
    accession: "ARO:3003274",
    name: "antibiotic sensitive cls",
    description:
      "Cardiolipin synthetase catalyzes the formation of cardiolipin from two phosphatidylglycerol molecules. Cardiolipin is important in membrane translocation and permeabilization. The inactivation of this enzyme by antibiotics leads to cell death.",
  },
  {
    accession: "ARO:3003275",
    name: "Daptomycin sensitive cardiolipin synthetase",
    description:
      "Cardiolipin synthetase catalyzes the formation of cardiolipin from two phosphatidylglycerol molecules. Cardiolipin is important in membrane translocation and permeabilization. Mutations on the enzyme confers resistance to Daptomycin",
  },
  {
    accession: "ARO:3003276",
    name: "antibiotic resistant rpoB",
    description:
      "RNA polymerase is a multisubunit enzyme that is necessary for transcription. The beta-subunit of RNA polymerase forms the active center of the enzyme and template/transcript binding sites. Mutations in rpoB gene confers antibiotic resistance",
  },
  {
    accession: "ARO:3003277",
    name: "antibiotic sensitive beta-subunit of RNA polymerase (rpoB)",
    description:
      "RNA polymerase is a multisubunit enzyme that is necessary for transcription. The beta-subunit of RNA polymerase forms the active center of the enzyme and template/transcript binding sites. This enzyme can be inactivated by antibiotics leading to cell death.",
  },
  {
    accession: "ARO:3003278",
    name: "daptomycin sensitive beta-subunit of RNA polymerase (rpoB)",
    description:
      "Daptomycin binds to the cytoplasmic membrane of Gram-positive. It's lipopeptide tail servers as a ion channel for potassium. This causes a rapid depolarization that result in failures in DNA/RNA/Protein synthesis. Eventually cell death.",
  },
  {
    accession: "ARO:3003279",
    name: "antibiotic resistant liaFSR system",
    description:
      "the liaFSR system regulates the cell envelope stress response. It is transcriptionally activated by exposure to alkaline shock, detergents, and particularly antibiotics with lipid II inhibition properties.",
  },
  {
    accession: "ARO:3003280",
    name: "antibiotic sensitive mprF",
    description:
      "MprF is a integral membrane protein that modifies the negatively-charged phosphatidylglycerol on the membrane surface of both Gram-positive and Gram-negative bacteria. This confers resistance to cationic peptides that disrupt the cell membrane, including defensins. Mutations in this enzyme can confer additional resistances.",
  },
  {
    accession: "ARO:3003281",
    name: "antibiotic sensitive pgsA",
    description:
      "pgsA or phosphatidylglycerophosphate synthetase is an integral membrane protein involved in phospholipid biosynthesis. Inhibition of this enzyme leads to cell death via the inability to form the plasma membrane.",
  },
  {
    accession: "ARO:3003283",
    name:
      "Mycobacterium tuberculosis rpoB mutants conferring resistance to rifampicin",
    description:
      "Point mutations that occurs in Mycobacterium tuberculosis rpoB resulting in resistance to rifampicin.",
  },
  {
    accession: "ARO:3003284",
    name:
      "Mycobacterium leprae rpoB mutations conferring resistance to rifampicin",
    description:
      "Point mutations that occurs in Mycobacterium leprae rpoB resulting in resistance to rifampicin",
  },
  {
    accession: "ARO:3003285",
    name:
      "Staphylococcus aureus rpoB mutants conferring resistance to rifampicin",
    description:
      "Point mutations that occurs in Staphylococcus aureus rpoB resulting in resistance to rifampicin",
  },
  {
    accession: "ARO:3003287",
    name:
      "Staphylococcus aureus rpoB mutants conferring resistance to daptomycin",
    description:
      "Point mutations that occurs in Staphylococcus aureus rpoB resulting in resistance to daptomycin",
  },
  {
    accession: "ARO:3003288",
    name: "Escherichia coli rpoB mutants conferring resistance to rifampicin",
    description:
      "Point mutations that occurs in Escherichia coli rpoB resulting in resistance to rifampicin",
  },
  {
    accession: "ARO:3003289",
    name: "antibiotic resistant rpoC",
    description:
      "RNA polymerase is a multisubunit enzyme that is necessary for transcription. The beta prime subunit of RNA polymerase forms the active center of the enzyme and template/transcript binding sites. Mutations in rpoC gene confers antibiotic resistance",
  },
  {
    accession: "ARO:3003290",
    name: "daptomycin resistant beta prime subunit of RNA polymerase (rpoC)",
    description:
      "Daptomycin resistant RNA polymerases include amino acids substitutions which alter the binding affinity of daptomycin to the protein, resulting in antibiotic resistance.",
  },
  {
    accession: "ARO:3003291",
    name: "Staphylococcus aureus rpoC conferring resistance to daptomycin",
    description:
      "Point mutations that occurs in Staphylococcus aureus rpoC resulting in resistance to daptomycin",
  },
  {
    accession: "ARO:3003292",
    name: "fluoroquinolone resistant gyrA",
    description:
      "DNA gyrase is responsible for DNA supercoiling and consists of two alpha and two beta subunits. GyrA point mutations confer resistance by preventing fluoroquinolone antibiotics from binding the alpha-subunit.",
  },
  {
    accession: "ARO:3003294",
    name: "Escherichia coli gyrA conferring resistance to fluoroquinolones",
    description:
      "Point mutation of Escherichia coli gyrA resulted in the lowered affinity between fluoroquinolones and gyrA. Thus, conferring resistance",
  },
  {
    accession: "ARO:3003295",
    name:
      "Mycobacterium tuberculosis gyrA conferring resistance to fluoroquinolones",
    description:
      "Point mutation of Mycobacterium tuberculosis gyrA resulted in the lowered affinity between fluoroquinolones and gyrA. Thus, conferring resistance.",
  },
  {
    accession: "ARO:3003296",
    name:
      "Staphylococcus aureus gyrA conferring resistance to fluoroquinolones",
    description:
      "Point mutation of Staphylococcus aureus gyrA resulted in the lowered affinity between fluoroquinolones and gyrA. Thus, conferring resistance.",
  },
  {
    accession: "ARO:3003297",
    name:
      "Bartonella bacilliformis gyrA conferring resistance to fluoroquinolones",
    description:
      "Point mutation of Bartonella bacilliformis gyrA resulted in the lowered affinity between fluoroquinolones and gyrA. Thus, conferring resistance.",
  },
  {
    accession: "ARO:3003298",
    name: "Mycobaterium leprae gyrA conferring resistance to fluoroquinolones",
    description:
      "Point mutation of Mycobaterium leprae gyrA resulted in the lowered affinity between fluoroquinolones and gyrA. Thus, conferring resistance.",
  },
  {
    accession: "ARO:3003301",
    name: "Staphylococcus aureus gyrB conferring resistance to aminocoumarin",
    description:
      "Point mutation in Staphylococcus aureus resulting in aminocoumarin resistance",
  },
  {
    accession: "ARO:3003302",
    name:
      "Bartonella bacilliformis gyrB conferring resistance to aminocoumarin",
    description:
      "Point mutation in Bartonella bacilliformis resulting in aminocoumarin resistance",
  },
  {
    accession: "ARO:3003303",
    name: "Escherichia coli gyrB conferring resistance to aminocoumarin",
    description:
      "Point mutation in Escherichia coli resulting in aminocoumarin resistance",
  },
  {
    accession: "ARO:3003304",
    name: "Mycobacterium leprae gyrB conferring resistance to fluoroquinolone",
    description:
      "Point mutation in Mycobacterium leprae gyrB resulting in fluoroquinolone resistance",
  },
  {
    accession: "ARO:3003305",
    name:
      "Ureaplasma urealyticum  gyrB conferring resistance to fluoroquinolone",
    description:
      "Point mutation in Ureaplasma urealyticum resulting in fluoroquinolone resistance",
  },
  {
    accession: "ARO:3003306",
    name: "Morganella morganii gyrB conferring resistance to fluoroquinolone",
    description:
      "Point mutation in Morganella morganii resulting in fluoroquinolone resistance",
  },
  {
    accession: "ARO:3003307",
    name: "Salmonella serovars gyrB conferring resistance to fluoroquinolone",
    description:
      "Point mutation in Salmonella serovars resulting in fluoroquinolone resistance",
  },
  {
    accession: "ARO:3003308",
    name: "Escherichia coli parC conferring resistance to fluoroquinolone",
    description:
      "Point mutation in Escherichia coli parC resulting in fluoroquinolone resistance",
  },
  {
    accession: "ARO:3003309",
    name:
      "Ureaplasma urealyticum parC conferring resistance to fluoroquinolone",
    description:
      "Point mutation in Ureaplasma urealyticum parC resulting in fluoroquinolone resistance",
  },
  {
    accession: "ARO:3003310",
    name: "Mycoplasma hominis parC conferring resistance to fluoroquinolone",
    description:
      "Point mutation in Mycoplasma hominis parC resulting in fluoroquinolone resistance",
  },
  {
    accession: "ARO:3003311",
    name:
      "Streptococcus pneumoniae parC conferring resistance to fluoroquinolone",
    description:
      "Point mutation in Streptococcus pneumoniae parC resulting in fluoroquinolone resistance",
  },
  {
    accession: "ARO:3003312",
    name: "Staphylococcus aureus parC conferring resistance to fluoroquinolone",
    description:
      "Point mutation in Staphylococcus aureus parC resulting in fluoroquinolone resistance",
  },
  {
    accession: "ARO:3003313",
    name: "fluoroquinolone resistant parE",
    description:
      "ParE is a subunit of topoisomerase IV, necessary for cell survival. Point mutations in ParE prevent fluoroquinolones from inhibiting DNA synthesis, thus conferring resistance.",
  },
  {
    accession: "ARO:3003314",
    name: "Staphylococcus aureus parE conferring resistance to aminocoumarin",
    description:
      "Point mutation in Staphylococcus aureus parE resulting in aminocoumarin resistance. Please note: the reference sequence associated with this model corresponds to Staphylococcus aureus MRSA parE, while the SNPs correspond to Staphylococcus aureus RN4220 (for which a complete annotated genome does not exist in genbank at present).",
  },
  {
    accession: "ARO:3003315",
    name:
      "Staphylococcus aureus parE conferring resistance to fluoroquinolones",
    description:
      "Point mutation in Staphylococcus aureus parE resulting in fluoroquinolones resistance",
  },
  {
    accession: "ARO:3003316",
    name: "Escherichia coli parE conferring resistance to fluoroquinolones",
    description:
      "Point mutation in Escherichia coli parE resulting in fluoroquinolones resistance",
  },
  {
    accession: "ARO:3003317",
    name: "Salmonella serovars parE conferring resistance to fluoroquinolones",
    description:
      "Point mutation in Salmonella serovars parE resulting in fluoroquinolones resistance",
  },
  {
    accession: "ARO:3003318",
    name:
      "Streptomyces rishiriensis parY mutant conferring resistance to aminocoumarin",
    description:
      "point mutation on the Streptomyces rishiriensis parY resulting in aminocoumarin resistance",
  },
  {
    accession: "ARO:3003319",
    name:
      "Staphylococcus aureus mprF with mutation conferring resistance to daptomycin",
    description:
      "Point mutations that occur within Staphylococcus aureus mprF gene resulting in resistance to daptomycin",
  },
  {
    accession: "ARO:3003323",
    name:
      "Staphylococcus aureus pgsA mutations conferring resistance to daptomycin",
    description:
      "Point mutations that occur within Staphylococcus aureus pgsA gene resulting in resistance to daptomycin",
  },
  {
    accession: "ARO:3003324",
    name: "Bacillus subtilis mprF",
    description:
      "MprF is a integral membrane protein that modifies the negatively-charged phosphatidylglycerol on the membrane surface. This confers resistance to cationic peptides that disrupt the cell membrane, including defensins. Additionally, large-scale mutations causing loss of function of the gene result in increased susceptibility to daptomycin.",
  },
  {
    accession: "ARO:3003325",
    name:
      "Mycobacterium bovis embB with mutation conferring resistance to ethambutol",
    description:
      "Point mutations that occur within Mycobacterium bovis embB gene resulting in resistance to ethambutol",
  },
  {
    accession: "ARO:3003326",
    name:
      "Mycobacterium tuberculosis embB with mutation conferring resistance to ethambutol",
    description:
      "Point mutations that occur within Mycobacterium tuberculosis embB gene resulting in resistance to ethambutol",
  },
  {
    accession: "ARO:3003327",
    name:
      "Mycobacterium tuberculosis mutant embC conferring resistance to ethambutol",
    description:
      "Point mutations in Mycobacterium tuberculosis embC that result in lower affinity between ethambutol and EmbC, resulting in resistance",
  },
  {
    accession: "ARO:3003333",
    name:
      "Escherichia coli 16S rRNA (rrsC) mutation conferring resistance to kasugamicin",
    description:
      "Point mutations in the 3' minor, 3' major, and central domains in the rrsC 16S rRNA gene of Escherichia coli can confer resistance to kasugamicin",
  },
  {
    accession: "ARO:3003356",
    name: "antibiotic resistant EF-Tu",
    description:
      "Sequence variants of elongation factor Tu that confer resistance to different classes of antibiotics.",
  },
  {
    accession: "ARO:3003357",
    name:
      "Clostridium difficile EF-Tu mutants conferring resistance to elfamycin",
    description:
      "Sequence variants of Clostridium difficile elongation factor Tu that confer resistance to elfamycin antibiotics.",
  },
  {
    accession: "ARO:3003358",
    name: "Escherichia coli EF-Tu mutants conferring resistance to elfamycin",
    description:
      "Sequence variants of Escherichia coli elongation factor Tu that confer resistance to elfamycin antibiotics.",
  },
  {
    accession: "ARO:3003359",
    name:
      "Streptomyces cinnamoneus EF-Tu mutants conferring resistance to elfamycin",
    description:
      "Sequence variants of Streptomyces cinnamoneus elongation factor Tu that confer resistance to elfamycin antibiotics.",
  },
  {
    accession: "ARO:3003360",
    name: "Planobispora rosea EF-Tu mutants conferring resistance to elfamycin",
    description:
      "Sequence variants of Planobispora rosea elongation factor Tu that confer resistance to elfamycin antibiotics.",
  },
  {
    accession: "ARO:3003361",
    name:
      "Planobispora rosea EF-Tu mutants conferring resistance to inhibitor GE2270A",
    description:
      "Sequence variants of Planobispora rosea elongation factor Tu that confer resistance to inhibitor GE2270A",
  },
  {
    accession: "ARO:3003368",
    name: "Escherichia coli EF-Tu mutants conferring resistance to kirromycin",
    description:
      "Sequence variants of Escherichia coli elongation factor Tu that confer resistance to kirromycin",
  },
  {
    accession: "ARO:3003369",
    name: "Escherichia coli EF-Tu mutants conferring resistance to Pulvomycin",
    description:
      "Sequence variants of Escherichia coli elongation factor Tu that confer resistance to Pulvomycin",
  },
  {
    accession: "ARO:3003370",
    name:
      "Escherichia coli EF-Tu mutants conferring resistance to Enacyloxin IIa",
    description:
      "Sequence variants of Escherichia coli elongation factor Tu that confer resistance to Enacyloxin IIa",
  },
  {
    accession: "ARO:3003372",
    name:
      "Escherichia coli 16S rRNA (rrsH) mutation conferring resistance to spectinomycin",
    description:
      "Point mutations in the 3' major domain of the rrsH 16S rRNA gene of Escherichia coli can confer resistance to spectinomycin",
  },
  {
    accession: "ARO:3003373",
    name:
      "Klebsiella pneumoniae acrR with mutation conferring multidrug antibiotic resistance",
    description:
      "AcrR is a repressor of the AcrAB-TolC multidrug efflux complex. AcrR mutations result in high level antibiotic resistance.",
  },
  {
    accession: "ARO:3003374",
    name:
      "Enterobacter aerogenes acrR with mutation conferring multidrug antibiotic resistance",
    description:
      "AcrR is a repressor of the AcrAB-TolC multidrug efflux complex. AcrR mutations result in high level antibiotic resistance.",
  },
  {
    accession: "ARO:3003376",
    name:
      "Escherichia coli 16S rRNA (rrsB) mutation conferring resistance to spectinomycin",
    description:
      "Point mutations in the 3' major domain of the rrsB 16S rRNA gene of Escherichia coli can confer resistance to spectinomycin",
  },
  {
    accession: "ARO:3003377",
    name:
      "Escherichia coli 16S rRNA (rrnB) mutation conferring resistance to spectinomycin",
    description:
      "Point mutations in the 3' major domain of helix 35, in the rrnB gene operon for 16S rRNA of Escherichia coli can confer resistance to spectinomycin",
  },
  {
    accession: "ARO:3003378",
    name: "Escherichia coli marR mutant conferring antibiotic resistance",
    description:
      "MarR is a repressor of the mar operon marRAB, thus regulating the expression of marA, the activator of multidrug efflux pump AcrAB.",
  },
  {
    accession: "ARO:3003379",
    name: "Salmonella enterica ramR mutants",
    description:
      "RamR is a repressor that regulates RamA expression. Mutations lead to the upregulation of AcrAB, which is positively regulated by RamA.",
  },
  {
    accession: "ARO:3003380",
    name: "Klebsiella pneumoniae ramR mutants",
    description:
      "RamR is a repressor that regulates RamA expression. Mutations lead to the upregulation of AcrAB, which is positively regulated by RamA.",
  },
  {
    accession: "ARO:3003381",
    name:
      "Escherichia coli soxR with mutation conferring antibiotic resistance",
    description:
      "SoxR is a sensory protein that upregulates soxS expression in the presence of redox-cycling drugs. This stress response leads to the expression many multidrug efflux pumps.",
  },
  {
    accession: "ARO:3003382",
    name:
      "Salmonella enterica soxR with mutation conferring antibiotic resistance",
    description:
      "SoxR is a sensory protein that upregulates soxS expression in the presence of redox-cycling drugs. This stress response leads to the expression many multidrug efflux pumps.",
  },
  {
    accession: "ARO:3003383",
    name:
      "Salmonella serovars soxS with mutation conferring antibiotic resistance",
    description:
      "SoxS is a global regulator that up-regulates the expression of AcrAB efflux genes. It also reduces OmpF expression to decrease cell membrane permeability.",
  },
  {
    accession: "ARO:3003385",
    name: "Enterobacter aerogenes Omp36",
    description:
      "Mutant forms of the porin Omp36 result in reduced permeability to antibiotics.",
  },
  {
    accession: "ARO:3003386",
    name:
      "Escherichia coli folP with mutation conferring resistance to sulfonamides",
    description:
      "Point mutations in Escherichia coli dihydropteroate synthase folP prevent sulfonamide antibiotics from inhibiting its role in folate synthesis, thus conferring sulfonamide resistance.",
  },
  {
    accession: "ARO:3003387",
    name:
      "Streptococcus pyogenes folP with mutation conferring resistance to sulfonamides",
    description:
      "Point mutations in Streptococcus pyogenes dihydropteroate synthase folP prevent sulfonamide antibiotics from inhibiting its role in folate synthesis, thus conferring sulfonamide resistance.",
  },
  {
    accession: "ARO:3003388",
    name: "dapsone resistant dihydropteroate synthase folP",
    description:
      "Dapsone inhibits bacterial synthesis of dihydrofolic acid by competing with with para-aminobenzoate for the active site of dihydropteroate synthetase. Thus acts as a competitive inhibitor of folP. Point mutation within the folP gene results in lowered affinity of dapsone for folP",
  },
  {
    accession: "ARO:3003389",
    name:
      "Mycobacterium leprae folP with mutation conferring resistance to dapsone",
    description:
      "Dapsone inhibits bacterial synthesis of dihydrofolic acid by competing with with para-aminobenzoate for the active site of dihydropteroate synthetase. Point mutation within the Mycobacterium leprae folP gene results in lowered affinity of dapsone for folP.",
  },
  {
    accession: "ARO:3003390",
    name:
      "Escherichia coli ompF with mutation conferring resistance to beta-lactam antibiotics",
    description:
      "The Escherichia coli ompF (oprF) is a nonspecific porin channel involved in the membrane translocation of small hydrophilic molecules, including and especially beta-lactam antibiotics. Mutations in ompF can decrease diffusion of antibiotics across the cellular membrane, thereby decreasing overall susceptibility through absence of porin function.",
  },
  {
    accession: "ARO:3003392",
    name:
      "Mycobacterium tuberculosis katG mutations conferring resistance to isoniazid",
    description:
      "katG is a catalase-peroxidase that catalyzes the activation of isoniazid. Isoniazid inhibits mycolic acid synthesis, which prevents cell wall synthesis in mycobacteria. Mutations in katG results in inability to activate isoniazid. Over 280 different mutations have been documented in PubMed for katG, with mutations to Ser315 being the most prevalent.",
  },
  {
    accession: "ARO:3003393",
    name:
      "Mycobacterium tuberculosis inhA mutations conferring resistance to isoniazid",
    description:
      "inhA is a enoyl-acyl carrier reductase used in lipid metabolism and farry acid biosynthesis. It is inhibited by isoniazid. mutations in the promoter region or multiple copies of the inhA shows marked resistance to isoniazid mediated inhibition of mycolic acid biosynthesis.",
  },
  {
    accession: "ARO:3003394",
    name:
      "Mycobacterium tuberculosis pncA mutations conferring resistance to pyrazinamide",
    description:
      "pncA is a pyrazinamidase/nicotinamidase. It catalyzes the activation of pyrazinamide. Some mutation within pncA are associated with loss of enzyme activity, resulting in pyrazinamide resistance.",
  },
  {
    accession: "ARO:3003395",
    name:
      "Mycobacterium tuberculosis rpsL mutations conferring resistance to Streptomycin",
    description:
      "Ribosomal protein S12 stabilizes the highly conserved\npseudoknot structure formed by 16S rRNA. Amino acid\nsubstitutions in RpsL affect the higher-order structure of 16S\nrRNA and confer streptomycin resistance by disrupting interactions between 16S rRNA and streptomycin.",
  },
  {
    accession: "ARO:3003396",
    name:
      "Escherichia coli 16S rRNA (rrsB) mutation conferring resistance to gentamicin C",
    description:
      "Point mutations in the 3' minor domain of helix 44, in the rrsB 16S rRNA gene of Escherichia coli can confer resistance to gentamicin C",
  },
  {
    accession: "ARO:3003397",
    name:
      "Escherichia coli 16S rRNA (rrsB) mutation conferring resistance to G418",
    description:
      "Point mutations in the 3' minor domain of helix 44, in the rrsB 16S rRNA gene of Escherichia coli can confer resistance to G418",
  },
  {
    accession: "ARO:3003398",
    name: "undecaprenyl pyrophosphate related proteins",
    description:
      "Undecaprenyl phosphate is a universal lipid carrier of glycan biosynthetic intermediates for carbohydrate polymers that are exported to the bacterial cell envelope. Antibiotics that targets this compound or proteins associated with the production of this compound leads to cell death.",
  },
  {
    accession: "ARO:3003399",
    name:
      "Escherichia coli 16S rRNA (rrsB) mutation conferring resistance to kanamycin A",
    description:
      "Point mutations in the 3' minor domain of helix 44, in the rrsB 16S rRNA gene of Escherichia coli can confer resistance to kanamycin A",
  },
  {
    accession: "ARO:3003401",
    name: "sulfone antibiotic",
    description:
      "A sulfone active against a wide range of bacteria but mainly employed for its actions against mycobacterium laprae. Its mechanism of action  involves inhibition of folic acid synthesis in susceptible organisms.",
  },
  {
    accession: "ARO:3003402",
    name:
      "Escherichia coli 16S rRNA (rrsB) mutation conferring resistance to neomycin",
    description:
      "Point mutations in the 3' minor domain of helix 44, in the rrsB 16S rRNA gene of Escherichia coli can confer resistance to neomycin",
  },
  {
    accession: "ARO:3003403",
    name:
      "Escherichia coli 16S rRNA (rrsB) mutation conferring resistance to paromomycin",
    description:
      "Point mutations in the 3' minor domain of helix 44, in the rrsB 16S rRNA gene of Escherichia coli can confer resistance to paromomycin",
  },
  {
    accession: "ARO:3003405",
    name:
      "Escherichia coli 16S rRNA (rrsB) mutation conferring resistance to streptomycin",
    description:
      "Point mutations in the 5' domain of helix 18, in the rrsB 16S rRNA gene of Escherichia coli can confer resistance to streptomycin",
  },
  {
    accession: "ARO:3003406",
    name:
      "Escherichia coli 16S rRNA (rrnB) mutation conferring resistance to streptomycin",
    description:
      "Point mutations in the 5' domain of helix 18, in the rrnB 16S rRNA gene of Escherichia coli can confer resistance to streptomycin",
  },
  {
    accession: "ARO:3003408",
    name:
      "Escherichia coli 16S rRNA (rrsB) mutation conferring resistance to tobramycin",
    description:
      "Point mutations in the 3' minor domain of the rrnB 16S rRNA gene of Escherichia coli can confer resistance to tobramycin",
  },
  {
    accession: "ARO:3003410",
    name:
      "Escherichia coli 16S rRNA (rrsB) mutation conferring resistance to tetracycline",
    description:
      "Point mutations in the 3' major domain of the rrsB 16S rRNA gene of Escherichia coli can confer resistance to tetracycline",
  },
  {
    accession: "ARO:3003411",
    name:
      "Escherichia coli 16S rRNA (rrnB) mutation conferring resistance to tetracycline",
    description:
      "Point mutations in the 3' major domain of the rrnB 16S rRNA gene of Escherichia coli can confer resistance to tetracycline",
  },
  {
    accession: "ARO:3003412",
    name: "dapsone",
    description:
      "Dapsone is a sulfone in which it inhibits folic acid synthesis, such as the dihydropteroate synthase.",
  },
  {
    accession: "ARO:3003413",
    name: "pyrazinamide",
    description:
      "Pyrazinamide is an Antimycobacterial. It is highly specific and active only against Mycobacterium tuberculosis. This compound is a prodrug and needs to be activated inside the cell. It interferes with the bacterium's ability to synthesize new fatty acids, causing cell death.",
  },
  {
    accession: "ARO:3003414",
    name: "LFF571",
    description:
      "LFF571 is a novel semi-synthetic thiopeptide antibiotic derived from GE2270. It has been shown to possess potent in vitro and in vivo activity against Gram-positive bacteria. It is hypothesized that it a translation inhibitor leading to cell death.",
  },
  {
    accession: "ARO:3003415",
    name: "sulfonamide resistant dihydropteroate synthase folP",
    description:
      "Point mutations in dihydropteroate synthase folP prevent sulfonamide antibiotics from inhibiting its role in folate synthesis, thus conferring sulfonamide resistance",
  },
  {
    accession: "ARO:3003416",
    name: "isoniazid resistant katG",
    description:
      "Bifunctional enzyme with both catalase and broad-spectrum peroxidase activity. It is a catalase-peroxidases that catalyzes the activation of isoniazid. Mutations that arises within this protein cause changes that results in the inability for katG to activate antibiotics, conferring resistance",
  },
  {
    accession: "ARO:3003417",
    name: "antibiotic resistant inhA",
    description:
      "inhA is a enoyl-acyl carrier reductase used in lipid metabolism and fatty acid biosynthesis. It is inhibited by isoniazid. Mutations in the promoter region or multiple copies of the inhA shows marked resistance to isoniazid mediated inhibition of mycolic acid biosynthesis.",
  },
  {
    accession: "ARO:3003418",
    name: "pyrazinamide resistant pncA",
    description:
      "pncA is a pyrazinamidase/nicotinamidase. It catalyzes the activation of pyrazinamide to pyrazinoic acid. Mutations arise within the pncA gene that caused the loss of pyrazinamidase activity is the major mechanism of antibiotic resistance",
  },
  {
    accession: "ARO:3003419",
    name: "antibiotic resistant rpsL",
    description:
      "Ribosomal protein S12 stabilizes the highly conserved pseudoknot structure formed by 16S rRNA. Amino acid substitutions in RpsL affect the higher-order structure of 16S rRNA and confer antibiotic resistance",
  },
  {
    accession: "ARO:3003420",
    name: "antibiotic resistant pgsA",
    description:
      "pgsA or phosphatidylglycerophosphate synthetase is an integral membrane protein involved in phospholipid biosynthesis. It is a CDP-diacylglycerol-glycerol-3-phosphate 3-phosphatidyltransferase.",
  },
  {
    accession: "ARO:3003421",
    name: "antibiotic resistant mprF",
    description:
      "Catalyzes the transfer of a lysyl group from L-lysyl-tRNA(Lys) to membrane-bound phosphatidylglycerol (PG), which produces lysylphosphatidylglycerol (LPG), a major component of the bacterial membrane with a positive net charge. LPG synthesis contributes to bacterial virulence as it is involved in the resistance mechanism against cationic antimicrobial peptides (CAMP) produces by the host's immune system (defensins, cathelicidins) and by the competing microorganisms (bacteriocins)",
  },
  {
    accession: "ARO:3003422",
    name: "antibiotic resistant arabinosyltransferase",
    description:
      "Arabinosyl transferases allow for the polymerization of arabinose to form arabinan. Arabanan is required for formation of mycobacterial cell walls",
  },
  {
    accession: "ARO:3003425",
    name: "antibiotic resistant dihydrofolate reductase",
    description:
      "Key enzyme in folate metabolism. Catalyzes an essential reaction for de novo glycine and purine synthesis, and for DNA precursor synthesis",
  },
  {
    accession: "ARO:3003426",
    name: "antibiotic sensitive inhA",
    description:
      "inhA is a enoyl-acyl carrier reductase used in lipid metabolism and fatty acid biosynthesis.",
  },
  {
    accession: "ARO:3003428",
    name: "porin proteins permeable to antibiotics",
    description: "Porin proteins permeable to antibiotics.",
  },
  {
    accession: "ARO:3003429",
    name: "antibiotic sensitive beta-prime subunit of RNA polymerase (rpoC)",
    description:
      "RNA polymerase is a multisubunit enzyme that is necessary for transcription. The beta-prime subunit of RNA polymerase forms the active center of the enzyme and template/transcript binding sites. This enzyme can be inactivated by antibiotics leading to cell death.",
  },
  {
    accession: "ARO:3003430",
    name: "fluoroquinolone sensitive parE",
    description:
      "ParE is a subunit of topoisomerase IV, necessary for cell survival. Point mutations in ParE prevent fluoroquinolones from inhibiting DNA synthesis, thus conferring resistance.",
  },
  {
    accession: "ARO:3003431",
    name: "antibiotic sensitive Ribosomal protein S12 (rpsL)",
    description:
      "Ribosomal protein S12 stabilizes the highly conserved pseudoknot structure formed by 16S rRNA. It is targeted and inhibited by streptomycin",
  },
  {
    accession: "ARO:3003432",
    name: "determinant of isoniazid resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to the antibiotic drug, isoniazid.",
  },
  {
    accession: "ARO:3003433",
    name: "determinant of pyrazinamide resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to the antibiotic drug pyrazinamide.",
  },
  {
    accession: "ARO:3003436",
    name:
      "Mycobacterium tuberculosis 16S rRNA mutation conferring resistance to kanamycin",
    description:
      "Point mutations in the 3' domain of 16S rRNA of Mycobacterium tuberculosis can confer resistance to kanamycin.",
  },
  {
    accession: "ARO:3003437",
    name:
      "Mycobacterium tuberculosis 16S rRNA mutation conferring resistance to viomycin",
    description:
      "Peptide antibiotics inhibit protein synthesis. Resistance to viomycin can be caused by an altered RNA molecule in the 16S ribosomal subunit.",
  },
  {
    accession: "ARO:3003438",
    name: "Enterococcus faecium EF-Tu mutants conferring resistance to GE2270A",
    description:
      "Sequence variants of Enterococcus faecium elongation factor Tu that confer resistance to GE2270A",
  },
  {
    accession: "ARO:3003439",
    name: "antibiotic sensitive RNA polymerase switch region",
    description:
      "The switch region is a structural element that mediates conformational changes and contacts required for RNAP to load DNA into the RNAP active-center cleft during transcription initiation. The switch region is located at the base of the RNAP clamp and serves as the hinge that mediates opening of the RNAP clamp to load DNA into the RNAP active-center cleft and mediates closing of the RNAP clamp to retain DNA in the RNAP active-center cleft.",
  },
  {
    accession: "ARO:3003440",
    name: "mecB",
    description:
      "A foreign PBP2 acquired by lateral gene transfer that able to perform peptidoglycan synthesis in the presence of beta-lactams.",
  },
  {
    accession: "ARO:3003441",
    name: "cfrA",
    description:
      "CfrA is a chloramphenicol-florfenicol resistance gene and methyltransferase enzyme. Methylation of position 8 of A2503 in 23S rRNA confers resistance to chloramphenicol antibiotics. Additional Oxazolidinone resistance mediated by the cfr gene in a human isolated was first reported from Colombia in linezolid- and methicillin-resistant staphylocci. Described by Arias et al. 2008.",
  },
  {
    accession: "ARO:3003443",
    name: "Antibiotic resistant tlyA",
    description:
      "tlyA encodes for hemolysin. It Catalyzes the 2'-O-methylation at nucleotides C1409 in 16S rRNA and C1920 in 23S rRNA. Mutation that arise within this gene reduces the binding affinity of aminoglycosides to rRNA",
  },
  {
    accession: "ARO:3003445",
    name:
      "Mycobacterium tuberculosis tlyA mutations conferring resistance to aminoglycosides",
    description:
      "Specific mutations that arise in Mycobacterium tuberculosis tlyA resulting in aminoglycosides resistance",
  },
  {
    accession: "ARO:3003446",
    name: "Antibiotic resistant iniA",
    description:
      "iniA, an isoniazid-induced protein, is part of the iniBAC operon that participates in the development of tolerance to both isoniazid and ethambutol. May function through a MDR-pump like mechanism, although it does not appear to directly transport isoniazid from the cell",
  },
  {
    accession: "ARO:3003447",
    name: "Ethambutol resistant iniA",
    description:
      "Mutations that occurs on the iniA genes resulting in the resistance to ethambutol",
  },
  {
    accession: "ARO:3003448",
    name:
      "Mycobacterium tuberculosis iniA mutant conferring resistance to Ethambutol",
    description:
      "Specific mutations in Mycobacterium tuberculosis iniA resulting in resistance to ethambutol",
  },
  {
    accession: "ARO:3003449",
    name: "Antibiotic resistant iniC",
    description:
      "iniC, an isoniazid-induced protein, is part of the iniBAC operon that participates in the development of tolerance to both isoniazid and ethambutol. May function through a MDR-pump like mechanism, although it does not appear to directly transport isoniazid from the cell",
  },
  {
    accession: "ARO:3003450",
    name: "Ethambutol resistant iniC",
    description:
      "Mutations that occurs on the iniC genes resulting in the resistance to ethambutol",
  },
  {
    accession: "ARO:3003451",
    name:
      "Mycobacterium tuberculosis iniC mutant conferring resistance to ethambutol",
    description:
      "Specific mutations that occurs on Mycobacterium tuberculosis iniC causing it to be ethambutol resistant.",
  },
  {
    accession: "ARO:3003452",
    name: "ethambutol resistant embA",
    description:
      "Mutations that occurs on the embA genes resulting in the resistance to ethambutol",
  },
  {
    accession: "ARO:3003453",
    name:
      "Mycobacterium tuberculosis embA mutant conferring resistance to ethambutol",
    description:
      "Specific mutations that occurs on Mycobacterium tuberculosis embA leading to the lowered affinity of ethambutol to embA",
  },
  {
    accession: "ARO:3003454",
    name: "ethambutol resistant embR",
    description:
      "Mutations that occurs on the embR genes resulting in the resistance to ethambutol",
  },
  {
    accession: "ARO:3003455",
    name:
      "Mycobacterium tuberculosis embR mutant conferring resistance to ethambutol",
    description:
      "Point mutation in the M.tuberculosis embR results in increased resistance to ethambutol",
  },
  {
    accession: "ARO:3003456",
    name: "antibiotic resistant ethA",
    description:
      "Monooxygenase able to convert a wide range of ketones to the corresponding esters or lactones via a Baeyer-Villiger oxidation reaction. It is responsible for the activation of several thiocarbamide-containing pro-drugs into cytotoxic species",
  },
  {
    accession: "ARO:3003457",
    name: "ethionamide resistant ethA",
    description:
      "Mutations that occurs on the ethA genes resulting in the inability to catalyzes the oxidation of ethionamide (ETH) to the corresponding sulfoxide (the active drug)",
  },
  {
    accession: "ARO:3003458",
    name:
      "Mycobacterium tuberculosis ethA with mutation conferring resistance to ethionamide",
    description:
      "Specific mutations that occurs on Mycobacterium tuberculosis ethA causing it to be ethionamide resistant",
  },
  {
    accession: "ARO:3003459",
    name:
      "Mycobacterium tuberculosis gyrB mutant conferring resistance to fluoroquinolone",
    description:
      "Point mutation in Mycobacterium tuberculosis gyrB resulting in fluoroquinolone resistance",
  },
  {
    accession: "ARO:3003460",
    name: "antibiotic resistant ndh",
    description:
      "ndh is a NADH oxidase. It participates in antibiotic resistance by diminishing NADH oxidation and consequently causes an increase in NADH concentration and depletion of NAD+. This alteration of the NADH/NAD+ ratio prevents the peroxidation reactions required for the activation of INH, as well as the displacement of the NADH-isonicotinic acyl complex from InhA enzyme binding site.",
  },
  {
    accession: "ARO:3003461",
    name:
      "Mycobacterium tuberculosis ndh with mutation conferring resistance to isoniazid",
    description:
      "Point mutations in the Mycobacterium tuberculosis ndh gene shown clinically to confer resistance to isoniazid.",
  },
  {
    accession: "ARO:3003462",
    name: "antibiotic resistant kasA",
    description:
      "kasA is a  ketoacyl acyl carrier protein synthase that catalyzes the condensation reaction of fatty acid synthesis by the addition to an acyl acceptor of two carbons from malonyl-ACP. It is involved in elongation of fatty acids intermediate in the biosynthetic pathway of mycolic acids.",
  },
  {
    accession: "ARO:3003463",
    name:
      "Mycobacterium tuberculosis kasA mutant conferring resistance to isoniazid",
    description:
      "Specific mutations on the Mycobacterium tuberculosis kasA gene resulting in lowered affinity of isoniazid, resulting in resistance",
  },
  {
    accession: "ARO:3003464",
    name: "rifamycin-resistant arabinosyltransferase",
    description:
      "Arabinosyl transferases allow for the polymerization of arabinose to form arabinan. Arabanan is required for formation of mycobacterial cell walls and arabinosyltransferases are targets of the drug ethambutol. Mutations in these genes can confer resistance to rifampicin.",
  },
  {
    accession: "ARO:3003465",
    name:
      "Mycobacterium tuberculosis embB with mutation conferring resistance to rifampicin",
    description:
      "Specific mutations that occurs on Mycobacterium tuberculosis embB causing it to be rifampicin resistant",
  },
  {
    accession: "ARO:3003466",
    name: "antibiotic resistant gidB",
    description:
      "GidB is a m7G methyltransferase specific for 16S rRNA. Mutations within the gidB gene causes changes in the structure or 16s rRNA, leading to resistance to aminoglycosides",
  },
  {
    accession: "ARO:3003470",
    name:
      "Mycobacterium tuberculosis gidB mutation conferring resistance to streptomycin",
    description:
      "Specific mutations that occurs on Mycobacterium tuberculosis gidB causing it to be streptomycin resistant",
  },
  {
    accession: "ARO:3003473",
    name: "antibiotic sensitive embR",
    description:
      "wildtype embR sensitive to antibiotics. It is an arabinosyltransferase involved in lipoarabinomannan synthesis.",
  },
  {
    accession: "ARO:3003474",
    name: "ethionamide",
    description:
      "ethionamide is a second-line antitubercular agent that inhibits mycolic acid synthesis",
  },
  {
    accession: "ARO:3003475",
    name: "antibiotic sensitive ethA",
    description:
      "ethA is a Monooxygenase that is able to convert a wide range of ketones to the corresponding esters or lactones via a Baeyer-Villiger oxidation reaction. Is responsible for the activation of several thiocarbamide-containing pro-drugs into cytotoxic species.",
  },
  {
    accession: "ARO:3003476",
    name: "antibiotic sensitive tlyA",
    description:
      "It Catalyzes the 2'-O-methylation at nucleotides C1409 in 16S rRNA and C1920 in 23S rRNA.",
  },
  {
    accession: "ARO:3003477",
    name: "antibiotic sensitive gidB",
    description: "GidB is a m7G methyltransferase specific for 16S rRNA.",
  },
  {
    accession: "ARO:3003478",
    name: "antibiotic sensitive kasA",
    description:
      "kasA is a ketoacyl acyl carrier protein synthase that catalyzes the condensation reaction of fatty acid synthesis by the addition to an acyl acceptor of two carbons from malonyl-ACP. It is involved in elongation of fatty acids intermediate in the biosynthetic pathway of mycolic acids.",
  },
  {
    accession: "ARO:3003479",
    name: "tetR",
    description:
      "TetR is the repressor of the tetracycline resistance element; its N-terminal region forms a helix-turn-helix structure and binds DNA. Binding of tetracycline to TetR reduces the repressor affinity for the tetracycline resistance gene (tetA) promoter operator sites. Mutations arise within tetR results in lower affinity for tetracyclin",
  },
  {
    accession: "ARO:3003480",
    name:
      "Mycobacterium tuberculosis 16S rRNA mutation conferring resistance to streptomycin",
    description:
      "Point mutations within the 530 loop of the 16S rRNA of Mycobacterium tuberculosis can result in resistance against streptomycin. These mutations do not directly inhibit the binding of streptomycin, but is hypothesized that they interfere with conformational perturbations that account for streptomycin's action on the ribosome.",
  },
  {
    accession: "ARO:3003481",
    name:
      "Mycobacterium tuberculosis 16S rRNA mutation conferring resistance to amikacin",
    description:
      "Point mutations in the 3' domain of 16S rRNA of Mycobacterium tuberculosis can confer resistance to amikacin.",
  },
  {
    accession: "ARO:3003485",
    name:
      "Chlamydophila psittaci 16S rRNA mutation conferring resistance to spectinomycin",
    description:
      "Point mutation in the 16S rRNA helix 34 region of Chlamydophila psittaci can confer resistance against spectinomycin",
  },
  {
    accession: "ARO:3003493",
    name:
      "Pasteurella multocida 16S rRNA mutation conferring resistance to spectinomycin",
    description:
      "Point mutations in the helix 34 region of 16S rRNA of Pasteurella multocida can confer resistance to spectinomycin",
  },
  {
    accession: "ARO:3003495",
    name:
      "Neisseria gonorrhoeae 16S rRNA mutation conferring resistance to spectinomycin",
    description:
      "Point mutations in the helix 34 region of 16S rRNA of Neisseria gonorrhoeae can confer resistance against spectinomycin",
  },
  {
    accession: "ARO:3003497",
    name:
      "Neisseria meningitidis 16S rRNA mutation conferring resistance to spectinomycin",
    description:
      "Point mutations in the 16S rRNA of Neisseria meningitidis can confer resistance to spectinomycin",
  },
  {
    accession: "ARO:3003499",
    name:
      "Propionibacterium acnes 16S rRNA mutation conferring resistance to tetracycline",
    description:
      "Tetracycline binds tightly to the helix 34 domain in 16S rRNA, where it interferes sterically with the binding of aminoacyl-tRNA to the ribosome A site to block protein synthesis. Mutations in the nucleotide sequence in this domain for Propionibacterium acnes can result in resistance against tetracycline.",
  },
  {
    accession: "ARO:3003502",
    name:
      "Borrelia burgdorferi 16S rRNA mutation conferring resistance to spectinomycin",
    description:
      "Point mutations in the 3' major domain of the 16S rRNA gene of Borrelia burgdorferi can confer resistance to spectinomycin",
  },
  {
    accession: "ARO:3003503",
    name:
      "Borrelia burgdorferi 16S rRNA mutation conferring resistance to kanamycin",
    description:
      "Point mutations in the 3' minor domain of the 16S rRNA gene of Borrelia burgdorferi can confer resistance to kanamycin",
  },
  {
    accession: "ARO:3003504",
    name:
      "Borrelia burgdorferi 16S rRNA mutation conferring resistance to gentamicin",
    description:
      "Point mutations in the 3' minor domain of the 16S rRNA gene of Borrelia burgdorferi can confer resistance to gentamicin",
  },
  {
    accession: "ARO:3003510",
    name:
      "Helicobacter pylori 16S rRNA mutation conferring resistance to tetracycline",
    description:
      "Tetracycline binds tightly to the helix 34 domain in 16S rRNA, where it interferes sterically with the binding of aminoacyl-tRNA to the ribosome A site to block protein synthesis. Mutations in the nucleotide sequence in this domain of Helicobacter pylori can result in resistance against tetracycline.",
  },
  {
    accession: "ARO:3003511",
    name:
      "Escherichia coli soxS with mutation conferring antibiotic resistance",
    description:
      "SoxS is a global regulator that up-regulates the expression of AcrAB efflux genes. It also reduces OmpF expression to decrease cell membrane permeability.",
  },
  {
    accession: "ARO:3003512",
    name:
      "Salmonella enterica 16S rRNA (rrsD) mutation conferring resistance to spectinomycin",
    description:
      "Point mutations in the helix 34 region of the rrsD 16S rRNA gene of Salmonella enterica serovar Typhimurium can confer resistance to spectinomycin",
  },
  {
    accession: "ARO:3003514",
    name:
      "Mycobacterium chelonae 16S rRNA mutation conferring resistance to amikacin",
    description:
      "Point mutations in the 16S rRNA of Mycobacterium chelonae can cause resistance to amikacin.",
  },
  {
    accession: "ARO:3003515",
    name:
      "Mycobacterium chelonae 16S rRNA mutation conferring resistance to kanamycin A",
    description:
      "Point mutations in the 16S rRNA of Mycobacterium chelonae can cause resistance to kanamycin A.",
  },
  {
    accession: "ARO:3003516",
    name:
      "Mycobacterium chelonae 16S rRNA mutation conferring resistance to tobramycin",
    description:
      "Point mutations in the 16S rRNA of Mycobacterium chelonae can cause resistance to tobramycin.",
  },
  {
    accession: "ARO:3003517",
    name:
      "Mycobacterium chelonae 16S rRNA mutation conferring resistance to gentamicin C",
    description:
      "Point mutations in the 16S rRNA of Mycobacterium chelonae can confer resistance to gentamicin C",
  },
  {
    accession: "ARO:3003518",
    name:
      "Mycobacterium chelonae 16S rRNA mutation conferring resistance to neomycin",
    description:
      "Point mutations in the 16S rRNA of Mycobacterium chelonae can cause resistance to neomycin.",
  },
  {
    accession: "ARO:3003520",
    name: "Antibiotic sensitive Borrelia burgdorferi 16S rRNA",
    description: "Sensitive form of 16s susceptable to antibiotics",
  },
  {
    accession: "ARO:3003521",
    name: "Antibiotic sensitive Chlamydophila psittaci 16S rRNA",
    description: "Sensitive form of 16s susceptable to antibiotics",
  },
  {
    accession: "ARO:3003522",
    name: "Antibiotic sensitive Helicobacter pylori 16S rRNA",
    description: "Sensitive form of 16s susceptable to antibiotics",
  },
  {
    accession: "ARO:3003523",
    name: "Antibiotic sensitive Mycobacterium chelonae 16S rRNA",
    description: "Sensitive form of 16s susceptable to antibiotics",
  },
  {
    accession: "ARO:3003524",
    name: "Antibiotic sensitive Mycobacterium smegmatis 16S rRNA",
    description: "Sensitive form of 16s susceptable to antibiotics",
  },
  {
    accession: "ARO:3003525",
    name: "Antibiotic sensitive Neisseria gonorrhoeae 16S rRNA",
    description: "Sensitive form of 16s susceptable to antibiotics",
  },
  {
    accession: "ARO:3003526",
    name: "Antibiotic sensitive Neisseria meningitidis 16S rRNA",
    description: "Sensitive form of 16s susceptable to antibiotics",
  },
  {
    accession: "ARO:3003527",
    name: "Antibiotic sensitive Pasteurella multocida 16S rRNA",
    description: "Sensitive form of 16s susceptable to antibiotics",
  },
  {
    accession: "ARO:3003528",
    name: "Antibiotic sensitive Propionibacterium acnes 16S rRNA",
    description: "Sensitive form of 16s susceptable to antibiotics",
  },
  {
    accession: "ARO:3003529",
    name:
      "Antibiotic sensitive Salmonella enterica serovar Typhimurium 16S rRNA",
    description: "Sensitive form of 16s susceptable to antibiotics",
  },
  {
    accession: "ARO:3003530",
    name: "Antibiotic sensitive Mycobacterium tuberculosis 16S rRNA",
    description: "Sensitive form of 16s susceptable to antibiotics",
  },
  {
    accession: "ARO:3003532",
    name: "determinant of resistance to polyamine antibiotics",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to polyamine antibiotics.",
  },
  {
    accession: "ARO:3003539",
    name:
      "Mycobacterium smegmatis 16S rRNA (rrsA) mutation conferring resistance to hygromycin B",
    description:
      "Point mutations in the highly conserved helix 44 of the 16S rrsA rRNA gene of Mycobacterium smegmatis can confer resistance to hygromycin B. Resistance against hygromycin B is the result of conformational alterations that distorts a strong hydrogen bond leading to a change in the local geometry of the hygromycin B binding site.",
  },
  {
    accession: "ARO:3003540",
    name:
      "Mycobacterium smegmatis 16S rRNA (rrsB) mutation conferring resistance to hygromycin B",
    description:
      "Point mutations in the highly conserved helix 44 of the 16S rrsB rRNA gene of Mycobacterium smegmatis can confer resistance to hygromycin B. Resistance against hygromycin B is the result of conformational alterations that distorts a strong hydrogen bond leading to a change in the local geometry of the hygromycin B binding site.",
  },
  {
    accession: "ARO:3003541",
    name:
      "Mycobacterium smegmatis 16S rRNA (rrsB) mutation conferring resistance to streptomycin",
    description:
      "Point mutations in the helix 44 region of the 16S rRNA rrsB gene of Mycobacterium smegmatis can confer resistance to streptomycin",
  },
  {
    accession: "ARO:3003542",
    name:
      "Mycobacterium smegmatis 16S rRNA (rrsB) mutation conferring resistance to kanamycin A",
    description:
      "Point mutations in the helix 44 region of the 16S rRNA rrsB gene of Mycobacterium smegmatis can confer resistance to kanamycin A",
  },
  {
    accession: "ARO:3003543",
    name:
      "Mycobacterium smegmatis 16S rRNA (rrsA) mutation conferring resistance to kanamycin A",
    description:
      "Point mutations in the helix 44 region of the 16S rRNA rrsA gene of Mycobacterium smegmastis can confer resistance to kanamycin A",
  },
  {
    accession: "ARO:3003544",
    name:
      "Mycobacterium smegmatis 16S rRNA (rrsA) mutation conferring resistance to neomycin",
    description:
      "Point mutations in the helix 44 region of the 16S rRNA rrsA gene of Mycobacterium smegmatis can confer resistance to neomycin",
  },
  {
    accession: "ARO:3003545",
    name:
      "Mycobacterium smegmatis 16S rRNA (rrsB) mutation conferring resistance to neomycin",
    description:
      "Point mutations in the helix 44 region of the 16S rRNA rrsB gene of Mycobacterium smegmatis can confer resistance to neomycin",
  },
  {
    accession: "ARO:3003546",
    name:
      "Mycobacterium smegmatis 16S rRNA (rrsA) mutation conferring resistance to viomycin",
    description:
      "Point mutations in the helix 44 region of the 16S rRNA rrsA gene of Mycobacterium smegmatis can confer resistance to viomycin.",
  },
  {
    accession: "ARO:3003547",
    name:
      "Mycobacterium smegmatis 16S rRNA (rrsB) mutation conferring resistance to viomycin",
    description:
      "Point mutations in the helix 44 region of the 16S rRNA rrsB gene of Mycobacterium smegmatis can confer resistance to viomycin.",
  },
  {
    accession: "ARO:3003548",
    name: "mdtN",
    description:
      "Multidrug resistance efflux pump. Could be involved in resistance to puromycin, acriflavine and tetraphenylarsonium chloride.",
  },
  {
    accession: "ARO:3003549",
    name: "mdtO",
    description:
      "Multidrug resistance efflux pump. Could be involved in resistance to puromycin, acriflavine and tetraphenylarsonium chloride",
  },
  {
    accession: "ARO:3003550",
    name: "mdtP",
    description:
      "Multidrug resistance efflux pump. Could be involved in resistance to puromycin, acriflavine and tetraphenylarsonium chloride",
  },
  {
    accession: "ARO:3003551",
    name: "emeA",
    description:
      "A multidrug efflux pump from Enterococcus faecalis. There exist efflux activity of several antimicrobial agents such as DAPI, Hoechst 33342 and acriflavine. Efflux of DAPI via EmeA was strongly inhibited by reserpine.",
  },
  {
    accession: "ARO:3003552",
    name: "fusB",
    description:
      "FusB encodes a 2-domain zinc-binding protein that binds the ribosomal translocase EF-G, causing it to dissociate from the ribosome. This action increases the ribosomal turnover rate and confers resistance to fusidic acid.",
  },
  {
    accession: "ARO:3003553",
    name: "CepS",
    description:
      "CEPS is a typical class C cephalosporinase found in Aeromonas sobria, first isolated from strain 163a.",
  },
  {
    accession: "ARO:3003554",
    name: "tcmA",
    description:
      "Major facilitator superfamily transporter. Resistance to tetracenomycin C by an active tetracenomycin C efflux system which is probably energized by transmembrane electrochemical gradients.",
  },
  {
    accession: "ARO:3003555",
    name: "SHW beta-lactamase",
    description:
      "This family of sublcass B1 beta-lactamases were discovered in species of the Shewanella genus.",
  },
  {
    accession: "ARO:3003556",
    name: "SLB-1",
    description:
      "This enzyme breaks the beta-lactam antibiotic ring open and deactivates the molecule's antibacterial properties.",
  },
  {
    accession: "ARO:3003557",
    name: "SFB-1",
    description:
      "This enzyme breaks the beta-lactam antibiotic ring open and deactivates the molecule's antibacterial properties.",
  },
  {
    accession: "ARO:3003558",
    name: "y56 beta-lactamase",
    description:
      "Class A beta-lactamase found in Yersinia enterocolitica biovar 1A",
  },
  {
    accession: "ARO:3003559",
    name: "cepA",
    description:
      "A beta-lactamase found in Bacteroides fragilis producing either low or high levels of the endogenous cephalosporinase activity",
  },
  {
    accession: "ARO:3003561",
    name: "Sed-1",
    description:
      "Sed-1 is a chromosomal-encoded class A beta-lactamase identified in Citrobacter sedlakii.",
  },
  {
    accession: "ARO:3003562",
    name: "blaF",
    description: "Class A beta-lactamase found in Mycobacterium fortuitum",
  },
  {
    accession: "ARO:3003563",
    name: "RCP-1",
    description:
      "RCP is a class A beta-lactamase found in Rhodopseudomonas capsulata.",
  },
  {
    accession: "ARO:3003564",
    name: "EXO-1",
    description: "Class A beta-lactamase found in Streptomyces albus G",
  },
  {
    accession: "ARO:3003565",
    name: "R39",
    description: "Class A beta-lactamase found in Actinomadura R39",
  },
  {
    accession: "ARO:3003567",
    name: "magainin A",
    description:
      "Magainins A are antimicrobial peptides produced by amphibians as part of their innate immune system.",
  },
  {
    accession: "ARO:3003568",
    name: "magainin B",
    description:
      "magainin B is is a synthetic analog of magainin 1, an antimicrobial peptide",
  },
  {
    accession: "ARO:3003569",
    name: "magainin G",
    description:
      "magainin G is is a synthetic analog of magainin 1, an antimicrobial peptide. It differs from magainin B by the addition of an beta-alanine at the N terminal",
  },
  {
    accession: "ARO:3003570",
    name: "magainin H",
    description:
      "magainin H is is a synthetic analog of magainin 1, an antimicrobial peptide. It differs from magainin B with the substitution of D-alanine instead of L-alanine.",
  },
  {
    accession: "ARO:3003571",
    name: "magainin 1",
    description:
      "Magainin1 is an antimicrobial peptides produced by amphibians as part of their innate immune system. It is derived from the Xenopus laevis.",
  },
  {
    accession: "ARO:3003572",
    name: "magainin 2",
    description:
      "Magainin 2 is an antimicrobial peptides produced from Xenopus laevis skin. It induce osmotic lysis of bacteria",
  },
  {
    accession: "ARO:3003573",
    name: "LpxA",
    description:
      "The LpxA gene is widely known to be involved in the biosynthesis of lipid A in Gram-negative bacteria and mutations to this gene may cause resistance to antimicrobial peptides that target the outer membrane",
  },
  {
    accession: "ARO:3003574",
    name: "LpxC",
    description:
      "The LpxC gene is widely known to be involved in the biosynthesis of lipid A in Gram-negative bacteria and mutations to this gene may cause resistance to antimicrobial peptides that target the outer membrane",
  },
  {
    accession: "ARO:3003575",
    name: "LpxD",
    description:
      "The LpxD gene is widely known to be involved in the biosynthesis of lipid A in Gram-negative bacteria and mutations to this gene may cause resistance to antimicrobial peptides that target the outer membrane",
  },
  {
    accession: "ARO:3003576",
    name: "eptA",
    description:
      "PmrC mediates the modification of Lipid A by the addition of 4-amino-4-deoxy-L-arabinose (L-Ara4N) and phosphoethanolamine, resulting in a less negative cell membrane and decreased binding of polymyxin B.",
  },
  {
    accession: "ARO:3003577",
    name: "ugd",
    description:
      "PmrE is required for the synthesis and transfer of 4-amino-4-deoxy-L-arabinose (Ara4N) to Lipid A, which allows gram-negative bacteria to resist the antimicrobial activity of cationic antimicrobial peptides and antibiotics such as polymyxin",
  },
  {
    accession: "ARO:3003578",
    name: "PmrF",
    description:
      "PmrF is required for the synthesis and transfer of 4-amino-4-deoxy-L-arabinose (Ara4N) to Lipid A, which allows gram-negative bacteria to resist the antimicrobial activity of cationic antimicrobial peptides and antibiotics such as polymyxin. pmrF corresponds to 1 locus in Pseudomonas aeruginosa PAO1 and 1 locus in Pseudomonas aeruginosa LESB58.",
  },
  {
    accession: "ARO:3003579",
    name: "basRS",
    description:
      "BasRS is a two-component regulatory system. BasR is phosphorylated by BasS at high[Fe3+] levels. The BasRS system regulates the expression of pmrE, pmrF and pmrC, leading to antibiotic resistance",
  },
  {
    accession: "ARO:3003580",
    name: "gene altering cell wall charge",
    description:
      "Genes involved in alteration of the cell wall overall charge, leading to antimicrobial resistance due to reduced binding",
  },
  {
    accession: "ARO:3003581",
    name: "Acinetobacter mutant Lpx gene conferring resistance to colistin",
    description:
      "These genes are involved in the biosynthesis of lipid A in Gram-negative bacteria and mutations to this gene may cause resistance to antimicrobial peptides that target the outer membrane. Mutation by absence or insertion of ISAba11 sequence is a known cause of resistance in Acinetobacter baumannii\u25bf.",
  },
  {
    accession: "ARO:3003582",
    name: "basR",
    description:
      "Response regulator for Lipid A modification genes; two-component system involved in polymyxin resistance that senses high extracellular Fe(2+)",
  },
  {
    accession: "ARO:3003583",
    name: "basS",
    description:
      "Histidine protein kinase sensor Lipid A modification gene; part of a two-component system involved in polymyxin resistance that senses high extracellular Fe(2+)",
  },
  {
    accession: "ARO:3003585",
    name: "Klebsiella mutant PhoP conferring antibiotic resistance to colistin",
    description:
      "A mutant phoP activates pmrHFIJKLM expression responsible for L-aminoarabinose synthesis and polymyxin resistance, by way of alteration of negative charge",
  },
  {
    accession: "ARO:3003588",
    name: "charge alteration conferring antibiotic resistance",
    description:
      "The loss or reduction of the net negative charge within the cell wall of gram negative bacteria is a mechanism of resistance for cationic antimicrobials that depend on the negative charge for binding to the surface",
  },
  {
    accession: "ARO:3003589",
    name: "NmcA",
    description:
      "NmcA is a class A serine beta-lactamase isolated from Enterobacter cloacae.",
  },
  {
    accession: "ARO:3003590",
    name: "SHV-190",
    description: "From the Lahey list of beta-lactamases, not yet released.",
  },
  {
    accession: "ARO:3003591",
    name: "SHV-191",
    description: "From the Lahey list of beta-lactamases, not yet released.",
  },
  {
    accession: "ARO:3003592",
    name: "TEM-223",
    description: "From Lahey's list of beta-lactamase, not yet released",
  },
  {
    accession: "ARO:3003593",
    name: "OXA-437",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003594",
    name: "OXA-438",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003595",
    name: "OXA-439",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003596",
    name: "OXA-440",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003597",
    name: "OXA-441",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003598",
    name: "OXA-442",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003599",
    name: "OXA-443",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003600",
    name: "OXA-444",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003601",
    name: "OXA-445",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003602",
    name: "OXA-446",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003603",
    name: "OXA-447",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003604",
    name: "OXA-448",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003605",
    name: "OXA-449",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003606",
    name: "OXA-450",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003607",
    name: "OXA-451",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003608",
    name: "OXA-452",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003609",
    name: "OXA-453",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003610",
    name: "OXA-454",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003611",
    name: "OXA-455",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003612",
    name: "OXA-456",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003613",
    name: "OXA-457",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003614",
    name: "OXA-458",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003615",
    name: "OXA-459",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003616",
    name: "OXA-460",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003617",
    name: "OXA-461",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003618",
    name: "OXA-462",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003619",
    name: "OXA-463",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003620",
    name: "OXA-464",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003621",
    name: "OXA-465",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003622",
    name: "OXA-466",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003623",
    name: "OXA-467",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003624",
    name: "OXA-468",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003625",
    name: "OXA-469",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003626",
    name: "OXA-470",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003627",
    name: "OXA-471",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003628",
    name: "OXA-472",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003629",
    name: "OXA-473",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003631",
    name: "OXA-474",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003632",
    name: "OXA-475",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003633",
    name: "OXA-476",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003634",
    name: "OXA-477",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003635",
    name: "OXA-478",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003636",
    name: "OXA-479",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003637",
    name: "OXA-480",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003638",
    name: "OXA-481",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003639",
    name: "OXA-482",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003640",
    name: "OXA-483",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003641",
    name: "OXA-484",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003642",
    name: "OXA-485",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003643",
    name: "OXA-486",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003644",
    name: "OXA-487",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003645",
    name: "OXA-488",
    description:
      "Assigned by Lahey's list of beta-lactamases, no accessions or other information available",
  },
  {
    accession: "ARO:3003646",
    name: "CTX-M-162",
    description:
      "Assigned by Lahey's list of beta-lactamase, no other information available",
  },
  {
    accession: "ARO:3003647",
    name: "CTX-M-163",
    description:
      "Assigned by Lahey's list of beta-lactamase, no other information available",
  },
  {
    accession: "ARO:3003648",
    name: "CTX-M-164",
    description:
      "Assigned by Lahey's list of beta-lactamase, no other information available",
  },
  {
    accession: "ARO:3003649",
    name: "CTX-M-165",
    description:
      "Assigned by Lahey's list of beta-lactamase, no other information available",
  },
  {
    accession: "ARO:3003650",
    name: "CTX-M-166",
    description:
      "Assigned by Lahey's list of beta-lactamase, no other information available",
  },
  {
    accession: "ARO:3003652",
    name: "CMY-132",
    description:
      "From Lahey's list of beta-lactamases, no additional information available",
  },
  {
    accession: "ARO:3003653",
    name: "CMY-133",
    description:
      "From Lahey's list of beta-lactamases, no additional information available",
  },
  {
    accession: "ARO:3003654",
    name: "CMY-134",
    description:
      "From Lahey's list of beta-lactamases, no additional information available",
  },
  {
    accession: "ARO:3003655",
    name: "ACT-38",
    description:
      "From Lahey's list of beta-lactamases, no additional information available",
  },
  {
    accession: "ARO:3003657",
    name: "IMP-49",
    description:
      "From Lahey's list of beta-lactamases, no additional information available",
  },
  {
    accession: "ARO:3003658",
    name: "IMP-50",
    description:
      "From Lahey's list of beta-lactamases, no additional information available",
  },
  {
    accession: "ARO:3003659",
    name: "IMP-51",
    description:
      "From Lahey's list of beta-lactamases, no additional information available",
  },
  {
    accession: "ARO:3003660",
    name: "VIM-44",
    description:
      "From Lahey's list of beta-lactamases, no additional information available",
  },
  {
    accession: "ARO:3003661",
    name: "VIM-45",
    description:
      "From Lahey's list of beta-lactamases, no additional information available",
  },
  {
    accession: "ARO:3003662",
    name: "VIM-46",
    description:
      "From Lahey's list of beta-lactamases, no additional information available",
  },
  {
    accession: "ARO:3003663",
    name: "NDM-15",
    description:
      "From Lahey's list of beta-lactamases, no additional information available",
  },
  {
    accession: "ARO:3003664",
    name: "NDM-16",
    description:
      "From Lahey's list of beta-lactamases, no additional information available",
  },
  {
    accession: "ARO:3003665",
    name: "NmcR",
    description:
      "NmcR is a homolog of the LysR regulator found in Enterobacter cloacae that contribute to the regulation of NmcA beta-lactamase",
  },
  {
    accession: "ARO:3003666",
    name:
      "16s rRNA with mutation conferring resistance to aminoglycoside antibiotics",
    description:
      "Point mutations in the 16S rRNA of bacteria can confer resistance to aminoglycosides.",
  },
  {
    accession: "ARO:3003667",
    name: "16s rRNA with mutation conferring resistance to peptide antibiotics",
    description:
      "Point mutations in the 16S rRNA of bacteria can confer resistance to peptide antibiotics.",
  },
  {
    accession: "ARO:3003668",
    name:
      "16s rRNA with mutation conferring resistance to polyamine antibiotics",
    description:
      "Point mutations in the 16S rRNA of bacteria can confer resistance to polyamine antibiotics.",
  },
  {
    accession: "ARO:3003669",
    name:
      "16S rRNA with mutation conferring resistance to tetracycline derivatives",
    description:
      "Point mutations in the bacterial 16S rRNA region shown clinically to confer resistance to tetracycline and tetracycline derivatives (polyketide antibiotics).",
  },
  {
    accession: "ARO:3003670",
    name: "PEDO-1",
    description:
      "PEDO-1 is a subclass B3 metallo-beta lactamase isolated from Pedobacter roseus exhibiting resistance to carbapenems.",
  },
  {
    accession: "ARO:3003675",
    name: "plazomicin",
    description:
      "Plazomicin is a neoglycoside, or next-generation, aminoglycoside, that has been identified as a potentially useful agent to combat drug-resistant bacteria, such as Acinetobacter baumannii and Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3003676",
    name: "AAC(6')-Ib'",
    description:
      "AAC(6')-Ib' is an aminoglycoside acetyltransferase encoded by plasmids, transposons, integrons in P. aeruginosa and P. fluorescens.",
  },
  {
    accession: "ARO:3003677",
    name: "AAC(6')-Iaj",
    description:
      "AAC(6')-Iaj is a functional acetyltransferase that modifies the amino groups at the 6' positions of aminoglycosides and contributes to aminoglycoside resistance of P. aeruginosa",
  },
  {
    accession: "ARO:3003678",
    name: "TriABC-OpmH",
    description:
      "TriABC-OpmH is a triclosan-specific efflux pump expressed in the Gram-negative Pseudomonas aeruginosa. TriABC is the only P. aeruginosa resistance nodulation cell division (RND) pump which contains two membrane fusion proteins, TriA and TriB, and both are required for efflux pump function. TriABC associated with OpmH assemble a functional triclosan efflux pump.",
  },
  {
    accession: "ARO:3003679",
    name: "TriA",
    description:
      "TriA is a membrane protein that is fused to TriB and both are required for the triclosan efflux pump function of TriABC-OpmH in P. aeruginosa.",
  },
  {
    accession: "ARO:3003680",
    name: "TriB",
    description:
      "TriB is a membrane protein that is fused to TriA and both are required for the triclosan efflux pump function of TriABC-OpmH in P. aeruginosa.",
  },
  {
    accession: "ARO:3003681",
    name: "TriC",
    description:
      "TriC is a resistance nodulation cell division (RND) transporter that is a part of TriABC-OpmH, a triclosan-specific efflux protein.",
  },
  {
    accession: "ARO:3003682",
    name: "OpmH",
    description:
      "OpmH is an outer membrane efflux protein required for triclosan-specific efflux pump function.",
  },
  {
    accession: "ARO:3003684",
    name:
      "Pseudomonas aeruginosa gyrA conferring resistance to fluoroquinolones",
    description:
      "Point mutation of Pseudomonas aeruginosa gyrA resulted in the lowered affinity between fluoroquinolones and gyrA. Thus, conferring resistance.",
  },
  {
    accession: "ARO:3003685",
    name:
      "Pseudomonas aeruginosa parE conferring resistance to fluoroquinolones",
    description:
      "Point mutation in Pseudomonas aeruginosa parE resulting in sensitivity to fluoroquinolones (ciprofloxacin). In combination with a gyrase mutation (gyrA or gyrB), it confers a high level of resistance to ciprofloxacin.",
  },
  {
    accession: "ARO:3003686",
    name:
      "Pseudomonas aeruginosa oprD with mutation conferring resistance to imipenem",
    description:
      "oprD is an outer membrane porin which facilitates the uptake of basic amino acids and imipenem in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3003687",
    name: "APH(3')-VI",
    description:
      "APH(3')-VI is an aminoglycoside phosphoryltransferase that acts on the 3-OH of target of aminoglycosides.",
  },
  {
    accession: "ARO:3003688",
    name: "PvrR",
    description:
      "PvrR is a response regulator that controls the conversion between antibiotic-resistant and antibiotic-susceptible forms of Pseudomonas aeruginosa biofilms through porin deletion/gene absence",
  },
  {
    accession: "ARO:3003689",
    name: "MCR-1",
    description:
      "MCR-1 is a plasmid-borne phosphoethanolamine transferase that interferes with binding of colistin to the cell membrane via addition of phosphoethanolamine to lipid A, resulting reduction in negative charge of the cell membrane.",
  },
  {
    accession: "ARO:3003690",
    name: "sitafloxacin",
    description:
      "Sitafloxacin is a fluoroquinolone active against multi-resistant Gram-positive and negative pathogens. Sitafloxacin shows inhibitory activity against DNA gyrase and topoisomerase IV, which blocks bacterial DNA replication, thereby causing double-stranded breaks in the bacterial chromosome.",
  },
  {
    accession: "ARO:3003691",
    name: "MexJK-OpmH",
    description:
      "MexJK-OpmH is a triclosan efflux protein expressed in the Gram-negative Pseudomonas aeruginosa. MexJ is the membrane fusion protein, MexK is the inner membrane resistance-nodulation-cell division (RND) transporter, and OpmH is the outer membrane efflux protein. MexJK is constitutively expressed in mexL mutants.",
  },
  {
    accession: "ARO:3003692",
    name: "MexJ",
    description:
      "mexJ is the membrane fusion protein of the MexJK multidrug efflux protein.",
  },
  {
    accession: "ARO:3003693",
    name: "MexK",
    description:
      "mexK is the inner membrane resistance-nodulation-cell division (RND) transporter in the MexJK multidrug efflux protein.",
  },
  {
    accession: "ARO:3003694",
    name: "MexJK-OprM",
    description:
      "MexJK-OprM is a multidrug efflux protein expressed in the Gram-negative Pseudomonas aeruginosa. MexJ is the membrane fusion protein, MexK is the inner membrane resistance-nodulation-cell division (RND) transporter, and OprM is the outer membrane factor protein.",
  },
  {
    accession: "ARO:3003696",
    name: "determinant of triclosan resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to the antibacterial biocide triclosan.",
  },
  {
    accession: "ARO:3003697",
    name: "MexPQ-OpmE",
    description:
      "MexPQ-OpmE multidrug efflux pump expressed in Pseudomonas aeruginosa. MexP is the membrane fusion protein; MexQ is the inner membrane transporter; and OpmE is the outer membrane channel.",
  },
  {
    accession: "ARO:3003698",
    name: "mexP",
    description:
      "MexP is the membrane fusion protein of the MexPQ-OpmE multidrug efflux complex",
  },
  {
    accession: "ARO:3003699",
    name: "mexQ",
    description:
      "MexQ is the inner membrane transporter of the multidrug efflux pump MexPQ-OpmE.",
  },
  {
    accession: "ARO:3003700",
    name: "opmE",
    description:
      "opmE is an outer membrane factor protein that is part of the multidrug efflux pump MexPQ-OpmE.",
  },
  {
    accession: "ARO:3003701",
    name: "rokitamycin",
    description:
      "Rokitamycin is a macrolide antibiotic. Synthesized from strains of Streptomyces kitasatoensis.",
  },
  {
    accession: "ARO:3003702",
    name:
      "Pseudomonas aeruginosa gyrA and parC conferring resistance to fluoroquinolone",
    description:
      "Point mutation in Pseudomonas aeruginosa parC resulting in fluoroquinolone resistance also requiring a gyrA mutation.",
  },
  {
    accession: "ARO:3003703",
    name: "MexMN-OprM",
    description:
      "MexMN-OprM is a multidrug efflux pump expressed in Pseudomonas aeruginosa. MexM is the membrane fusion protein; MexN is the inner membrane transporter; and OprM is the outer membrane channel.",
  },
  {
    accession: "ARO:3003704",
    name: "mexM",
    description:
      "mexM is the membrane fusion protein of the MexMN-OprM multidrug efflux complex.",
  },
  {
    accession: "ARO:3003705",
    name: "mexN",
    description:
      "MexN is the inner membrane transporter of the MexMN-OprM multidrug efflux complex.",
  },
  {
    accession: "ARO:3003706",
    name: "penem",
    description:
      "Penems are a class of unsaturated beta-lactam antibiotics with a broad spectrum of antibacterial activity and have a structure which renders them highly resistant to beta-lactamases. All penems are all synthetically made and act by inhibiting the synthesis of the peptidoglycan layer of bacterial cell walls. They are structurally similar to carbapenems, however, where carbapenems have a carbon, penems have a sulfur.",
  },
  {
    accession: "ARO:3003707",
    name: "faropenem",
    description:
      "Faropenem is an orally active beta-lactam antibiotic belonging to the penem group.",
  },
  {
    accession: "ARO:3003708",
    name: "panipenem",
    description:
      "Panipenem is a carbapenem antibacterial agent with a broad spectrum of in vitro activity covering a wide range of Gram-negative and Gram-positive aerobic and anaerobic bacterial.  It is used in combination with betamipron to inhibit panipenem uptake into the renal tubule and prevent nephrotoxicity.",
  },
  {
    accession: "ARO:3003709",
    name: "MexZ",
    description:
      "MexZ is a transcriptional regulator that downregulates the mexXY multidrug transporter operon, which confers to aminoglycoside resistance on Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3003710",
    name: "MexL",
    description:
      "MexL is a specific repressor of mexJK transcription and autoregulates its own expression.",
  },
  {
    accession: "ARO:3003712",
    name: "VEB-1b",
    description:
      "VEB-1b is a beta-lactamase that is found in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3003713",
    name: "VCC-1",
    description:
      "VCC-1 is a Class A carbapenemase isolated from Vibrio cholerae and exhibits resistance to penicillins, carbapenems and monobactam antibiotics.",
  },
  {
    accession: "ARO:3003714",
    name: "PEDO-2",
    description:
      "PEDO-2 is a subclass B3 metallo-beta lactamase isolated from Pedobacter borealis exhibiting carbapenem resistance.",
  },
  {
    accession: "ARO:3003715",
    name: "PEDO-3",
    description:
      "PEDO-3 is a class B1 metallo-beta lactamase isolated from Pedobacter kyungheensis exhibiting resistance to carbapenems.",
  },
  {
    accession: "ARO:3003716",
    name: "CPS-1",
    description:
      "CPS-1 is a subclass B3 metallo-beta lactamase isolated from Chyseobacterium piscium exhibiting carbapenem resistance.",
  },
  {
    accession: "ARO:3003717",
    name: "ESP-1",
    description:
      "ESP-1 is a subclass B3 metallo-beta lactamase isolated from Epilithonimonas tenax conferring resistance to carbapenems.",
  },
  {
    accession: "ARO:3003718",
    name: "MSI-1",
    description:
      "MSI-1 is a subclass B3 metallo-beta lactamase isolated from Massilia oculi conferring resistance to carbapenems.",
  },
  {
    accession: "ARO:3003719",
    name: "MSI-OXA",
    description:
      "MSI-OXA is an OXA-85-like class D beta lactamase isolated from Massilia oculi.",
  },
  {
    accession: "ARO:3003720",
    name: "SPG-1",
    description:
      "SPG-1 is a subclass B3 metallo-beta lactamase isolated from Sphingomonas sp.",
  },
  {
    accession: "ARO:3003722",
    name: "glycopeptide resistance gene cluster VanI",
    description:
      "This inducible cluster confers high resistance to vancomycin and intermediate resistance to teicoplanin. It achieves this through peptidoglycan restructuring with its VanA ligase homologue, VanI, causing the precursurs to end in D-Ala-D-Lac. The cluster uniquely contains murF, the final enzyme in the synthesis of UDP-N-acetylmuramoyl-pentapeptide, the precursor of murein. The VanI gene cluster has been found in chromosomal DNA. Gene orientation: vanSRIWK-murFvanX",
  },
  {
    accession: "ARO:3003723",
    name: "vanI",
    description:
      "VanI is a D-Ala-D-Lac ligase that reduces vancomycin binding affinity, helping to confer vancomycin resistance as part of the VanI resistance gene cluster in Desulfitobacterium spp. and Desulfosporosinus spp.",
  },
  {
    accession: "ARO:3003724",
    name: "vanWI",
    description: "VanWI is a vanW variant found in the vanI gene cluster",
  },
  {
    accession: "ARO:3003725",
    name: "vanXI",
    description:
      "VanXI is a VanX variant found in the VanI glycopeptide resistance gene cluster. It is a D-Ala-D-Ala dipeptidase.",
  },
  {
    accession: "ARO:3003726",
    name: "vanSI",
    description:
      "VanSI is a vanS variant found in the vanI gene cluster; it is a histidine kinase.",
  },
  {
    accession: "ARO:3003727",
    name: "vanKI",
    description:
      "VanKI is a peptidoglycan bridge formation protein also known as FemAB that is part of the VanI glycopeptide resistance gene cluster.",
  },
  {
    accession: "ARO:3003728",
    name: "vanRI",
    description:
      "VanRI is the regulatory transcriptional activator in the VanSR regulator within the VanI glycopeptide resistance gene cluster.",
  },
  {
    accession: "ARO:3003729",
    name:
      "Staphylococcus aureus ileS with mutation conferring resistance to mupirocin",
    description:
      "Point mutations to the isoleucyl-tRNA synthetase (ileS) in Staphylococcus aureus that confer resistance to mupirocin.",
  },
  {
    accession: "ARO:3003730",
    name: "Bifidobacterium ileS conferring resistance to mupirocin",
    description:
      "Bifidobacteria have an intrinsically resistant form of ileS (isoleucyl-tRNA synthetase) that confers resistance to mupirocin.",
  },
  {
    accession: "ARO:3003731",
    name: "fusD",
    description:
      "A specific fusidic acid resistance gene conferring intrinsic resistance in the bacteria Staphylococcus saprophyticus.",
  },
  {
    accession: "ARO:3003733",
    name: "fusC",
    description:
      "FusC is a fusidic acid resistance gene enabling ribosomal translocase EF-G dissociation from the ribosome that has been detected in Staphylococcus aureus and Staphylococcus intermedius. Its mechanism is believed to be similar to fusB due to its high level of sequence homology.",
  },
  {
    accession: "ARO:3003734",
    name: "antibiotic resistant fusA",
    description:
      "Antibiotic resistant fusA is caused by mutations to the elongation factor G (EF-G) and confers resistance to fusidic acid.",
  },
  {
    accession: "ARO:3003735",
    name:
      "Staphylococcus aureus fusA with mutation conferring resistance to fusidic acid",
    description:
      "The mutations to this gene are involved in altering the translation elongation factor G (EF-G) in association with the ribosome to prevent fusidic acid from binding EF-G and preventing translation.",
  },
  {
    accession: "ARO:3003736",
    name: "antibiotic resistant fusE",
    description:
      "Antibiotic resistant fusE is caused by mutations in a region of the rplF gene encoding riboprotein L6, and confers resistance to fusidic acid.",
  },
  {
    accession: "ARO:3003737",
    name:
      "Staphylococcus aureus fusE with mutation conferring resistance to fusidic acid",
    description:
      "The mutations to the rplF gene encoding riboprotein L6 have been shown to cause fusidic acid resistance, demonstrating a potential secondary site of action of the antibiotic that is blocked through these mutations. Several types of mutations have been identified that cause resistance, including SNPs, frameshift mutations and early stop codons.",
  },
  {
    accession: "ARO:3003741",
    name: "mphE",
    description:
      "mphE is a macrolide phosphotransferase and resistance gene identified on a plasmid, pRSB105",
  },
  {
    accession: "ARO:3003742",
    name: "mphG",
    description:
      "The mphG gene encodes a macrolide 2'-phosphotransferase found in Photobacterium damselae sharing sequence similarity to mphA in E. coli.",
  },
  {
    accession: "ARO:3003744",
    name: "vatF",
    description:
      "vatF is a streptogramin A acetyl transferase gene isolated from the chromosome of Yersinia enterocolitica.",
  },
  {
    accession: "ARO:3003745",
    name: "mefC",
    description:
      "mefC is a macrolide efflux gene isolated from a plasmid in Photobacterium damselae.",
  },
  {
    accession: "ARO:3003746",
    name: "optrA",
    description:
      "OptrA is a member of the ABC-F protein subfamily that confers resistance to oxazolidinones. The gene encoding the protein was originally isolated from a plasmid in Enterococcus faecalis and Enterococcus faecium.",
  },
  {
    accession: "ARO:3003747",
    name: "determinant of oxazolidinone resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to oxazolidinone (ie., linezolid) antibiotics.",
  },
  {
    accession: "ARO:3003748",
    name: "oleC",
    description:
      "oleC is an ABC transporter isolated from Streptomyces antibioticus and is involved in oleandomycin secretion.",
  },
  {
    accession: "ARO:3003749",
    name: "salA",
    description:
      "salA is an ABC-F subfamily protein gene isolated from the chromosome of Staphylococcus sciuri conferring resistance to lincosamides and streptogramins.",
  },
  {
    accession: "ARO:3003751",
    name:
      "Escherichia coli nfsA mutations conferring resistance to nitrofurantoin",
    description:
      "nfsA encodes the major oxygen-insesitive nitroreductase in E. coli. The first step of resistance to nitrofurazone is mutation of nfsA.",
  },
  {
    accession: "ARO:3003753",
    name: "determinant of pleuromutilin resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to pleuromutilin antibiotics.",
  },
  {
    accession: "ARO:3003754",
    name: "antibiotic resistant nfsA",
    description:
      "The nfsA-encoded nitroreductase is the major oxygen-insensitive nitroreductase present in E. coli. NfsA uses only NADPH and has broad electron acceptor specificity. Mutations in nfsA cause resistance to nitrofurazone and furazolidone. Resistance to nitrofurantoin via mutation of nfsA reduces the fitness of clinical isolates of E. coli.",
  },
  {
    accession: "ARO:3003755",
    name: "antibiotic resistant nfsB",
    description:
      "nsfB encodes a minor oxygen-insensitive nitroreductase. NfsB reduces a broad range of nitroaromatic compounds including the antibiotics nitrofurazone and nitrofurantoin. NfsB is a flavin mononucleotide (FMN)-containing protein and uses both NADH and NADPH as a source of reducing equivalents. Mutations in nfsB lead to increased resistance to nitrofurazone and furazolidone in an nfsA mutant background.",
  },
  {
    accession: "ARO:3003756",
    name:
      "Escherichia coli nfsB with mutation conferring resistance to nitrofurantoin",
    description:
      "nfsB encodes the minor oxygen-insesitive nitroreductase in E. coli. The first step of resistance to nitrofurazone is mutation of nfsA, while the increased resistance associated with second step mutants is a consequence of nfsB mutation.",
  },
  {
    accession: "ARO:3003760",
    name:
      "Enterococcus faecalis cls with mutation conferring resistance to daptomycin",
    description:
      "cls or cardiolipin synthase is an inner membrane protein involved in membrane synthesis and phosopholipid metabolism, with mutations to the gene being capable of conferring daptomycin resistance.",
  },
  {
    accession: "ARO:3003761",
    name: "eatAv",
    description:
      "eatAv is a mutated form of the wildtype eatA ABC-F subfamily protein isolated from Enterococcus faecium conferring resistance to lincosamides, streptogramin A's and pleuromutilins (LSaP phenotype).",
  },
  {
    accession: "ARO:3003762",
    name: "lnuE",
    description:
      "lnuE encodes a lincosamide resistance gene isolated from Streptococcus suis that was truncated by an ISEnfa5-cfr-ISEnfa5 segment insertion. It shares the closest sequence similarity to lnuA.",
  },
  {
    accession: "ARO:3003763",
    name: "determinant of nitrofuran resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to the antibiotic drug, nitrofurantoin",
  },
  {
    accession: "ARO:3003764",
    name: "resistance by absence",
    description:
      "Mechanism of antibiotic resistance conferred by deletion of gene (usually a porin)",
  },
  {
    accession: "ARO:3003767",
    name: "mphM",
    description:
      "mphM is a chromosomally-encoded macrolide phosphotransferases that inactivate 14-, 15- and 16-membered macrolides.",
  },
  {
    accession: "ARO:3003768",
    name: "gene conferring resistance via absence",
    description:
      "Deletion of gene or gene product results in resistance. For example, deletion of a porin gene blocks drug from entering the cell.",
  },
  {
    accession: "ARO:3003769",
    name: "Staphylococcus aureus mprF",
    description:
      "MprF is a integral membrane protein that modifies the negatively-charged phosphatidylglycerol on the membrane surface. A point mutation in the putative mprF synthase domain is associated with enhanced expression, increased synthesis and reduced daptomycin surface binding. This confers resistance to cationic peptides that disrupt the cell membrane, including defensins.",
  },
  {
    accession: "ARO:3003770",
    name: "Listeria monocytogenes mprF",
    description:
      "MprF is a integral membrane protein that modifies the negatively-charged phosphatidylglycerol on the membrane surface. This confers resistance to cationic peptides that disrupt the cell membrane, including defensins.",
  },
  {
    accession: "ARO:3003772",
    name: "Brucella suis mprF",
    description:
      "MprF is a integral membrane protein that modifies the negatively-charged phosphatidylglycerol on the membrane surface. This confers resistance to cationic peptides that disrupt the cell membrane, including defensins.",
  },
  {
    accession: "ARO:3003773",
    name: "Clostridium perfringens mprF",
    description:
      "MprF is a integral membrane protein that modifies the negatively-charged phosphatidylglycerol on the membrane surface. This confers resistance to cationic peptides that disrupt the cell membrane, including defensins.",
  },
  {
    accession: "ARO:3003774",
    name: "Streptococcus agalactiae mprF",
    description:
      "MprF is a integral membrane protein that modifies the negatively-charged phosphatidylglycerol on the membrane surface. This confers resistance to cationic peptides that disrupt the cell membrane, including defensins.",
  },
  {
    accession: "ARO:3003775",
    name:
      "Escherichia coli murA with mutation conferring resistance to fosfomycin",
    description:
      "murA or UDP-N-acetylglucosamine enolpyruvyl transferase catalyses the initial step in peptidoglycan biosynthesis and is inhibited by fosfomycin. Overexpression of murA through mutations such as Asp369Asn and Leu370Ile confers fosfomycin resistance. Extensive evidence has shown the significance of C115 mutations in conferring fosfomycin resistance since this residue represents a primary binding site for the antibiotic across many species.",
  },
  {
    accession: "ARO:3003776",
    name:
      "Staphylococcus aureus murA with mutation conferring resistance to fosfomycin",
    description:
      "murA or UDP-N-acetylglucosamine enolpyruvyl transferase catalyses the initial step in peptidoglycan biosynthesis and is inhibited by fosfomycin. Overexpression of murA through mutations confers fosfomycin resistance.",
  },
  {
    accession: "ARO:3003777",
    name:
      "Borrelia burgdorferi murA with mutation conferring resistance to fosfomycin",
    description:
      "murA or UDP-N-acetylglucosamine enolpyruvyl transferase catalyses the initial step in peptidoglycan biosynthesis and is inhibited by fosfomycin. Mutations to the murA enzyme confers resistance to the antibiotic.",
  },
  {
    accession: "ARO:3003778",
    name:
      "Mycobacterium smegmatis ndh with mutation conferring resistance to isoniazid",
    description:
      "Mutations in the Mycobacterium smegmatis ndh gene that results in increased resistance to isoniazid.",
  },
  {
    accession: "ARO:3003779",
    name:
      "Mycobacterium bovis ndh with mutation conferring resistance to isoniazid",
    description:
      "Mutations in the Mycobacterium bovis ndh gene that results in increased resistance to isoniazid.",
  },
  {
    accession: "ARO:3003780",
    name: "nybomycin",
    description:
      "A  heterocyclic antibiotic that targets mutant gyrA (type II topoisomerase) containing an S84L substitution, counteracting acquired quinolone resistance. It is effective against quinolone-resistant Gram-positive bacteria including S. aureus and E. faecalis. Due to its ability to counteract quinolone resistance by targeting the mutant form of the gyrA protein, it is classified as a reverse antibiotic (RA).",
  },
  {
    accession: "ARO:3003784",
    name:
      "Mycobacterium tuberculosis intrinsic murA conferring resistance to fosfomycin",
    description:
      "Mycobacterium tuberculosis murA confers intrinsic resistance to fosfomycin. The presence of an aspartic acid residue in place of the critical cysteine at position 117 that enables fosfomycin binding is believed to be responsible for this intrinsic resistance.",
  },
  {
    accession: "ARO:3003785",
    name:
      "Chlamydia trachomatis intrinsic murA conferring resistance to fosfomycin",
    description:
      "Chlamydia murA confers intrinsic resistance to fosfomycin. The presence of an aspartic acid residue in place of the critical cysteine at position 119 that enables fosfomycin binding is believed to be responsible for this intrinsic resistance.",
  },
  {
    accession: "ARO:3003786",
    name: "fluoroquinolone self resistant parC",
    description:
      "Inherent parC resistance to fluoroquinolone from an antibiotic producer. The presence of these genes confers self-resistance to the antibiotic it produces.",
  },
  {
    accession: "ARO:3003787",
    name: "aminocoumarin self resistant parY",
    description:
      "Inherent ParY resistant to aminocoumarin from an antibiotic producer. The presence of these genes confers self resistance to the antibiotic it produces",
  },
  {
    accession: "ARO:3003788",
    name:
      "Bacillus subtilis pgsA with mutation conferring resistance to daptomycin",
    description:
      "Point mutations that occur within the Bacillus subtilis pgsA gene resulting in resistance to daptomycin.",
  },
  {
    accession: "ARO:3003789",
    name: "Campylobacter jejuni gyrA conferring resistance to fluoroquinolones",
    description:
      "Campylobacter jejuni is a major bacterial infectious agent associated with gastroenteritis. Quinolone resistance is reportedly conferred by a single C-257-T nucleotide substitution in the gyrA gene.",
  },
  {
    accession: "ARO:3003790",
    name: "Enterococcus faecium liaF mutant conferring daptomycin resistance",
    description:
      "liaF is an accessory protein that acts as a negative regulator of liaRS signal transduction pathway. Mutations confer daptomycin resistance.",
  },
  {
    accession: "ARO:3003791",
    name: "Enterococcus faecalis liaS mutant conferring daptomycin resistance",
    description:
      "liaS is a histidine kinase found in the liaFSR signal transduction pathway. Mutations confer daptomycin resistance.",
  },
  {
    accession: "ARO:3003792",
    name: "Enterococcus faecalis liaR mutant conferring daptomycin resistance",
    description:
      "liaR is a response regulator found in the liaFSR signal transduction pathway. Mutations confer daptomycin resistance.",
  },
  {
    accession: "ARO:3003793",
    name: "SPM-1",
    description:
      "Plasmid-mediated SPM metallo-beta-lactamase conferring resistance to carbapenem. Originally isolated from Pseudomonas aeruginosa. Responsible for carbapenem-resistant Pseudomonas aeruginosa (CRPA) outbreaks in Brazil.",
  },
  {
    accession: "ARO:3003794",
    name:
      "Staphylococcus aureus walK with mutation conferring resistance to daptomycin",
    description:
      "walK is the histidine kinase sensor of a two-component regulatory system controlling peptidoglycan metabolism through regulation of the expression of most of the peptidoglycan hydrolase genes. Mutations in the gene have been found that confer daptomycin resistance.",
  },
  {
    accession: "ARO:3003795",
    name: "daptomycin resistant walK",
    description:
      "Mutations to the walK gene, part of a cell wall metabolism 2-component regulatory system, confers resistance to antibiotics, specifically daptomycin.",
  },
  {
    accession: "ARO:3003796",
    name: "Acinetobacter baumannii ampC beta-lactamase",
    description:
      "A class C beta-lactamase found in Acinetobacter baumannii that confers resistance to piperacillin and cefepime.",
  },
  {
    accession: "ARO:3003797",
    name:
      "Enterococcus faecalis YybT with mutation conferring daptomycin resistance",
    description:
      "YybT has phosphodiesterase activity towards cyclic dinucleotides using a c-di-GMP hydrolyzing phosphodiesterase domain. Mutations to the gene confer resistance to daptomycin.",
  },
  {
    accession: "ARO:3003798",
    name: "daptomycin resistant YybT",
    description: "Mutations to the YybT gene confers daptomycin resistance",
  },
  {
    accession: "ARO:3003800",
    name:
      "Enterococcus faecalis gdpD with mutation conferring daptomycin resistance",
    description:
      "gdpD is a glycerolphosphodiesterase whose mutations confer resistance to daptomycin.",
  },
  {
    accession: "ARO:3003801",
    name: "bcr-1",
    description:
      "Transmembrane protein which expels bicyclomycin from the cell, leading to bicyclomycin resistance. Identified in Pseudomonas aeruginosa strains responsible for outbreaks in Brazil, often appearing with blaSPM-1, another bicyclomycin resistance gene",
  },
  {
    accession: "ARO:3003802",
    name: "daptomycin resistant gdpD",
    description:
      "Mutations to the gdpD glycerolphosphodiesterase confer resistance to daptomycin.",
  },
  {
    accession: "ARO:3003803",
    name:
      "Staphylococcus aureus agrA with mutation conferring resistance to daptomycin",
    description:
      "agrA is an autoinducer in a peptide-quorum two-component regulatory system whose mutations confer daptomycin resistance.",
  },
  {
    accession: "ARO:3003804",
    name: "daptomycin resistant agrA",
    description:
      "Mutations to the regulatory gene agrA confer daptomycin resistance.",
  },
  {
    accession: "ARO:3003805",
    name:
      "Enterococcus faecalis gshF with mutation conferring daptomycin resistance",
    description:
      "gshF is a bifunctional glutamate-cysteine ligase/ glutathione synthetase that when mutated, confers daptomycin resistance.",
  },
  {
    accession: "ARO:3003806",
    name: "daptomycin resistant gshF",
    description:
      "Mutations to the glutathione synthetase gshF confers daptomycin resistance.",
  },
  {
    accession: "ARO:3003807",
    name:
      "Escherichia coli acrR with mutation conferring multidrug antibiotic resistance",
    description:
      "AcrR is a repressor of the AcrAB-TolC multidrug efflux complex. AcrR mutations result in high level antibiotic resistance. The mutations associated with this model are specific to E. coli.",
  },
  {
    accession: "ARO:3003808",
    name: "carO",
    description:
      "carO is a transmembrane beta-barrel involved in the influx of carbapenem antibiotics in Acinetobacter baumannii. Disruption of the carO gene by distinct insertion elements results in a loss of carO expression causing resistance to carbapenem antibiotics. Homologs of carO have been identified in genera Acinetobacter, Moraxella and Psychrobacter.",
  },
  {
    accession: "ARO:3003809",
    name: "Acinetobacter baumannii OprD conferring resistance to imipenem",
    description:
      "An outer member protein (OMP) found in Acinetobacter baumannii involved in the uptake of imipenem and basic amino acids. This porin is homologous to Pseudomonas aeruginosa OprD, which performs an identical function.",
  },
  {
    accession: "ARO:3003810",
    name: "kirromycin self resistant EF-Tu",
    description:
      "Natural producers of kirromycin and kirromycin-like antibiotics (i.e., kirrothrycin) possess self-resistance, which is classified here",
  },
  {
    accession: "ARO:3003811",
    name: "adeC",
    description:
      "AdeC is the outer membrane factor of the AdeABC multidrug efflux complex. It can be replaced by other outer membrane channels, and is not essential for antibiotic resistance.",
  },
  {
    accession: "ARO:3003813",
    name:
      "Enterococcus faecalis drmA with mutation conferring daptomycin resistance",
    description:
      "drmA is an uncharacterized 6-pass membrane protein, with mutations to the protein causing modest resistance to daptomycin.",
  },
  {
    accession: "ARO:3003814",
    name: "daptomycin resistant drmA",
    description: "Mutations to drmA confer daptomycin resistance.",
  },
  {
    accession: "ARO:3003815",
    name:
      "Enterococcus faecalis YvlB with mutation conferring daptomycin resistance",
    description:
      "YvlB is a putative target of the liaFSR signalling pathway, whose mutations confer daptomycin resistance.",
  },
  {
    accession: "ARO:3003816",
    name: "daptomycin resistant YvlB",
    description:
      "Mutations to the putative liaFSR target YvlB confer daptomycin resistance.",
  },
  {
    accession: "ARO:3003817",
    name:
      "Acinetobacter baumannii gyrA conferring resistance to fluoroquinolones",
    description:
      "Mutations in the A subunit of DNA gyrase reduce its affinity for fluoroquinolones, thereby conferring resistance.",
  },
  {
    accession: "ARO:3003818",
    name:
      "Acinetobacter baumannii parC conferring resistance to fluoroquinolone",
    description:
      "Mutations in Acinetobacter baumannii parC that result in resistance to fluoroquinolones.",
  },
  {
    accession: "ARO:3003820",
    name: "mgrB",
    description:
      "mgrB is a small transmembrane protein produced in the PhoPQ signalling system. It acts as a negative regulator in this system. Inactivation or down-regulation of mgrB confers colistin resistance by absence as shown in Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3003821",
    name: "BAL30072",
    description:
      "BAL30072 is a monocyclic beta-lactam antibiotic belonging to the sulfactams. BAL30072 was found to trigger the spheroplasting and lysis of Escherichia coli rather than the formation of extensive filaments.",
  },
  {
    accession: "ARO:3003822",
    name: "battacin",
    description:
      "Battacin is a cationic lipopeptide antibiotic produced from a Paenibacillus tianmuensis soil isolate. Battacin kills bacteria in vitro and has potent activity against Gram-negative bacteria, including multidrug-resistant and extremely drug-resistant clinical isolates. The bactericidal kinetics of battacin correlate with the depolarization of the cell membrane, suggesting that battacin kills bacteria by disrupting the cytoplasmic membrane.",
  },
  {
    accession: "ARO:3003823",
    name: "benzalkonium chloride",
    description:
      "Benzalkonium chloride is a type of cationic surfactant. It is an organic salt called a quaternary ammonium compound. It has three main categories of use: as a biocide, a cationic surfactant, and as a phase transfer agent.",
  },
  {
    accession: "ARO:3003827",
    name: "rhodamine",
    description:
      "Rhodamine is a flurone dye that is often used as a tracer due to determine the rate and direction of flow and transport. It permeates the cell membrane of gram negative organisms E. coli and P. aeruginosa.",
  },
  {
    accession: "ARO:3003830",
    name: "aminocoumarin resistant alaS",
    description:
      "alanyl-tRNA synthetase conferring resistance to novobiocin in Escherichia coli",
  },
  {
    accession: "ARO:3003831",
    name: "temocillin",
    description:
      "Temocillin is a beta-lactamase resistant carboxypenicillin. It is primarily used for the treatment of multiple drug resistant, Gram-negative bacteria, specifically Enterobacteriaceae.",
  },
  {
    accession: "ARO:3003832",
    name: "ticarcillin",
    description:
      "Ticarcillin is a carboxypenicillin used for the treatment of Gram-negative bacteria, particularly P. aeruginosa. Ticarcillin's antibiotic properties arise from its ability to prevent cross-linking of peptidoglycan during cell wall synthesis, when the bacteria try to divide, causing cell death.",
  },
  {
    accession: "ARO:3003835",
    name: "cdeA",
    description:
      "Clostridium difficile and Escherichia coli multidrug efflux transporter with antiporter function. Confers resistance to fluoroquinolones in E. coli and acriflavin in C.difficile.",
  },
  {
    accession: "ARO:3003836",
    name: "qacH",
    description:
      "qacH is a subunit of the qac multidrug efflux pump in Vibrio cholerae",
  },
  {
    accession: "ARO:3003838",
    name: "gadW",
    description:
      "GadW is an AraC-family regulator that promotes mdtEF expression to confer multidrug resistance. GadW inhibits GadX-dependent activation. GadW clearly represses gadX and, in situations where GadX is missing, activates gadA and gadBC.",
  },
  {
    accession: "ARO:3003840",
    name: "aminocoumarin resistant cysB",
    description:
      "Positive regulator of gene expression in the cysteine regulon. cysB mutants confer resistance to novobiocin, an aminocoumarin antibiotic, in Escherichia coli.",
  },
  {
    accession: "ARO:3003841",
    name: "kdpE",
    description:
      "kdpE is a transcriptional activator that is part of the two-component system KdpD/KdpE that is studied for its regulatory role in potassium transport and has been identified as an adaptive regulator involved in the virulence and intracellular survival of pathogenic bacteria. kdpE regulates a range of virulence loci through direct promoter binding.",
  },
  {
    accession: "ARO:3003842",
    name: "MUS-2",
    description:
      "MUS-2 is a chromosome-encoded beta-lactamase from Myroides odoratimimus.",
  },
  {
    accession: "ARO:3003843",
    name: "leuO",
    description:
      "leuO, a LysR family transcription factor, exists in a wide variety of bacteria of the family Enterobacteriaceae and is involved in the regulation of as yet unidentified genes affecting the stress response and pathogenesis expression. LeuO is also an activator of the MdtNOP efflux pump.",
  },
  {
    accession: "ARO:3003844",
    name: "mfd",
    description:
      "mfd dissociates transcription elongation complexes blocked at nonpairing lesions and mediates recruitment of DNA repair proteins. The Mfd (mutation frequency decline) protein, also known as transcription-repair coupling factor, is responsible for ATP-dependent removal of stalled RNA polymerase from DNA lesions by inducing forward movement of the RNA polymerase and subsequent recruitment of nucleotide excision repair machinery to the sites of the lesions.",
  },
  {
    accession: "ARO:3003846",
    name: "ADC beta-lactamase without carbapenemase activity",
    description:
      "ADC beta-lactamases, also known as AmpC beta-lactamases, are cephalosporinases with extended-spectrum resistance to cephalosporins but not to carbapenems. ADC beta-lactamases are found in Acinetobacter sp. and Oligella urethralis.",
  },
  {
    accession: "ARO:3003847",
    name: "ADC-1",
    description: "ADC-1 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003848",
    name: "ADC-2",
    description: "ADC-2 is a beta-lactamase found in Oligella urethralis.",
  },
  {
    accession: "ARO:3003849",
    name: "ADC-3",
    description: "ADC-3 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003850",
    name: "ADC-4",
    description: "ADC-4 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003851",
    name: "ADC-5",
    description: "ADC-5 is a beta-lactamase found in Acinetobacter pittii.",
  },
  {
    accession: "ARO:3003852",
    name: "ADC-6",
    description: "ADC-6 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003853",
    name: "ADC-7",
    description:
      "ADC-7 is a beta-lactamase that is found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003854",
    name: "ADC-8",
    description: "ADC-8 is a beta-lactamase found in Acinetobacter sp.",
  },
  {
    accession: "ARO:3003856",
    name: "ADC-12",
    description: "ADC-12 is a beta-lactamase found in Acinetobacter pittii.",
  },
  {
    accession: "ARO:3003857",
    name: "ADC-13",
    description: "ADC-13 is a beta-lactamase found in Acinetobacter pittii.",
  },
  {
    accession: "ARO:3003858",
    name: "ADC-14",
    description: "ADC-14 is a beta-lactamase found in Acinetobacter pittii.",
  },
  {
    accession: "ARO:3003859",
    name: "ADC-15",
    description: "ADC-15 is a beta-lactamase found in Acinetobacter pittii.",
  },
  {
    accession: "ARO:3003860",
    name: "ADC-16",
    description: "ADC-16 is a beta-lactamase found in Acinetobacter pittii.",
  },
  {
    accession: "ARO:3003861",
    name: "ADC-17",
    description: "ADC-17 is a beta-lactamase found in Acinetobacter pittii.",
  },
  {
    accession: "ARO:3003862",
    name: "ADC-18",
    description: "ADC-18 is a beta-lactamase found in Acinetobacter pittii.",
  },
  {
    accession: "ARO:3003863",
    name: "ADC-19",
    description: "ADC-19 is a beta-lactamase found in Acinetobacter pittii.",
  },
  {
    accession: "ARO:3003864",
    name: "ADC-20",
    description: "ADC-20 is a beta-lactamase found in Acinetobacter pittii.",
  },
  {
    accession: "ARO:3003865",
    name: "ADC-21",
    description: "ADC-21 is a beta-lactamase found in Acinetobacter pittii.",
  },
  {
    accession: "ARO:3003866",
    name: "ADC-22",
    description: "ADC-22 is a beta-lactamase found in Acinetobacter pittii.",
  },
  {
    accession: "ARO:3003867",
    name: "ADC-23",
    description: "ADC-23 is a beta-lactamase found in Acinetobacter pittii.",
  },
  {
    accession: "ARO:3003868",
    name: "ADC-25",
    description: "ADC-25 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003870",
    name: "ADC-31",
    description: "ADC-31 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003871",
    name: "ADC-39",
    description: "ADC-39 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003872",
    name: "ADC-41",
    description: "ADC-41 is a beta-lactamase found in Acinetobacter pittii.",
  },
  {
    accession: "ARO:3003873",
    name: "ADC-42",
    description: "ADC-42 is a beta-lactamase found in Acinetobacter pittii.",
  },
  {
    accession: "ARO:3003874",
    name: "ADC-43",
    description: "ADC-43 is a beta-lactamase found in Acinetobacter pittii.",
  },
  {
    accession: "ARO:3003875",
    name:
      "Enterococcus faecium EF-Tu mutants conferring resistance to elfamycin",
    description:
      "Sequence variants of Enterococcus faecium elongation factor Tu that confer resistance to elfamycin antibiotics.",
  },
  {
    accession: "ARO:3003876",
    name: "ADC-44",
    description: "ADC-44 is a beta-lactamase found in Acinetobacter pittii.",
  },
  {
    accession: "ARO:3003878",
    name: "ADC-56",
    description: "ADC-56 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003879",
    name: "ADC-61",
    description: "ADC-61 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003880",
    name: "ADC-74",
    description: "ADC-74 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003881",
    name: "ADC-75",
    description: "ADC-75 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003882",
    name: "ADC-76",
    description: "ADC-76 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003883",
    name: "ADC-77",
    description: "ADC-77 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003884",
    name: "ADC-78",
    description: "ADC-78 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003885",
    name: "ADC-79",
    description: "ADC-79 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003886",
    name: "ADC-80",
    description: "ADC-80 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003887",
    name: "ADC-81",
    description: "ADC-81 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003888",
    name: "ADC-82",
    description: "ADC-82 is a beta-lactamase found in Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3003889",
    name:
      "Escherichia coli GlpT with mutation conferring resistance to fosfomycin",
    description:
      "Point mutations to the active importer GlpT, which is involved with the uptake of many phosphorylated sugars, confer resistance to fosfomycin by reducing import of the drug into the bacteria.",
  },
  {
    accession: "ARO:3003890",
    name:
      "Escherichia coli UhpT with mutation conferring resistance to fosfomycin",
    description:
      "Mutations to the active importer UhpT, which is involved with the uptake of many phosphorylated sugars, confer resistance to fosfomycin by reducing import of the drug into the bacteria.",
  },
  {
    accession: "ARO:3003893",
    name:
      "Escherichia coli UhpA with mutation conferring resistance to fosfomycin",
    description:
      "uhpA is a positive activator of the fosfomycin importer uhpT, thus mutations to uhpA confer fosfomycin resistance by reducing uhpT expression. Both knockout and amino acid substitution mutations have been found that confer resistance, with the Protein Knockout model describing the large, knockout mutations causing loss of function of the gene, and the Protein Variant model describing the amino acid substitutions.",
  },
  {
    accession: "ARO:3003894",
    name: "Rm3",
    description:
      "Rm3 is a class B3 metallo-beta-lactamase isolated from a metagenomic soil sample in Yorkshire, United Kingdom, shown to confer resistance to clinically used penicillins, cephalosporins, and carbapenems.",
  },
  {
    accession: "ARO:3003895",
    name: "Pseudomonas mutant PhoP conferring resistance to colistin",
    description:
      "Mutations in Pseudomonas aeruginosa PhoP of the two-component PhoPQ regulatory system. Presence of mutation confers resistance to colistin.",
  },
  {
    accession: "ARO:3003896",
    name: "Pseudomonas mutant PhoQ conferring resistance to colistin",
    description:
      "Mutations in Pseudomonas aeruginosa PhoQ of the two-component PhoPQ regulatory system. Presence of mutation confers resistance to colistin",
  },
  {
    accession: "ARO:3003897",
    name: "tedizolid",
    description:
      "Tedizolid is a synthetic oxazolidinone antibiotic that is administered to the patient as a prodrug that is activated in vivo. It acts as an inhibitor of protein synthesis by binding the 50S subunit of the ribosome and preventing formation of the 70S initiation complex. It is active against Gram-positive pathogens.",
  },
  {
    accession: "ARO:3003898",
    name: "lugdunin",
    description:
      "Lugdunin is a thiazolidine-containing cyclic peptide antibiotic isolated from nasal Staphylococcus lugnunensis that prohibits Staphylococcus aureus colony formation. It is a human commensal bacteria, indicating the potential for further antibiotic discoveries from human microbiota.",
  },
  {
    accession: "ARO:3003899",
    name:
      "Escherichia coli PtsI with mutation conferring resistance to fosfomycin",
    description:
      "PtsI (phosphoenolpyruvate-protein phosphotransferase) is involved in cyclic AMP synthesis, which regulates glpT expression. As a result, mutations to ptsI confer resistance to fosfomycin by affecting the regulation of fosfomycin import.",
  },
  {
    accession: "ARO:3003900",
    name:
      "Escherichia coli CyaA with mutation conferring resistance to fosfomycin",
    description:
      "CyaA (adenylate cyclase) is involved with the synthesis of cyclic AMP which regulates the fosfomycin transporter glpT. As a result, mutations to cyaA confer resistance to fosfomycin.",
  },
  {
    accession: "ARO:3003901",
    name:
      "Staphylococcus aureus GlpT with mutation conferring resistance to fosfomycin",
    description:
      "Mutations to the active importer GlpT, which is involved with the uptake of many phosphorylated sugars, confer resistance to fosfomycin by reducing import of the drug into the bacteria.",
  },
  {
    accession: "ARO:3003902",
    name:
      "Staphylococcus aureus UhpT with mutation conferring resistance to fosfomycin",
    description:
      "Mutations to the active importer UhpT, which is involved with the uptake of many phosphorylated sugars, confer resistance to fosfomycin by reducing import of the drug into the bacteria.",
  },
  {
    accession: "ARO:3003903",
    name: "thusin",
    description:
      "Thusin is a two-component lantibiotic isolated from B. thuringiensis strain BGSC 4BT1, exhibiting antimicrobial activity against several Gram-positive pathogens. Bacillus cereus, Listeria monocytogenes, Staphylococcus aureus (MRSA), Staphylococcus sciuri, Enterococcus faecalis, and Streptococcus pneumoniae have all shown susceptibility to Thusin, with inhibition activities greater than Vancomycin or other lantibiotics.",
  },
  {
    accession: "ARO:3003904",
    name: "formicin",
    description:
      "Formicin is a two-component broad-spectrum lantibiotic produced by Bacillus paralicheniformis APC 1576, exhibiting antimicrobial activity against several Gram-positive species, including Staphylococcus aureus, Clostridium difficile and Listeria monocytogenes.",
  },
  {
    accession: "ARO:3003905",
    name: "ANT(4')-Ib",
    description:
      "Kanamycin nucleotidyltransferase sequence from Staphylococcus aureus plasmid. Confers resistance to kanamycin, neomycin and other aminoglycosides",
  },
  {
    accession: "ARO:3003907",
    name: "cipA",
    description:
      "Cfr-like methyltransferase enzyme conferring resistance to multiple clinically relevant antibiotic classes.",
  },
  {
    accession: "ARO:3003908",
    name: "Erm(47)",
    description:
      "Chromosome-encoded gene conferring MLSB resistance. Identified from Helcococcus kunzii.",
  },
  {
    accession: "ARO:3003914",
    name: "cycB",
    description:
      "Cytochrome c5 protein. Pseudomonas PA14 mutant cycB confers resistance to aminoglycoside Tobramycin. Resistance only detected in vitro.",
  },
  {
    accession: "ARO:3003915",
    name: "Lysocin E",
    description:
      "A drug which binds to menaquinone in the bacterial cell membrane. Binding of Lysocin to menaquinone facilitates lysis of the cell.",
  },
  {
    accession: "ARO:3003916",
    name: "menaquinone",
    description:
      "A molecule involved in electron transport in bacterial cell membrane.",
  },
  {
    accession: "ARO:3003917",
    name:
      "Staphylococcus aureus menA with mutation conferring resistance to lysocin",
    description:
      "menA encodes a 1,4-dihydroxy-2-naphthoate octaprenyltransferase, with mutations to the protein conferring resistance to lysocin E.",
  },
  {
    accession: "ARO:3003918",
    name: "apmA",
    description:
      "Plasmid-borne apramycin-resistant aminocyclitol acetyltransferase gene identified from bovine MRSA.",
  },
  {
    accession: "ARO:3003919",
    name: "lysocin",
    description:
      "Lysocin family of peptide antibiotics isolated from Lysobacter spp.",
  },
  {
    accession: "ARO:3003920",
    name: "pgpB",
    description: "A gene that produces the protein lipid A 4\u2019-phosphatase",
  },
  {
    accession: "ARO:3003921",
    name: "oqxAB",
    description:
      "oqxAB encodes for OqxAB, a plasmid-encoded efflux pump, which confer resistance to multiple agents including fluoroquinolones.",
  },
  {
    accession: "ARO:3003922",
    name: "oqxA",
    description: "RND efflux pump conferring resistance to fluoroquinolone",
  },
  {
    accession: "ARO:3003923",
    name: "oqxB",
    description: "RND efflux pump conferring resistance to fluoroquinolone",
  },
  {
    accession: "ARO:3003924",
    name:
      "Haemophilus parainfluenzae gyrA conferring resistance to fluoroquinolones",
    description:
      "Point mutation of Haemophilus parainfluenzae gyrA resulted in the lowered affinity between fluoroquinolones and GyrA. Thus, conferring resistance. Both Ser and Tyr are hydrophilic amino acids, however Tyr has an additional bulky hydrophobic group which could affect the interaction between the DNA gyrase with quinolones",
  },
  {
    accession: "ARO:3003925",
    name:
      "Haemophilus parainfluenzae parC conferring resistance to fluoroquinolones",
    description:
      "Point mutation of Haemophilus parainfluenzae parC resulted in the lowered affinity between fluoroquinolones and ParC. Thus, conferring resistance.",
  },
  {
    accession: "ARO:3003926",
    name: "Salmonella enterica gyrA conferring resistance to fluoroquinolones",
    description:
      "Point mutations in Salmonella gyrA that confer resistance to Nalidixic acid, a fluoroquinolone antibiotic. These mutations have also been shown to reduce susceptibility to ciprofloxacin.",
  },
  {
    accession: "ARO:3003927",
    name: "ceftolozane",
    description: "a 5th generation cephalosporin antibiotic",
  },
  {
    accession: "ARO:3003928",
    name:
      "Neisseria gonorrhoeae gyrA conferring resistance to fluoroquinolones",
    description:
      "Point mutation in Neisseria gonorrhoeae DNA gyrase subunit A. Decreases affinity between fluoroquinolone antibiotic molecule and gyrA, thereby conferring resistance to fluoroquinolone",
  },
  {
    accession: "ARO:3003929",
    name: "Neisseria gonorrhoeae parC conferring resistance to fluoroquinolone",
    description:
      "Point mutations in Neisseria gonorrhoeae parC protein that confer resistance to fluoroquinolone by reducing affinity to antibiotic binding site",
  },
  {
    accession: "ARO:3003930",
    name: "rpsJ",
    description:
      "rpsJ is a tetracycline resistance protein identified in Neisseria gonorrhoeae. Tetracycline resistance is conferred by binding to the ribosome as a 30S ribosomal protection protein.",
  },
  {
    accession: "ARO:3003931",
    name:
      "Capnocytophaga gingivalis gyrA conferring resistance to fluoroquinolones",
    description:
      "Point mutation in Capnocytophaga gingivalis that decreases binding affinity of fluoroquinolone antibiotics to gyrA, thereby conferring resistance.",
  },
  {
    accession: "ARO:3003932",
    name: "antibiotic efflux inhibitor",
    description:
      "Compounds and molecules that inhibit the action of antibiotic efflux proteins, thereby decreasing resistance caused by efflux pumps",
  },
  {
    accession: "ARO:3003933",
    name: "Berberine",
    description:
      "Berberine is a benzylisoquinoline alkaloid shown to reduce MexXY-dependent antibiotic resistance of cephalosporins, macrolides, lincosamides and aminoglycosides in Pseudomonas, Burkholderia and Achromobacter",
  },
  {
    accession: "ARO:3003936",
    name: "omadacycline",
    description:
      "a aminomethylcycline antibiotic that targets 16s rRNA in gram positive and gram negative bacteria",
  },
  {
    accession: "ARO:3003937",
    name: "Neisseria meningititis PBP2 conferring resistance to beta-lactam",
    description:
      "Point mutation in Neisseria meningititis PBP2 (penA) decreases affinity between beta-lactam antibiotic molecule and PBP2, thereby conferring resistance to beta-lactam.",
  },
  {
    accession: "ARO:3003938",
    name:
      "Penicillin-binding protein mutations conferring resistance to beta-lactam antibiotics",
    description:
      "Mutations in PBP transpeptidases that change the affinity for penicillin thereby conferring resistance to penicillin antibiotics",
  },
  {
    accession: "ARO:3003939",
    name: "Salmonella enterica parC conferring resistance to fluoroquinolones",
    description:
      "Point mutations in Salmonella parC gene implicated in decreased susceptibility to fluoroquinolone antibiotics, primarily ciprofloxacin and nalidixic acid.",
  },
  {
    accession: "ARO:3003940",
    name: "Shigella flexneri gyrA conferring resistance to fluoroquinolones",
    description:
      "Point mutations in Shigella flexneri gyrA observed to confer resistance to fluoroquinolone antibiotics.",
  },
  {
    accession: "ARO:3003941",
    name: "Shigella flexneri parC conferring resistance to fluoroquinolones",
    description:
      "Point mutation in parC conferring resistance to fluoroquinolone antibiotics in Shigella flexneri,",
  },
  {
    accession: "ARO:3003942",
    name: "abcA",
    description:
      "AbcA is a multidrug resistant ABC transporter that confers resistance to methicillin, daptomycin, cefotaxime, and moenomycin.",
  },
  {
    accession: "ARO:3003947",
    name: "EfrAB",
    description:
      "EfrAB is a heterodimeric ABC transporter efflux pump found in Enterococcus faecalis and Enterococcus faecium that confers resistance to ciprofloxacin, erythromycin, rifampicin, quinupristine.",
  },
  {
    accession: "ARO:3003948",
    name: "efrA",
    description:
      "efrA is a part of the EfrAB efflux pump, and both efrA and efrB are necessary to confer drug resistance.",
  },
  {
    accession: "ARO:3003949",
    name: "efrB",
    description:
      "efrB is a part of the EfrAB efflux pump, and both efrA and efrB are necessary to confer multidrug resistance.",
  },
  {
    accession: "ARO:3003950",
    name: "msbA",
    description:
      "MsbA is a multidrug resistance transporter homolog from E. coli and belongs to a superfamily of transporters that contain an adenosine triphosphate (ATP) binding cassette (ABC) which is also called a nucleotide-binding domain (NBD). MsbA is a member of the MDR-ABC transporter group by sequence homology. MsbA transports lipid A, a major component of the bacterial outer cell membrane, and is the only bacterial ABC transporter that is essential for cell viability.",
  },
  {
    accession: "ARO:3003951",
    name: "microcin J25",
    description:
      "microcin J25 is a peptide antibiotic that inhibits transcription by bacterial RNA polymerase. MccJ25 is produced by Escherichia coli strains that harbor a plasmid-borne antibiotic-synthesis and antibiotic-export cassette, consisting of a gene for MccJ25 precursor (a 58 residue linear peptide), two genes for factors that process MccJ25 precursor into MccJ25, and one gene for export of MccJ25.",
  },
  {
    accession: "ARO:3003952",
    name: "YojI",
    description:
      "YojI mediates resistance to the peptide antibiotic microcin J25 when it is expressed from a multicopy vector. YojI is capable of pumping out microcin molecules.  The outer membrane protein TolC in addition to YojI is required for export of microcin J25 out of the cell. Microcin J25 is thus the first known substrate for YojI.",
  },
  {
    accession: "ARO:3003953",
    name: "hmrM",
    description:
      "hmrM is a multidrug efflux pump belonging to the MATE family and functions as a  Na+/drug antiporter.",
  },
  {
    accession: "ARO:3003954",
    name: "efmA",
    description: "efmA is an MFS transporter permease in  E. faecium.",
  },
  {
    accession: "ARO:3003955",
    name: "efpA",
    description:
      "efpA is an MFS transporter found in Mycobacterium tuberculosis.",
  },
  {
    accession: "ARO:3003956",
    name: "antibacterial free fatty acids",
    description:
      "Amongst the diverse and potent biological activities of free fatty acids (FFAs) is the ability to kill or inhibit the growth of bacteria. The antibacterial properties of FFAs are used by many organisms to defend against parasitic or pathogenic bacteria. The prime target of FFA action is the cell membrane, where FFAs disrupt the electron transport chain and oxidative phosphorylation. Besides interfering with cellular energy production, FFA action may also result from the inhibition of enzyme activity, impairment of nutrient uptake, generation of peroxidation and auto-oxidation degradation products or direct lysis of bacterial cells.",
  },
  {
    accession: "ARO:3003957",
    name: "palmitic acid",
    description:
      "Palmitic acid is the most common saturated fatty acid found in animals, plants, and microorganisms. Palmitic acid is found to have antibacterial properties.",
  },
  {
    accession: "ARO:3003958",
    name: "oleic acid",
    description:
      "Oleic acid is a fatty acid that occurs naturally in various animal and vegetable fats and oils. Oleic acid is found to have antibacterial activity, particularly in inhibiting the growth of several Gram-positive bacterial species.",
  },
  {
    accession: "ARO:3003959",
    name: "linoleic acid",
    description:
      "Linoleic acid is a polyunsaturated omega-6 fatty acid. Linoleic acid has been found to have antibacterial activity, particularly in inhibiting the growth of Gram-positive bacterial species.",
  },
  {
    accession: "ARO:3003960",
    name: "farAB",
    description:
      "farAB is an MFS efflux pump found in gonococci that is involved in antibacterial fatty acid resistance. Moreover, expression of the tandemly linked farA and farB genes was positively associated with the presence of the MtrR transcriptional regulatory protein that normally downregulates the expression of mtrCDE.",
  },
  {
    accession: "ARO:3003961",
    name: "farA",
    description:
      "farA is the membrane fusion protein that is part of the farAB efflux pump.",
  },
  {
    accession: "ARO:3003962",
    name: "farB",
    description:
      "farB is the cytoplasmic transporter protein that is part of the farAB efflux pump. farB corresponds to 3 loci in Pseudomonas aeruginosa PAO1 and 3 loci in Pseudomonas aeruginosa LESB58.",
  },
  {
    accession: "ARO:3003963",
    name: "flo",
    description:
      "The Flo transporters can be plasmid- or chromosome-encoded. flo is an important determinant of florfenicol resistance in animal isolates of E. coli but is also found in human pathogens (e.g. Salmonella enterica and V. cholerae).",
  },
  {
    accession: "ARO:3003964",
    name: "hp1181",
    description:
      "hp1181 is a translocase that is part of the MFS efflux pump family. It is found in H. pylori and has role in the active efflux of antibiotics.",
  },
  {
    accession: "ARO:3003965",
    name: "hp1184",
    description:
      "hp1184 is a translocase that belongs to the MATE efflux pump family. It is found in H. pylori and is involved in the active efflux of antibiotics.",
  },
  {
    accession: "ARO:3003966",
    name: "Klebsiella pneumoniae OmpK35",
    description:
      "Klebsiella pneumoniae outer membrane porin protein. In beta-lactam-resistant Klebsiella, this porin is often deleted, limiting diffusion of the antibiotic into the cell. Klebsiella strains expressing OmpK35 are often beta-lactam sensitive even in the presence of beta-lactamases because of an inefficient mechanism.",
  },
  {
    accession: "ARO:3003967",
    name: "lrfA",
    description:
      "lfrA is involved in the active efflux of quinolones and is found in Mycobacterium smegmatis. lfrA corresponds to 2 loci in Pseudomonas aeruginosa PAO1 and 2 loci in Pseudomonas aeruginosa LESB58.",
  },
  {
    accession: "ARO:3003968",
    name: "Klebsiella pneumoniae OmpK36",
    description:
      "Klebsiella pneumoniae outer membrane porin protein. In beta-lactam-resistant Klebsiella, this porin is often deleted, limiting diffusion of the antibiotic into the cell. Klebsiella strains expressing OmpK36 are often beta-lactam sensitive even in the presence of beta-lactamases because of an inefficient mechanism.",
  },
  {
    accession: "ARO:3003969",
    name: "lmrP",
    description:
      "lmrP is a proton motive force-dependent drug transporter that is part of the MFS efflux pump family.",
  },
  {
    accession: "ARO:3003970",
    name: "D-Ala-D-Ala ligase",
    description:
      "Non-van ligases that synthesize D-Ala-D-Ala, the default cell wall precursor that makes a cell vulnerable to glycopeptide antibiotics. Mutations in the ddl gene can cause the production of nonfunctional/inactivated D-Ala-D-Ala ligases, which can render bacteria glycopeptide dependent depending on the presence of vancomycin resistance clusters.",
  },
  {
    accession: "ARO:3003971",
    name: "Erm(44)",
    description:
      "Variant of Erm(44) isolated from Staphylococcus saprophyticus, confers resistance to lincosamide and macrolide antibiotics but not streptogramins.",
  },
  {
    accession: "ARO:3003972",
    name: "Solithromycin",
    description:
      "A clinically developed Macrolide antibiotic targeting gram positive and gram negative bacteria.",
  },
  {
    accession: "ARO:3003974",
    name:
      "Propionibacterium acnes gyrA conferring resistance to fluoroquinolones",
    description:
      "Point mutations in P. acnes gyrA protein known to confer resistance to fluoroquinolone antibiotics.",
  },
  {
    accession: "ARO:3003975",
    name: "pactamycin",
    description:
      "Antibiotic produced by Streptomyces pactum, considered a universal translation inhibitor",
  },
  {
    accession: "ARO:3003976",
    name: "16S rRNA with mutation conferring resistance to pactamycin",
    description:
      "Point mutations in bacterial 16S rRNA that confer resistance to antibiotic pactamycin",
  },
  {
    accession: "ARO:3003977",
    name:
      "Halobacterium salinarum 16S rRNA mutation conferring resistance to pactamycin",
    description:
      "Point mutations in Halobacterium 16S rRNA that confer resistance to pactamycin antibiotic",
  },
  {
    accession: "ARO:3003978",
    name:
      "Chlamydomonas reinhardtii 16S rRNA (rrnS) mutation conferring resistance to streptomycin",
    description:
      "Point mutation in C. reinhardtii rrnS conferring resistance to streptomycin antibiotic.",
  },
  {
    accession: "ARO:3003979",
    name: "Eravacycline",
    description: "A fluorocycline that acts to inhibit the bacterial ribosome.",
  },
  {
    accession: "ARO:3003980",
    name: "tetA(48)",
    description:
      "TetA(48) is a Tetracycline efflux pump described in Paenibacillus sp. LC231, a strain of Paenibacillus isolated from Lechuguilla Cave, NM, USA. Described by Pawlowski et al. 2016.",
  },
  {
    accession: "ARO:3003981",
    name: "tetB(48)",
    description:
      "Tetracycline resistant TetB(48) efflux pump found in Paenibacillus sp. LC231, a strain of Paenibacillus isolated from Lechuguilla Cave, NM, USA. Described by Pawlowski et al. 2016.",
  },
  {
    accession: "ARO:3003982",
    name: "LlmA 23S ribosomal RNA methyltransferase",
    description:
      "Lincosamide (Clindamycin) resistant (putative) ribosomal methyltransferase related to the RlmK 23S rRNA methyltransferase COG family. Detected in Paenibacillus sp. LC231, an isolated strain of Paenibacillus found in Lechuguilla Cave, NM, USA. Described by Pawlowski et al. 2016.",
  },
  {
    accession: "ARO:3003983",
    name: "CatU",
    description:
      "Chloramphenicol acetyltransferase conferring resistance to phenicol antibiotics detected in Paenibacillus sp. LC231, an isolated strain of Paenibacillus from Lechuguilla Cave, NM, USA. Described by Pawlowski et al. 2016.",
  },
  {
    accession: "ARO:3003984",
    name: "BahA",
    description:
      "Bacitracin amidohydrolase found in Paenibacillus sp. LC231, an isolated strain of Paenibacillus from Lechuguilla Cave, NM, USA. Confers resistance by bacitracin inactivation through amidohydrolysis. Described by Pawlowski et al. 2016.",
  },
  {
    accession: "ARO:3003985",
    name: "amidohydrolysis of bacitracin undecaprenyl pyrophosphate",
    description:
      "Hydrolysis of amido side-chain of asparagine-12 forming hydrogen bond with undecaprenyl pyrophosphate in bacitracin leading to antibiotic inactivation.",
  },
  {
    accession: "ARO:3003986",
    name: "TaeA",
    description:
      "Pleuromutilin (Tiamulin) ABC efflux pump found in Paenibacillus sp. LC231, a strain of Paenibacillus isolated from Lechuguilla Cave, NM, USA. Confers resistance to pleuromutilin antibiotics. Described by Pawlowski et al. 2016.",
  },
  {
    accession: "ARO:3003987",
    name: "VatI",
    description:
      "Streptogramin A acetyltransferase found in Paenibacillus sp. LC231, isolated from Lechuguilla Cave, NM, USA. Confers resistance to streptogramin A antibiotics. Described by Pawlowski et al. 2016.",
  },
  {
    accession: "ARO:3003988",
    name: "AAC(2')-IIb",
    description:
      "AAC(2')-IIb is an intrinsic Kasugamycin 2' acetyltransferase protein found in Paenibacillus sp. LC231, a strain of Paenibacillus isolated from Lechuguilla Cave, NM, USA. Confers resistance to aminoglycoside antibiotics, including kasugamycin. Described by Pawlowski et al. 2016.",
  },
  {
    accession: "ARO:3003989",
    name: "AAC(6')-34",
    description:
      "AAC(6')-34 is an aminoglycoside 6'-phosphotransferase found in Paenibacillus sp. LC231, a strain of Paenibacillus from Lechuguilla Cave, NM, USA. Confers resistance to aminoglycoside antibiotics. Described by Pawlowski et al. 2016.",
  },
  {
    accession: "ARO:3003990",
    name: "vgbC",
    description:
      "VgbC is a streptogramin B-type lyase found in Paenibacillus sp. LC231, a strain of Paenibacillus isolated from Lechuguilla Cave, NM, USA. Confers resistance to streptogramin B-type antibiotics by linearization of the lactone ring on an ester bond, resulting in antibiotic inactivation. Described by Pawlowski et al. 2016.",
  },
  {
    accession: "ARO:3003991",
    name: "mphI",
    description:
      "mphI is a macrolide phosphotransferase protein found in Paenibacillus sp. LC231, a strain of Paenibacillus isolated from Lechuguilla Cave, NM, USA. Confers resistance to macrolide antibiotics. Described by Pawlowski et al. 2016.",
  },
  {
    accession: "ARO:3003992",
    name: "rphB",
    description:
      "rphB is a rifampin phosphotransferase protein found in Paenibacillus sp. LC231, a strain of Paenibacillus isolated from Lechuguilla Cave, NM, USA. Confers resistance to rifamycin antibiotics, specifically rifampin, through rifampin inactivation. Described by Pawlowski et al. 2016.",
  },
  {
    accession: "ARO:3003993",
    name: "capreomycin",
    description:
      "Capreomycin is an aminoglycoside antibiotic, capable of treating a large number of infections but in particular used for killing bacteria causing tuberculosis.",
  },
  {
    accession: "ARO:3003994",
    name: "cpaA",
    description:
      "cpaA is a capreomycin acetyltransferase protein found in Paenibacillus sp. LC231, an isolated strain of Paenibacillus in Lechuguilla Cave, NM, USA. Confers resistance to capreomycin, an aminoglycoside antibiotic. Described by Pawlowski et al. 2016.",
  },
  {
    accession: "ARO:3003995",
    name:
      "Clostridium difficile gyrA conferring resistance to fluoroquinolones",
    description:
      "Amino acid substitutions in Clostridium difficile gyrase subunit A which when present confer functional resistance to fluoroquinolone antibiotics.",
  },
  {
    accession: "ARO:3003997",
    name: "amoxicillin-clavulanic acid",
    description:
      "An antibiotic cocktail containing the beta-lactam antibiotic Amoxicillin and the beta-lactamase inhibitor Clavulanic Acid (potassium clavulanate).",
  },
  {
    accession: "ARO:3003998",
    name: "ampicillin-sulbactam",
    description:
      "An antibiotic mixture containing the penicillin-derived beta-lactam antibiotic Ampicillin and the beta-lactamase inhibitor Sulbactam. Also known as Unasyn and/or Ampictam.",
  },
  {
    accession: "ARO:3003999",
    name: "cefamandole",
    description:
      "Cefamandole is a second-generation cephalosporin-class beta-lactam antibiotic.",
  },
  {
    accession: "ARO:3004000",
    name: "cefetamet",
    description:
      "Cefetamet is a third-generation semi-synthetic beta-lactam antibiotic that inactivates inner cell wall PBPs through targeting binding, resulting in cell lysis.",
  },
  {
    accession: "ARO:3004001",
    name: "cefmetazole",
    description:
      "Cefmetazole is a semi-synthetic cephamycin antibiotic with broad spectrum antibiotic activity against both gram-positive and gram-negative bacteria, that disrupt cell wall synthesis through binding to PBPs causing cell lysis.",
  },
  {
    accession: "ARO:3004002",
    name: "cefonicid",
    description:
      "Cefonicid is a second-generation cephalosporin-class beta-lactam antibiotic with broad spectrum activity. Particularly used against urinary tract infections and lower respiratory infections. Causes cell lysis by inactivation of PBPs through binding, inhibiting peptidoglycan synthesis.",
  },
  {
    accession: "ARO:3004003",
    name: "cefoperazone",
    description:
      "Cefoperazone is a semi-synthetic cephalosporin containing a tetrazolyl moiety that is resistant to beta-lactamases. Particularly used for treatment of Pseudomonas spp. infections.",
  },
  {
    accession: "ARO:3004004",
    name: "cefotetan",
    description:
      "Cefotetan is a cephamycin-class beta-lactam antibiotic that is highly resistant to beta-lactamases and effective against a wide range of gram-negative and gram-positive bacteria.",
  },
  {
    accession: "ARO:3004005",
    name: "cefprozil",
    description:
      "Cefprozil is a cephalosporin and beta-lactam antibiotic with bactericidal activity. It selectively binds to PBPs and inhibits peptidoglycan synthesis, a major cell wall component, resulting in cell lysis.",
  },
  {
    accession: "ARO:3004006",
    name: "ceftiofur",
    description:
      "Ceftiofur is a third-generation broad spectrum cephalosporin and beta-lactam antibiotic. It causes cell lysis by disrupting peptidoglycan cross-linkage and cell wall formation by binding to PBPs.",
  },
  {
    accession: "ARO:3004007",
    name: "ceftizoxime",
    description:
      "Ceftizoxime is a third-generation cephalosporin and broad spectrum beta-lactam antibiotic. Ceftizoxime causes bacterial cell lysis through peptidoglycan cross-linking inhibition by binding to PBPs.",
  },
  {
    accession: "ARO:3004008",
    name: "cephapirin",
    description:
      "Cephapirin is a first-generation cephalosporin and broad spectrum beta-lactam antibiotic. Inactivation of penicillin-binding proteins through cephapirin binding disrupts peptidoglycan cross-linking, resulting in cell lysis.",
  },
  {
    accession: "ARO:3004009",
    name: "cefradine",
    description:
      "Cefradine is a first-generation cephalosporin and broad spectrum beta-lactam antibiotic. Cefradine binding to penicillin-binding proteins disrupts cell wall peptidoglycan cross-linkage, resulting in cell lysis.",
  },
  {
    accession: "ARO:3004010",
    name: "cinoxacin",
    description:
      "Cinoxacin is a fluoroquinolone antibiotic primarily used for the treatment of urinary tract infections in adults. Cinoxacin binds to DNA gyrase, resulting in double-stranded DNA breaks and cell death.",
  },
  {
    accession: "ARO:3004011",
    name: "clinafloxacin",
    description:
      "Clinafloxacin is a fluoroquinolone antibiotic and gyrase (DNA topoisomerase II) inhibitor. It binds to DNA gyrase and disrupts replication by causing double-stranded DNA breaks, resulting in cell death.",
  },
  {
    accession: "ARO:3004012",
    name: "Clofazimine",
    description:
      "Clofazimine is a fluoroquinolone-class phenazine dye used for the treatment of leprosy. Clofazimine binds to DNA and disrupts bacterial DNA gyrase, thereby causing double-stranded DNA breaks, and subsequent cell death.",
  },
  {
    accession: "ARO:3004013",
    name: "fleroxacin",
    description:
      "Fleroxacin is a broad spectrum fluoroquinolone antibiotic that inhibits the DNA supercoiling activity of bacterial DNA gyrase, resulting in double-stranded DNA breaks and subsequent cell death.",
  },
  {
    accession: "ARO:3004014",
    name: "flomoxef",
    description:
      "Flomoxef is a second-generation cephamycin (oxacephem) and beta-lactam antibiotic.",
  },
  {
    accession: "ARO:3004015",
    name: "gentamicin A",
    description:
      "Gentamicin A is part of a complex of broad spectrum aminoglycoside antibiotics. Gentamicin inhibits protein synthesis, resulting in bacterial cell death.",
  },
  {
    accession: "ARO:3004016",
    name: "loracarbef",
    description:
      "Loracarbef is a second-generation cephalosporin (carbacephem) and broad spectrum beta-lactam antibiotic. Loracarbef inhibits PBPs through binding, disrupting peptidoglycan cell wall cross-linkage and resulting in cell death.",
  },
  {
    accession: "ARO:3004017",
    name: "moxalactam",
    description:
      "Moxalactam (Latamoxef) is a broad spectrum cephalosporin (oxacephem) and beta-lactam antibiotic. Moxalactam binding to PBPs inhibits peptidoglycan cross-linkage in the cell wall, resulting in cell death. Moxalactam is proposed to be effective against meningitides as it passes the blood-brain barrier.",
  },
  {
    accession: "ARO:3004018",
    name: "nafcillin",
    description:
      "Nafcillin is a penicillin-class narrow-spectrum beta-lactam antibiotic. It inhibits cell wall synthesis and peptidoglycan cross-linkage through binding to PBPs, similarly to other penicillin-like antibiotics.",
  },
  {
    accession: "ARO:3004019",
    name: "para-aminosalicylic acid",
    description:
      "para-aminosalicylic acid (PAS) is an anti-tubercular antibiotic agent, often used in conjunction with Isoniazid for treatment of M. tuberculosis infections. PAS diminishes bacterial cell growth by limiting folic acid production.",
  },
  {
    accession: "ARO:3004020",
    name: "piperacillin-sulbactam",
    description:
      "An antibiotic cocktail containing the beta-lactam penam antibiotic Piperacillin and the beta-lactamase inhibitor sulbactam.",
  },
  {
    accession: "ARO:3004021",
    name: "piperacillin-tazobactam",
    description:
      "An antibiotic cocktail containing the penam beta-lactam antibiotic Piperacillin and the beta-lactamase inhibitor Tazobactam.",
  },
  {
    accession: "ARO:3004022",
    name: "quinupristin-dalfopristin",
    description:
      "An antibiotic cocktail of the streptogramin A dalfopristin and the streptogramin B quinupristin antibiotics. Used particularly to treat MRSA and vancomycin-resistant Enterococcus.",
  },
  {
    accession: "ARO:3004023",
    name: "ticarcillin-clavulanic acid",
    description:
      "An antibiotic cocktail containing the beta-lactam antibiotic ticarcillin and the beta-lactamase inhibitor clavulanic acid (clavulanate).",
  },
  {
    accession: "ARO:3004024",
    name: "trimethoprim-sulfamethoxazole",
    description:
      "An antibiotic cocktail containing the diaminopyrimidine antibiotic Trimethoprim and the sulfonamide antibiotic sulfamethoxazole (1 TMP:5 SMX).",
  },
  {
    accession: "ARO:3004025",
    name: "prothionamide",
    description:
      "Prothionamide is a thioamide derivative with antibacterial properties. It increases cell wall permeability and decreases cell wall damage resistance by inhibition of mycolic acid synthesis, resulting in cell death. It is particularly used to treat M. tuberculosis and M. leprae infections.",
  },
  {
    accession: "ARO:3004026",
    name: "tinidazole",
    description:
      "Tinidazole is a nitroimidazole anti-parasitic drug. It is used in the treatment of protozoan infections.",
  },
  {
    accession: "ARO:3004027",
    name: "determinant of resistance to nucleoside antibiotics",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to nucleoside antibiotics, including aminonucleosides.",
  },
  {
    accession: "ARO:3004028",
    name: "determinant of diaminopyrimidine resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to diaminopyrimidine (incl. trimethoprim) antibiotics.",
  },
  {
    accession: "ARO:3004030",
    name: "RosAB",
    description:
      "An efflux pump/potassium antiporter system (RosAB) formed by the RosA and RosB proteins found in Yersinia.",
  },
  {
    accession: "ARO:3004031",
    name: "tetAB(46)",
    description:
      "tetAB(46) is a heterodimeric ABC transporter conferring tetracycline resistance in Streptococcus australis isolated from the oral cavity. Both tetA(46) and tetB(46) genes are required for resistance.",
  },
  {
    accession: "ARO:3004032",
    name: "tetA(46)",
    description:
      "tetA(46) is a subunit of tetAB(46), a heterodimeric ABC transporter, that is required for conferring tetracycline resistance in Streptococcus australis isolated from the oral cavity.",
  },
  {
    accession: "ARO:3004033",
    name: "tetB(46)",
    description:
      "tetB(46) is a subunit of tetAB(46), a heterodimeric ABC transporter, that is required for conferring tetracycline resistance in Streptococcus australis isolated from the oral cavity.",
  },
  {
    accession: "ARO:3004034",
    name: "tetAB(60)",
    description:
      "tetAB(60) is an ABC transporter that confers resistance to tetracycline and tigercycline identified by screening a human saliva metagenomic library in Escherichia coli. Both tetA(60) and tetB(60) are required for resistance.",
  },
  {
    accession: "ARO:3004035",
    name: "tetA(60)",
    description:
      "tetA(60) is a subunit of tetAB(60), an ABC transporter that confers resistance to tetracycline and tigercycline identified by screening a human saliva metagenomic library in Escherichia coli, which is required for resistance.",
  },
  {
    accession: "ARO:3004036",
    name: "tetB(60)",
    description:
      "tetB(60) is a subunit of tetAB(60), an ABC transporter that confers resistance to tetracycline and tigercycline identified by screening a human saliva metagenomic library in Escherichia coli, which is required for resistance.",
  },
  {
    accession: "ARO:3004038",
    name: "Pseudomonas aeruginosa emrE",
    description:
      "EmrE is a small multidrug transporter that functions as a homodimer and that couples the efflux of small polyaromatic cations from the cell with the import of protons down an electrochemical gradient. Confers resistance to tetraphenylphosphonium, methyl viologen, gentamicin, kanamycin, and neomycin.",
  },
  {
    accession: "ARO:3004039",
    name: "Escherichia coli emrE",
    description:
      "Member of the small MDR (multidrug resistance) family of transporters; in Escherichia coli this protein provides resistance against a number of positively charged compounds including ethidium bromide and erythromycin; proton-dependent secondary transporter which exchanges protons for compound translocation",
  },
  {
    accession: "ARO:3004040",
    name: "rifampin phosphotransferase",
    description:
      "Enzymes, protein or other gene products that inactivate rifampin (rifamycin) antibiotics through phosphorylation of the antibiotic at the 21-OH position.",
  },
  {
    accession: "ARO:3004041",
    name: "Klebsiella pneumoniae acrA",
    description:
      "AcrA is a subunit of the AcrAB multidrug efflux system that in K. pneumoniae, which is encoded by the acrRAB operon.",
  },
  {
    accession: "ARO:3004042",
    name: "Enterobacter cloacae acrA",
    description:
      "AcrA is a subunit of the AcrAB-TolC multidrug efflux system that in E. cloacae.",
  },
  {
    accession: "ARO:3004043",
    name: "Escherichia coli acrA",
    description:
      "AcrA is a subunit of the AcrAB-TolC multidrug efflux system that in E. coli.",
  },
  {
    accession: "ARO:3004044",
    name: "antibiotic sensitive fabI",
    description:
      "fabI is a bacterial enoyl-acyl carrier reductase (ENR) isozyme. As with other ENRs, mutations in fabI can confer resistance to the biocide triclosan.",
  },
  {
    accession: "ARO:3004045",
    name:
      "Escherichia coli fabI mutations conferring resistance to isoniazid and triclosan",
    description:
      "fabI is a enoyl-acyl carrier reductase used in lipid metabolism and fatty acid biosynthesis. The bacterial biocide Triclosan blocks the final reduction step in fatty acid elongation, inhibiting biosynthesis. Point mutations in fabI can confer resistance to Triclosan and Isoniazid",
  },
  {
    accession: "ARO:3004046",
    name: "kdpDE",
    description:
      "kdpDE is a two-component regulatory system in Escherichia coli, well studied for its role in potassium transport and homeostasis. kdpE is also implicated in virulence loci regulation and overexpression of kdpE is shown to confer resistance to aminoglycoside antibiotics.",
  },
  {
    accession: "ARO:3004047",
    name: "kdpD",
    description:
      "kdpD is a component of the two-component regulatory system, kdpDE. This system is involved in potassium transport and homeostasis, and has been implicated in virulence factor regulation along with kdpE.",
  },
  {
    accession: "ARO:3004048",
    name:
      "MexAB-OprM with prematurely terminated MexR conferring resistance to meropenem and ciprofloxacin",
    description:
      "MexAB-OprM efflux pump system with a prematurely terminated MexR confers resistance to meropenem and ciprofloxacin.",
  },
  {
    accession: "ARO:3004049",
    name: "Escherichia coli fabG mutations conferring resistance to triclosan",
    description:
      "fabG is a 3-oxoacyl-acyl carrier protein reductase involved in lipid metabolism and fatty acid biosynthesis.The bacterial biocide Triclosan blocks the final reduction step in fatty acid elongation, inhibiting biosynthesis. Point mutations in fabG can confer resistance to Triclosan.",
  },
  {
    accession: "ARO:3004050",
    name: "antibiotic sensitive fabG",
    description:
      "fabG is a bacterial 3-oxoacyl enoyl-acyl carrier reductase (ENR) isozyme. As with other ENRs, mutations in fabG are shown to confer resistance to the biocide triclosan.",
  },
  {
    accession: "ARO:3004051",
    name: "MexAB-OprM with NalC mutations conferring resistance to aztreonam",
    description:
      "MexAB-OprM efflux pump system with NalC mutations conferring resistance to aztreonam. While efflux gene hyperexpression typically results from mutations in the linked mexR repressor gene, it also occurs independently of mexR mutations in so-called nalC mutants that demonstrate more modest mexAB-oprM expression and, thus, more modest multidrug resistance than do mexR strains.",
  },
  {
    accession: "ARO:3004052",
    name:
      "MexAB-OprM with NalD mutations conferring resistance to multiple antibiotics",
    description:
      "MexAB-OprM efflux pump system with a NalD mutation conferring resistance to multiple antibiotics. This NalD mutation was found in multidrug-resistant clinical strains lacking mutations in MexR or NalC.",
  },
  {
    accession: "ARO:3004053",
    name:
      "MexAB-OprM with CpxR regulator conferring resistance to ciprofloxacin, ceftazidime, and aztreonam",
    description:
      "CpxR is directly involved in activation of expression of RND efflux pump MexAB-OprM in P. aeruginosa. CpxR is required to enhance mexAB-oprM expression and drug resistance, in the absence of repressor MexR.",
  },
  {
    accession: "ARO:3004054",
    name: "Pseudomonas aeruginosa CpxR",
    description:
      "CpxR is directly involved in activation of expression of RND efflux pump MexAB-OprM in P. aeruginosa. CpxR is required to enhance mexAB-oprM expression and drug resistance, in the absence of repressor MexR.",
  },
  {
    accession: "ARO:3004055",
    name: "Escherichia coli CpxR",
    description:
      "The Escherichia coli CpxR/CpxA system responds to general periplasmic stress and has been used as a model two-component system for transcription regulation in bacteria. CpxR/CpxA system facilitates Salmonella and E. coli resistance to cationic antimicrobial peptides.The CpxR/CpxA system can activate transcription of marRAB operon, thus facilitating multidrug-resistant regulator to enhance expression of TolC-dependent tripartite multidrug transporters.",
  },
  {
    accession: "ARO:3004056",
    name: "ArmR",
    description:
      "ArmR, a 53-amino-acid antirepressor, allosterically inhibits MexR dimer-DNA binding by occupying a hydrophobic binding cavity within the center of the MexR dimer. ArmR up-regulation and MexR-ArmR complex formation have previously been shown to upregulate MexAB-OprM.",
  },
  {
    accession: "ARO:3004057",
    name:
      "23S rRNA with mutation conferring resistance to linezolid antibiotics",
    description:
      "Point mutations in the 23S rRNA subunit may confer resistance to linezolid and other oxazolidinone antibiotics.",
  },
  {
    accession: "ARO:3004058",
    name:
      "Staphylococcus aureus 23S rRNA with mutation conferring resistance to linezolid",
    description:
      "Point mutations in the 23S rRNA subunit of the large ribosomal bacterial subunit in Staphylococcus aureus, which confer resistance to linezolid by disrupting antibiotic target binding",
  },
  {
    accession: "ARO:3004059",
    name: "Type A NfxB",
    description:
      "Type A NfxB mutants are four to eight times more resistant to ofloxacin, erythromycin, and new zwitterionic cephems, i.e., cefpirome, cefclidin, cefozopran, and cefoselis, than the parent strain, PAO1. nfxB corresponds to 2 loci in Pseudomonas aeruginosa PAO1 (gene name: esrC/nfxB) and 2 loci in Pseudomonas aeruginosa LESB58 (gene name: nfxB).",
  },
  {
    accession: "ARO:3004060",
    name: "Type B NfxB",
    description:
      "Type B NfxB mutants are more resistant to tetracycline and chloramphenicol, as well as ofloxacin, erythromycin, and the new zwitterionic cephems, than was PAO1, and they are four to eight times more susceptible to carbenicillin, sulbenicillin, imipenem, panipenem, biapenem, moxalactam, aztreonam, gentamicin, and kanamycin than PAO1. The mutation at the 46th amino acid position is sufficient for overproduction of OprJ and the multidrug resistance. nfxB corresponds to 2 loci in Pseudomonas aeruginosa PAO1 (gene name: esrC/nfxB) and 2 loci in Pseudomonas aeruginosa LESB58 (gene name: nfxB).",
  },
  {
    accession: "ARO:3004061",
    name: "MexCD-OprJ with type A NfxB mutation",
    description:
      "MexCD\u2013OprJ with type A NfxB phenotype are four to eight times more resistant to ofloxacin, erythromycin, and new zwitterionic cephems, i.e., cefpirome, cefclidin, cefozopran, and cefoselis, than the parent strain, PAO1.",
  },
  {
    accession: "ARO:3004062",
    name: "MexCD-OprJ with type B NfxB mutation",
    description:
      "MexCD-OprJ with Type B NfxB mutions are more resistant to tetracycline and chloramphenicol, as well as ofloxacin, erythromycin, and the new zwitterionic cephems, than was PAO1, and they are four to eight times more susceptible to carbenicillin, sulbenicillin, imipenem, panipenem, biapenem, moxalactam, aztreonam, gentamicin, and kanamycin than PAO1.",
  },
  {
    accession: "ARO:3004063",
    name: "EdeQ",
    description:
      "EdeQ is an N-acetyltransferase enzyme that confers high-level self-resistance to edeine in Brevibacillus brevis, a natural edeine producer. EdeQ converts active edeine to N-acetyledeine, which is ineffective in vivo.",
  },
  {
    accession: "ARO:3004064",
    name: "Edeine acetyltransferase",
    description:
      "Edeine acetyltransferase enzymes catalyze the transfer of an acetyl group to active edeine, converting it to an inactive form in vivo. This mechanism is used for high-level self-resistance in edeine-producing Brevibacillus spp.",
  },
  {
    accession: "ARO:3004066",
    name:
      "MexEF-OprN with MexT mutation conferring resistance to chloramphenicol, ciprofloxacin, and trimethoprim",
    description:
      "The MexEF\u2013OprN efflux pump in P. aeruginosa is overexpressed with MexT mutation conferring resistance to chloramphenicol and ciprofloxacin",
  },
  {
    accession: "ARO:3004067",
    name: "MUS beta-lactamase",
    description:
      "Subclass B1 (metallo-) beta-lactamases found in Myroides spp., which confer resistance to carbapenam class beta-lactamase antibiotics.",
  },
  {
    accession: "ARO:3004068",
    name:
      "MexEF-OprN with MexS mutations conferring resistance to chloramphenicol, ciprofloxacin, and trimethoprim",
    description:
      "The MexEF\u2013OprN efflux pump with MexS mutations conferring resistance to chloramphenicol and ciprofloxacin. The model includes the MexT regulator, as MexS is suggested to inactivate MexT.",
  },
  {
    accession: "ARO:3004069",
    name: "MvaT",
    description:
      "MvaT, a global regulator of virulence genes in P. aeruginosa, has also shown to be able to repress the expression of the MexEF-OprN pump.",
  },
  {
    accession: "ARO:3004070",
    name:
      "MexEF-OprN with MvaT deletion conferring resistance to chloramphenicol and norfloxacin",
    description:
      "A deletion of MvaT results in the overexpression of MexEF-OprN conferring resistance to chloramphenicol and norfloxacin.",
  },
  {
    accession: "ARO:3004071",
    name: "determinant of ethionamide resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to the second-line anti-tubercular agent, ethionamide.",
  },
  {
    accession: "ARO:3004072",
    name: "OpmB",
    description:
      "OpmB is an outer membrane efflux protein in Pseudomonas aeruginosa that shows functional cooperation with MuxABC, to form the efflux pump system MuxABC-OpmB.",
  },
  {
    accession: "ARO:3004073",
    name: "MuxA",
    description:
      "MuxA is a membrane fusion protein component of the efflux pump system MuxABC-OpmB in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3004074",
    name: "MuxB",
    description:
      "MuxB is one of the two necessary RND components in the Pseudomonas aeruginosa efflux pump system MuxABC-OpmB.",
  },
  {
    accession: "ARO:3004075",
    name: "MuxC",
    description:
      "MuxC is one of the two necessary RND components of the MuxABC-OpmB efflux pumps system in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3004076",
    name: "MuxABC-OpmB",
    description:
      "MuxABC-OpmB is an RND-type multidrug efflux pump in Pseudomonas aeruginosa. This efflux pump confers resistance to aztreonam, novobiocin, tetracycline, erythromycin, kitasamycin and rokitamycin.",
  },
  {
    accession: "ARO:3004077",
    name: "PmpM",
    description:
      "PmpM is a multidrug efflux pump belonging to the MATE family of Pseudomonas aeruginosa. PmpM is an H+ drug antiporter and is the first reported case of an H+ coupled efflux pump in the MATE family. PmpM confers resistance to fluoroquinolones, fradiomycin, benzalkonium chloride, chlorhexidine gluconate, ethidium bromide, tetraphenylphosphonium chloride (TPPCl), and rhodamine 6G.",
  },
  {
    accession: "ARO:3004078",
    name:
      "Escherichia coli AcrAB-TolC with AcrR mutation conferring resistance to ciprofloxacin, tetracycline, and ceftazidime",
    description:
      "AcrR is the repressor of the AcrAB operon, where a mutation in AcrR allows AcrAB-TolC to confer resistance to ciprofloxacin, tetracycline, and ceftazidime.",
  },
  {
    accession: "ARO:3004079",
    name:
      "Escherichia coli AcrAB-TolC with MarR mutations conferring resistance to ciprofloxacin and tetracycline",
    description:
      "The Escherichia coli AcrAB-TolC with MarR mutation (Y137H) conferring resistance to ciprofloxacin and tetracycline.",
  },
  {
    accession: "ARO:3004080",
    name:
      "MexAB-OprM with MexR mutations confers resistance to multiple antibiotics",
    description:
      "MexAB-OprM efflux pump system with MexR mutations confers resistance to multiple antibiotics in P. aeruginosa.",
  },
  {
    accession: "ARO:3004081",
    name:
      "AcrAD-TolC confers resistance to amikacin, gentamicin, and tobramycin",
    description:
      "The AcrAD-TolC efflux pump system in E. coli confers resistance to amikacin, gentamicin, and tobramycin.",
  },
  {
    accession: "ARO:3004082",
    name: "AcrAD-TolC",
    description:
      "AcrAD-TolC efflux pump system in E. coli from the resistance-nodulation-division family, was shown to participate in the efflux of aminoglycosides.",
  },
  {
    accession: "ARO:3004083",
    name: "AcrEF-TolC confers resistance to ciprofloxacin",
    description:
      "AcrEF-TolC efflux pump system of E. coli confers resistance to fluoroquinolones (ciprofloxacin).",
  },
  {
    accession: "ARO:3004085",
    name: "lnuG",
    description:
      "lnuG is a transposon-mediated lincosamide nucleotidyltransferase found in Enterococcus faecalis on Tn6260.",
  },
  {
    accession: "ARO:3004086",
    name: "APH(3')-VIII",
    description:
      "APH(3')-VIII is an aminoglycoside phosphoryltransferase that acts on the 3-OH target of aminoglycosides.",
  },
  {
    accession: "ARO:3004087",
    name: "APH(3')-IX",
    description:
      "APH(3')-IX is an aminoglycoside phosphoryltransferase that acts on the 3-OH target of aminoglycosides.",
  },
  {
    accession: "ARO:3004088",
    name: "lefamulin",
    description:
      "Lefamulin is a semi-synthetic pleuromutilin compound highly active against multi-resistant pathogens.",
  },
  {
    accession: "ARO:3004089",
    name: "ANT(3'')-IIa",
    description:
      "ANT(3'')-IIa is a aminoglycoside nucleotidyltransferase identified in Acinetobacter spp. via horizontal gene transfer mechanisms.",
  },
  {
    accession: "ARO:3004090",
    name: "ANT(3'')-IIb",
    description:
      "ANT(3'')-IIb is a aminoglycoside nucleotidyltransferase identified in Acinetobacter spp. via horizontal gene transfer mechanisms.",
  },
  {
    accession: "ARO:3004091",
    name: "ANT(3'')-IIc",
    description:
      "ANT(3'')-IIc is a aminoglycoside nucleotidyltransferase identified in Acinetobacter spp. via horizontal gene transfer mechanisms.",
  },
  {
    accession: "ARO:3004092",
    name: "HMB-1",
    description:
      "Shotgun cloning of beta-lactam resistant P. aeruginosa NRZ-03096 yielded a clone producing a novel subclass B1 enzyme with only 74.3% identity to the next nearest relative, KHM-1. The novel MBL was named HMB-1 (for Hamburg MBL). HMB-1 gene was chromosomally located as part of a Tn 3 family transposon that was named Tn 6345, where expression of bla HMB-1 in E. coli TOP10 led to increased resistance to \u03b2-lactams.",
  },
  {
    accession: "ARO:3004093",
    name: "NDM-17",
    description:
      "NDM-17 is a metallo-beta-lactamase isolated from Escherichia coli with enhanced carbapenemase activity compared to NDM-1.",
  },
  {
    accession: "ARO:3004096",
    name: "daptomycin resistant CdsA",
    description:
      "Mutations to the CdsA phosphatidate cytidylyltransferase conferring resistance to daptomycin.",
  },
  {
    accession: "ARO:3004097",
    name:
      "Streptococcus mitis CdsA with mutation conferring daptomycin resistance",
    description:
      "CdsA is a phosphatidate cytidylyltransferase which plays a role in the production of membrane phosphatidylglycerol and cardiolipin.",
  },
  {
    accession: "ARO:3004098",
    name: "LpeAB",
    description:
      "LpeAB are components of an efflux pump, which is a macrolide resistance determinant in L. pneumophila Paris strain. Mutations observed in the upstream sequence of lpeAB operon in resistant lineages led to an overexpression of this efflux pump. Sub-inhibitory concentrations of macrolides themselves participated in upregulating this efflux and could constitute a first step in the acquisition of a high macrolide resistance level. As Lpp2879\u2013Lpp2880 is homologous to Escherichia coli AcrAB, lpp2879\u2013lpp2880 was renamed to lpeAB.",
  },
  {
    accession: "ARO:3004099",
    name: "LpeA",
    description:
      "LpeA is a subunit of the LpeAB efflux pump in Legionella pneumophila, which is homologous to AcrA in E. coli.",
  },
  {
    accession: "ARO:3004100",
    name: "LpeB",
    description:
      "LpeB is a subunit of the LpeAB efflux pump in Legionella pneumophila, which is homologous to AcrB in E. coli.",
  },
  {
    accession: "ARO:3004101",
    name: "MdtNOP",
    description:
      "MdtNOP is a MFS efflux pump protein found in E. coli. The deletion of mdtP from strain W3110 resulted in increased susceptibility to acriflavin, puromycin, and tetraphenylarsonium chloride. An E. coli mdtN null mutant is more sensitive to sulfur drugs than wild type",
  },
  {
    accession: "ARO:3004102",
    name: "kamB",
    description:
      "Methyltransferase enzyme first described in Streptoalloteichus tenebrarius. Confers resistance to aminoglycoside antibiotics (esp. apramycin) through methylation of the 16S rRNA at A1408, thereby modifying the antibiotic target.",
  },
  {
    accession: "ARO:3004103",
    name: "QepA2",
    description:
      "QepA2 is a plasmid-mediated quinolone resistance pump found in an Escherichia coli isolate from France",
  },
  {
    accession: "ARO:3004104",
    name: "TMB beta-lactamase",
    description:
      "TMB-1, the first known member of the Tripoli metallo-beta-lactamase family (TMB) was isolated from Achromobacter xylosoxidans in a Tripoli central hospital. TMB-1 was located on a class 1 integron and is a chromosomally-encoded beta-lactamase capable of hydrolyzing multiple antibiotics.",
  },
  {
    accession: "ARO:3004105",
    name: "TMB-1",
    description:
      "TMB-1 is a TMB metallo-beta-lactamase found in Achromobacter sp.",
  },
  {
    accession: "ARO:3004106",
    name: "TMB-2",
    description:
      "TMB-2 is a TMB metallo-beta-lactamase found in Acinetobacter sp.",
  },
  {
    accession: "ARO:3004107",
    name: "Pseudomonas aeruginosa soxR",
    description:
      "SoxR is a redox-sensitive transcriptional activator that induces expression of a small regulon that includes the RND efflux pump-encoding operon mexGHI-opmD. SoxR was shown to be activated by pyocyanin.",
  },
  {
    accession: "ARO:3004108",
    name: "Enterobacter cloacae rob",
    description:
      "rob is a positive regulator for the acrAB efflux genes, and is structurally similar to SoxS and MarA.",
  },
  {
    accession: "ARO:3004109",
    name: "Escherichia coli rob",
    description:
      "rob is a positive regulator for the acrAB efflux genes, and is structurally similar to SoxS and MarA.",
  },
  {
    accession: "ARO:3004110",
    name: "MCR-2",
    description:
      "MCR-2 is a plasmid-borne phosphoethanolamine transferase that interferes with binding of colistin to the cell membrane via addition of phosphoethanolamine to lipid A, resulting in reduction of the negative charge of the cell membrane. Originally identified in a Belgian sample of Escherichia coli by Xavier et al, (2016).",
  },
  {
    accession: "ARO:3004111",
    name: "FosA6",
    description:
      "fosA6 is a plasmid-encoded enzyme that confers resistance to fosfomycin in Escherichia coli by breaking the epoxide ring of the molecule.",
  },
  {
    accession: "ARO:3004112",
    name: "phosphoethanolamine transferase conferring colistin resistance",
    description:
      "This group of enzymes catalyzes the addition of a phosphoethanolamine group to another molecule. The addition of this moiety to lipid A in bacterial species is often associated with polymyxin (otherwise known as colistin) resistance.",
  },
  {
    accession: "ARO:3004113",
    name: "FosA7",
    description:
      "fosA7 is an enzyme that confers resistance to fosfomycin in Escherichia coli by breaking the epoxide ring of the molecule.",
  },
  {
    accession: "ARO:3004114",
    name: "porin OmpC",
    description:
      "In the presence of antibiotic stress, there is a coupled down-regulation of the porin OmpC with the OmpF.  Mutants both lacking both OmpC and OmpF proteins are resistant to cephaloridine and cefazolin. Analyses of genes involved in the increased resistance to tetracycline suggest that the up-regulation of efflux pump genes is accompanied by a decrease of OmpF and OmpC synthesis. Homologs of OmpC have been identified in Escherichia coli, Salmonella enterica, Enterobacter aerogenes and Serratia marcescens. \n",
  },
  {
    accession: "ARO:3004115",
    name: "nitroimidazole antibiotic",
    description:
      "Nitroimidazoles are a group of drugs that have both antiprotozoal and antibacterial activity, classified with respect to the location of the nitro functional group.",
  },
  {
    accession: "ARO:3004116",
    name: "nitrofuran antibiotic",
    description:
      "Nitrofurans are chemotherapeutic agents with antibacterial and antiprotozoal activity.",
  },
  {
    accession: "ARO:3004117",
    name: "Vibrio cholerae OmpU",
    description:
      "A ToxR-Regulated outer membrane porin. In Vibrio cholerae, ToxR controls resistance to P2 (a BPI-derived antimicrobial peptide) by regulating the production of OmpU. OmpU also confers resistance to polymyxin B sulfate.",
  },
  {
    accession: "ARO:3004118",
    name: "Vibrio cholerae OmpT",
    description:
      "A ToxR negatively regulated outer membrane porin. In Vibrio cholerae, ToxR controls resistance to P2 (a BPI-derived antimicrobial peptide) by regulating the production of OmpT, especially in the absence of OmpU.",
  },
  {
    accession: "ARO:3004120",
    name: "CBP C1-aminomethylene vancomycin",
    description:
      "CBP C1-aminomethylene vancomycin is a synthetic vancomycin derivative which exhibits multiple synergistic mechanisms of action, allowing it to effectively inhibit otherwise-glycopeptide resistant Gram-positive bacteria (notably those with functioning VanA clusters).",
  },
  {
    accession: "ARO:3004121",
    name: "BPI",
    description:
      "Bactericidal/permeability-increasing (BPI) protein is a member of a new generation of proteins known as super-antibiotics that are implicated as endotoxin neutralising agents. The potent (nM) cytotoxicity of BPI is limited to gram-negative bacteria (GNB), reflecting the high affinity (<10 nM) of BPI for bacterial lipopolysaccharides (LPS). Binding of BPI to live bacteria via LPS causes immediate growth arrest, actual killing coincides with later damage to the inner membrane.",
  },
  {
    accession: "ARO:3004122",
    name: "Klebsiella pneumoniae OmpK37",
    description:
      "Klebsiella pneumoniae outer membrane porin protein. Is preferentially detected in porin-deficient strains. Functional characterization of this new porin revealed a narrower pore than those of porins OmpK35 and OmpK36, which did not allow penetration by certain \u03b2-lactams. Also, when a resistant strain expresses porin OmpK37 is less susceptible to cefotaxime and cefoxitin than when it is expressing either OmpK36 or OmpK35.",
  },
  {
    accession: "ARO:3004123",
    name: "Burkholderia pseudomallei Omp38",
    description:
      "Heterologous expression of Burkholderia pseudomallei Omp38 (BpsOmp38) in Omp-deficient E. coli host cells lowers their permeability and in consequence, their antimicrobial susceptibility to penicillin G, cefoxitin, ceftazidime and imipenem.",
  },
  {
    accession: "ARO:3004124",
    name: "Moraxella catarrhalis M35",
    description:
      "The outer membrane protein M35 of Moraxella catarrhalis is an antigenically conserved porin. Down-regulation of M35 significantly increases the MICs of aminopenicillins, specifically amoxicillin.",
  },
  {
    accession: "ARO:3004125",
    name:
      "23S rRNA with mutation conferring resistance to macrolide antibiotics",
    description:
      "Nucleotide point mutations in the 23S rRNA subunit may confer resistance to macrolide antibiotics.",
  },
  {
    accession: "ARO:3004126",
    name: "Escherichia coli LamB",
    description:
      "LamB is a negative regulator for antibiotic resistance, it serves as a porin to influx antibiotic. When down-regulated, it increases resistance to chlortetracycline, ciprofloxacin, balofloxacin and nalidixic acid. It also interacts with Odp1, an energy metabolic enzyme, creating a complex that decreases in antibiotic-resistant strains.",
  },
  {
    accession: "ARO:3004127",
    name: "Escherichia coli mipA",
    description:
      "MltA-interacting protein (mipA), is an antibiotic resistance-related outer membrane protein. Deletion of mipA increases kanamycin, nalidixic acid and streptomycin resistance.",
  },
  {
    accession: "ARO:3004128",
    name: "Serratia marcescens Omp1",
    description:
      "Omp1 is an outer membrane porin that confers resistance by absence in S. marcescens.  Knockout, deletion or other inhibition of the omp1 gene confers resistance to certain beta-lactamase antibiotics - including Cefoxitin, Ceftriaxone, Cefotaxime, and Moxalactam - as well as Ciprofloxacin, Tetracycline, and Chloramphenicol, by preventing passage of the antibiotic into the cell.",
  },
  {
    accession: "ARO:3004129",
    name: "cephaloridine",
    description:
      "Cephaloridine is a semisynthetic, broad-spectrum, first-generation cephalosporin with antibacterial activity. Cephaloridine binds to and inactivates penicillin-binding proteins (PBPs) located on the inner membrane of the bacterial cell wall. PBPs are enzymes involved in the terminal stages of assembling the bacterial cell wall and in reshaping the cell wall during growth and division. Inactivation of PBPs interferes with the cross-linkage of peptidoglycan chains necessary for bacterial cell wall strength and rigidity. This results in the weakening of the bacterial cell wall and causes cell lysis.",
  },
  {
    accession: "ARO:3004130",
    name: "balofloxacin",
    description:
      "Balofloxacin is an 8-methoxy fluoroquinolone antibiotic. It shows potent bactericidal activity and inhibits the supercoiling activity of DNA gyrase of S. aureus, E. coli, and P. aeruginosa.",
  },
  {
    accession: "ARO:3004131",
    name:
      "Escherichia coli 23S rRNA with mutation conferring resistance to erythromycin and telithromycin",
    description:
      "Point mutation in the 23S rRNA of Escherichia coli shown clinically to confer resistance to the macrolide-class antibiotics erythromycin and telithromycin.",
  },
  {
    accession: "ARO:3004132",
    name:
      "Chlamydomonas reinhardtii 23S rRNA with mutation conferring resistance to erythromycin",
    description:
      "Point mutation in 23S rRNA of C. reinhardtii chloroplast shown clinically to confer resistance to Erythromycin, a macrolide antibiotic.",
  },
  {
    accession: "ARO:3004133",
    name:
      "Brachyspira hyodysenteriae 23S rRNA with mutation conferring resistance to tylosin",
    description:
      "Point mutation in B. hyodysenteriae 23S rRNA shown to confer resistance to tylosin and erythromycin",
  },
  {
    accession: "ARO:3004134",
    name:
      "Helicobacter pylori 23S rRNA with mutation conferring resistance to clarithromycin",
    description:
      "Point mutations in Helicobacter pylori shown to confer resistance to clarithromycin, a macrolide antibiotic",
  },
  {
    accession: "ARO:3004135",
    name:
      "Mycobacterium tuberculosis iniB with mutation conferring resistance to ethambutol",
    description:
      "Point mutations in M. tuberculosis iniB shown to confer resistance to ethambutol.",
  },
  {
    accession: "ARO:3004136",
    name: "Ethambutol resistant iniB",
    description:
      "Point mutations occurring in the iniB region of the iniBAC operon shown to confer resistance to ethambutol",
  },
  {
    accession: "ARO:3004137",
    name: "Antibiotic resistant iniB",
    description:
      "iniB, an isoniazid-induced protein, is part of the iniBAC operon that participates in the development of tolerance to both isoniazid and ethambutol. May function through a MDR-pump like mechanism, although it does not appear to directly transport isoniazid from the cell.",
  },
  {
    accession: "ARO:3004138",
    name:
      "Moraxella catarrhalis 23S rRNA with mutation conferring resistance to macrolide antibiotics",
    description:
      "Point mutations in the 23S ribosomal RNA subunit of M. catarrhalis shown clinically to confer resistance to macrolide class antibiotics.",
  },
  {
    accession: "ARO:3004139",
    name: "MCR-3",
    description:
      "MCR-3 is a plasmid-borne phosphoethanolamine transferase that interferes with binding of colistin to the cell membrane via addition of phosphoethanolamine to lipid A, resulting reduction in negative charge of the cell membrane. Originally described by Yin et al. 2017, from a porcine Escherichia coli plasmid pWJ1.",
  },
  {
    accession: "ARO:3004140",
    name: "hydrolysis of fusidic acid to inactive lactone derivative",
    description:
      "Enzymes shown to inactivate fusidic acid by hydrolytic cleavage from the 16 beta-position of fusidic acid and its derivatives. This converts fusidic acid to an inactivate lactone derivative, thus conferring resistance to fusidic acid.",
  },
  {
    accession: "ARO:3004141",
    name: "AxyXY-OprZ",
    description:
      "An RND-type efflux system that confers innate aminoglycoside resistance to Achromobacter spp. It has a transcriptional regulator encoded by the axyZ gene located upstream of axyXY-oprZ.  In-frame axyZ gene deletion assay led to increased MICs of antibiotic substrates of the efflux system: aminoglycosides, cefepime, fluoroquinolones, tetracyclines and erythromycin, indicating that the product of axyZ negatively regulates expression of axyXY-oprZ. Moreover a 45 amino-acid substitution at position 29 of AxyZ (V29G) was identified in a clinical Achromobacter strain that occurred during the course of chronic respiratory tract colonization in a cystic fibrosis (CF) patient. This substitution, also detected in 3 other strains exposed in vitro to tobramycin, led to the increase in axyY transcription level (5 to 17-fold) together with the increase in antibiotic resistance level.",
  },
  {
    accession: "ARO:3004142",
    name: "OprZ",
    description:
      "OprZ is the outer membrane component of the AxyXY-OprZ efflux pump system in Achromobacter spp.",
  },
  {
    accession: "ARO:3004143",
    name: "AxyX",
    description:
      "AxyX is the inner membrane transporter of the AxyXY-OprZ efflux pump system in Achromobacter spp.",
  },
  {
    accession: "ARO:3004144",
    name: "AxyY",
    description:
      "AxyY is the periplasmic adaptor protein of the AxyXY-OprZ efflux pump system in Achromobacter spp.",
  },
  {
    accession: "ARO:3004145",
    name: "AxyZ",
    description:
      "AxyZ is a transcriptional regulator of the AxyXY-OprZ efflux pump system.",
  },
  {
    accession: "ARO:3004146",
    name: "cfrC",
    description:
      "Acfr-like 23S rRNA methyltransferase shown to confer resistance to linezolid and phenicol antibiotics, including florfenicol and chloramphenicol, in Clostridium.",
  },
  {
    accession: "ARO:3004149",
    name:
      "Escherichia coli 23S rRNA with mutation conferring resistance to clindamycin",
    description:
      "Point mutation in the 23S rRNA of Escherichia coli shown clinically to confer resistance to clindamycin, a lincosamide antibiotic",
  },
  {
    accession: "ARO:3004150",
    name:
      "Escherichia coli 23S rRNA with mutation conferring resistance to chloramphenicol",
    description:
      "Point mutation in the 23S rRNA of Escherichia coli shown clinically to confer resistance to chloramphenicol",
  },
  {
    accession: "ARO:3004151",
    name: "determinant of aminosalicylate resistance",
    description:
      "Enzymes, other proteins or other gene products shown clinically to confer resistance to aminosalicylates, esp. para-aminosalicylic acid, commonly used to treat Mycobacterium infections.",
  },
  {
    accession: "ARO:3004152",
    name: "aminosalicylate resistant thymidylate synthase",
    description:
      "Antibiotic resistant form of thymidylate synthase (synthetase), an enzyme that catalyzes the conversion of dUMP to dTMP in nucleotide biosynthesis. Loss-of-function mutations in thymidylate synthase confer resistance to p-aminosalicylic acid by disrupting the substrate-binding affinity and catalytic activity.",
  },
  {
    accession: "ARO:3004153",
    name:
      "Mycobacterium tuberculosis thyA with mutation conferring resistance to para-aminosalicylic acid",
    description:
      "Point mutations in the thymidylate synthetase thyA gene shown clinically to confer resistance to para-aminosalicylic acid. Loss-of-function mutations in thyA disrupt the catalytic activity and substrate-binding affinity, thus conferring resistance.",
  },
  {
    accession: "ARO:3004154",
    name: "antibiotic sensitive thymidylate synthase",
    description:
      "Thymidylate synthase (synthetase) catalyzes the conversion of dUMP to dTMP in the nucleic acid biosynthesis pathway. Inhibition of thyA imbalances the cellular concentration of dUMP causing DNA damage.",
  },
  {
    accession: "ARO:3004155",
    name: "aminosalicylate resistant dihydrofolate synthase",
    description:
      "Dihydrofolate synthase (synthetase) enzymes resistant to aminosalicylates (inc. para-aminosalicylic acid) caused by mutation. Dihydrofolate synthase is required for bioactivation of p-aminosalicylic acid, and mutation in dihydrofolate synthase inhibits production of the dihydrofolate analog hydroxyl-dihydrofolate, thus preventing activation and conferring resistance.",
  },
  {
    accession: "ARO:3004156",
    name: "antibiotic sensitive dihydrofolate synthase",
    description:
      "Wild-type form of dihydrofolate synthase (synthetase) enzymes. Bioactivation of p-aminosalicylic acid occurs during the folate pathway when dihydrofolate and its analog hydroxyl-dihydrofolate are produced by dihydrofolate synthase. Disruption of this pathway by PAS results in cell damage and death.",
  },
  {
    accession: "ARO:3004157",
    name:
      "Mycobacterium tuberculosis folC with mutation conferring resistance to para-aminosalicylic acid",
    description:
      "Point mutations in the dihydrofolate synthetase folC gene shown clinically to confer resistance to p-aminosalicylic acid or other aminosalicylates. Mutations in folC inhibit bioactivation of PAS and thus confer resistance.",
  },
  {
    accession: "ARO:3004159",
    name:
      "Mycoplasma gallisepticum 23S rRNA mutation conferring resistance to pleuromutilin antibiotics",
    description:
      "Point mutations in the 23S rRNA (domain V) of Mycoplasma gallisepticum can confer resistance to pleuromutilin antibiotics (tiamulin and valnemulin). The mutants also show cross-resistance to lincomycin, chloramphenicol and florfenicol.",
  },
  {
    accession: "ARO:3004160",
    name:
      "Escherichia coli 23S rRNA with mutation conferring resistance to clarithromycin",
    description:
      "Point mutation in the 23S rRNA of Escherichia coli shown clinically to confer resistance to clarithromycin, a macrolide antibiotic",
  },
  {
    accession: "ARO:3004161",
    name:
      "Propionibacteria 23S rRNA with mutation conferring resistance to macrolide antibiotics",
    description:
      "Point mutation in the 23S rRNA of Propionibacteria shown clinically to confer resistance to macrolides",
  },
  {
    accession: "ARO:3004163",
    name:
      "Mycobacterium abscessus 23S rRNA with mutation conferring resistance to clarithromycin",
    description:
      "Point mutation in the 23S rRNA of Mycobacterium abscessus shown to confer resistance to clarithromycin, a macrolide type antibiotic.",
  },
  {
    accession: "ARO:3004164",
    name:
      "Mycobacterium avium 23S rRNA with mutation conferring resistance to clarithromycin",
    description:
      "Point mutation in the 23S rRNA of Mycobacterium avium shown to confer resistance to clarithromycin, a macrolide type antibiotic.",
  },
  {
    accession: "ARO:3004165",
    name:
      "Mycobacterium chelonae 23S rRNA with mutation conferring resistance to clarithromycin",
    description:
      "Point mutation in the 23S rRNA of Mycobacterium chelonae shown to confer resistance to clarithromycin, a macrolide type antibiotic.",
  },
  {
    accession: "ARO:3004166",
    name:
      "Mycobacterium intracellulare 23S rRNA with mutation conferring resistance to clarithromycin",
    description:
      "Point mutation in the 23S rRNA of Mycobacterium intracellulare shown to confer resistance to clarithromycin, a macrolide type antibiotic.",
  },
  {
    accession: "ARO:3004167",
    name:
      "Mycobacterium intracellulare 23S rRNA with mutation conferring resistance to azithromycin",
    description:
      "Point mutation in the 23S rRNA of Mycobaccterium avium shown to confer resistance to azithromycin, a macrolide type antibiotic.",
  },
  {
    accession: "ARO:3004168",
    name:
      "Mycobacterium kansasii 23S rRNA with mutation conferring resistance to clarithromycin",
    description:
      "Point mutation in the 23S rRNA of Mycobacterium kansasii shown to confer resistance to clarithromycin, a macrolide type antibiotic.",
  },
  {
    accession: "ARO:3004169",
    name:
      "Mycobacterium smegmatis 23S rRNA with mutation conferring resistance to clarithromycin",
    description:
      "Point mutation in the 23S rRNA of Mycobaccterium smegmatis shown to confer resistance to clarithromycin, a macrolide type antibiotic.",
  },
  {
    accession: "ARO:3004170",
    name:
      "Streptococcus pneumoniae 23S rRNA with mutation conferring resistance to macrolide antibiotics",
    description:
      "Point mutation in the 23S rRNA of Streptococcus pneumoniae shown to confer resistance to macrolide type antibiotics",
  },
  {
    accession: "ARO:3004171",
    name:
      "Streptomyces ambofaciens 23S rRNA with mutation conferring resistance to macrolide antibiotics",
    description:
      "Point mutation in the 23S rRNA of Streptomyces ambofaciens shown to confer resistance to macrolide type antibiotics",
  },
  {
    accession: "ARO:3004172",
    name:
      "23S rRNA with mutation conferring resistance to oxazolidinone antibiotics",
    description:
      "Point mutations in the 23S rRNA subunit may confer resistance to oxazolidinone antibiotics",
  },
  {
    accession: "ARO:3004173",
    name:
      "Escherichia coli 23S rRNA with mutation conferring resistance to oxazolidinone antibiotics",
    description:
      "Point mutation in the 23S rRNA of Escherichia coli shown to confer resistance to oxazolidinone type antibiotics",
  },
  {
    accession: "ARO:3004174",
    name:
      "Chlamydia trachomatis 23S rRNA with mutation conferring resistance to macrolide antibiotics",
    description:
      "Point mutation in the 23S rRNA of Chlamydia trachomatis shown to confer resistance to macrolide type antibiotics",
  },
  {
    accession: "ARO:3004176",
    name:
      "Mycoplasma hominis 23S rRNA with mutation conferring resistance to macrolide antibiotics",
    description:
      "Point mutation in the 23S rRNA of Mycoplasma hominis shown to confer resistance to macrolide type antibiotics",
  },
  {
    accession: "ARO:3004177",
    name:
      "Mycoplasma fermentans 23S rRNA with mutation conferring resistance to macrolide antibiotics",
    description:
      "Point mutation in the 23S rRNA of Mycoplasma fermentans shown to confer resistance to macrolide type antibiotics",
  },
  {
    accession: "ARO:3004178",
    name:
      "23S rRNA with mutation conferring resistance to pleuromutilin antibiotics",
    description:
      "Point mutations in the 23S rRNA subunit may confer resistance to pleuromutilin antibiotics",
  },
  {
    accession: "ARO:3004179",
    name:
      "Mycoplasma pneumoniae 23S rRNA mutation conferring resistance to erythromycin",
    description:
      "Point mutation in the 23S rRNA of Mycoplasma pneumoniae shown to confer resistance to erythromycin",
  },
  {
    accession: "ARO:3004180",
    name:
      "Halobacterium halobium 23S rRNA mutation conferring resistance to chloramphenicol",
    description:
      "Point mutation in the 23S rRNA of Halobacterium halobium shown to confer resistance to chloramphenicol",
  },
  {
    accession: "ARO:3004181",
    name:
      "Streptococcus pneumoniae 23S rRNA mutation conferring resistance to macrolides and streptogramins antibiotics",
    description:
      "Point mutation in the 23S rRNA of Streptococcus pneumoniae shown to confer resistance to macrolides and streptogramins types of antibiotics",
  },
  {
    accession: "ARO:3004182",
    name:
      "23S rRNA with mutation conferring resistance to streptogramins antibiotics",
    description:
      "Nucleotide point mutations in the 23S rRNA subunit may confer resistance to streptogramins antibiotics.",
  },
  {
    accession: "ARO:3004183",
    name: "aminosalicylate resistant dihydrofolate reductase",
    description:
      "Antibiotic target replacement dihydrofolate reductase enzymes or domains with catalytic activity that confer resistance to aminosalicylates, esp. p-aminosalicylic acid.",
  },
  {
    accession: "ARO:3004184",
    name:
      "Mycobacterium tuberculosis ribD with mutation conferring resistance to para-aminosalicylic acid",
    description:
      "ribD is a Mycobacterium tuberculosis riboflavin biosynthesis enzyme. Point mutations in ribD cause enzyme overexpression, which allows the C-terminal reductase domain to act as an alternative dihydrofolate reductase. Thus, mutations in ribD confer resistance to DHFR inhibitors such as para-aminosalicylic acid.",
  },
  {
    accession: "ARO:3004185",
    name: "mecD",
    description:
      "mecD is a PBP2a variant identified on a genomic resistance island in Macrococcus caseolyticus. MecD confers resistance to methicillin and other beta-lactam antibiotics through the production of an alternative low-affinity PBP. First described by Schwendener et al. 2017 and identified from canine and bovine sources.",
  },
  {
    accession: "ARO:3004187",
    name:
      "23S rRNA with mutation conferring resistance to lincosamide antibiotics",
    description:
      "Point mutations in the 23S rRNA subunit may confer resistance to lincosamide antibiotics by reducing antibiotic binding-site affinity.",
  },
  {
    accession: "ARO:3004188",
    name:
      "23S rRNA with mutation conferring resistance to phenicol antibiotics",
    description:
      "Point mutations in the 23S rRNA subunit shown clinically to confer resistance to phenicol class antibiotics, including chloramphenicol and florfenicol, by disrupting antibiotic binding-site affinity",
  },
  {
    accession: "ARO:3004189",
    name: "Bla2",
    description:
      "Bla2 is a chromosomal-encoded beta-lactamase, found in Bacillus anthracis, which has penicillin, cephalosporin, and carbapenem-hydrolizing abilities.",
  },
  {
    accession: "ARO:3004190",
    name: "BlaA beta-lactamase",
    description:
      "BlaA beta-lactamases are Class A beta-lactamases first identified in Yersinia enterocolitica and have the ability to hydrolize penicilins and cephalosporins.",
  },
  {
    accession: "ARO:3004191",
    name: "APH(2'')-If",
    description:
      "Aminoglycoside 2''-phosphotransferase identified from the gram-negative pathogen Campylobacter jejuni. APH(2'')-If was shown to confer resistance to 4,6-disubstituted antibiotics kanamycin, tobramycin, dibekacin, gentamicin and sisomicin through antibiotic phosphorylation. Described by Toth et al. 2013.",
  },
  {
    accession: "ARO:3004192",
    name: "cepA beta-lactamase",
    description:
      "cepA beta-lactamases are Class A beta-lactamases found in Bateroides fragilis and have the ability to hydrolyze cephalosporin.",
  },
  {
    accession: "ARO:3004193",
    name: "Sed beta-lactamase",
    description:
      "Sed beta-lactamases are Class A beta-lactamases that are capable of hydrolyzing benzypenicillin, cephalothin, and cloxacillin.",
  },
  {
    accession: "ARO:3004194",
    name: "MCR-1.2",
    description:
      "A plasmid-borne novel MCR-1 functional variant detected in a Klebsiella pneumoniae isolate collected from a rectal swab in Italy. MCR-1.2 differs from MCR-1 by a single amino acid substitution from Glutamine to Leucine at position 3 in the N-terminal region. Described by DiPilato et al. 2016",
  },
  {
    accession: "ARO:3004195",
    name: "NmcA beta-lactamase",
    description:
      "nmcA beta-lactamases are chromosomal-encoded Class A beta-lactamases first isolated from Enterobacter cloacae, specifically a clinical strain known as NOR-1. nmcA beta-lactamases have been shown to hydrolyze carbapenems.",
  },
  {
    accession: "ARO:3004196",
    name: "VCC beta-lactamase",
    description: "VCC beta-lactamases are Class A beta-lactamases.",
  },
  {
    accession: "ARO:3004197",
    name: "blaZ beta-lactamase",
    description:
      "blaZ beta-lactamases are Class A beta-lactamases. These beta-lactamases are responsible for penicillin resistance in Staphylococcus aureus.",
  },
  {
    accession: "ARO:3004198",
    name: "R39 beta-lactamase",
    description:
      "R39 beta-lactamases are Class A beta-lactamases encoded in Actinomadura R39 with the ability to hydrolyze penicillins.",
  },
  {
    accession: "ARO:3004199",
    name: "CepS beta-lactamase",
    description:
      "CepS beta-lactamases are Class C beta-lactamases capable of hydrolyzing cephalosporin.",
  },
  {
    accession: "ARO:3004200",
    name: "CcrA beta-lactamase",
    description:
      "CcrA beta-lactamases are chromosomal-encoded carbapenemase commonly found in Bacteroides fragilis isolates.",
  },
  {
    accession: "ARO:3004201",
    name: "BlaB beta-lactamase",
    description:
      "BlaB beta-lactamases are class B beta-lactamases that are found in a variety of species and have the ability to hydrolyze penams and carbapenems.",
  },
  {
    accession: "ARO:3004202",
    name: "JOHN beta-lactamase",
    description:
      "JOHN beta-lactamases hydrolyse penicillins, narrow- and expanded-spectrum cephalosporins, and carbapenems.",
  },
  {
    accession: "ARO:3004203",
    name: "CGB beta-lactamase",
    description:
      "CGB beta-lactamases are Class B beta-lactamases found in Chryseobacterium gleu that can hydrolyze penicillins; narrow- and expanded-spectrum cephalosporins; and carbapenems.",
  },
  {
    accession: "ARO:3004204",
    name: "EBR beta-lactamase",
    description:
      "EBR beta-lactamases are Class B beta-lactamases first isolated from Empedobacter brevis and are able to hydrolyze penicillins, cephalosporins, and carbapenems.",
  },
  {
    accession: "ARO:3004205",
    name: "TUS beta-lactamase",
    description:
      "TUS beta-lactamases are Class B beta-lactamases that can hydrolyze a variety of beta-lactams, such as cephems and carbapenems",
  },
  {
    accession: "ARO:3004206",
    name: "SIM beta-lactamase",
    description:
      "SIM beta-lactamases are Class B beta-lactamases that are capable of hydrolyzing a wide variety of beta-lactams, including penicillins, narrow- to expanded-spectrum cephalosporins, and carbapenem. The SIM family of beta-lactamases appear to be transferable through integrons.",
  },
  {
    accession: "ARO:3004207",
    name: "KHM beta-latamase",
    description:
      "KHM beta-lactmases are Class B beta-lactamases that can confer resistance to all classes of beta-lactams, except the monobactams.",
  },
  {
    accession: "ARO:3004208",
    name: "DIM beta-lactamase",
    description:
      "DIM type beta-lactamases were first identified from a carbapenem-resistant Pseudomonas stutzeri strain isolated from a Dutch patient. Encoded in mobile elements, these MBLs significantly hydrolyze broad-spectrum cephalosporins and carbapenems.",
  },
  {
    accession: "ARO:3004209",
    name: "HMB beta-lactamase",
    description:
      "First identified from a multi-drug resistant Pseudomonas aeruginosa clinical isolate in 2012, HMB type beta-lactamases can be encoded in transposons and hydrolyze carbapenems.",
  },
  {
    accession: "ARO:3004210",
    name: "SFH beta-lactamase",
    description:
      "This type of Subclass B2 beta-lactamases was first identified from a Serratia fonticola environmental isolate.",
  },
  {
    accession: "ARO:3004211",
    name: "FEZ beta-lactamase",
    description:
      "The FEZ family of beta-lactamases are subclass B3 beta-lactamases that hydrolyze penicillins, cephalosporins, and carbapenems.",
  },
  {
    accession: "ARO:3004212",
    name: "GOB beta-lactamase",
    description:
      "The GOB family of beta-lactamases have been discovered in the Elizabethkingia meningoseptica and are classified as subclass B3 beta-lactamase. They confer resistance to cephalosporins, penicillins, and carbapenems.",
  },
  {
    accession: "ARO:3004213",
    name: "GOB-18",
    description:
      "GOB-18 is a member of subclass B3 GOB family of beta-lactamases isolated from Elizabethkingia meningoseptica. GOB-18 is unlike other subclass B3 beta-lactamases as it is fully active against a broad range of beta-lactam substrates using a single Zn(II) ion in its active site.",
  },
  {
    accession: "ARO:3004214",
    name: "THIN-B beta-lactamase",
    description:
      "Beta-lactamases that are part of the THIN-B family, which is a subclass B3 beta-lactamase family and hydrolyze a broad spectrum of beta-lactams including penicillins, cephalosporins, and carbapenems.",
  },
  {
    accession: "ARO:3004215",
    name: "L1 family beta-lactamase",
    description:
      "This subclass B3 of beta-lactamses have the ability to hydrolyze cephalosporins.",
  },
  {
    accession: "ARO:3004216",
    name: "AIM beta-lactamase",
    description:
      "A subclass B3 family of beta-lactamases that confer resistance to a range of beta-lactam antibiotics including penams, cephamycins, and cephalosporins.",
  },
  {
    accession: "ARO:3004217",
    name: "SMB beta-lactamase",
    description:
      "SMB beta-lactamases are a subclass B3 beta-lactamases that hydrolyze a variety of beta-lactams, including penicillins, cephalosporins, and carbapenems.",
  },
  {
    accession: "ARO:3004218",
    name: "CAU beta-lactamase",
    description: "CAU beta-lactamases are a subclass B3 family.",
  },
  {
    accession: "ARO:3004219",
    name: "BJP beta-lactamase",
    description: "BJP beta-lactamases are a subclass B3 family.",
  },
  {
    accession: "ARO:3004220",
    name: "subclass B3 PEDO beta-lactamase",
    description:
      "PEDO family enzymes that are classified under subclass B3 (metallo-) beta-lactamases.",
  },
  {
    accession: "ARO:3004221",
    name: "CPS beta-lactamase",
    description: "CPS beta-lactamases are a subclass B3 family.",
  },
  {
    accession: "ARO:3004222",
    name: "ESP beta-lactamase",
    description:
      "ESP family beta-lactamases are subclass B3 (metallo-) beta-lactamases.",
  },
  {
    accession: "ARO:3004223",
    name: "MSI beta-lactamase",
    description:
      "MSI beta-lactamases are a family of subclass B3 (metallo-) beta-lactamases.",
  },
  {
    accession: "ARO:3004224",
    name: "SPG beta-lacatamase",
    description:
      "SPG beta-lactamases are a family of subclass B3 (metallo-) beta-lactamases.",
  },
  {
    accession: "ARO:3004225",
    name: "Rm3 family beta-lactamase",
    description: "A family encompassing subclass B3 Rm3-like beta-lactamases.",
  },
  {
    accession: "ARO:3004226",
    name: "subclass B3 LRA beta-lactamase",
    description:
      "Beta-lactamases that are part of the LRA gene family and are classified as B3 (metallo-) beta-lactamases.",
  },
  {
    accession: "ARO:3004227",
    name: "subclass B1 PEDO beta-lactamase",
    description:
      "Beta-lactamases that are part of the PEDO gene family and are classified as subclass B1 (metallo-) beta-lactamases.",
  },
  {
    accession: "ARO:3004228",
    name: "class A LRA beta-lactamase",
    description:
      "Beta-lactamases that are part of the LRA gene family and are classified as Class A beta-lactamases.",
  },
  {
    accession: "ARO:3004229",
    name: "class A Bacillus anthracis Bla beta-lactamase",
    description:
      "Beta-lactamases belonging to the Bla genes from Bacillus anthracis that are classified as class A beta-lactamases.",
  },
  {
    accession: "ARO:3004230",
    name: "subclass B1 Bacillus anthracis Bla beta-lactamase",
    description:
      "Beta-lactamases belonging to the Bla genes from Bacillus anthracis that are classified as subclass B1 beta-lactamases.",
  },
  {
    accession: "ARO:3004231",
    name: "class C LRA beta-lactamase",
    description:
      "Beta-lactamases that are part of the LRA gene family and are classified as Class C beta-lactamases.",
  },
  {
    accession: "ARO:3004232",
    name: "ampC-type beta-lactamase",
    description:
      "AmpC beta-lactamases are clinically important class C beta-lactamase enzymes which confer resistance to cephalosporins and penicillin-like antibiotics. AmpC beta-lactamases are typically found in Enterobacteriaceae, and were described in Escherichia coli in 1940 as the first reported enzymatic deactivation of penicillin. The name AmpC connects these enzymes functionally across many species, however these enzymes are generally unnamed and not phylogenetically related.",
  },
  {
    accession: "ARO:3004233",
    name: "blaF family beta-lactamase",
    description:
      "Class A Beta-lactamases first isolated from Mycobacterium fortuitum.",
  },
  {
    accession: "ARO:3004234",
    name: "EXO beta-lactamase",
    description:
      "Beta-lactamases part of this family discovered in Streptomyces albus G.",
  },
  {
    accession: "ARO:3004235",
    name: "RCP beta-lactamase",
    description:
      "A family of class A beta-lactamases that have been discovered in the Rhodobacter genus.",
  },
  {
    accession: "ARO:3004236",
    name: "gimA family macrolide glycosyltransferase",
    description:
      "This family of macrolide glycosyltransferases derive from gimA, which was discovered in Streptomyces ambofaciens.",
  },
  {
    accession: "ARO:3004237",
    name: "mgt macrolide glycotransferase",
    description:
      "The mgt family encompasses macrolide glycotransferases of the Streptomyces genus.",
  },
  {
    accession: "ARO:3004238",
    name: "sulfonamide resistant sul",
    description:
      "The sul genes encode forms of dihydropteroate synthase that confer resistance to sulfonamide.",
  },
  {
    accession: "ARO:3004239",
    name: "NPS-1",
    description:
      "NPS-1 is a plasmid-encoded class D beta-lactamases identified from two Pseudomonas aeruginosa isolates in 1986.",
  },
  {
    accession: "ARO:3004240",
    name: "NPS beta-lactamase",
    description:
      "NPS beta-lactamases are class D beta-lactamases that have partial hydrolyzing abilities against penicillins and cephalosporin.",
  },
  {
    accession: "ARO:3004241",
    name: "class D LRA beta-lactamase",
    description:
      "Beta-lactamases that are part of the LRA gene family and are classified as Class D beta-lactamases.",
  },
  {
    accession: "ARO:3004242",
    name: "MSI-OXA family beta-lactamase",
    description:
      "Members of the MSI-OXA family are class D beta-lactamases that encompass hybrids of MSI-1 and putative OXA homologues.",
  },
  {
    accession: "ARO:3004243",
    name: "RbpA bacterial RNA polymerase-binding protein",
    description:
      "RbpA is a family of bacterial RNA polymerase-binding proteins, which acts as a transcription factor and binds to the sigma subunit of RNA polymerase.",
  },
  {
    accession: "ARO:3004244",
    name: "DnaA chromosomal replication initiation protein",
    description:
      "The DnaA family of replication initiation proteins interact with RNA polymerase to confer resistance against rifampicin anitibiotics.",
  },
  {
    accession: "ARO:3004245",
    name: "fosC phosphotransferase family",
    description:
      "The fosC family of phosphotransferases phosphorylate fosfomycin to confer resistance and have been found in various bacterial isolates.",
  },
  {
    accession: "ARO:3004246",
    name: "Fom phosphotransferase family",
    description:
      "Two members of the Fom family have been identified, FomA and FomB. FomB must interact with FomA confer resistance to fosfomycin, however FomA is capable of conferring resistance alone.",
  },
  {
    accession: "ARO:3004247",
    name: "GlpT",
    description:
      "Fosfomycin is transported bacterial cells through transporters, one of them being glycerol-3-phosphate, which is encoded by the GlpT gene. Mutations in the GlpT gene can confer resistance to fosfomycin.",
  },
  {
    accession: "ARO:3004248",
    name: "UhpT",
    description:
      "UhpT encodes a transporter that can import fosfomycin-type drugs into bacterial cells. Mutations to UhpT confer resistance.",
  },
  {
    accession: "ARO:3004249",
    name: "UhpA",
    description:
      "UhpA acts as a positive regulator of UhpT, which is a transporter to bring fosfomycin drugs into bacterial cells. Mutations in UhpA that negatively impact the expression of UhpT can confer resistance.",
  },
  {
    accession: "ARO:3004250",
    name: "PtsI phosphotransferase",
    description:
      "PtsI family phosphotransferases are involved in cyclic AMP synthesis, which regulates glpT expression. Mutations in PtsI family genes can negatively affect expression of UhpT, which is needed for fosfomycin import.",
  },
  {
    accession: "ARO:3004251",
    name: "cya adenylate cyclase",
    description:
      "Adenylate cyclases encoded by cya genes, which are involved in the synthesis cyclic AMP which regulates the fosfomycin transporter glpT. Mutations in cya genes can confer resistance to fosfomycin.",
  },
  {
    accession: "ARO:3004253",
    name: "vanUG",
    description: "vanUG is a vanG variant found in the vanG gene cluster.",
  },
  {
    accession: "ARO:3004254",
    name: "vanVB",
    description:
      "vanVB is a vanV variant found in the vanB gene cluster. It is found in some but not all vanB operons in E. faecalis, suggesting the existence of varied gene compositions in vanB operons.",
  },
  {
    accession: "ARO:3004255",
    name: "vanJ membrane protein",
    description:
      "vanJ and vanJ homologue proteins confer resistance to the teicoplanin.",
  },
  {
    accession: "ARO:3004256",
    name: "Bleomycin resistant protein",
    description:
      "Bleomycin resistant proteins (BRP) confer resistance to bleomycin and to bleomycin-like molecules.",
  },
  {
    accession: "ARO:3004257",
    name: "cpa acetyltransferase",
    description:
      "Acetyltransferases of the cpa family confer resistance to capreomycin, an aminoglycoside antibiotic",
  },
  {
    accession: "ARO:3004258",
    name: "amp acetyltransferase",
    description:
      "A family of acetyltransferases that confer resistance to apramycin.",
  },
  {
    accession: "ARO:3004259",
    name: "AAC(1)",
    description:
      "Acetylation of the aminoglycoside antibiotic on the amino group at position 1.",
  },
  {
    accession: "ARO:3004260",
    name: "Bah amidohydrolase",
    description:
      "Bah amidohydrolases are membrane proteins that inactivate bacitracin.",
  },
  {
    accession: "ARO:3004261",
    name: "viomycin phosphotransferase",
    description:
      "Viomycin family of phosphotransferases confer resistance to viomycin antibiotics.",
  },
  {
    accession: "ARO:3004262",
    name: "daptomycin resistant liaF",
    description:
      "Mutations to the liaF accessory protein that confer resistance to daptomycin.",
  },
  {
    accession: "ARO:3004263",
    name: "daptomycin resistant liaR",
    description:
      "Mutations to the liaR response regulator that confer resistance to daptomycin.",
  },
  {
    accession: "ARO:3004264",
    name: "daptomycin resistant liaS",
    description:
      "Mutations in the liaS histidine kinase that confer daptomycin resistance.",
  },
  {
    accession: "ARO:3004265",
    name: "lysocin resistant menA",
    description: "Mutations to demA confer lycosin resistance.",
  },
  {
    accession: "ARO:3004266",
    name: "antibiotic resistant katG",
    description:
      "Point mutations in katG confer antibiotic resistance by preventing katG from catalyzing the activation of antibiotics.",
  },
  {
    accession: "ARO:3004267",
    name: "antibiotic resistant pncA",
    description:
      "Point mutations in pncA prevent the enzyme from activating antibiotics, such as pyrazinamide.",
  },
  {
    accession: "ARO:3004268",
    name: "MCR phosphoethanolamine transferase",
    description:
      "A group of mobile colistin resistance genes encode the MCR family of phosphoethanolamine transferases, which catalyze the addition of phosphoethanolamine onto lipid A, thus interfering with the binding of colistin to the cell membrane.",
  },
  {
    accession: "ARO:3004269",
    name: "pmr phosphoethanolamine transferase",
    description:
      "This family of phosphoethanolamine transferase catalyze the addition of 4-amino-4-deoxy-L-arabinose (L-Ara4N) and phosphoethanolamine to lipid A, which impedes the binding of colistin to the cell membrane.",
  },
  {
    accession: "ARO:3004270",
    name: "antibiotic resistant fabI",
    description:
      "fabI is a enoyl-acyl carrier reductase used in lipid metabolism and fatty acid biosynthesis. The bacterial biocide Triclosan blocks the final reduction step in fatty acid elongation, inhibiting biosynthesis. Point mutations in fabI can confer resistance to Triclosan and Isoniazid.",
  },
  {
    accession: "ARO:3004271",
    name: "16S rRNA methyltransferase (G1405)",
    description:
      "Methyltransferases that methylate the G1405 position of 16S rRNA, which is part of an aminoglycoside binding site.",
  },
  {
    accession: "ARO:3004272",
    name: "16S rRNA methyltransferase (A1408)",
    description:
      "Methyltransferases that methylate the A1408 position of 16S rRNA, which is part of an aminoglycoside binding site.",
  },
  {
    accession: "ARO:3004273",
    name: "Llm 23S ribosomal RNA methyltransferase",
    description:
      "A family of lincosamide resistant 23S rRNA methyltransferases. The only member of the family discovered so far was isolated from Paenibacillus sp. LC231, a strain found in Lechuguilla Cave, NM, USA.",
  },
  {
    accession: "ARO:3004274",
    name: "23S ribosomal RNA methyltransferase",
    description:
      "Methyltransferases that modify the 23S rRNA of the 50S subunit of bacterial ribosomes, conferring resistance to drugs that target 23S rRNA.",
  },
  {
    accession: "ARO:3004275",
    name: "ANT(3'')",
    description:
      "Nucleotidylylation of streptomycin at the hydroxyl group at position 3''",
  },
  {
    accession: "ARO:3004276",
    name: "ANT(2'')",
    description:
      "Nucelotidylylation of streptomycin at the hydroxyl group at position 2''.",
  },
  {
    accession: "ARO:3004277",
    name: "General Bacterial Porin (GBP)",
    description:
      "General bacterial porins are a family of porins found in the outer membrane of Gram-negative bacteria. They catalyze the energy-independent transport of small molecules across the outer membranes of bacteria. Most of the porins of these family are homotrimeric and have variable selectivity, ranging from being cation-selective, anion-selective or selective for different compounds. The most well characterized members of this family are the OmpF and OmpC porins of Escherichia coli. This family is also known as the classical porins.",
  },
  {
    accession: "ARO:3004278",
    name: "Outer Membrane Porin (Opr)",
    description:
      "The Opr family consists of porins in Pseudomonas species, and other Gram-negative bacteria, that exhibit a variety of substrate selectivities.",
  },
  {
    accession: "ARO:3004279",
    name: "Sugar Porin (SP)",
    description:
      "Members of the Sugar Porin family tend to facilitate the transport of maltodextrins and other sugars across the outer membrane of Gram-negative bacteria. These porins form a homotrimeric structure.",
  },
  {
    accession: "ARO:3004280",
    name: "MipA-interacting Protein",
    description:
      "The MltA-interacting Protein (MipA) family consists mainly of homologs to MipA and OmpV proteins. Proteins of this family, are predicted to form a \u00df-barrel.",
  },
  {
    accession: "ARO:3004281",
    name: "General Bacterial Porin with reduced permeability to beta-lactams",
    description:
      "These are GBPs that are associated with decreased susceptibility to beta-lactams either through mutations in the porin protein, absence of the porin protein, or expression of the porin protein.",
  },
  {
    accession: "ARO:3004282",
    name:
      "General Bacterial Porin with reduced permeability to peptide antibiotics",
    description:
      "These are GBPs that are associated with decreased susceptibility to peptide antibiotics either through mutations in the porin protein, absence of the porin protein, or expression of the porin protein.",
  },
  {
    accession: "ARO:3004283",
    name: "CarO porin",
    description:
      "The imipenum resistance-associated CarO porin family is composed of the CarO porin originally identified in Acinetobacter baumannii. The loss of these porins is associated with imipenem and meropenem multi-drug resistance. The channels formed by CarO porins show slight cation selectivity.",
  },
  {
    accession: "ARO:3004284",
    name: "antibiotic resistance fabG",
    description:
      "fabG is a 3-oxoacyl-acyl carrier protein reductase involved in lipid metabolism and fatty acid biosynthesis. The bacterial biocide Triclosan blocks the final reduction step in fatty acid elongation, inhibiting biosynthesis. Point mutations in fabG can confer resistance to Triclosan.",
  },
  {
    accession: "ARO:3004285",
    name: "tunicamycin resistance protein",
    description:
      "A grouping of tunicamycin resistance proteins that are homologous to tmrB found in Bacillus subtilis.",
  },
  {
    accession: "ARO:3004286",
    name: "phenotypic variant regulator",
    description:
      "Phenotypic variant regulator proteins play a role in controlling the switch between antibiotic-susceptible and antibiotic-resistant forms of bacteria. The characterized member of this family is the PvrR protein in Pseudomonas aeruginosa, which when absent, confers antibiotic resistance.",
  },
  {
    accession: "ARO:3004287",
    name: "lipid A phosphatase",
    description:
      "The antimicrobial activity of certain antibiotics, such as peptide antibiotics, is proposed to be initiated through binding to the lipid A moiety of lipopolysaccharides. Thus, covalent modification of Gram-negative bacterial lipid A by phosphatases is a mechanism to reduce the susceptibility of the bacteria to antibiotics.",
  },
  {
    accession: "ARO:3004288",
    name: "subclass B1 Vibrio cholerae varG beta-lactamase",
    description:
      "varG is an Ambler class B metallo-beta-lactamase found on the antibiotic resistance var regulon in Vibrio cholerae, along with an antibiotic efflux pump varABCDEF. These genes are organized as a regulon under the control of VarR transcriptional activator. VarG was shown to have beta-lactamase activity against penicillins, carbapenems and cephalosporins in-vitro.",
  },
  {
    accession: "ARO:3004289",
    name: "Vibrio cholerae varG",
    description:
      "varG is an Ambler class B metallo-beta-lactamase found on the antibiotic resistance var regulon in Vibrio cholerae, along with an antibiotic efflux pump varABCDEF. These genes are organized as a regulon under the control of VarR transcriptional activator. VarG was shown to have beta-lactamase activity against penicillins, carbapenems and cephalosporins in-vitro. Described by Lin HV et al. 2017.",
  },
  {
    accession: "ARO:3004290",
    name: "Escherichia coli ampC beta-lactamase",
    description:
      "A class C ampC beta-lactamase (cephalosporinase) enzyme described in Escherichia coli shown clinically to confer resistance to penicillin-like and cephalosporin-class antibiotics.",
  },
  {
    accession: "ARO:3004291",
    name: "Rhodobacter sphaeroides ampC beta-lactamase",
    description:
      "A periplasmic cephalosporinase described in Rhodobacter sphaeroides shown to contribute to resistance of beta-lactam antibiotics.",
  },
  {
    accession: "ARO:3004292",
    name: "Laribacter hongkongensis ampC beta-lactamase",
    description:
      "A chromosome-encoded class C beta-lactamase described in the pathogen Laribacter hongkongensis, often associated with gastroenteritis. Laribacter ampC has been shown to confer resistance to non-carbapenem beta-lactamases, as described by Lau et al. 2005.",
  },
  {
    accession: "ARO:3004293",
    name: "BUT beta-lactamase",
    description:
      "A class C beta-lactamase family of chromosome-encoded antibiotic resistance genes originally described from Buttiauxella spp.",
  },
  {
    accession: "ARO:3004294",
    name: "BUT-1",
    description:
      "A chromosome-encoded  class C cephalosporinase and penicillinase from Buttiauxella spp. shown clinically to confer resistance to beta-lactam antibiotics. Described by Fihman et al. 2002.",
  },
  {
    accession: "ARO:3004295",
    name: "lassomycin",
    description:
      "A ribosomally synthesized cyclic peptide antibiotic. Lassomycin has been shown to be effective against drug-resistant mycobacteria but shows little activity against other bacteria. Lassomycin targets and binds to a region of the ClpC1 ATPase complex and uncouples ATPase from proteolytic activity, an essential function for mycobacterial viability, causing cell death.",
  },
  {
    accession: "ARO:3004296",
    name: "bulgecin A",
    description:
      "Bulgecin A is a lytic transglycosylase inhibitor and antibiotic adjuvant produced from Pseudomonas mesoacidophila. In the presence of some beta-lactam antibiotics, including ampicillin and meropenem, Bulgecin A has been shown to restore susceptibility to carbapenems in carbapenem-resistant pathogens.",
  },
  {
    accession: "ARO:3004299",
    name: "antimicrobial phenotype",
    description:
      "Qualitative description of a microbe\u2019s response to antimicrobial agents and the expected clinical outcome. Bacteria may be susceptible or resistant to a broad range of antibiotic drugs or drug classes, with several intermediate states or phases.",
  },
  {
    accession: "ARO:3004300",
    name: "intermediate antimicrobial phenotype",
    description:
      "Intermediate sensitivity of a bacterial strain to a given antibiotic occurs when it is inhibited by a concentration of the drug that is associated with uncertain therapeutic effect. The elimination of intermediate phenotype bacteria may depend on the site of infection and/or the concentration of the administered drug.",
  },
  {
    accession: "ARO:3004301",
    name: "resistant antimicrobial phenotype",
    description:
      "A bacterial strain is said to be resistant to a given antibiotic when it is inhibited in vitro by a concentration of this drug that is associated with a high likelihood of therapeutic failure.",
  },
  {
    accession: "ARO:3004302",
    name: "susceptible antimicrobial phenotype",
    description:
      "A bacterial strain is said to be susceptible to a given antibiotic when it is inhibited in vitro by a concentration of a drug that is associated with a high likelihood of therapeutic success",
  },
  {
    accession: "ARO:3004303",
    name: "nonsusceptible antimicrobial phenotype",
    description:
      "A bacterial strain is said to be nonsusceptible to a given antibiotic drug if it demonstrates either an intermediate or resistant phenotype when exposed to a concentration of a drug that is associated with a high likelihood of therapeutic failure, i.e. if the bacterial strain is not susceptible to the antibiotic drug but not necessarily totally resistant under all conditions.",
  },
  {
    accession: "ARO:3004304",
    name: "susceptible dose dependent antimicrobial phenotype",
    description:
      "A bacterial strain is said to be susceptible-dose dependent (SDD) to a given antibiotic when growth is inhibited with a high likelihood of therapeutic success, but when multiple approved dosing options exist. When a higher or more frequent dose is used, the predicted therapeutic result is the same as for susceptible individuals.",
  },
  {
    accession: "ARO:3004305",
    name: "multidrug resistance antimicrobial phenotype",
    description:
      "Multidrug-resistant organisms are defined as bacterial strains that have become resistant to multiple classes of antibacterial drugs or other agents. The phenotype of these organisms is defined by the spectrum of antibiotic drug classes to which that organism is resistant. Often, these organisms are susceptible to a narrow range of available drugs, making treatment difficult and urgent.",
  },
  {
    accession: "ARO:3004306",
    name: "methicillin-resistant Staphylococcus aureus",
    description:
      "Any strain of Staphylococcus aureus that has tested resistant (R) to at least one of methicillin, oxacillin or cefoxitin by standard susceptibility testing methods.",
  },
  {
    accession: "ARO:3004307",
    name: "Linezolid-resistant MRSA",
    description:
      "Methicillin-resistant Staphylococcus aureus that has additionally tested resistant to the antibiotic drug linezolid",
  },
  {
    accession: "ARO:3004308",
    name: "Fluoroquinolone-resistant MRSA",
    description:
      "Methicillin-resistant Staphylococcus aureus that has additionally tested resistant to the fluoroquinolone antibiotics ciprofloxacin and/or levofloxacin",
  },
  {
    accession: "ARO:3004309",
    name: "Vancomycin-intermediate MRSA",
    description:
      "Methicillin-resistant Staphylococcus aureus that has additionally tested intermediate to vancomycin antibiotics.",
  },
  {
    accession: "ARO:3004310",
    name: "Daptomycin-resistant MRSA",
    description:
      "Methicillin-resistant Staphylococcus aureus that has additionally tested non-susceptible (NS, resistant or intermediate) to daptomycin antibiotics",
  },
  {
    accession: "ARO:3004311",
    name: "carbapenem-resistant Acinetobacter",
    description:
      "Any strain of Acinetobacter spp. that has tested either Intermediate or Resistant to at least one of the following carbapenem-class antibiotics by standard testing using approved breakpoints: imipenem, meropenem, doripenem",
  },
  {
    accession: "ARO:3004312",
    name: "multidrug-resistant Acinetobacter",
    description:
      "Any strain of an Acinetobacter spp. that has either tested Intermediate or Resistant to at least one antibiotic drug in at least 3 of the 6 following categories: 1) Extended-spectrum cephalosporins, 2) Fluoroquinolone, 3) Aminoglycosides, 4) Carbapenems, 5) Piperacillin group, and 6) Ampicillin/sulbactam",
  },
  {
    accession: "ARO:3004313",
    name: "carbapenem-resistant Escherichia coli",
    description:
      "Any strain of E. coli that has tested resistant to one of the following carbapenem antibiotics: imipenem, meropenem, doripenem, ertapenem.",
  },
  {
    accession: "ARO:3004314",
    name: "extended-spectrum cephalosporin-resistant Escherichia coli",
    description:
      "Any E.coli that has tested resistant (R) to at least one of the following: ceftriaxone, ceftazidime, cefepime, and cefotaxime",
  },
  {
    accession: "ARO:3004315",
    name: "fluoroquinolone-resistant Escherichia coli",
    description:
      "Any E.coli that has tested resistant (R) to at least one of the following: ciprofloxacin, levofloxacin, and moxifloxacin.",
  },
  {
    accession: "ARO:3004316",
    name: "multidrug-resistant Escherichia coli",
    description:
      "Any E.coli that has tested either intermediate (I) or resistant (R) to at least one drug in at least 3 of the following 5 categories: 1) Extended-spectrum cephalosporins, 2) Fluoroquinolones (ciprofloxacin, levofloxacin, moxifloxacin), 3) Aminoglycosides (amikacin, gentamicin, tobramycin), 4) Carbapenems (imipenem, meropenem, doripenem, ertapenem), 5) Piperacillin Group (piperacillin, piperacillin/tazobactam)",
  },
  {
    accession: "ARO:3004317",
    name: "carbapenem-resistant Enterobacter",
    description:
      "Any Enterobacter spp. that has tested resistant to at least one of the following antibiotic drugs: imipenem, meropenem, doripenem, ertapenem.",
  },
  {
    accession: "ARO:3004318",
    name: "extended-spectrum cephalosporin-resistant Enterobacter",
    description:
      "Any Enterobacter spp. that has tested resistant to at least one of the following antibiotics: ceftriaxone, ceftazidime, cefepime, and/or cefotaxime.",
  },
  {
    accession: "ARO:3004319",
    name: "multidrug-resistant Enterobacter",
    description:
      "Any Enterobacter spp. that has tested either Intermediate or Resistant to at least one drug in at least 3 of the following categories: extended-spectrum cephalosporins, fluoroquinolones, aminoglycosides, carbapenems, and/or pipericillin/tazobactam",
  },
  {
    accession: "ARO:3004320",
    name: "carbapenem-resistant Klebsiella",
    description:
      "Any Klebsiella oxytoca or Klebsiella pneumoniae that has tested resistant to at least one of the following antibiotics: imipenem, meropenem, doripenem, ertapenem.",
  },
  {
    accession: "ARO:3004321",
    name: "extended-spectrum cephalosporin-resistant Klebsiella",
    description:
      "Any Klebsiella oxytoca or Klebsiella pneumoniae that has tested resistant to at least one of the following antibiotics: ceftriaxone, ceftazidime, cefepime, and/or cefotaxime.",
  },
  {
    accession: "ARO:3004322",
    name: "multidrug-resistant Klebsiella",
    description:
      "Any Klebsiella oxytoca or Klebsiella pneumoniae that has tested Intermediate or Resistant in at least 3 of the following categories: extended-spectrum cephalosporins, fluoroquinolones, aminoglycosides, carbapenems, or pipericillin/tazobactam",
  },
  {
    accession: "ARO:3004323",
    name: "carbapenem-resistant Pseudomonas aeruginosa",
    description:
      "Any strain of Pseudomonas aeruginosa that has tested either Intermediate or Resistant to at least one of the following antibiotics: imipenem, meropenem and/or doripenem",
  },
  {
    accession: "ARO:3004324",
    name: "extended-spectrum cephalosporin-resistant Pseudomonas aeruginosa",
    description:
      "Any strain of Pseudomonas aeruginosa that has tested Resistant to at least one of the following antibiotics: cefepime and/or ceftazidime",
  },
  {
    accession: "ARO:3004325",
    name: "MCR-4",
    description:
      "MCR-4 is a plasmid-borne phosphoethanolamine transferase variant of MCR-1, isolated from Salmonella enterica serovar Typhimurium of porcine origin in Italy, Spain and Belgium through 2013 and 2015-2016. MCR-4 confers resistance to collistin via addition of a phosphoethanolamine group to lipid A, reducing the negative charge of the cell membrane. Described by Carattoli et al. 2017.",
  },
  {
    accession: "ARO:3004326",
    name: "fluoroquinolone-resistant Pseudomonas aeruginosa",
    description:
      "Any Pseudomonas aeruginosa that has tested resistant (R) to at least one of the following: ciprofloxacin, levofloxacin",
  },
  {
    accession: "ARO:3004327",
    name: "aminoglycoside-resistant Pseudomonas aeruginosa",
    description:
      "Any Pseudomonas aeruginosa that has tested resistant (R) to at least one of the following: amikacin, gentamicin, tobramycin",
  },
  {
    accession: "ARO:3004328",
    name: "pipericillin-resistant Pseudomonas aeruginosa",
    description:
      "Any Pseudomonas aeruginosa that has tested resistant (R) to at least one of the following: piperacillin, piperacillin/tazobactam",
  },
  {
    accession: "ARO:3004329",
    name: "vancomycin-resistant Enterococcus",
    description:
      "Any Enterococcus faecium or Enterococcus faecalis that has tested resistant (R) to vancomycin",
  },
  {
    accession: "ARO:3004330",
    name: "daptomycin-resistant Enterococcus",
    description:
      "Any Enterococcus faecium or Enterococcus faecalis that has tested non-susceptible (NS) to daptomycin",
  },
  {
    accession: "ARO:3004331",
    name: "vancomycin-resistant coagulase-negative Staphylococci",
    description:
      "Coagulase-Negative Staphylococci that has tested resistant(R) or intermediate(I) to vancomycin",
  },
  {
    accession: "ARO:3004332",
    name: "MCR-5",
    description:
      "MCR-5 is a transposon-associated phosphoethanolamine transferase gene, identified in Salmonella Paratyphi B dTa+ (d-tartrate fermenting Salmonella enterica subsp. enterica serovar Paratyphi B) isolates from food-producing animals. The isolates were collected between 2011 and 2013, and retrieved from the German National Reference Laboratory for Salmonella. MCR-5 confers resistance to collistin through the addition of a phosphoethanolamine group to lipid A, causing a reduction in negative charge of the cell membrane. Described by Borowiak et al, 2017.",
  },
  {
    accession: "ARO:3004333",
    name: "triclosan resistant gyrA",
    description:
      "DNA gyrase is responsible for DNA supercoiling and consists of two alpha and two beta subunits. Point mutations in gyrA have been shown to decrease susceptibility to the antibiotic triclosan. Although the mechanism is unclear, it is hypothesized that changes in supercoiling activity of mutant DNA gyrase proteins alters expression of stress response pathways thereby indirectly decreasing triclosan susceptibility. It has been shown that triclosan does not interact directly with gyrA.",
  },
  {
    accession: "ARO:3004334",
    name:
      "Salmonella enterica gyrA with mutation conferring resistance to triclosan",
    description:
      "Point mutations in Salmonella enterica serovar Typhimurium which have been shown to increase the minimum inhibitory concentration of the antibiotic triclosan. It is hypothesized that decreased susceptibility to triclosan in Salmonella gyrA mutants occurs indirectly due to alterations in the stress response pathways.",
  },
  {
    accession: "ARO:3004335",
    name:
      "Escherichia coli gyrA with mutation conferring resistance to triclosan",
    description:
      "Point mutations in Escherichia coli which have been shown to increase the minimum inhibitory concentration of the antibiotic triclosan. It is hypothesized that decreased susceptibility to triclosan in E. coli gyrA mutants occurs indirectly due to alterations in the stress response pathways.",
  },
  {
    accession: "ARO:3004336",
    name: "PDC-73",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3004337",
    name: "PDC-74",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004338",
    name: "PDC-75",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004339",
    name: "PDC-76",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004340",
    name: "PDC-77",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004341",
    name: "PDC-78",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004342",
    name: "PDC-79",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004343",
    name: "PDC-80",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004344",
    name: "PDC-81",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004345",
    name: "PDC-82",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004346",
    name: "PDC-83",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004347",
    name: "PDC-84",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004348",
    name: "PDC-85",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004349",
    name: "PDC-86",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004350",
    name: "PDC-87",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004351",
    name: "PDC-88",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004352",
    name: "PDC-89",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004353",
    name: "PDC-90",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004354",
    name: "PDC-91",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004355",
    name: "PDC-92",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004356",
    name: "PDC-93",
    description: "An AmpC-like beta-lactamase found in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004357",
    name: "catV",
    description:
      "A chloramphenicol acetyltransferase identified in Brevibacillus brevis and shown to confer resistance to chloramphenicol antibiotics. Described by Pawlowski et al. 2017.",
  },
  {
    accession: "ARO:3004358",
    name: "ACI beta-lactamase",
    description:
      "A family of class A beta-lactamase enzymes described in the gram-negative cocci Acidaminococcus fermentans.",
  },
  {
    accession: "ARO:3004359",
    name: "ACI-1",
    description:
      "A class A beta-lactamase described in Acidaminococcus fermentans. ACI-1 was the first description of a beta-lactamase enzyme in a gram-negative anaerobic cocci. Described by Galan et al 2000.",
  },
  {
    accession: "ARO:3004360",
    name: "antimicrobial resistance testing reference standard",
    description:
      "Uses breakpoints to create standards for interpreting the clinical outcome of an MIC value for a particular microbial species and antimicrobial agent",
  },
  {
    accession: "ARO:3004361",
    name: "sul4",
    description:
      "sul4 is a dihydropteroate synthase gene and mobile sulfonamide resistance gene shown to confer resistance when expressed in E. coli.",
  },
  {
    accession: "ARO:3004362",
    name: "OXA-535",
    description:
      "A chromosome-carried OXA-48-like beta-lactamase, OXA-535, identified in Shewanella spp. JAB-1. Likely the progenitor to the plasmid-carried OXA-436. Expression of OXA-535 in E. coli demonstrated carbapenem-hydrolyzing activity. Described by Jousset et al. 2017.",
  },
  {
    accession: "ARO:3004363",
    name: "lipid A acyltransferase",
    description:
      "Lipid A acyltransferase genes confer resistance to certain types of peptide antibiotics such as polymyxins through the aminoacylation of lipopolysaccharide, thereby decreasing the negative charge of the outer membrane surface.",
  },
  {
    accession: "ARO:3004364",
    name: "almG",
    description:
      "A member of the AlmEFG operon, responsible for polymyxin resistance in Vibrio cholerae. AlmG is a glycyltransferase closely related to lipid A acyltransferases. Described by Henderson JC et al. 2017.",
  },
  {
    accession: "ARO:3004365",
    name: "British Society for Antimicrobial Chemotherapy (BSAC)",
    description:
      "BSAC is a British inter-professional organisation involved in antibiotic education, research and leadership. Baquero F. found that the BSAC system recommends lower breakpoints than the consensus standard of all systems for antimicrobial susceptibility testing used in Europe (BSAC, DIN, SFM, SIR, NCCLS and WRG).",
  },
  {
    accession: "ARO:3004366",
    name: "Clinical and Laboratory Standards Institute (CLSI)",
    description:
      "The Clinical and Laboratory Standards Institute develops and implements clinical laboratory testing standards.",
  },
  {
    accession: "ARO:3004367",
    name: "Deutsches Institut f\u00fcr Normung (DIN)",
    description:
      "DIN is recognized by the Federal Government of Germany as the competent standards organization for Germany and as the national standards body representing Germany in non-governmental international standards organizations. Baquero F. found that the DIN system recommends lower breakpoints than the consensus standard of all systems for antimicrobial susceptibility testing used in Europe (BSAC, DIN, SFM, SIR, NCCLS and WRG).",
  },
  {
    accession: "ARO:3004368",
    name: "European Committee on Antimicrobial Susceptibility Testing (EUCAST)",
    description:
      "EUCAST is a standing committee jointly organized by ESCMID, ECDC and European national breakpoint committees; it deals with breakpoints and technical aspects of phenotypic in-vitro antimicrobial susceptibility testing.",
  },
  {
    accession: "ARO:3004369",
    name: "Soci\u00e9t\u00e9 Fran\u00e7aise de Microbiologie (SFM)",
    description:
      "The French Society of Microbiology (SFM) is a non-profit association which aims to bring together microbiologists from French-speaking countries, working in the domains of bacteria, viruses, Fungi and parasites, and related medical, industrial and environmental microbiology, physiology, genetics, taxonomy, hygiene, and antimicrobial agents. Baquero F. found that the SFM system recommends higher breakpoints than the consensus standard of all systems for antimicrobial susceptibility testing used in Europe (BSAC, DIN, SFM, SIR, NCCLS and WRG).",
  },
  {
    accession: "ARO:3004370",
    name: "minimum inhibitory concentration (MIC)",
    description:
      "Minimum inhibitory concentration (MIC), in microbiology, is the lowest concentration of an antimicrobial that will inhibit the visible growth of a microorganism after overnight incubation.",
  },
  {
    accession: "ARO:3004371",
    name: "MIC value",
    description: "The measured or quantified MIC",
  },
  {
    accession: "ARO:3004372",
    name: "MIC unit",
    description: "mg/L",
  },
  {
    accession: "ARO:3004374",
    name: "malacidin A",
    description:
      "A calcium-dependent cyclic lipopeptide found to be effective against multidrug-resistant Staphylococcus aureus in cutaneous wound infections.",
  },
  {
    accession: "ARO:3004375",
    name: "malacidin biosynthesis",
    description:
      "Biosynthesis of malacidin A and malacidin B calcium-dependent cyclic lipopeptide antibiotics from uncultured soil bacteria.",
  },
  {
    accession: "ARO:3004376",
    name: "bicyclomycin biosynthesis",
    description:
      "Biosynthesis of the broad-spectrum Rho transcription factor inhibitor bicyclomycin by hydroxylation of a tRNA-dependent cyclodipeptide synthase core.",
  },
  {
    accession: "ARO:3004378",
    name: "QepA3",
    description:
      "A plasmid-mediate quinolone efflux pump described in Enterobacter aerogenes and Escherichia coli. QepA3 confers resistance to fluoroquinolone antibiotics by expelling the antibiotic molecules from the cell. This is distinct from the usual chromosomal mutation mechanism of quinolone resistance. Described by Wang et al. 2017.",
  },
  {
    accession: "ARO:3004379",
    name: "QepA4",
    description:
      "A plasmid-mediate quinolone efflux pump variant described in Escherichia coli. QepA4 confers resistance to quinolone and fluoroquinolone antibiotics by expulsion from the cell. Described by Manageiro et al. 2017.",
  },
  {
    accession: "ARO:3004380",
    name: "vaborbactam",
    description:
      "Antibiotic adjuvant and beta-lactamase inhibitor, often paired with meropenem for clinical use.",
  },
  {
    accession: "ARO:3004387",
    name: "in-vivo microbial susceptibility test",
    description:
      "An experimental test conducted inside of a living organism, none of which currently exist for antimicrobial resistance testing",
  },
  {
    accession: "ARO:3004388",
    name: "in-vitro microbial susceptibility test",
    description:
      "An experimental test conducted outside of a living organism, usually in a tube or plate.",
  },
  {
    accession: "ARO:3004389",
    name: "reagent or protocol used for testing AMR",
    description:
      "Reagents and procedures used during microbial susceptibility testing.",
  },
  {
    accession: "ARO:3004390",
    name: "antimicrobial resistance test platform",
    description:
      "System of analysis for determining antimicrobial susceptibility.",
  },
  {
    accession: "ARO:3004391",
    name: "diffusion method",
    description:
      "Microbial species are inoculated in a growth medium onto which antimicrobial agents are transferred by diffusion.",
  },
  {
    accession: "ARO:3004392",
    name: "thin-layer chromatography (TLC)-bioautography",
    description:
      "Components of complex mixtures such as plant extracts are separated across a TLC plate and tested for anti-microbial activity.",
  },
  {
    accession: "ARO:3004393",
    name: "time-kill test",
    description:
      "Allows for time-dependent or concentration-dependent analysis of antimicrobial effects. Several tubes of broth culture medium containing a suspension of the test microbe are treated with varying concentrations of the antimicrobial agent and the percentage of dead cells determined at various time points.",
  },
  {
    accession: "ARO:3004394",
    name: "ATP bioluminescence assay",
    description:
      "Relies on the indirect measurement of microbe populations by quantifying microbial ATP production.",
  },
  {
    accession: "ARO:3004395",
    name: "flow cytofluorometric method",
    description:
      "Microbes exposed to antimicrobial agents are then stained with the intercalating agent propidium iodide. Damaged cells emit a positive signal that is detected by flow cytometry analysis. Flow cytometry can also be used to discriminate between dead, damaged and live cells.",
  },
  {
    accession: "ARO:3004396",
    name: "mechanism-specific test",
    description:
      "Test that directly detect the presence of a particular resistance mechanism.",
  },
  {
    accession: "ARO:3004397",
    name: "broth dilution method",
    description:
      "The microbial species are inoculated in liquid growth medium containing incremental dilutions (usually 2-fold) of the antimicrobial agent.",
  },
  {
    accession: "ARO:3004398",
    name: "automated testing platform",
    description:
      "Automated system of analysis for determining antimicrobial susceptibility.",
  },
  {
    accession: "ARO:3004399",
    name: "manual testing platform",
    description:
      "Non-automated system of analysis for determining antimicrobial susceptibility.",
  },
  {
    accession: "ARO:3004400",
    name: "MicroScan",
    description:
      "Specific automated system of analysis for determining antimicrobial susceptibility developed by Omron.",
  },
  {
    accession: "ARO:3004401",
    name: "Phoenix",
    description:
      "Specific automated system of analysis for determining antimicrobial susceptibility developed by Becton Dickinson.",
  },
  {
    accession: "ARO:3004402",
    name: "Sensitire",
    description:
      "Specific automated system of analysis for determining antimicrobial susceptibility developed by Trek, Thermo Fisher Scientific.",
  },
  {
    accession: "ARO:3004403",
    name: "Vitek System",
    description:
      "Specific automated system of analysis for determining antimicrobial susceptibility developed by bioM\u00e9rieux.",
  },
  {
    accession: "ARO:3004404",
    name: "automated testing platform vendor",
    description:
      "Vendor of automated system of analysis for determining antimicrobial susceptibility.",
  },
  {
    accession: "ARO:3004405",
    name: "Becton Dickinson",
    description:
      "Specific vendor of automated system of analysis for determining antimicrobial susceptibility headquartered in New Jersey, USA. Becton Dickinson is the developer of the Phoenix automated testing platform.",
  },
  {
    accession: "ARO:3004406",
    name: "bioM\u00e9rieux",
    description:
      "Specific vendor of automated system of analysis for determining antimicrobial susceptibility headquartered in France. bioM\u00e9rieux is the developer of the Vitek automated testing platform.",
  },
  {
    accession: "ARO:3004407",
    name: "Siemens",
    description:
      "Specific vendor of automated system of analysis for determining antimicrobial susceptibility headquartered in Berlin and Munich.",
  },
  {
    accession: "ARO:3004408",
    name: "Omron",
    description:
      "Specific vendor of automated system of analysis for determining antimicrobial susceptibility based in Japan. Omron is the developer of the MicroScan automated testing platform.",
  },
  {
    accession: "ARO:3004409",
    name: "Trek",
    description:
      "Specific vendor of automated system of analysis for determining antimicrobial susceptibility that was acquired by Thermo Fisher Scientific Inc. Thermo Fisher Scienfiic Inc. is headquartered in Massachusetts, USA.  Trek is the developer of the Sensitire automated testing platform.",
  },
  {
    accession: "ARO:3004410",
    name: "micro broth dilution method",
    description:
      "Broth dilution conducted on 96-well microtiter plates (\u2264200 \u00b5L per well).",
  },
  {
    accession: "ARO:3004411",
    name: "agar well diffusion method",
    description:
      "This test is a common method of evaluating the antimicrobial activity of plants or microbial extracts. In this method, bacteria are inoculated onto agar plates and the antimicrobial agent is injected into the plate and allowed to diffuse.",
  },
  {
    accession: "ARO:3004412",
    name: "agar plug diffusion method",
    description:
      "This method is commonly used to study the antagonism between microorganisms. The first bacterial strain is inoculated onto agar plates in tight streaks. The bacteria will secrete molecules that diffuse in the agar medium; this medium is cut and placed on another agar plate inoculated with another microorganism.",
  },
  {
    accession: "ARO:3004413",
    name: "cross streak method",
    description:
      "This method is commonly used to study the antagonism between microorganisms. The microbial strain of interest is seeded by a single streak in the center of the agar plate. After incubation, the plate is seeded with the second microorganism by single streak perpendicular to the central streak.",
  },
  {
    accession: "ARO:3004414",
    name: "agar diffusion bioautography",
    description:
      "The antimicrobial agent is transferred from a chromatogram to an agar plate previously inoculated with the test microorganism.",
  },
  {
    accession: "ARO:3004415",
    name: "direct bioautography",
    description:
      "A developed TLC plate is dipped into or sprayed with a microbial suspension and then incubated. Live microbial cells are visualized using tetrazolium salts.",
  },
  {
    accession: "ARO:3004416",
    name: "agar overlay bioautography",
    description:
      "The TLC plate is covered with a molten agar medium seeded with the test microbe and the antimicrobial compounds are allowed to diffuse onto the agar medium. Tetrazolium dye staining allows for visualization.",
  },
  {
    accession: "ARO:3004417",
    name: "inoculum",
    description:
      "Concentration at which bacterial cells in suspension, calculated with respect to final volume (colony forming units/milliliter (cfu/mL)). The standard for broth dilution is 5 x 105cfu/mL. The standard for agar dilution is 104cfu/mL.",
  },
  {
    accession: "ARO:3004418",
    name: "growth medium",
    description: "Medium that supports microorganism growth.",
  },
  {
    accession: "ARO:3004419",
    name: "chromogenic cephalosporinase test",
    description: "Assay for beta lactamase detection",
  },
  {
    accession: "ARO:3004420",
    name: "chloramphenicol acetyltransferase assay (CAT)",
    description:
      "Commercial colorimetric assay for detection of chloramphenicol modifying enzyme chloramphenicol acetyltransferase.",
  },
  {
    accession: "ARO:3004421",
    name: "greater than standard inoculum",
    description:
      "Greater than 5 x 105 cfu/mL for broth dilution and 104 cfu/mL for agar dilution.",
  },
  {
    accession: "ARO:3004422",
    name: "less than standard inoculum",
    description:
      "Less than 5 x 105 cfu/mL for broth dilution and 104 cfu/mL for agar dilution.",
  },
  {
    accession: "ARO:3004423",
    name: "= standard inoculum",
    description:
      "Around equal to 5 x 105 cfu/mL for broth dilution and 104 cfu/mL for agar dilution.",
  },
  {
    accession: "ARO:3004424",
    name: "Mueller-Hinton Broth",
    description:
      "General purpose medium used for culture of nonfastidious microorganisms.",
  },
  {
    accession: "ARO:3004425",
    name: "CO2 supplemented medium",
    description: "Incubation atmosphere supplemented with CO2.",
  },
  {
    accession: "ARO:3004426",
    name: "divalent cation supplemented medium",
    description:
      "Medium supplementation with 20\u201325 mg Ca2+ and 10\u201312.5 mg Mg2+ per liter, which is representative of the divalent cation concentration in blood. Divalent cation supplemented medium is correct to use for susceptibility testing of tetracyclines, daptomycin for gram-positive bacteria and aminoglycosides for Pseudomonas aeruginosarequires.",
  },
  {
    accession: "ARO:3004427",
    name: "NaCl supplemented medium",
    description:
      "Growth medium supplemented with 2% (wt/vol) sodium chloride for testing susceptibility of Staphylococcus aureus to methicillin, oxacillin and nafcillin.",
  },
  {
    accession: "ARO:3004428",
    name: "Difco Mueller-Hinton Broth",
    description: "Low thymine and thymidine content.",
  },
  {
    accession: "ARO:3004429",
    name: "BBL Mueller-Hinton Broth",
    description: "High thymine and thymidine content.",
  },
  {
    accession: "ARO:3004430",
    name: "antimicrobial resistance tissue specificity",
    description:
      "The tissue type used to select breakpoints from a particular standard, for the interpretation of MIC results.",
  },
  {
    accession: "ARO:3004431",
    name: "EUCAST antimicrobial phenotype",
    description:
      "Qualitative description of a microbe\u2019s response to antimicrobial agents as described by EUCAST using breakpoints determined through in-vitro testing.  These breakpoints are determined irrespective of tissue specificity of infection.",
  },
  {
    accession: "ARO:3004432",
    name: "wild-type",
    description:
      "A microbe whose susceptibility to antibiotics is comparable to that of a reference susceptible microbe, meaning that the microbe is sensitive to the particular antibiotic.",
  },
  {
    accession: "ARO:3004433",
    name: "non wild-type",
    description:
      "A microbe whose susceptibility to antibiotics differs from that of a reference susceptible microbe, meaning that the microbe is resistant to the particular antibiotic.",
  },
  {
    accession: "ARO:3004438",
    name: "AMR testing vendor",
    description: "Vendor of AMR testing equipment and reagents.",
  },
  {
    accession: "ARO:3004439",
    name: "component of AMR phenotype measurement",
    description: "Details required in reporting antimicrobial resistance.",
  },
  {
    accession: "ARO:3004440",
    name: "CLSI antimicrobial phenotype",
    description:
      "Qualitative description of a microbe\u2019s response to antimicrobial agents as described by CLSI using breakpoints that take into consideration tissue specificity of infection.",
  },
  {
    accession: "ARO:3004441",
    name: "tet(59)",
    description:
      "Tet(59) is a chromosome-encoded tetracycline efflux pump described from a Chinese pig manure sample.",
  },
  {
    accession: "ARO:3004442",
    name: "tet(W/N/W)",
    description:
      "tet(W/N/W) is a mosaic tetracycline resistance gene and ribosomal protection protein.",
  },
  {
    accession: "ARO:3004443",
    name: "RSA beta-lactamase",
    description:
      "A family of class A beta-lactamase enzymes, RSA beta-lactamases show carbapenemase and cephalosporinase activity",
  },
  {
    accession: "ARO:3004444",
    name: "RSA-1",
    description:
      "RSA-1 is a class A beta-lactamase resistance enzyme identified from a functional metagenomic study of contaminated river sediments.",
  },
  {
    accession: "ARO:3004445",
    name: "RSA-2",
    description:
      "RSA-2 is a class A beta-lactamase enzyme identified from a functional metagenomics study of contaminated river sediments in India.",
  },
  {
    accession: "ARO:3004446",
    name:
      "Haemophilus influenzae PBP3 conferring resistance to beta-lactam antibiotics",
    description:
      "PBP3 is a penicillin-binding protein and beta-lactam resistance enzyme encoded by the ftsI gene in Haemophilus influenzae. Mutations in ftsI confer resistance to beta-lactam antibiotics.",
  },
  {
    accession: "ARO:3004447",
    name: "HERA beta-lactamase",
    description:
      "A class A beta-lactamase family shown to hydrolyze penicillin but without cephalosporinase activity.",
  },
  {
    accession: "ARO:3004448",
    name: "HERA-1",
    description:
      "A class A beta-lactamase with penicillinase activity described in Atlantibacter (Escherichia) hermannii",
  },
  {
    accession: "ARO:3004449",
    name: "TRU beta-lactamase",
    description:
      "A class C beta-lactamase endogenous to Aeromonas enteropelogenes (tructi)",
  },
  {
    accession: "ARO:3004450",
    name: "TRU-1",
    description:
      "A class C beta-lactamase enzyme identified from Aeromonas enteropelogenes. TRU-1 confers resistance to penicillin and cephalosporin antibiotics.",
  },
  {
    accession: "ARO:3004451",
    name: "Agrobacterium fabrum chloramphenicol acetyltransferase",
    description:
      "A chloramphenicol resistance determinant described in the Gram-negative bacterium Agrobacterium fabrum.",
  },
  {
    accession: "ARO:3004452",
    name: "Bacillus clausii chloramphenicol acetyltransferase",
    description:
      "A chloramphenicol resistance determinant described in Bacillus clausii.",
  },
  {
    accession: "ARO:3004453",
    name: "Proteus mirabilis chloramphenicol acetyltransferase",
    description:
      "A chloramphenicol resistance determinant described in Proteus mirabilis.",
  },
  {
    accession: "ARO:3004454",
    name: "Campylobacter coli chloramphenicol acetyltransferase",
    description:
      "A chloramphenicol resistance determinant described in Campylobacter coli.",
  },
  {
    accession: "ARO:3004455",
    name: "Streptococcus suis chloramphenicol acetyltransferase",
    description:
      "A chloramphenicol resistance determinant described in Streptococcus suis.",
  },
  {
    accession: "ARO:3004456",
    name: "Enterococcus faecium chloramphenicol acetyltransferase",
    description:
      "A chloramphenicol resistance determinant described in an Enterococcus faecium plasmid",
  },
  {
    accession: "ARO:3004457",
    name: "Staphylococcus intermedius chloramphenicol acetyltransferase",
    description:
      "A chloramphenicol resistance determinant described in Staphylococcus intermedius",
  },
  {
    accession: "ARO:3004458",
    name: "Enterococcus faecalis chloramphenicol acetyltransferase",
    description:
      "A chloramphenicol acetyltransferase and resistance determinant described in Enterococcus faecalis",
  },
  {
    accession: "ARO:3004459",
    name: "Shigella flexneri chloramphenicol acetyltransferase",
    description:
      "A chloramphenicol acetyltransferase and chloramphenicol resistance determinant described in Shigella flexneri",
  },
  {
    accession: "ARO:3004460",
    name: "Vibrio anguillarum chloramphenicol acetyltransferase",
    description:
      "A chloramphenicol resistance determinant described in a Vibrio anguillarum plasmid sequence",
  },
  {
    accession: "ARO:3004462",
    name: "delafloxacin",
    description:
      "Delafloxacin is a novel fluoroquinolone antibiotic, differentiated from other on-market fluoroquinolones by the absence of a protonatable substituent conferring a weakly acidic character to the molecule. This property results in increased intracellular penetration and enhanced bactericidal activity under acidic conditions.",
  },
  {
    accession: "ARO:3004463",
    name: "EBR-2",
    description:
      "EBR-2 is a novel EBR-like class B beta-lactamase isolated from an extensively drug-resistant strain of Empedobacter falsenii",
  },
  {
    accession: "ARO:3004464",
    name: "CFE-2",
    description:
      "CFE-2 is a plasmid-encoded AmpC beta-lactamase first described from Citrobacter freundii by Chen et al. 2018.",
  },
  {
    accession: "ARO:3004465",
    name: "intrinsic colistin resistant phosphoethanolamine transferase",
    description:
      "Chromosomally-encoded phosphoethanolamine (PEtN) transferases shown to confer resistance to polymyxin antibiotics, including colistin.",
  },
  {
    accession: "ARO:3004466",
    name: "ICR-Mc",
    description:
      "A chromosomally-encoded colistin resistance phosphoethanolamine (PEtN) transferase of Moraxella catarrhalis. ICR-Mc is the protein that represents the closest known ortholog to the colistin resistance MCR-1 and MCR-2 PEtN transferases.",
  },
  {
    accession: "ARO:3004467",
    name: "ciprofloxacin phosphotransferase",
    description:
      "A class of phosphotransferase enzymes shown to confer resistance to ciprofloxacin by antibiotic inactivation through phosphorylation.",
  },
  {
    accession: "ARO:3004468",
    name: "CrpP",
    description:
      "A plasmid-encoded ciprofloxacin resistance protein identified from a Pseudomonas aeruginosa plasmid. CrpP confers resistance to ciprofloxacin by antibiotic inactivation through phosphorylation. Described by Chavez-Jacobo et al. 2018.",
  },
  {
    accession: "ARO:3004469",
    name: "ABC-F ATP-binding cassette ribosomal protection protein",
    description:
      "A subfamily of the ATP-binding cassette protein superfamily. Unlike other ABC proteins, ABC-F genes are not fused to a transmembrane domain nor associated with transport. It has been shown that ABC-F proteins confer antibiotic resistance via ribosomal protection and not antibiotic efflux as in other ABC proteins.",
  },
  {
    accession: "ARO:3004470",
    name: "poxtA",
    description:
      "PoxtA is an ABC-F subfamily ATP-binding cassette protein that confers resistance to tetracycline, -phenicol, and oxazolidinone via modification of the bacterial ribosome. The encoding gene was isolated from a methicillin-resistant Staphylococcus aureus strain.",
  },
  {
    accession: "ARO:3004471",
    name: "msr-type ABC-F protein",
    description:
      "msr-type ABC-F subfamily ribosomal protection proteins expression in Staphylococci species and confer resistance to erythromycin and streptogramin B antibiotics through antibiotic target protection mechanisms.",
  },
  {
    accession: "ARO:3004472",
    name: "lsa-type ABC-F protein",
    description:
      "A subgroup of the ABC-F protein subfamily of ATP-binding cassette proteins. lsa-type ABC-F proteins confer resistance to streptogramin, lincosamide and pleuromutilin antibiotics through antibiotic target protection of the ribosome.",
  },
  {
    accession: "ARO:3004473",
    name: "tebipenem",
    description:
      "A carbapenem antibiotic, administered as the oral drug tebipenem-pivoxil, shown to be effective at treating bacterial infections.",
  },
  {
    accession: "ARO:3004474",
    name: "cefiderocol",
    description:
      "A siderophore cephalosporin with antibiotic activity against Gram-negative pathogens including and specifically Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3004476",
    name: "vmlR",
    description:
      "vmlR is an ABC-F ATPase ribosomal protection protein identified in Bacillus subtilus. Shown to confer resistance to lincomycin and streptogramin A virginiamycin. Described by Crowe-McAuliffe et al. 2018.",
  },
  {
    accession: "ARO:3004477",
    name: "eremomycin pyrrolidide",
    description:
      "Eremomycin pyrrolidide (5) is a semisynthetic glycopeptides 3-6 that is derived from vancomycin or eremomycin. It has high activity against staphylococci and enterococci, including vancomycin-resistant strains .",
  },
  {
    accession: "ARO:3004478",
    name: "OXA-665",
    description:
      "Beta-lactamase found in Acinetobacter rudis efficiently inactivating carbapenems and amoxicillin conferring resistance to cephalosporins",
  },
  {
    accession: "ARO:3004479",
    name: "OXA-664",
    description:
      "Beta-lactamase found in Acinetobacter spp. efficiently inactivating carbapenems and amoxicillin conferring resistance to cephalosporins",
  },
  {
    accession: "ARO:3004480",
    name:
      "Bifidobacterium adolescentis rpoB mutants conferring resistance to rifampicin",
    description:
      "Bifidobacterium are antibiotic resistant probiotics are prescribed to upkeep the population beneficial bacteria in the gut microbiome. However, horizontal gene transfer among gut microbes could create harmful antibiotic-resistant pathogenic bacteria, such as Mycobacterium tuberculosis. Lokesh et al. analyzed Bifidobacterium antitubercular drug resistance and mutations in rpo\u03b2. They found that B. animalis, B. longum and B. adolescentis showed considerable resistance to pyrazinamide, isoniazid, and streptomycin, while B. adolescentis had mutations both in the rifampicin (RIF) pocket and in regions outside the pockets, and also showed considerable resistance to RIF.",
  },
  {
    accession: "ARO:3004489",
    name: "OXA-663",
    description:
      "An OXA-10 family class beta-lactamase identified from a Klebsiella pneumoniae BIDMC 35 isolate.",
  },
  {
    accession: "ARO:3004490",
    name: "delamanid",
    description:
      "A novel nitroimidazole antibiotic for treating Mycobacterium tuberculosis infection. Delamanid inhibits bacterial cell wall growth by mycolic acid synthesis disruption and is particularly effective in combination therapies against multidrug-resistant tuberculosis.",
  },
  {
    accession: "ARO:3004491",
    name: "diarylquinoline antibiotic",
    description:
      "A class of antibiotics used to treat specifically Mycobacterium tuberculosis infection; therefore, referred to as an antimycobacterial. Diarylquinoline antibiotics inhibit ATP synthesis in tuberculosis cells by disruption of mycobacterial ATP synthase.",
  },
  {
    accession: "ARO:3004492",
    name: "bedaquiline",
    description:
      "A diarylquinoline antibiotic drug sold under the brand name Sirturo, used to treat infection from Mycobacterium spp., particularly multidrug-resistant tuberculosis. Bedaquiline disrupts ATP synthase by proton pump blockage, inhibiting ATP synthesis.",
  },
  {
    accession: "ARO:3004493",
    name: "ADC-68",
    description:
      "A class C carbapenemase and extended-spectrum beta-lactamase identified from Acinetobacter baumannii",
  },
  {
    accession: "ARO:3004494",
    name: "IMP-55",
    description:
      "An IMP class B metallo-beta-lactamase enzyme identified from Acinetobacter baumannii.",
  },
  {
    accession: "ARO:3004495",
    name: "IMP-56",
    description:
      "An IMP class B metallo-beta-lactamase enzyme identified from carbapenem-resistant Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004496",
    name: "KPC-24",
    description:
      "A KPC class A beta-lactamase and carbapenemase identified from Chilean isolates of Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3004498",
    name: "dfrB4",
    description:
      "A trimethoprim-resistant dihydrofolate reductase characterized on a class I integron from an E. coli isolate.",
  },
  {
    accession: "ARO:3004499",
    name: "Tosufloxacin",
    description:
      "Tosufloxacin is a fluoroquinolone antibiotic with an extended spectrum of activity.",
  },
  {
    accession: "ARO:3004500",
    name: "MCR-3.11",
    description:
      "An MCR-3 phosphoethanolamine transferase and polymyxin resistance gene variant differing by 2 amino acid substitutions, identified from an Escherichia coli isolate.",
  },
  {
    accession: "ARO:3004501",
    name: "MCR-6.1",
    description:
      "An MCR-2 phosphoethanolamine transferase and polymyxin resistance gene variant identified in Moraxella isolated from pigs in the United Kingdom.",
  },
  {
    accession: "ARO:3004502",
    name: "MCR-2.2",
    description:
      "An MCR-2 phosphoethanolamine transferase and polymyxin resistance gene variant identified in Moraxella isolated from pigs in Great Britain",
  },
  {
    accession: "ARO:3004503",
    name: "MCR-3.6",
    description:
      "An MCR-3 phosphoethanolamine transferase and polymyxin resistance gene variant identified in Aeromonas.",
  },
  {
    accession: "ARO:3004504",
    name: "MCR-3.10",
    description:
      "An MCR-3 phosphoethanolamine transferase and polymyxin (colistin) resistance gene variant identified from Aeromonas, Proteus and Escherichia coli.",
  },
  {
    accession: "ARO:3004505",
    name: "MCR-3.5",
    description:
      "An MCR-3 phosphoethanolamine transferase and polymyxin (colistin) resistance gene variant identified from an extensively-resistant Escherichia coli clinical isolate.",
  },
  {
    accession: "ARO:3004506",
    name: "MCR-1.10",
    description:
      "An MCR-1 phosphoethanolamine transferase and polymyxin (colistin) resistance gene variant identified from an Escherichia porcine isolate in Great Britain.",
  },
  {
    accession: "ARO:3004507",
    name: "MCR-1.9",
    description:
      "An MCR-1 phosphoethanolamine transferase and polymyxin resistance gene variant identified from an enterotoxigenic Escherichia coli clinical isolate.",
  },
  {
    accession: "ARO:3004508",
    name: "MCR-3.9",
    description:
      "An MCR-3 polymyxin (incl. colistin) resistance gene variant isolated from an Aeromonas isolate",
  },
  {
    accession: "ARO:3004509",
    name: "MCR-3.8",
    description:
      "An MCR-3 polymyxin (incl. colistin) resistance gene variant identified from an Aeromonas isolate",
  },
  {
    accession: "ARO:3004510",
    name: "MCR-3.7",
    description:
      "An MCR polymyxin (colistin) resistance gene variant identified in Aeromonas",
  },
  {
    accession: "ARO:3004511",
    name: "MCR-3.2",
    description:
      "An MCR-3 polymyxin resistance gene variant identified in a colistin-resistant Salmonella isolate from Canada, located on an IncHI-2 plasmid",
  },
  {
    accession: "ARO:3004512",
    name: "MCR-3.1",
    description:
      "An MCR-3 polymyxin resistance gene variant identified in colistin-resistance Enterobacteriaceae",
  },
  {
    accession: "ARO:3004513",
    name: "MCR-1.6",
    description:
      "An MCR-1 polymyxin resistance gene variant identified from colistin-resistant Salmonella enterica (serovar typhimurium) isolate, carried by an IncP plasmid.",
  },
  {
    accession: "ARO:3004514",
    name: "MCR-1.3",
    description:
      "An MCR-1 polymyxin resistance gene variant identified from a poultry Raoultella planticola isolate",
  },
  {
    accession: "ARO:3004515",
    name: "MCR-1.4",
    description:
      "An MCR-1 polymyxin resistance gene variant identified from a colistin-resistant Escherichia coli isolate.",
  },
  {
    accession: "ARO:3004516",
    name: "MCR-8",
    description:
      "A novel phosphoethanolamine transferase and mobile colistin resistance gene identified from carbapenem-resistant NDM-1-producing Klebsiella pneumoniae.",
  },
  {
    accession: "ARO:3004517",
    name: "MCR-7.1",
    description:
      "A novel plasmid-mediated colistin-resistant phosphoethanolamine transferase identified from a poultry isolate of Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3004519",
    name: "MIC sign",
    description: "MIC sign (>, <, =) of the MIC value",
  },
  {
    accession: "ARO:3004520",
    name: ">",
    description: "Greater than MIC value.",
  },
  {
    accession: "ARO:3004521",
    name: "<",
    description: "Less than MIC value.",
  },
  {
    accession: "ARO:3004522",
    name: "=",
    description: "Equal to MIC value.",
  },
  {
    accession: "ARO:3004523",
    name: "MIC range",
    description:
      "MIC is a range of two MIC values (e.g., given 10 clinical isolates with a resistance determinant, there are 5 that have an MIC of 25ug/mL and 5 that have an MIC of 100ug/mL agaisnt tobramycin)",
  },
  {
    accession: "ARO:3004524",
    name: "strain type",
    description: "Strain type used for antibiotic susceptibility testing.",
  },
  {
    accession: "ARO:3004525",
    name: "resistance determinant in strain",
    description:
      "Resistant determinant is identified in strain without any experimental methods.",
  },
  {
    accession: "ARO:3004526",
    name: "deletion of resistance determinant",
    description:
      "Experimental deletion of resistance determinant in the chromosome of the strain initially with this resistance determinant.",
  },
  {
    accession: "ARO:3004527",
    name:
      "deletion of resistance determinant with plasmid harbouring resistance determinant",
    description:
      "Experimental protocol where there is a deletion of resistance determinant and the transformation of a plasmid harbouring resistance determinant into the same strain.",
  },
  {
    accession: "ARO:3004528",
    name: "plasmid harbouring resistance determinant",
    description:
      "Experimental procedure where plasmid harbouring a resistance determinant is transformed into a strain.",
  },
  {
    accession: "ARO:3004529",
    name: "resistance determinant not in strain",
    description:
      "Without any experimental protocol, the resistance determinant is absent from the strain.",
  },
  {
    accession: "ARO:3004530",
    name: "insertion of resistance determinant",
    description:
      "Experimental insertion of resistance determinant in the chromosome of the strain initially without this resistance determinant.",
  },
  {
    accession: "ARO:3004531",
    name: "strain",
    description:
      "Specific strain of bacteria that was used to perform the antibiotic susceptibility test.",
  },
  {
    accession: "ARO:3004532",
    name: "strain source",
    description:
      "Source of the strain used in the antibiotic susceptibility test (e.g., clinical, environmental, agriculture)",
  },
  {
    accession: "ARO:3004533",
    name: "clinical source",
    description: "Strain isolated in a clinical setting (e.g., from a patient)",
  },
  {
    accession: "ARO:3004534",
    name: "environmental source",
    description: "Strain collected from environment (e.g., soil, water)",
  },
  {
    accession: "ARO:3004535",
    name: "agricultural source",
    description:
      "Strain collected from agricultural setting (e.g., chicken, pig)",
  },
  {
    accession: "ARO:3004536",
    name: "antimicrobial susceptibility test control",
    description:
      "The control of the antibiotic susceptibility testing, whether there was an experimental control (same strain, with and without resistance determinant) or if there was a group of clinical isolates found to have a resistance determinant compared against a strain without that resistance determinant.",
  },
  {
    accession: "ARO:3004537",
    name: "experimental control",
    description:
      "Antimicrobial susceptibility testing was performed on the same strain with and without a resistance determinant.",
  },
  {
    accession: "ARO:3004538",
    name: "correlative control",
    description:
      "Antibiotic susceptibility testing is performed on a group of clinical isolates with a resistance determinant compared to different strains that do not have this resistance determinant. No experiment was performed to delete / insert the resistance determinant to determine antibiotic susceptibility using the same genetic background.",
  },
  {
    accession: "ARO:3004539",
    name: "mphH",
    description:
      "A chromosomal macrolide 2'-phosphotransferase and resistance gene identified from a Brachybacterium faecium cave isolate",
  },
  {
    accession: "ARO:3004540",
    name: "experimental source",
    description:
      "Source is an experimental, commonly used lab strain (e.g., P. aeruginosa PAO1).",
  },
  {
    accession: "ARO:3004541",
    name: "mphK",
    description:
      "A chromosomal macrolide phosphotransferase identified from Bacillus subtilis",
  },
  {
    accession: "ARO:3004542",
    name: "mphN",
    description:
      "A macrolide phosphotransferase identified on a plasmid in Exiguobacterium",
  },
  {
    accession: "ARO:3004543",
    name: "mphO",
    description:
      "A chromosomal macrolide phosphostransferase identified in Brachybacterium paraconglomeratum",
  },
  {
    accession: "ARO:3004544",
    name: "mphJ",
    description:
      "A chromosomal macrolide phosphotransferase identified in Brevibacillus brevis VM4.",
  },
  {
    accession: "ARO:3004545",
    name: "ADC beta-lactamase with carbapenemase activity",
    description:
      "ADC beta-lactamase enzymes with demonstrable carbapenemase activity, an exception amongst these enzymes.",
  },
  {
    accession: "ARO:3004546",
    name:
      "Campylobacter jejuni 23S rRNA with mutation conferring resistance to erythromycin",
    description:
      "Point mutations in the 23S ribosomal RNA domain of Campylobacter jejuni which confer resistance to macrolide antibiotics, including erythromycin.",
  },
  {
    accession: "ARO:3004547",
    name: "dfrA6",
    description:
      "A dihydrofolate reductase and antibiotic resistance gene identified from an integron in Vibrio cholerae",
  },
  {
    accession: "ARO:3004548",
    name: "dfrA9",
    description:
      "A dihydrofolate reductase and trimethoprim resistance gene identified from porcine isolates of Esherichia coli",
  },
  {
    accession: "ARO:3004549",
    name: "dfrB5",
    description:
      "A dihydrofolate reductase and trimethoprim resistance gene identified from an integron in Pseudomonas aeruginosa",
  },
  {
    accession: "ARO:3004550",
    name: "dfrA27",
    description:
      "A dihydrofolate reductase and trimethoprim resistance gene from non-O1, non-O139 Vibrio cholerae",
  },
  {
    accession: "ARO:3004551",
    name: "dfrA28",
    description:
      "A dihydrofolate reductase and trimethoprim resistance gene from Aeromonas hydrophilia",
  },
  {
    accession: "ARO:3004552",
    name: "dfrA30",
    description:
      "A dihydrofolate reductase and trimethoprim resistance gene from Klebsiella pneumoniae",
  },
  {
    accession: "ARO:3004553",
    name: "dfr-lie",
    description: "A dihydrofolate reductase and trimethoprim resistance gene",
  },
  {
    accession: "ARO:3004554",
    name: "dfrA29",
    description:
      "A dihydrofolate reductase and trimethoprim resistance gene isolated from canine Escherichia coli infection",
  },
  {
    accession: "ARO:3004555",
    name: "dfrA32",
    description:
      "A dihydrofolate reductase and trimethoprim resistance gene identified from non-typhoidal multidrug-resistant Salmonella enterica",
  },
  {
    accession: "ARO:3004556",
    name: "dfrB7",
    description:
      "A dihydrofolate reductase and trimethoprim resistance gene identified on an integron in multidrug-resistant gram-negative bacteria",
  },
  {
    accession: "ARO:3004558",
    name: "CAM beta-lactamase",
    description:
      "CAM (Central Alberta Metallo) beta-lactamases are class B metallo-beta-lactamases and carbapenemases found to confer resistance to broad spectrum antibiotics in Pseudomonas aeruginosa.",
  },
  {
    accession: "ARO:3004559",
    name: "CAM-1",
    description:
      "CAM-1 is a CAM (Central Alberta Metallo) beta-lactamase and carbapenemase identified from 4 clinical isolates of Pseudomonas aeruginosa from a Canadian hospital",
  },
  {
    accession: "ARO:3004560",
    name: "murG transferase",
    description:
      "murG acetylglucosaminyltransferase enzymes inhibited by vancomycin. Mutations in murG confer resistance to vancomycin through antibiotic target alteration.",
  },
  {
    accession: "ARO:3004561",
    name:
      "Clostridium difficile murG with mutation conferring resistance to vancomycin",
    description:
      "Mutations in murG which confer resistance to vancomycin in Clostridium difficile",
  },
  {
    accession: "ARO:3004562",
    name: "Clostridium difficile gyrB conferring resistance to fluoroquinolone",
    description:
      "Point mutations in gyrB of Clostridium (Clostridioides) difficile conferring resistance to fluoroquinolone antibiotics",
  },
  {
    accession: "ARO:3004563",
    name:
      "Clostridium difficile rpoB with mutation conferring resistance to rifampicin",
    description:
      "Point mutations in the rpoB region of Clostridium (Clostridioides) difficile which confer resistance to rifampin and rifampicin antibiotics",
  },
  {
    accession: "ARO:3004564",
    name: "beta-lactam potentiator",
    description:
      "Beta-lactam potentiators are a type of resistance-modifying agent (RMA) which do not impose antibacterial or beta-lactamase inhibitory ability on their own, but improve the performance of antibiotic drugs, specifically beta-lactams.",
  },
  {
    accession: "ARO:3004565",
    name: "tetracyclic indoline",
    description:
      "A class of resistance-modifying agents and beta-lactam potentiators which selectively potentiate beta-lactam antibiotics, specifically Amoxicillin-clavulanic acid, in methicillin-resistant Staphylococcus aureus (MRSA) without their own antimicrobial activity.",
  },
  {
    accession: "ARO:3004566",
    name: "Rifalazil",
    description:
      "Rifalazil is a rifamycin derivative antibiotic, sometimes used for treating Clostridioides difficile infections.",
  },
  {
    accession: "ARO:3004567",
    name: "Ramoplanin",
    description:
      "Ramoplanin is a glycolipodepsipeptide antibiotic derived from gram-positive Actinoplanes used to treat infection with Clostridioides difficile",
  },
  {
    accession: "ARO:3004568",
    name: "dfrA18",
    description:
      "A trimethoprim-resistant dihydrofolate reductase identified from Vibrio cholerae",
  },
  {
    accession: "ARO:3004569",
    name: "ICR-Mo",
    description:
      "A chromosomally-encoded colistin resistance phosphoethanolamine (PEtN) transferase of Moraxella osloensis. ICR-Mo represents the closest known ortholog to the colistin resistance MCR-1 and MCR-2 PEtN transferases",
  },
  {
    accession: "ARO:3004570",
    name: "PNGM beta-lactamase",
    description:
      "PNGM is a family of metallo-beta-lactamases first identified from a deep-sea sediment metagenome and shown to reduce antibiotic susceptibility by hydrolysis of penicillins, carbapenems and extended- and broad-spectrum cephalosporins.",
  },
  {
    accession: "ARO:3004571",
    name: "PNGM-1",
    description:
      "PNGM-1 is a subclass B3 metallo-beta-lactamase first identified from a deep-sea sediment meta-genome. PNGM-1 is shown to reduce susceptibility to extended- and broad-spectrum cephalosporins, carbapenems, and penicillins. PNGM-1 is a novel metallo-beta-lactamase which predates the current antibiotic era.",
  },
  {
    accession: "ARO:37",
    name: "skin disease",
    description: "An integumentary system disease that is located in the skin.",
  },
  {
    accession: "ARO:850",
    name: "lung disease",
    description:
      "A lower respiratory tract disease in which the function of the lungs is adversely affected by narrowing or blockage of the airways resulting in poor air flow, a loss of elasticity in the lungs that produces a decrease in the total volume of air that the lungs are able to hold, and clotting, scarring, or inflammation of the blood vessels that affect the ability of the lungs to take up oxygen and to release carbon dioxide.",
  },
  {
    accession: "ARO:85540",
    name: "microbial susceptibility test",
    description:
      "A method to determine whether microbial growth is affected by an antimicrobial compound",
  },
  {
    accession: "ARO:85555",
    name: "agar dilution method",
    description:
      "A method to determine microbial susceptibility to antibiotics in which test bacteria is inoculated onto agar plates with serial dilutions of know quantities of antibiotics. The minimum inhibitory concentration of antibiotic is assessed.",
  },
  {
    accession: "ARO:85595",
    name: "disk-diffusion method",
    description:
      "A method to determine microbial susceptibility to antibiotics wherein bacteria are inoculated onto agar plates and the antimicrobial agent diffused onto the agar plate from a filter paper disk. An approximate, but not exact, MIC is determined by the diameter of growth inhibition zones. This method has been found to have good correlation between in-vivo data and in-vitro data.",
  },
  {
    accession: "ARO:85596",
    name: "antimicrobial gradient method (E-test)",
    description:
      "A method to determine microbial susceptibility to antibiotics in which a plastic strip impregnated with the antibiotic of interest is placed on an agar plate that has been inoculated with bacteria. The antibiotic diffuses from the strip producing a concentration gradient of drug on the agar. The point at which the elliptical shaped area of growth inhibition meets the strip is the minimum inhibitory concentration of the drug of interest.",
  },
  {
    accession: "ARO:85701",
    name: "macro broth dilution method",
    description:
      "A method to determine the minimum inhibitory concentration (MIC) of antibiotics in which growth media is pipetted into a multi-well plate. Each well contains a known concentration of drug which is inoculated with a known quantity of bacteria. The plate is incubated for a specified period of time and each well is assessed for bacterial cell growth. The drug concentration in the first well showing no growth is known as the MIC for that antibiotic. Minimum 2 mL total broth dilution volume.",
  },
];
